#########################
#
# R code for the study:
# Next generation restoration metrics: Using soil eDNA bacterial community data
# to measure trajectories towards rehabilitation targets
#
# Craig Liddicoat, Siegfried L. Krauss, Andrew Bissett, Ryan J. Borrett,
# Luisa C. Ducki, Shawn D. Peddle, Paul Bullock, Mark P. Dobrowolski, Andrew Grigg,
# Mark Tibbett, Martin F. Breed
#
# Outline: this code applies 16S rRNA gene amplicon bacterial community data 
# to develop and explore microbiota-based 'rehabilitation trajectory assessments'.
#
# Case study minesites: 
# 1. Alcoa - Huntly (sampled in 2016; rehab 2-29 yr old)
# 2. Iluka - Eneabba (sampled in 2019; rehab 7-38 yr old)
# 3. South32 - Worsley (sampled in 2019; rehab 2-28 yr old)
#
#########################

# record library and version info
.libPaths() # "/Library/Frameworks/R.framework/Versions/4.0/Resources/library"

R.Version()
# "R version 4.0.3 (2020-10-10)"
citation()
# R Core Team (2020). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. URL
# https://www.R-project.org/.

library(readxl); packageVersion("readxl") # '1.3.1'
library(plyr); packageVersion("plyr") # '1.8.6'
library(dplyr); packageVersion("dplyr") # '1.0.2'
library(vegan);packageVersion("vegan") # '2.5.7'
library(phyloseq); packageVersion("phyloseq") # '1.34.0'
library(ggplot2); packageVersion("ggplot2") # '3.3.2'
library(grid); packageVersion("grid") #  '4.0.3'
library(reshape2); packageVersion("reshape2") # '1.4.4'
library(tidyr); packageVersion("tidyr") # '1.1.2'
library(FSA); packageVersion("FSA") # '0.8.31'
library(rcompanion); packageVersion("rcompanion") # '2.3.26'
library(decontam); packageVersion("decontam") #  '1.10.0'
library(biomformat); packageVersion("biomformat") # '1.18.0'
library(seqinr); packageVersion("seqinr") # '4.2.5'
library(dada2); packageVersion("dada2") # '1.18.0'
library(geodist); packageVersion("geodist") # '0.0.6'
library(DECIPHER); packageVersion("DECIPHER") # ‘2.18.1’
library(speedyseq); packageVersion("speedyseq") # ‘0.5.3.9001’
library(drc); packageVersion("drc") # ‘3.0.1’
library(aomisc); packageVersion("aomisc") # ‘0.64’
library(caret); packageVersion("caret") # ‘6.0.86’
library(corrplot); packageVersion("corrplot") #  '0.84'
library(sp); packageVersion("sp") #  '1.4.4'
library(parallel); packageVersion("parallel") #  '4.0.3'
library(doParallel); packageVersion("doParallel") #  ‘1.0.16’


#########################
##save.image("/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling/WORKSPACE-vFinal.RData")
##load("/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling/WORKSPACE-vFinal.RData")
#########################


workdir <- "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"
setwd(workdir)
getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"

datadir <- "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/data"

par.default <- par()


#### 1a. Alcoa post-mining restoration chronosequence
#    - create phyloseq object from metadata, OTU abundances, taxonomy
#    - data cleaning
#    - build and add phylogenetic tree
#-------------------------

### metadata

context <- read_excel(path= paste0(datadir,"/alcoa/","Alcoa-BASE-Metadata-April2018.xlsx"),
                      sheet=1, range="A1:BR73", col_names = TRUE)

context <- as.data.frame(context)
str(context)

names(context)
# [1] "FULL_ID"                                               "Sample_id"                                            
# [3] "Date sampled"                                          "lat raw"                                              
# [5] "lon raw"                                               "lat (-) in decimal degrees"                           
# [7] "long in decimal degrees"                               "Depth"                                                
# [9] "Horizon controlled vocab (1)"                          "Notes"                                                
# [11] "Location description"                                  "Current land-use controlled vocab (2)"                
# [13] "General Ecological Zone controlled vocab (3)"          "Vegetation Type (descriptive)"                        
# [15] "Vegetation Type Controlled vocab (4)"                  "Vegetation Total cover (%)"                           
# [17] "Vegetation Dom. Trees (%)"                             "Vegetation Dom. Shrubs (%)"                           
# [19] "Vegetation Dom. Grasses (%)"                           "Elevation (m)"                                        
# [21] "Slope (%)"                                             "Slope Aspect (Direction or degrees; e.g., NW or 315?)"
# [23] "Profile Position controlled vocab (5)"                 "Other comments"                                       
# [25] "Australian Soil Classification controlled vocab (6)"   "Method of Australian soil classification"             
# [27] "FAO soil classification controlled vocab (7)"          "Method of FAO soil classification"                    
# [29] "Immediate Previous Land Use controlled vocab (2)"      "Date since change in Land Use"                        
# [31] "1yr since present"                                     "2yrs since present"                                   
# [33] "3yrs since present"                                    "4yrs since present"                                   
# [35] "5yrs since present"                                    "Agrochemical Additions"                               
# [37] "Tillage controlled vocab (9)"                          "Fire"                                                 
# [39] "Fire intensity if known"                               "Flooding"                                             
# [41] "Extreme Events"                                        "Sample ID"                                            
# [43] "Soil moisture (%)"                                     "Color controlled vocab (10)"                          
# [45] "Gravel (%) - ( >2.0 mm)"                               "Texture ()"                                           
# [47] "Course Sand (%) (200-2000 ?m)"                         "Fine Sand (%) - (20-200 ?m)"                          
# [49] "Sand (%)"                                              "Silt  (%) (2-20 ?m)"                                  
# [51] "Clay (%) (<2 ?m)"                                      "Ammonium Nitrogen (mg/Kg)"                            
# [53] "Nitrate Nitrogen (mg/Kg)"                              "Phosphorus Colwell (mg/Kg)"                           
# [55] "Potassium Colwell (mg/Kg)"                             "Sulphur (mg/Kg)"                                      
# [57] "Organic Carbon (%)"                                    "Conductivity (dS/m)"                                  
# [59] "pH Level (CaCl2) (pH)"                                 "pH Level (H2O) (pH)"                                  
# [61] "DTPA Copper (mg/Kg)"                                   "DTPA Iron (mg/Kg)"                                    
# [63] "DTPA Manganese (mg/Kg)"                                "DTPA Zinc (mg/Kg)"                                    
# [65] "Exc. Aluminium (meq/100g)"                             "Exc. Calcium (meq/100g)"                              
# [67] "Exc. Magnesium (meq/100g)"                             "Exc. Potassium (meq/100g)"                            
# [69] "Exc. Sodium (meq/100g)"                                "Boron Hot CaCl2 (mg/Kg)"

row.names(context) <- context$Sample_id
str(context)

row.names(context)
# [1] "39041" "39042" "39043" "39044" "39045" "39046" "39047" "39048" "39049" "39050" "39051" "39052" "39053" "39054" "39055" "39056"
# [17] "39057" "39058" "39059" "39060" "39061" "39062" "39063" "39064" "39065" "39066" "39067" "39068" "39069" "39070" "39071" "39072"
# [33] "39073" "39074" "39075" "39076" "39077" "39078" "39079" "39080" "39081" "39082" "39083" "39084" "39085" "39086" "39087" "39088"
# [49] "39089" "39090" "39091" "39092" "39093" "39094" "39095" "39096" "39097" "39098" "39099" "39100" "39161" "39162" "39163" "39164"
# [65] "39165" "39166" "39191" "39192" "39193" "39194" "39195" "39196"

table(context$Depth)
# 10 20 
# 36 36

sel <- which(context$Depth == 10) # qty 36

context <- context[sel, ]


samp.16s <- context
row.names(samp.16s)
row.names(samp.16s) <- paste0("X",row.names(samp.16s))

write.csv(context, file = "alcoa-metadata.csv")


head(context$FULL_ID)
# [1] "102.100.100/39041" "102.100.100/39043" "102.100.100/39045" "102.100.100/39047" "102.100.100/39049"
# [6] "102.100.100/39051"

AMI_IDs <- context$FULL_ID


### OTU raw data matrix -- use recent AMI download

# Australian Microbiome OTU Database - tabular export
# 
#   Amplicon filter:
#   amplicon is '27f519r_bacteria'
# 
# Taxonomy filter:
#   kingdom is 'Bacteria'
# 
# Contextual filter:
#   environment is 'Soil'
# Ncbi Bioproject Accession contains "PRJNA317932"
# 116.03<=Longitude<=116.23
# 
# Australian Microbiome Database Metadata:
#   Dataset methodology=v1
# Dataset revision date=2020-04-15



raw.otu.16s <- read.csv(file = "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/data/alcoa/AustralianMicrobiome-2021-02-01T144346-csv/Bacteria.csv",header = TRUE )
raw.otu.16s <- as.data.frame(raw.otu.16s)

str(raw.otu.16s)
# 'data.frame':	588917 obs. of  11 variables:
# $ Sample.ID: chr  "102.100.100/39041" "102.100.100/39041" "102.100.100/39041" "102.100.100/39041" ...
# $ OTU      : chr  "AAAGAACGCTGGCGGCATGCTTAACACATGCAAGTCGAACGAGTAGTAGCAATACTATTAGTGGCAGACGGGTGAGTAATACGTAGGAATCTCCCCTATAGTACGGAACAA"| __truncated__ "AAAGAACGCTGGCGGCATGCTTAACACATGCAAGTCGAACGGATAGTAGCAATACTATCAGTGGCAAACGGGTGAGTAATACGTAGGAATCTACCCTGTAGTACGGAACAA"| __truncated__ "AAAGAACGCTGGCGGCATGCTTAACACATGCAAGTCGAACGGATGGTAGCAATACCATCAGTGGCAAACGGGTGAGTAATACGTAGGAATCTACCCTGTAGTACGGAACAA"| __truncated__ "AAATAACGCTGGCGGCATGCTTAACACATGCAAGTCGAACGAATAGTAGCAATACTATTAGTGGCAAACGGGTGAGTAACACGTAGGAATCTCCCCTATAGTATGGAATAA"| __truncated__ ...
# $ OTU.Count: int  12 1 4 5 2 2 2 1 1 1 ...
# $ Amplicon : chr  "27f519r_bacteria" "27f519r_bacteria" "27f519r_bacteria" "27f519r_bacteria" ...
# $ Kingdom  : chr  "Bacteria" "Bacteria" "Bacteria" "Bacteria" ...
# $ Phylum   : chr  "Proteobacteria" "Proteobacteria" "Proteobacteria" "Proteobacteria" ...
# $ Class    : chr  "Alphaproteobacteria" "Alphaproteobacteria" "Alphaproteobacteria" "Alphaproteobacteria" ...
# $ Order    : chr  "Rickettsiales" "Rickettsiales" "Rickettsiales" "Rickettsiales" ...
# $ Family   : chr  "Rickettsiaceae" "Rickettsiaceae" "Rickettsiaceae" "Rickettsiaceae" ...
# $ Genus    : chr  "uncultured" "uncultured" "uncultured" "uncultured" ...
# $ Species  : logi  NA NA NA NA NA NA ...

# this is in long format, not wide format !!

names(raw.otu.16s)
# [1] "Sample.ID" "OTU"       "OTU.Count" "Amplicon"  "Kingdom"   "Phylum"    "Class"     "Order"     "Family"
# [10] "Genus"     "Species"

head(raw.otu.16s$Sample.ID)
# [1] "102.100.100/39041" "102.100.100/39041" "102.100.100/39041" "102.100.100/39041" "102.100.100/39041"
# [6] "102.100.100/39041"

dim(raw.otu.16s) # 588917     11


## only need rows corresponding to Alcoa samples

dim(context) # 36 70

sel.keep <- which(raw.otu.16s$Sample.ID %in% AMI_IDs )
length(unique(raw.otu.16s$Sample.ID[sel.keep])) # 36
sort( unique(raw.otu.16s$Sample.ID) )

# [1] "102.100.100/39041" "102.100.100/39042" "102.100.100/39043" "102.100.100/39044" "102.100.100/39045"
# [6] "102.100.100/39046" "102.100.100/39047" "102.100.100/39048" "102.100.100/39049" "102.100.100/39050"
# [11] "102.100.100/39051" "102.100.100/39052" "102.100.100/39053" "102.100.100/39054" "102.100.100/39055"
# [16] "102.100.100/39056" "102.100.100/39057" "102.100.100/39058" "102.100.100/39059" "102.100.100/39060"
# [21] "102.100.100/39061" "102.100.100/39062" "102.100.100/39063" "102.100.100/39064" "102.100.100/39065"
# [26] "102.100.100/39066" "102.100.100/39067" "102.100.100/39068" "102.100.100/39069" "102.100.100/39070"
# [31] "102.100.100/39071" "102.100.100/39072" "102.100.100/39073" "102.100.100/39074" "102.100.100/39075"
# [36] "102.100.100/39076" "102.100.100/39077" "102.100.100/39078" "102.100.100/39079" "102.100.100/39080"
# [41] "102.100.100/39081" "102.100.100/39082" "102.100.100/39083" "102.100.100/39084" "102.100.100/39085"
# [46] "102.100.100/39086" "102.100.100/39087" "102.100.100/39088" "102.100.100/39089" "102.100.100/39090"
# [51] "102.100.100/39091" "102.100.100/39092" "102.100.100/39093" "102.100.100/39094" "102.100.100/39095"
# [56] "102.100.100/39096" "102.100.100/39097" "102.100.100/39098" "102.100.100/39099" "102.100.100/39100"
# [61] "102.100.100/39161" "102.100.100/39162" "102.100.100/39163" "102.100.100/39164" "102.100.100/39165"
# [66] "102.100.100/39166" "102.100.100/39191" "102.100.100/39192" "102.100.100/39193" "102.100.100/39194"
# [71] "102.100.100/39195" "102.100.100/39196"

AMI_IDs
# [1] "102.100.100/39041" "102.100.100/39043" "102.100.100/39045" "102.100.100/39047" "102.100.100/39049"
# [6] "102.100.100/39051" "102.100.100/39053" "102.100.100/39055" "102.100.100/39057" "102.100.100/39059"
# [11] "102.100.100/39061" "102.100.100/39063" "102.100.100/39065" "102.100.100/39067" "102.100.100/39069"
# [16] "102.100.100/39071" "102.100.100/39073" "102.100.100/39075" "102.100.100/39077" "102.100.100/39079"
# [21] "102.100.100/39081" "102.100.100/39083" "102.100.100/39085" "102.100.100/39087" "102.100.100/39089"
# [26] "102.100.100/39091" "102.100.100/39093" "102.100.100/39095" "102.100.100/39097" "102.100.100/39099"
# [31] "102.100.100/39161" "102.100.100/39163" "102.100.100/39165" "102.100.100/39191" "102.100.100/39193"
# [36] "102.100.100/39195"

raw.otu.16s <- raw.otu.16s[sel.keep, ]
dim(raw.otu.16s) # 300921     11


## translate from long to wide format

names(raw.otu.16s)
# [1] "Sample.ID"    "OTU"          "OTU.Count"    "Amplicon"     "Kingdom"      "Phylum"       "Class"
# [8] "Order"        "Family"       "Genus"        "Species"

head(raw.otu.16s)

raw.otu.16s$sampID_short <- gsub(pattern = "102.100.100/", replacement = "", x = raw.otu.16s$Sample.ID )


otu.16s <- dcast(raw.otu.16s, formula = OTU ~ sampID_short, value.var = "OTU.Count")
dim(otu.16s) # 36061    37

length(unique(otu.16s$OTU)) # 36061
sel.na <- which(is.na(otu.16s),arr.ind = TRUE)
otu.16s[sel.na] <- 0

row.names(otu.16s) <- otu.16s$OTU

names(otu.16s)

# now remove 1st column (it is now stored as the row.name)
otu.16s <- otu.16s[ , -1]

rep_seq <- row.names(otu.16s)
head(rep_seq)
length(rep_seq) # 36061
names(rep_seq) <- paste0("otu_",1:length(rep_seq))
head(rep_seq)
dim(otu.16s) #  36061    36


rep_seq.alcoa <- rep_seq


# substitute otu names for rep sequences
row.names(otu.16s) <- names(rep_seq)


names(otu.16s)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065" "39067"
# [15] "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091" "39093" "39095"
# [29] "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

names(otu.16s) <- paste0("X",names(otu.16s))

identical( context$Sample_id, names(otu.16s)) # FALSE

row.names(context) <- paste0("X",row.names(context))

identical( row.names(context), names(otu.16s) ) # TRUE
# i.e. sample context and otu data are in the same order

dim(otu.16s) # 36061    36

sum(colSums(otu.16s)) # 1791794 = 1,791,794 reads





### Taxonomy table

## get representative sequences

head( rep_seq.alcoa ) # from above

length(rep_seq.alcoa) # 36061
dim(otu.16s) # 36061    36

identical(names(rep_seq.alcoa),row.names(otu.16s)) # TRUE

### Assign taxonomy using dada2
# https://benjjneb.github.io/dada2/tutorial_1_8.html

# SILVA taxonomy training databases from here 
# https://zenodo.org/record/3986799#.X5kj-aozaUl

# assignTaxonomy(seqs, refFasta, minBoot = 50, tryRC = FALSE,
#                outputBootstraps = FALSE, taxLevels = c("Kingdom", "Phylum", "Class",
#                                                        "Order", "Family", "Genus", "Species"), multithread = FALSE,
#                verbose = FALSE)



head(as.character(rep_seq.alcoa))

taxa <- assignTaxonomy(seqs = as.character(rep_seq.alcoa), 
                       refFasta = "/Users/lidd0026/WORKSPACE/DATA/SILVA_DB/silva_nr99_v138_train_set.fa.gz",
                       tryRC = TRUE,
                       multithread=TRUE, verbose = TRUE)

temp.taxa.alcoa <- taxa
#taxa <- temp.taxa.alcoa

str(taxa)
# Large matrix

tax.16s <- as.data.frame(taxa)
head(row.names(tax.16s)) # these are sequences ... swap for OTU names

identical(names(rep_seq.alcoa),row.names(otu.16s)) # TRUE
identical(toupper(as.character(rep_seq.alcoa)),row.names(tax.16s)) # TRUE

# swap labels
row.names(tax.16s) <- names(rep_seq.alcoa)




# tidy  up taxa table

tax <- tax.16s
names(tax)
# "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

dim(tax) # 36061     6

temp <- tax
#tax <- temp

#taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__", "s__")
taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__")

for (i in 1:length(names(tax))) {
  #i<-1
  tax[ , names(tax)[i] ] <- paste0(taxa_prefix[i],tax[ , names(tax)[i] ])
  
  sel <- which( is.na(tax[, i]) | tax[, i] == "" | tax[, i] == taxa_prefix[i] | tax[, i] == paste0(taxa_prefix[i],NA) )
  if (length(sel)>=1) {
    tax[ sel, names(tax)[i] ] <- paste0(taxa_prefix[i],"Unclassified")
  }
  print(paste0("completed ",i))
}

write.csv(tax, file = "alcoa-surf-16s-Taxonomy-Genus-level-ASV-UPDATE.csv", row.names = TRUE )


# confirm matching row names between Taxonomy and OTUs
identical(row.names(tax),row.names(otu.16s)) # TRUE


## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax )
TAX.16s <- tax_table( tax.m )





## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( otu.16s )
dim(otu.m)
# 36061    36

OTU.16s <- otu_table(otu.16s, taxa_are_rows = TRUE)





## Create a phyloseq object, merging OTU & TAX tables
phy.16s = phyloseq(OTU.16s, TAX.16s)
phy.16s
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 36061 taxa and 36 samples ]
# tax_table()   Taxonomy Table:    [ 36061 taxa by 6 taxonomic ranks ]

sample_names(phy.16s)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(samp.16s) #  36  70
identical( row.names(samp.16s), sample_names(phy.16s) ) # TRUE


SAMP.16s <- sample_data(samp.16s)


### Combine SAMPDATA into phyloseq object
phy.16s <- merge_phyloseq(phy.16s, SAMP.16s)
phy.16s
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 36061 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 36061 taxa by 6 taxonomic ranks ]




### data cleaning

## remove taxa not assigned as Bacteria

phy_in <- phy.16s

sum(sample_sums(phy_in)) # 1791794 = 1,791,794

rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

levels(factor(tax_table(phy_in)[, "Kingdom"])) # "k__Bacteria"


## remove taxa not assigned at the phylum level
sort( unique( as.character(tax_table(phy_in)[, "Phylum"]) ) )
# [1] "p__Abditibacteriota"             "p__Acidobacteriota"              "p__Actinobacteriota"            
# [4] "p__Armatimonadota"               "p__Bacteroidota"                 "p__Bdellovibrionota"            
# [7] "p__Chloroflexi"                  "p__Cyanobacteria"                "p__Dadabacteria"                
# [10] "p__Deinococcota"                 "p__Dependentiae"                 "p__Desulfobacterota"            
# [13] "p__Elusimicrobiota"              "p__Entotheonellaeota"            "p__FCPU426"                     
# [16] "p__Fibrobacterota"               "p__Firmicutes"                   "p__GAL15"                       
# [19] "p__Gemmatimonadota"              "p__Hydrogenedentes"              "p__Latescibacterota"            
# [22] "p__MBNT15"                       "p__Methylomirabilota"            "p__Myxococcota"                 
# [25] "p__NB1-j"                        "p__Nitrospirota"                 "p__Patescibacteria"             
# [28] "p__Planctomycetota"              "p__Proteobacteria"               "p__RCP2-54"                     
# [31] "p__SAR324 clade(Marine group B)" "p__Spirochaetota"                "p__Sumerlaeota"                 
# [34] "p__Unclassified"                 "p__Verrucomicrobiota"            "p__WPS-2"                       
# [37] "p__WS2"                          "p__WS4" 


rem_taxa <- which(tax_table(phy_in)[, "Phylum"] == "p__Unclassified") # qty 63
phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-rem_taxa, ]) )
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 35998 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 35998 taxa by 6 taxonomic ranks ]

sum(sample_sums(phy_in)) # 1790321


rank_names(phy_in) #  "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"
sort( as.character( unique( tax_table(phy_in)[, "Phylum"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Class"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Order"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Family"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Genus"] ) ))


## remove taxa associated with chloroplast and mitochondria

rem_taxa1 <- which(tax_table(phy_in)[, "Order"] == "o__Chloroplast") # qty 47... ASV
rem_taxa2 <- which(tax_table(phy_in)[, "Family"] == "f__Mitochondria") # qty 42... ASV

length( c(rem_taxa1,rem_taxa2) ) # 89
length( unique(c(rem_taxa1,rem_taxa2)) ) # 89

# remove chloroplast and mitochondria
phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-unique(c(rem_taxa1,rem_taxa2)), ]) )
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 35909 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 35909 taxa by 6 taxonomic ranks ]

sum(sample_sums(phy_in)) # 1786302


## remove taxa that do not occur in at least two samples (e.g avoiding unrepresentative taxa only present in a single sample)
temp.phy_in <- phy_in
phy_in <- prune_taxa( taxa = apply( otu_table(phy_in), 1, function(x) {sum(x > 0) >= 2 }), x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 31746 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 31746 taxa by 6 taxonomic ranks ]

sum(sample_sums(phy_in)) # 1772249
min(taxa_sums(phy_in)) # 2


## check only surface samples
table(phy_in@sam_data$Depth)
# 10 
# 36 

sort(sample_sums(phy_in))
# X39195 X39193 X39045 X39059 X39061 X39099 X39041 X39051 X39161 X39071 X39053 X39095 X39067 X39063 X39075 X39077 
# 17485  19598  31768  31804  32412  34570  35628  39709  42102  42302  42476  42579  43797  43924  45554  45587 
# X39163 X39043 X39087 X39079 X39083 X39089 X39097 X39093 X39085 X39057 X39049 X39081 X39073 X39069 X39047 X39165 
# 46572  48218  51424  51794  52427  53532  57002  57338  58043  58093  58221  59390  60439  61056  61698  62357 
# X39191 X39065 X39091 X39055 
# 67370  67413  69298  79269

min(sample_sums(phy_in)) # 17485
sum(sample_sums(phy_in)) # 1772249


phy.alcoa <- phy_in



### build and add phylogenetic tree to the phyloseq object

ps <- phy.alcoa

head( taxa_names(ps) )
# "otu_1" "otu_2" "otu_3" "otu_4" "otu_5" "otu_6"

taxa_names_ps <- taxa_names(ps)
length(taxa_names_ps) # 31746


## read rep seq fasta file
#rep_seq <- read.fasta(file = paste0(datadir,"/mtbold/","Mt-Bold-AMD-16S-table-otus.fasta"), seqtype = "DNA", as.string = TRUE)
length( rep_seq.alcoa ) # 36061
rep_seq <- rep_seq.alcoa

head(rep_seq)

## subselect rep seqs
rep_seq <- rep_seq[taxa_names_ps]
length(rep_seq) # 31746
head(names(rep_seq))
# "otu_1" "otu_2" "otu_3" "otu_4" "otu_5" "otu_6"

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"
#write.fasta(sequences = rep_seq, names = names(rep_seq), file.out = "alcoa_clean_surf_16s_rep_seq.fasta", open = "w",nbchar = 80 )
write.fasta(sequences = rep_seq[1], names = names(rep_seq)[1], file.out = "alcoa_clean_surf_16s_rep_seq.fasta", open = "w",nbchar = 80 )
for (i in 2:length(rep_seq)) {
  write.fasta(sequences = rep_seq[i], names = names(rep_seq)[i], file.out = "alcoa_clean_surf_16s_rep_seq.fasta", open = "a",nbchar = 80 )
}



## run MAFFT multiple sequence alignment on DeepThought HPC ...

# # # # # # # # # # # 
# # # # MAFFT # # # #
# # # # # # # # # # # 
# # https://mafft.cbrc.jp/alignment/software/
# # A fast option (FFT-NS-2) for a larger sequence alignment:
# # mafft input > output
# 
# $ whereis mafft   #:mafft: /home/lidd0026/miniconda3/bin/mafft
# 
# # run mafft using slurm script called 'run_mafft_alcoa_slurm.sh'
# - - - - - - - - - - - - - - - 
# #!/bin/bash
#   
# # How many tasks and processes
# #SBATCH --ntasks=1
#   
# #SBATCH --cpus-per-task=32
#   
# # How much memory
# #SBATCH --mem=64G
#   
# # Put your code to run here:
# mafft --thread -1 alcoa_clean_surf_16s_rep_seq.fasta > alcoa_clean_surf_16s_mafft_aln.fasta
# 
# - - - - - - - - - - - - - - - 
#   
# # RUN USING THIS COMMAND:
# $ cd /scratch/user/lidd0026/resto_traj
# $ ls
# 
# # more detailed view of the cluster with sinfo:
# $ sinfo -N -o "%N %C %e %m"
# 
# $ sbatch run_mafft_alcoa_slurm.sh

# Strategy:
#   FFT-NS-2 (Fast but rough)
# Progressive method (guide trees were built 2 times.)

## output of MAFFT is: 'alcoa_clean_surf_16s_mafft_aln.fasta'


## run Gblocks on Mac to clean noisy alignments
# set -b2 to 60% of sequences = 0.6*31746 = 19048
## run 'relaxed' cleaning with Gblocks
# /Users/lidd0026/opt/Gblocks_0.91b/Gblocks alcoa_clean_surf_16s_mafft_aln.fasta -t=d -b2=19048 -b3=10 -b4=5 -b5=h -e=gbrel 
# 31746 sequences and 1755 positions in the first alignment file:
# alcoa_clean_surf_16s_mafft_aln.fasta
# 
# alcoa_clean_surf_16s_mafft_aln.fasta
# Original alignment: 1755 positions
# Gblocks alignment:  336 positions (19 %) in 20 selected block(s)

## output of Gblocks is: 'alcoa_clean_surf_16s_mafft_aln_Gblocks_relaxed.fasta'

gblocks_aln <- read.fasta( paste0(getwd(),"/deepthoughtHPC/","alcoa_clean_surf_16s_mafft_aln_Gblocks_relaxed.fasta"),
                           as.string = FALSE)
head(gblocks_aln)
unique(unlist(gblocks_aln)) # "a" "g" "c" "t" " " "-"
gblocks_aln <- read.fasta( paste0(getwd(),"/deepthoughtHPC/","alcoa_clean_surf_16s_mafft_aln_Gblocks_relaxed.fasta"),
                           as.string = TRUE)
names_gblocks_aln <- names(gblocks_aln)
head(names_gblocks_aln) # "otu_1" "otu_2" "otu_3" "otu_4" "otu_5" "otu_6"
gblocks_aln <- toupper(gblocks_aln)
gblocks_aln <- gsub(pattern = " ",replacement = "-", x = gblocks_aln)
names(gblocks_aln) <- names_gblocks_aln
head(gblocks_aln)


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"
write.fasta(sequences = gblocks_aln[1], names = names(gblocks_aln)[1], file.out = "alcoa_clean_16s_mafft_Gblocks_rel_tidy.fasta", open = "w",nbchar = 80 )
for (i in 2:length(gblocks_aln)) {
  write.fasta(sequences = gblocks_aln[i], names = names(gblocks_aln)[i], file.out = "alcoa_clean_16s_mafft_Gblocks_rel_tidy.fasta", open = "a",nbchar = 80 )
}

rm(gblocks_aln, names_gblocks_aln)


## run IQTREE2 to build phylogenetic tree on DeepThought HPC
# ... fit a GTR+G+I (Generalized time-reversible with Gamma rate variation) ?

# # # # # # # # # # #
# # # # IQTREE2 # # #
# # # # # # # # # # # 
# 
# # run iqtree2 using slurm script called 'run_iqtree2_alcoa_slurm.sh'
# 
# - - - - - - - - - - - - - - - 
# #!/bin/bash
#   
# # How many tasks and processes
# #SBATCH --ntasks=1
#   
# #SBATCH --cpus-per-task=32
#   
# # How much memory
# #SBATCH --mem=64G
#   
# # Put your code to run here:
# /home/lidd0026/opt/iqtree-2.1.2-Linux/bin/iqtree2 -s /scratch/user/lidd0026/resto_traj/alcoa_clean_surf_16s_mafft_aln.fasta --prefix iqtree2_alcoa_surf_16s_deepthought -m GTR+I+G -T 32 -n 200
# 
# - - - - - - - - - - - - - - - 
#   
# # RUN USING THIS COMMAND:
# $ cd /scratch/user/lidd0026/resto_traj
# $ ls
# 
# $ sinfo -N -o "%N %C %e %m"
# 
# $ sbatch run_iqtree2_alcoa_slurm.sh


# ... fit a GTR+G+I (Generalized time-reversible with Gamma rate variation)





### add phylo tree

phy_in <- phy.alcoa

phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 31746 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 31746 taxa by 6 taxonomic ranks ]

tree <- read_tree("iqtree2_alcoa_gblocks.treefile")

# need to match taxa names ...
head( taxa_names(phy_in) ) # "OTU_ 13" "OTU_ 14" "OTU_ 15" "OTU_ 16" "OTU_ 30" "OTU_ 32"
head( taxa_names(tree)) # "OTU_13"    "OTU_14"    "OTU_48062" "OTU_16"    "OTU_18566" "OTU_18570"

length(taxa_names(phy_in)) # 31746
length(taxa_names(tree)) # 31746

sel <- which(taxa_names(tree) %in% taxa_names(phy_in))
length(sel) # 31746
identical(sort(taxa_names(tree)),sort(taxa_names(phy_in))) # TRUE

phy_in <- merge_phyloseq(phy_in, tree)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 31746 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 31746 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 31746 tips and 31744 internal nodes ]

phy.alcoa.withTree <- phy_in


#
#-------------------------



#### Alcoa - form 99% , 97%, 95%, 90% identity OTUs
#-------------------------
phy_in <- phy.alcoa.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 31746 taxa and 36 samples ]:
# sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
# tax_table()   Taxonomy Table:     [ 31746 taxa by 6 taxonomic ranks ]:
# phy_tree()    Phylogenetic Tree:  [ 31746 tips and 31744 internal nodes ]:
# taxa are rows


str(rep_seq.alcoa)
# Named chr [1:36061] "AAAGAACGCTGGCGGCATGCTTAACACATGCAAGTCGAACGAGTAGTAGCAATACTATTAGTGGCAGACGGGTGAGTAATACGTAGGAATCTCCCCTATAGTACGGAACAA"| __truncated__ ...
# - attr(*, "names")= chr [1:36061] "otu_1" "otu_2" "otu_3" "otu_4" ...
length(rep_seq.alcoa)
# 36061

taxa_names_ps <- taxa_names(phy_in)
identical(taxa_names_ps, row.names(otu_table(phy_in))) # TRUE

rep_seq <- rep_seq.alcoa

## subselect rep seqs
rep_seq <- rep_seq[taxa_names_ps]
length(rep_seq) # 31746
head(names(rep_seq))
# "otu_1" "otu_2" "otu_3" "otu_5" "otu_4" "otu_6"

identical(names(rep_seq), row.names(otu_table(phy_in))) # TRUE

## place these in the order from highest abundance to lowest abundance ...

abund <- as.data.frame(otu_table(phy_in))
class(abund) # "data.frame"
head(row.names(abund)) # "otu_1" "otu_2" "otu_3" "otu_5" "otu_4" "otu_6"
names(abund) # [1] "X39041" "X39043" "X39045" "X39047"... etc.
abund$counts <- rowSums(abund)
abund <- abund[ order(abund$counts, decreasing = TRUE), ]

new_rep_seq_order <- row.names(abund)
head(new_rep_seq_order) # "otu_16952" "otu_16954" "otu_1827"  "otu_16932" "otu_16957" "otu_16967"

rep_seq <- rep_seq[new_rep_seq_order]
head(names(rep_seq)) # "otu_16952" "otu_16954" "otu_1827"  "otu_16932" "otu_16957" "otu_16967"

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"
#write.fasta(sequences = rep_seq, names = names(rep_seq), file.out = "iluka_eneabba_clean_surf_16s_rep_seq.fasta", open = "w",nbchar = 80 )
write.fasta(sequences = rep_seq[1], names = names(rep_seq)[1], file.out = "alcoa_clean_surf_16s_rep_seq.ORDERED.fasta", open = "w",nbchar = 80 )
for (i in 2:length(rep_seq)) {
  write.fasta(sequences = rep_seq[i], names = names(rep_seq)[i], file.out = "alcoa_clean_surf_16s_rep_seq.ORDERED.fasta", open = "a",nbchar = 80 )
}


### run VSEARCH from MacOS Terminal 
# VSEARCH manual is here - https://github.com/torognes/vsearch/releases/download/v2.17.0/vsearch_manual.pdf
# reordered representative sequence fasta file is here -
# /Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling/alcoa_clean_surf_16s_rep_seq.ORDERED.fasta
# 
# cd /Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling
# ls
# which vsearch # /Users/lidd0026/opt/anaconda3/bin/vsearch
# 
# # run VSEARCH clustering command

# 99% identity OTUs 
# vsearch --cluster_smallmem alcoa_clean_surf_16s_rep_seq.ORDERED.fasta --usersort --id 0.99 --centroids alcoa_otu99.fasta --uc alcoa_uclust99.uc
# vsearch v2.17.0_macos_x86_64, 32.0GB RAM, 8 cores
# https://github.com/torognes/vsearch
# Reading file alcoa_clean_surf_16s_rep_seq.ORDERED.fasta 100%  
# 14680011 nt in 31746 seqs, min 432, max 519, avg 462
# Masking 100% 
# Counting k-mers 100% 
# Clustering 100%  
# Sorting clusters 100%
# Writing clusters 100% 
# Clusters: 18062 Size min 1, max 37, avg 1.8
# Singletons: 11993, 37.8% of seqs, 66.4% of clusters

# 97% identity OTUs 
# vsearch --cluster_smallmem alcoa_clean_surf_16s_rep_seq.ORDERED.fasta --usersort --id 0.97 --centroids alcoa_otu97.fasta --uc alcoa_uclust97.uc
# vsearch v2.17.0_macos_x86_64, 32.0GB RAM, 8 cores
# https://github.com/torognes/vsearch
# Reading file alcoa_clean_surf_16s_rep_seq.ORDERED.fasta 100%  
# 14680011 nt in 31746 seqs, min 432, max 519, avg 462
# Masking 100% 
# Counting k-mers 100% 
# Clustering 100%  
# Sorting clusters 100%
# Writing clusters 100% 
# Clusters: 8663 Size min 1, max 153, avg 3.7
# Singletons: 4279, 13.5% of seqs, 49.4% of clusters

# 95% identity OTUs 
# vsearch --cluster_smallmem alcoa_clean_surf_16s_rep_seq.ORDERED.fasta --usersort --id 0.95 --centroids alcoa_otu95.fasta --uc alcoa_uclust95.uc
# vsearch v2.17.0_macos_x86_64, 32.0GB RAM, 8 cores
# https://github.com/torognes/vsearch
# Reading file alcoa_clean_surf_16s_rep_seq.ORDERED.fasta 100%  
# 14680011 nt in 31746 seqs, min 432, max 519, avg 462
# Masking 100% 
# Counting k-mers 100% 
# Clustering 100%  
# Sorting clusters 100%
# Writing clusters 100% 
# Clusters: 4868 Size min 1, max 491, avg 6.5
# Singletons: 1960, 6.2% of seqs, 40.3% of clusters

# 90% identity OTUs 
# vsearch --cluster_smallmem alcoa_clean_surf_16s_rep_seq.ORDERED.fasta --usersort --id 0.90 --centroids alcoa_otu90.fasta --uc alcoa_uclust90.uc
# vsearch v2.17.0_macos_x86_64, 32.0GB RAM, 8 cores
# https://github.com/torognes/vsearch
# Reading file alcoa_clean_surf_16s_rep_seq.ORDERED.fasta 100%  
# 14680011 nt in 31746 seqs, min 432, max 519, avg 462
# Masking 100% 
# Counting k-mers 100% 
# Clustering 100%  
# Sorting clusters 100%
# Writing clusters 100% 
# Clusters: 1512 Size min 1, max 1309, avg 21.0
# Singletons: 418, 1.3% of seqs, 27.6% of clusters


## loop through clusters ...

phy_otu_clust <- list()
cluster_names <- c("otu99","otu97","otu95","otu90")
#cluster_ids <- c(0.99, 0.97, 0.95, 0.90)

for (j in 1:length(cluster_names)) {
  #j<-2
  this_clust_id_label <- sub(pattern = "otu",replacement = "",x = cluster_names[j] )
  uc <- read.table(file = paste0("alcoa_uclust", this_clust_id_label,".uc"), header = FALSE, sep = "\t" )
  class(uc) # "data.frame"
  names(uc) <- c("record_type",  # S = centroid, H = hit, C = cluster
                 "cluster_no",   # (zero-based)
                 "size",         # Centroid length (S), query length (H), or cluster size (C)
                 "percent_similarity",
                 "match_orientation",
                 "not_used1",
                 "not_used2",
                 "cigar_string",
                 "label_query_sequence",
                 "Label_centroid_sequence")
  uc$cluster_no <- paste0(cluster_names[j],"_",uc$cluster_no)
  print(paste0("length of unique clusters for ",cluster_names[j]," is: ", length(unique(uc$cluster_no)))) # 18062

  # find XX% identity otu clusters
  sel <- which(uc$record_type == "C") # qty 18062
  uc <- uc[-sel, ]
  print( head(uc) )
  #   record_type cluster_no size percent_similarity match_orientation not_used1 not_used2 cigar_string label_query_sequence Label_centroid_sequence
  # 1           S    otu99_0  436                  *                 *         *         *            *            otu_16952                       *
  # 2           H    otu99_0  436               99.8                 +         0         0         436M            otu_16954               otu_16952
  # 3           S    otu99_1  432                  *                 *         *         *            *             otu_1827                       *
  # 4           H    otu99_0  436               99.5                 +         0         0         436M            otu_16932               otu_16952
  # 5           H    otu99_0  436               99.5                 +         0         0         436M            otu_16957               otu_16952
  # 6           H    otu99_0  436               99.3                 +         0         0         436M            otu_16967               otu_16952
  
  new.tax <- as.data.frame( tax_table(phy_in) )
  new.tax$temp <- NA
  sel.col <- which(names(new.tax)=="temp")
  names(new.tax)[sel.col] <- cluster_names[j]
  
  #length(row.names(new.tax)) # 31746

  # remove taxanomic ranks as grouping will only occur by OTUs now
  names(new.tax) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "otu99" 
  dim(new.tax) # 31746     7
  names(new.tax)[c(1,7)] # "Kingdom" "otu99"

  new.tax <- new.tax[ , c(1,7)]

  for (i in 1:length(row.names(new.tax))) {
    #i<-1
    this_asv <- row.names(new.tax)[i]
    
    sel.query.uc <- which(uc$label_query_sequence == this_asv)
    #identical(this_asv, uc$label_query_sequence[sel.query.uc])
    
    if (uc$Label_centroid_sequence[sel.query.uc] == "*" & uc$record_type[sel.query.uc] == "S") {
      new.tax[i, cluster_names[j] ] <- uc$label_query_sequence[sel.query.uc]
    } else {
      new.tax[i, cluster_names[j] ] <- uc$Label_centroid_sequence[sel.query.uc]
    }
    #print(paste0("completed ", i))
  }

  phy_new <- phy_in
  print( identical(taxa_names(phy_new),row.names(new.tax)) ) # TRUE
  tax_table(phy_new) <- tax_table(as.matrix( new.tax ))
  
  print( head( tax_table(phy_new) ) )
  # Taxonomy Table:     [ 6 taxa by 2 taxonomic ranks ]:
  #       Kingdom     otu99
  #       <chr>       <chr>
  # otu_1 k__Bacteria otu_1
  # otu_2 k__Bacteria otu_1
  # otu_3 k__Bacteria otu_3
  # otu_5 k__Bacteria otu_3
  # otu_4 k__Bacteria otu_4
  # otu_6 k__Bacteria otu_6

  rank_names(phy_new) # "Kingdom" "otu99" 
  ntaxa(phy_new) # 31746
  sum(sample_sums(phy_new)) # 1772249
  print(paste0("number of unique taxa for ",cluster_names[j]," is: ", length(unique(tax_table(phy_new)[ , cluster_names[j] ])) ))

  phy_otu_clust[[ cluster_names[j] ]] <- tax_glom(physeq = phy_new, taxrank = cluster_names[j])
  
  print(paste0( "number of taxa for ",cluster_names[j], " is: ",ntaxa( phy_otu_clust[[ cluster_names[j] ]] )))
  
  print(paste0("total number of sequences for ",cluster_names[j]," is: ", sum(sample_sums( phy_otu_clust[[ cluster_names[j] ]] ))) )

}


# [1] "number of unique taxa for otu99 is: 18062"
# [1] "number of taxa for otu99 is: 18062"
# [1] "total number of sequences for otu99 is: 1772249"
# [1] "length of unique clusters for otu97 is: 8663"
# record_type cluster_no size percent_similarity match_orientation not_used1 not_used2 cigar_string label_query_sequence Label_centroid_sequence
# 1           S    otu97_0  436                  *                 *         *         *            *            otu_16952                       *
#   2           H    otu97_0  436               99.8                 +         0         0         436M            otu_16954               otu_16952
# 3           S    otu97_1  432                  *                 *         *         *            *             otu_1827                       *
#   4           H    otu97_0  436               99.5                 +         0         0         436M            otu_16932               otu_16952
# 5           H    otu97_0  436               99.5                 +         0         0         436M            otu_16957               otu_16952
# 6           H    otu97_0  436               99.3                 +         0         0         436M            otu_16967               otu_16952
# [1] TRUE
# Taxonomy Table:     [ 6 taxa by 2 taxonomic ranks ]:
#   Kingdom     otu97
# otu_1 k__Bacteria otu_1
# otu_2 k__Bacteria otu_1
# otu_3 k__Bacteria otu_4
# otu_5 k__Bacteria otu_4
# otu_4 k__Bacteria otu_4
# otu_6 k__Bacteria otu_4
# 
# [1] "number of unique taxa for otu97 is: 8663"
# [1] "number of taxa for otu97 is: 8663"
# [1] "total number of sequences for otu97 is: 1772249"
# [1] "length of unique clusters for otu95 is: 4868"
# record_type cluster_no size percent_similarity match_orientation not_used1 not_used2 cigar_string label_query_sequence Label_centroid_sequence
# 1           S    otu95_0  436                  *                 *         *         *            *            otu_16952                       *
#   2           H    otu95_0  436               99.8                 +         0         0         436M            otu_16954               otu_16952
# 3           S    otu95_1  432                  *                 *         *         *            *             otu_1827                       *
#   4           H    otu95_0  436               99.5                 +         0         0         436M            otu_16932               otu_16952
# 5           H    otu95_0  436               99.5                 +         0         0         436M            otu_16957               otu_16952
# 6           H    otu95_0  436               99.3                 +         0         0         436M            otu_16967               otu_16952
# [1] TRUE
# Taxonomy Table:     [ 6 taxa by 2 taxonomic ranks ]:
#   Kingdom     otu95
# otu_1 k__Bacteria otu_1
# otu_2 k__Bacteria otu_1
# otu_3 k__Bacteria otu_4
# otu_5 k__Bacteria otu_4
# otu_4 k__Bacteria otu_4
# otu_6 k__Bacteria otu_4
# 
# [1] "number of unique taxa for otu95 is: 4868"
# [1] "number of taxa for otu95 is: 4868"
# [1] "total number of sequences for otu95 is: 1772249"
# [1] "length of unique clusters for otu90 is: 1512"
# record_type cluster_no size percent_similarity match_orientation not_used1 not_used2 cigar_string label_query_sequence Label_centroid_sequence
# 1           S    otu90_0  436                  *                 *         *         *            *            otu_16952                       *
#   2           H    otu90_0  436               99.8                 +         0         0         436M            otu_16954               otu_16952
# 3           S    otu90_1  432                  *                 *         *         *            *             otu_1827                       *
#   4           H    otu90_0  436               99.5                 +         0         0         436M            otu_16932               otu_16952
# 5           H    otu90_0  436               99.5                 +         0         0         436M            otu_16957               otu_16952
# 6           H    otu90_0  436               99.3                 +         0         0         436M            otu_16967               otu_16952
# [1] TRUE
# Taxonomy Table:     [ 6 taxa by 2 taxonomic ranks ]:
#   Kingdom     otu90
# otu_1 k__Bacteria otu_1
# otu_2 k__Bacteria otu_1
# otu_3 k__Bacteria otu_1
# otu_5 k__Bacteria otu_1
# otu_4 k__Bacteria otu_1
# otu_6 k__Bacteria otu_1
# 
# [1] "number of unique taxa for otu90 is: 1512"
# [1] "number of taxa for otu90 is: 1512"
# [1] "total number of sequences for otu90 is: 1772249"


#-------------------------




#### 1b. Alcoa distance plots

## Alcoa - with Tree - i. Bray-Curtis - ASV-level
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  30751
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30751 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 30751 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 30751 tips and 30749 internal nodes ]


table(r1.ps@sam_data$`Date since change in Land Use`)
# 1987 1991 1999 2002 2008 2014  N/A 
#    3    3    3    3    3    3   18 



head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1"  

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
#  2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"   
df_out$facet_x <- "Bray-Curtis\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 0
names(df_results)

df_results[["Alcoa-Bray-ASV-withTree"]] <- df_out
#df_out <- df_results[[1]]


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Reference (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
 
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 354.92, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group )
# [1] 2   Ref 8   25  29  14  17 
# Levels: 2 < 8 < 14 < 17 < 25 < 29 < Ref


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
#   Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14      a       a   
# 2    17      b        b  
# 4    25      b        b  
# 5    29     bd        b d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(22,30,38,48,48,51,55),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-Bray-ASV-withTree"]] <- anno



## Include significance letters in plot

set.seed(1234)
#p <- ggplot(data=df_out, aes(x=group, distance)) + # ggtitle("Alcoa - Bray Curtis Similarity") +
p <- ggplot(data=df_out, aes(x=group, similarity)) + #ggtitle("Huntly - Bray Curtis Similarity") +
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  annotate(geom="text", x= 1.2, y= 22, label = "c", hjust=0, vjust=0, size = 3 ) + # , col="blue") +
  annotate(geom="text", x= 2.2, y= 30, label = "ac", hjust=0, vjust=0, size = 3 ) + #, col="blue") +
  annotate(geom="text", x= 3.2, y= 38, label = "a", hjust=0, vjust=0, size = 3 ) + #, col="blue") +
  annotate(geom="text", x= 4.2, y= 48, label = "b", hjust=0, vjust=0, size = 3 ) + #, col="blue") +
  annotate(geom="text", x= 5.2, y= 48, label = "b", hjust=0, vjust=0, size = 3 ) + #, col="blue") +
  annotate(geom="text", x= 6.2, y= 51, label = "bd", hjust=0, vjust=0, size = 3 ) + #, col="blue") +
  annotate(geom="text", x= 7.2, y= 55, label = "d", hjust=0, vjust=0, size = 3 ) + #, col="blue") +

  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = "Rehabilitation age (years)", y = "Similarity to Reference (%)")
p

grid.text(label = "(B)", x = unit(0.03, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )

dev.print(tiff, file = paste0(workdir,"/plots/","Alcoa-ASV-Bray-Curtis-boxplots.tiff"), width = 14, height = 10, units = "cm", res=600, compression="lzw", type = "cairo")




## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1100346 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

temp <- r1.ps

cols.group.alcoa # defined below
#      2 yr      8 yr     14 yr     17 yr     25 yr     29 yr       Ref 
# "#9e0142" "#d53e4f" "#fdae61" "#e6f598" "#abdda4" "#66c2a5" "#5e4fa2" 

temp@sam_data$group
temp@sam_data$group <- factor(temp@sam_data$group, 
                              levels = c("2","8","14","17","25","29","Ref"),
                              labels = c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref"),
                              ordered=TRUE)

cols <- cols.group.alcoa

p <- plot_ordination(temp, ord, type="samples", color="group") +
  theme_bw()+
  scale_color_manual(values = cols, name = "Rehab\nage") +
  annotate(geom="text", x= -1.2, y= 1.3, label = paste0("Stress = ",round(ord$stress,digits=4)),size = 3, hjust=0, vjust=1) +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank())
p

grid.text(label = "(A)", x = unit(0.03, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )

dev.print(tiff, file = paste0(workdir,"/plots/","Alcoa-ASV-Bray-Curtis-NMDS-ordination.tiff"), width = 14, height = 10, units = "cm", res=600, compression="lzw", type = "cairo")



str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 0
names(ord_plots)
ord_plots[["Alcoa-Bray-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj)
ord_obj[["Alcoa-Bray-ASV-withTree"]] <- ord

#-------------------------

## Alcoa - with Tree - ii. Jaccard - ASV-level
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  30751
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30751 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 30751 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 30751 tips and 30749 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 1
names(df_results) #

df_results[["Alcoa-Jaccard-ASV-withTree"]] <- df_out





## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 239.7, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 

pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      d         d 
# 6     8     bd       b d 
# 1    14     ab      ab   
# 2    17     ac      a c  
# 4    25     ac      a c  
# 5    29     ce        c e
# 7   Ref      e          e

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  


p <- ggplot(data=df_out, aes(x=group, similarity)) +
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(20,25,27,29,30,35,40),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nASV-level", times = dim(sig)[1])
)
anno

# only make this once
#df_anno <- list()
length(df_anno) # 1
names(df_anno) # 

df_anno[["Alcoa-Jaccard-ASV-withTree"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)
# https://stats.stackexchange.com/questions/242110/nmds-from-jaccard-and-bray-curtis-identical-is-that-a-bad-thing

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1116162 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



#p <- plot_ordination(r1.16s, ord, type="samples", color="site", shape = "Sample_type__row_type")
p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) #
ord_obj[["Alcoa-Jaccard-ASV-withTree"]] <- ord

#-------------------------

## Alcoa - with Tree - iii. Weighted UniFrac - ASV-level
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  30751
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30751 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 30751 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 30751 tips and 30749 internal nodes ]


dist.wuni <- phyloseq::distance(r1.ps, "wunifrac") # weighted UniFrac
class(dist.wuni) # "dist"
dist.wuni <- as(dist.wuni, "matrix")
# express as similarity
sim.wuni <- 100*(1 - dist.wuni)
# list all sample names
colnames(sim.wuni)
identical(colnames(sim.wuni),rownames(sim.wuni)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1"  

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.wuni

df_out <- data.frame(samp= rep( x = colnames(sim.wuni), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.wuni))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Weighted UniFrac"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Weighted UniFrac\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results)

df_results[["Alcoa-WtUniFrac-ASV-withTree"]] <- df_out
#df_out <- df_results[[1]]


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Weighted UniFrac Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 327.36, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'
## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around

pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(93,95.5,96.5,97,97.5,98,98),
                   dist_measure = rep( "Weighted UniFrac", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Weighted UniFrac\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-WtUniFrac-ASV-withTree"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "unifrac", weighted = TRUE)

ord
# Call:
#   metaMDS(comm = ps.dist) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     ps.dist 
# Distance: user supplied 
# 
# Dimensions: 2 
# Stress:     0.09659549 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation 
# Species: scores missing

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 0
names(ord_plots)
ord_plots[["Alcoa-WtUniFrac-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj)
ord_obj[["Alcoa-WtUniFrac-ASV-withTree"]] <- ord

#-------------------------

## Alcoa - with Tree - iv. Unweighted UniFrac - ASV-level
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  30751
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30751 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 30751 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 30751 tips and 30749 internal nodes ]


dist.uuni <- phyloseq::distance(r1.ps, "uunifrac") # Unweighted UniFrac
class(dist.uuni) # "dist"
dist.uuni <- as(dist.uuni, "matrix")
# express as similarity
sim.uuni <- 100*(1 - dist.uuni)
# list all sample names
colnames(sim.uuni)
identical(colnames(sim.uuni),rownames(sim.uuni)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1"  

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.uuni

df_out <- data.frame(samp= rep( x = colnames(sim.uuni), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.uuni))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Unweighted UniFrac"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Unweighted UniFrac\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results)

df_results[["Alcoa-UnWtUniFrac-ASV-withTree"]] <- df_out
#df_out <- df_results[[1]]


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Unweighted UniFrac Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 196.58, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14     ab       ab  
# 2    17     ac       a c 
# 4    25      a       a   
# 5    29     bd        b d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(50,58,61,61,62,65,65),
                   dist_measure = rep( "Unweighted UniFrac", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Unweighted UniFrac\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-UnWtUniFrac-ASV-withTree"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "unifrac", weighted = FALSE)

ord
# Call:
#   metaMDS(comm = ps.dist) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     ps.dist 
# Distance: user supplied 
# 
# Dimensions: 2 
# Stress:     0.07536944 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation 
# Species: scores missing

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 0
names(ord_plots)
ord_plots[["Alcoa-UnWtUniFrac-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj)
ord_obj[["Alcoa-UnWtUniFrac-ASV-withTree"]] <- ord

#-------------------------




## Plot of 6: (ASV >) "Genus" > "Family" > "Order" > "Class" > "Phylum" - PRUNED to remove unclassified taxa at that level

## Alcoa - with Tree - v. BRAY-CURTIS - GENUS-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Genus")
phy.alcoa.withTree.GENUS <- phy_in
#phy_in <- phy.alcoa.withTree.GENUS

ntaxa(phy_in) # 771
sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Genus"]))) # 458
sel <- which(tax_table(phy_in)[ ,"Genus"]=="g__Unclassified") # qty 314

 
phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 457 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 457 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 457 tips and 456 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 42.31736 %


sample_sums(phy_in)
min(sample_sums(phy_in)) # 10378

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10378

ntaxa(r1.ps) #  454
sum(sample_sums(r1.ps)) # 373608

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 454 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 454 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 454 tips and 453 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Genus (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nGenus-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Genus-withTree-Pruned"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 334.91, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 3   Ref 9   26  30  15  18 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(55,72,80,81,83,86,87),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Genus (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nGenus-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Genus-withTree-Pruned"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1082902 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Genus-withTree-Pruned"]] <- p


#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Genus-withTree-Pruned"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - FAMILY-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Family")
phy.alcoa.withTree.FAMILY <- phy_in
#phy_in <- phy.alcoa.withTree.FAMILY

ntaxa(phy_in) # 404
sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Family"]))) # 242
sel <- which(tax_table(phy_in)[ ,"Family"]=="f__Unclassified") # qty 161


phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 243 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 243 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 243 tips and 242 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 20.23254 %


sample_sums(phy_in)
min(sample_sums(phy_in)) # 14109


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 14109

ntaxa(r1.ps) #  242
sum(sample_sums(r1.ps)) # 507924

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 242 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 242 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 242 tips and 241 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Family (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nFamily-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Family-withTree-Pruned"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 274.92, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 

pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8      b        b  
# 1    14     ab       ab  
# 2    17     ab       ab  
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(63,80,85,86,88,92,91),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Family (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nFamily-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Family-withTree-Pruned"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.09505967 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)




p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Family-withTree-Pruned"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Family-withTree-Pruned"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - ORDER-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Order")
phy.alcoa.withTree.ORDER <- phy_in
#phy_in <- phy.alcoa.withTree.ORDER

ntaxa(phy_in) # 266
sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Order"]))) # 203
sel <- which(tax_table(phy_in)[ ,"Order"]=="o__Unclassified") # qty 64


phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 202 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 202 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 202 tips and 201 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 3.157457 %


sample_sums(phy_in)
min(sample_sums(phy_in)) # 17299



# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17299

ntaxa(r1.ps) #  200
sum(sample_sums(r1.ps)) # 622764

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 200 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 200 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 200 tips and 199 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Order (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nOrder-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Order-withTree-Pruned"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 300.8, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) #  2   Ref 8   25  29  14  17 



pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	7 obs. of  3 variables:

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(69,83,86,88,89,92,96),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Order (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nOrder-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Order-withTree-Pruned"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.09700408 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Order-withTree-Pruned"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Order-withTree-Pruned"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - CLASS-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Class")
phy.alcoa.withTree.CLASS <- phy_in
#phy_in <- phy.alcoa.withTree.CLASS

ntaxa(phy_in) # 110
sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Class"]))) # 95
sel <- which(tax_table(phy_in)[ ,"Class"]=="c__Unclassified") # qty 16


phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 94 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 94 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 94 tips and 93 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 0.5679225 %


sample_sums(phy_in)
min(sample_sums(phy_in)) # 17448



# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17448

ntaxa(r1.ps) #  92
sum(sample_sums(r1.ps)) # 628128

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 92 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 92 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 92 tips and 91 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Class (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nClass-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Class-withTree-Pruned"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 331.28, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(79,87,91,92,94,96,98),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Class (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nClass-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Class-withTree-Pruned"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1023132 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Class-withTree-Pruned"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Class-withTree-Pruned"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - PHYLUM-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Phylum")
phy.alcoa.withTree.PHYLUM <- phy_in
#phy_in <- phy.alcoa.withTree.PHYLUM


ntaxa(phy_in) # 33
as.character( unique(tax_table(phy_in)[,"Phylum"]) )
# [1] "p__Proteobacteria"               "p__RCP2-54"                      "p__Desulfobacterota"            
# [4] "p__MBNT15"                       "p__NB1-j"                        "p__Bdellovibrionota"            
# [7] "p__Verrucomicrobiota"            "p__Planctomycetota"              "p__Abditibacteriota"            
# [10] "p__Latescibacterota"             "p__Sumerlaeota"                  "p__Spirochaetota"               
# [13] "p__FCPU426"                      "p__Cyanobacteria"                "p__Patescibacteria"             
# [16] "p__Chloroflexi"                  "p__WPS-2"                        "p__WS2"                         
# [19] "p__Bacteroidota"                 "p__Elusimicrobiota"              "p__Armatimonadota"              
# [22] "p__Gemmatimonadota"              "p__Fibrobacterota"               "p__Dependentiae"                
# [25] "p__Acidobacteriota"              "p__WS4"                          "p__Firmicutes"                  
# [28] "p__Methylomirabilota"            "p__Actinobacteriota"             "p__Nitrospirota"                
# [31] "p__Entotheonellaeota"            "p__Myxococcota"                  "p__SAR324 clade(Marine group B)"

sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Phylum"]))) # 458
sel <- which(tax_table(phy_in)[ ,"Phylum"]=="p__Unclassified") # empty

##phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33 tips and 32 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 0 % - at this stage, however reads were excluded during early cleaning



sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 7
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  33
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 33 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 33 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 33 tips and 32 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Phylum (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nPhylum-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Phylum-withTree-Pruned"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 292.2, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(86,92,93,96.5,97,98,98),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Phylum (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nPhylum-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Phylum-withTree-Pruned"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1382105 
# Stress type 1, weak ties
# No convergent solutions - best solution after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Phylum-withTree-Pruned"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Phylum-withTree-Pruned"]] <- ord


#-------------------------



## Alcoa - with Tree - v. JACCARD - GENUS-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Genus")
phy.alcoa.withTree.GENUS <- phy_in
#phy_in <- phy.alcoa.withTree.GENUS

ntaxa(phy_in) # 771
sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Genus"]))) # 458
sel <- which(tax_table(phy_in)[ ,"Genus"]=="g__Unclassified") # qty 314


phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 457 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 457 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 457 tips and 456 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 42.31736 %


sample_sums(phy_in)
min(sample_sums(phy_in)) # 10378



# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10378

ntaxa(r1.ps) #  454
sum(sample_sums(r1.ps)) # 373608

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 454 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 454 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 454 tips and 453 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Genus (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nGenus-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Genus-withTree-Pruned"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 236.52, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) #  2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17     ab       ab  
# 4    25     ab       ab  
# 5    29      a       a   
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(62,68,72,73,74,75,79),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Genus (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nGenus-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Genus-withTree-Pruned"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1180375 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Genus-withTree-Pruned"]] <- p



#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Genus-withTree-Pruned"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - FAMILY-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Family")
phy.alcoa.withTree.FAMILY <- phy_in
#phy_in <- phy.alcoa.withTree.FAMILY

ntaxa(phy_in) # 404
sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Family"]))) # 242
sel <- which(tax_table(phy_in)[ ,"Family"]=="f__Unclassified") # qty 161


phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 243 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 243 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 243 tips and 242 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 20.23254 %


sample_sums(phy_in)
min(sample_sums(phy_in)) # 14109


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 14109
ntaxa(r1.ps) #  242
sum(sample_sums(r1.ps)) # 507924

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 242 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 242 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 242 tips and 241 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
#  2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Family (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nFamily-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Family-withTree-Pruned"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 265.92, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest(  similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25     ab       ab  
# 5    29      b        b  
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(72,76,80,83,84,85,90),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Family (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nFamily-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Family-withTree-Pruned"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1123689 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Family-withTree-Pruned"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Family-withTree-Pruned"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - ORDER-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Order")
phy.alcoa.withTree.ORDER <- phy_in
#phy_in <- phy.alcoa.withTree.ORDER

ntaxa(phy_in) # 266
sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Order"]))) # 458
sel <- which(tax_table(phy_in)[ ,"Order"]=="o__Unclassified") # qty 64


phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 202 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 202 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 202 tips and 201 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 3.157457 %


sample_sums(phy_in)
min(sample_sums(phy_in)) # 17299


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17299

ntaxa(r1.ps) #  200
sum(sample_sums(r1.ps)) # 622764

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 200 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 200 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 200 tips and 199 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Order (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nOrder-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Order-withTree-Pruned"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 241.29, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) #  2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)


unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      b         b 
# 6     8      a        a  
# 1    14      a        a  
# 2    17      a        a  
# 4    25      a        a  
# 5    29      a        a  
# 7   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(70,80,82,83,82,84,88),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Order (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nOrder-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Order-withTree-Pruned"]] <- anno





## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1145402 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)




p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Order-withTree-Pruned"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Order-withTree-Pruned"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - CLASS-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Class")
phy.alcoa.withTree.CLASS <- phy_in
#phy_in <- phy.alcoa.withTree.CLASS

ntaxa(phy_in) # 110
sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Class"]))) # 95
sel <- which(tax_table(phy_in)[ ,"Class"]=="c__Unclassified") # qty 16


phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 94 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 94 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 94 tips and 93 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 0.5679225 %


sample_sums(phy_in)
min(sample_sums(phy_in)) # 17448


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17448

ntaxa(r1.ps) #  92
sum(sample_sums(r1.ps)) # 628128

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 92 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 92 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 92 tips and 91 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



table(df_out$group)
#  2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Class (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nClass-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Class-withTree-Pruned"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 234.01, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      b         b 
# 6     8      b         b 
# 1    14      a        a  
# 2    17      b         b 
# 4    25      b         b 
# 5    29      a        a  
# 7   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(75,81,83,81,80,85,93),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Class (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nClass-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Class-withTree-Pruned"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1102037 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Class-withTree-Pruned"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Class-withTree-Pruned"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - PHYLUM-level - PRUNED
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Phylum")
phy.alcoa.withTree.PHYLUM <- phy_in
#phy_in <- phy.alcoa.withTree.PHYLUM

ntaxa(phy_in) # 33

sort( as.character( unique(tax_table(phy_in)[,"Phylum"]) ))
# [1] "p__Proteobacteria"               "p__RCP2-54"                      "p__Desulfobacterota"             "p__MBNT15"                      
# [5] "p__NB1-j"                        "p__Bdellovibrionota"             "p__Verrucomicrobiota"            "p__Planctomycetota"             
# [9] "p__Abditibacteriota"             "p__Latescibacterota"             "p__Sumerlaeota"                  "p__Spirochaetota"               
# [13] "p__FCPU426"                      "p__Cyanobacteria"                "p__Patescibacteria"              "p__Chloroflexi"                 
# [17] "p__WPS-2"                        "p__WS2"                          "p__Bacteroidota"                 "p__Elusimicrobiota"             
# [21] "p__Armatimonadota"               "p__Gemmatimonadota"              "p__Fibrobacterota"               "p__Dependentiae"                
# [25] "p__Acidobacteriota"              "p__WS4"                          "p__Firmicutes"                   "p__Methylomirabilota"           
# [29] "p__Actinobacteriota"             "p__Nitrospirota"                 "p__Entotheonellaeota"            "p__Myxococcota"                 
# [33] "p__SAR324 clade(Marine group B)"

sum(sample_sums(phy_in)) # 1772249
sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485


( pre_prune <- sum(sample_sums(phy_in)) ) # 1772249

sort( unique(as.character( tax_table(phy_in)[ ,"Phylum"]))) # 33
sel <- which(tax_table(phy_in)[ ,"Phylum"]=="p__Unclassified") # empty

##phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-sel, ]))
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33 tips and 32 internal nodes ]:
#   taxa are rows

# % excluded reads due to unclassified GENERA
( percent_reads_excluded <- 100*(1 - sum(sample_sums(phy_in))/pre_prune) ) # 0 %


sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485



# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 7
sample_sums(r1.ps) # all 17485
ntaxa(r1.ps) #  33
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33 tips and 32 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Phylum (with Tree & pruned)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nPhylum-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Phylum-withTree-Pruned"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 75.138, df = 6, p-value = 3.595e-14


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      a        a  
# 6     8     bc         bc
# 1    14    abc        abc
# 2    17      a        a  
# 4    25     ab        ab 
# 5    29     bc         bc
# 7   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(88,93,94,92,96,97,98),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Phylum (with Tree & pruned)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nPhylum-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Phylum-withTree-Pruned"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1929036 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’  

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Phylum-withTree-Pruned"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Phylum-withTree-Pruned"]] <- ord


#-------------------------




## Plot of 6: (ASV >) "Genus" > "Family" > "Order" > "Class" > "Phylum" - NOT PRUNED, i.e. unclassified taxa are grouped into next available higher taxonomic group

## Alcoa - with Tree - v. BRAY-CURTIS - GENUS-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Genus")
phy.alcoa.withTree.GENUS <- phy_in
#phy_in <- phy.alcoa.withTree.GENUS

ntaxa(phy_in) # 771
#table(phy_in@tax_table[ , "Genus"])

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  767
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 767 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 767 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 767 tips and 766 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Genus (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nGenus-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Genus-withTree"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 326.68, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(58,72,80,81,82,86,87),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Genus (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nGenus-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Genus-withTree"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.09785614 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Genus-withTree"]] <- p


#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Genus-withTree"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - FAMILY-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Family")
phy.alcoa.withTree.FAMILY <- phy_in
#phy_in <- phy.alcoa.withTree.FAMILY

ntaxa(phy_in) # 404

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  400
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 400 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 400 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 400 tips and 399 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Family (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nFamily-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Family-withTree"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)



# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 302.39, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(61,78,85,86,88,89,91),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Family (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nFamily-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Family-withTree"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.08985758 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Family-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Family-withTree"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - ORDER-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Order")
phy.alcoa.withTree.ORDER <- phy_in
#phy_in <- phy.alcoa.withTree.ORDER

ntaxa(phy_in) # 266

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  265
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 265 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 265 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 265 tips and 264 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Order (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nOrder-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Order-withTree"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)




# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 311.7, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) #  2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(66,81,86,88,89,92,93),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Order (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nOrder-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Order-withTree"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.09746626 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Order-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Order-withTree"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - CLASS-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Class")
phy.alcoa.withTree.CLASS <- phy_in
#phy_in <- phy.alcoa.withTree.CLASS

ntaxa(phy_in) # 110

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  109
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 109 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 109 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 109 tips and 108 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Class (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nClass-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Class-withTree"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)



# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 332.58, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 3   Ref 9   26  30  15  18 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(78,88,91,92,93,95,96),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Class (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nClass-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Class-withTree"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1002544 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Class-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Class-withTree"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - PHYLUM-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Phylum")
phy.alcoa.withTree.PHYLUM <- phy_in
#phy_in <- phy.alcoa.withTree.PHYLUM

ntaxa(phy_in) # 33

as.character( unique(tax_table(phy_in)[,"Phylum"]) )
# [1] "p__Proteobacteria"               "p__RCP2-54"                      "p__Desulfobacterota"             "p__MBNT15"                      
# [5] "p__NB1-j"                        "p__Bdellovibrionota"             "p__Verrucomicrobiota"            "p__Planctomycetota"             
# [9] "p__Abditibacteriota"             "p__Latescibacterota"             "p__Sumerlaeota"                  "p__Spirochaetota"               
# [13] "p__FCPU426"                      "p__Cyanobacteria"                "p__Patescibacteria"              "p__Chloroflexi"                 
# [17] "p__WPS-2"                        "p__WS2"                          "p__Bacteroidota"                 "p__Elusimicrobiota"             
# [21] "p__Armatimonadota"               "p__Gemmatimonadota"              "p__Fibrobacterota"               "p__Dependentiae"                
# [25] "p__Acidobacteriota"              "p__WS4"                          "p__Firmicutes"                   "p__Methylomirabilota"           
# [29] "p__Actinobacteriota"             "p__Nitrospirota"                 "p__Entotheonellaeota"            "p__Myxococcota"                 
# [33] "p__SAR324 clade(Marine group B)"

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 7
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  33
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 33 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 33 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 33 tips and 32 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "Phylum (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nPhylum-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-Phylum-withTree"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)



# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 292.2, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(86,92,93,96.5,97,99,99),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "Phylum (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nPhylum-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-Phylum-withTree"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1382105 
# Stress type 1, weak ties
# No convergent solutions - best solution after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-Phylum-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-Phylum-withTree"]] <- ord


#-------------------------



## Alcoa - with Tree - v. JACCARD - GENUS-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Genus")
phy.alcoa.withTree.GENUS <- phy_in
#phy_in <- phy.alcoa.withTree.GENUS

ntaxa(phy_in) # 771
#table(phy_in@tax_table[ , "Genus"])

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  767
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 767 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 767 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 767 tips and 766 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Genus (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nGenus-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Genus-withTree"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)



# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 248.48, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14     ab       ab  
# 2    17     ac       a c 
# 4    25     ab       ab  
# 5    29      b        b  
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(63,69,75,76,77,78,80),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Genus (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nGenus-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Genus-withTree"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.08583889 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Genus-withTree"]] <- p


#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Genus-withTree"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - FAMILY-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Family")
phy.alcoa.withTree.FAMILY <- phy_in
#phy_in <- phy.alcoa.withTree.FAMILY

ntaxa(phy_in) # 404

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  400
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 400 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 400 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 400 tips and 399 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Family (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nFamily-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Family-withTree"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 279.58, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) #  "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14     ab       ab  
# 2    17     ac       a c 
# 4    25      a       a   
# 5    29      b        b  
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(65,73,79,80,81,83,86),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Family (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nFamily-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Family-withTree"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.09066587 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Family-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Family-withTree"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - ORDER-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Order")
phy.alcoa.withTree.ORDER <- phy_in
#phy_in <- phy.alcoa.withTree.ORDER

ntaxa(phy_in) # 266

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  265
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 265 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 265 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 265 tips and 264 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Order (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nOrder-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Order-withTree"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 240.23, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 

pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14     ab       ab  
# 2    17      c         c 
# 4    25     ac       a c 
# 5    29     bd        b d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(66,76,79,76,78,80,86),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Order (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nOrder-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Order-withTree"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1070148 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Order-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Order-withTree"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - CLASS-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Class")
phy.alcoa.withTree.CLASS <- phy_in
#phy_in <- phy.alcoa.withTree.CLASS

ntaxa(phy_in) # 110

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  109
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 109 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 109 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 109 tips and 108 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Class (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nClass-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Class-withTree"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 220.24, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	7 obs. of  3 variables:

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14     ab       ab  
# 2    17     ac       a c 
# 4    25     ac       a c 
# 5    29      b        b  
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(75,80,82,80,80,85,89),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Class (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nClass-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Class-withTree"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1121146 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Class-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Class-withTree"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - PHYLUM-level - NOT PRUNED !!!
#-------------------------

phy_in <- phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 31746

phy_in <- tax_glom(physeq = phy_in, taxrank = "Phylum")
phy.alcoa.withTree.PHYLUM <- phy_in
#phy_in <- phy.alcoa.withTree.PHYLUM

ntaxa(phy_in) # 33

as.character( unique(tax_table(phy_in)[,"Phylum"]) )
# [1] "p__Proteobacteria"               "p__RCP2-54"                      "p__Desulfobacterota"             "p__MBNT15"                      
# [5] "p__NB1-j"                        "p__Bdellovibrionota"             "p__Verrucomicrobiota"            "p__Planctomycetota"             
# [9] "p__Abditibacteriota"             "p__Latescibacterota"             "p__Sumerlaeota"                  "p__Spirochaetota"               
# [13] "p__FCPU426"                      "p__Cyanobacteria"                "p__Patescibacteria"              "p__Chloroflexi"                 
# [17] "p__WPS-2"                        "p__WS2"                          "p__Bacteroidota"                 "p__Elusimicrobiota"             
# [21] "p__Armatimonadota"               "p__Gemmatimonadota"              "p__Fibrobacterota"               "p__Dependentiae"                
# [25] "p__Acidobacteriota"              "p__WS4"                          "p__Firmicutes"                   "p__Methylomirabilota"           
# [29] "p__Actinobacteriota"             "p__Nitrospirota"                 "p__Entotheonellaeota"            "p__Myxococcota"                 
# [33] "p__SAR324 clade(Marine group B)"

sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 7
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  33
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 33 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 33 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 33 tips and 32 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "Phylum (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nPhylum-level"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-Phylum-withTree"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 75.138, df = 6, p-value = 3.595e-14


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) #  2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      a        a  
# 6     8     bc         bc
# 1    14    abc        abc
# 2    17      a        a  
# 4    25     ab        ab 
# 5    29     bc         bc
# 7   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(88,93,93,92,96,97,98),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "Phylum (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nPhylum-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-Phylum-withTree"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1929036 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-Phylum-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-Phylum-withTree"]] <- ord


#-------------------------





## Plot of 5: (ASV >) 99% > 97% > 95% > 90% identity OTUs

cluster_names # "otu99" "otu97" "otu95" "otu90"

## Alcoa - with Tree - v. BRAY-CURTIS - 99% OTUs
#-------------------------

phy_in <- phy_otu_clust[[ "otu99" ]]
rank_names(phy_in) # "Kingdom" "otu99" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 18062

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  17584
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 17584 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 17584 taxa by 2 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 17584 tips and 17583 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "99% OTUs"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\n99% OTUs"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-otu99"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 364.14, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) #  2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14      a       a   
# 2    17      b        b  
# 4    25      b        b  
# 5    29     bd        b d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(30,40,50,56,57,58,63),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "99% OTUs", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\n99% OTUs", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-otu99"]] <- anno




## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1044828 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-otu99"]] <- p


#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-otu99"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - 97% OTUs
#-------------------------

phy_in <- phy_otu_clust[[ "otu97" ]]  #phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "otu97" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 8663

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  8461
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 8461 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 8461 taxa by 2 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 8461 tips and 8460 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "97% OTUs"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\n97% OTUs"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-otu97"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 365.41, df = 6, p-value < 2.2e-16



## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14      a       a   
# 2    17      b        b  
# 4    25      b        b  
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(40,50,60,65,66,70,75),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "97% OTUs", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\n97% OTUs", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-otu97"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.09805841 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-otu97"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-otu97"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - 95% OTUs
#-------------------------

phy_in <- phy_otu_clust[[ "otu95" ]]  #phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "otu95" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 4868

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  4772
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 4772 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 4772 taxa by 2 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 4772 tips and 4771 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "95% OTUs"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\n95% OTUs"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-otu95"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 350.01, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(45,60,67,70,73,77,80),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "95% OTUs", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\n95% OTUs", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-otu95"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.08540078 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-otu95"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-otu95"]] <- ord


#-------------------------

## Alcoa - with Tree - v. BRAY-CURTIS - 90% OTUs
#-------------------------

phy_in <- phy_otu_clust[[ "otu90" ]]  #phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "otu90"  
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 1512

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  1489
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 1489 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 1489 taxa by 2 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 1489 tips and 1488 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "39041" "39043" "39045" "39047" "39049" "39051" "39053" "39055" "39057" "39059" "39061" "39063" "39065"
# [14] "39067" "39069" "39071" "39073" "39075" "39077" "39079" "39081" "39083" "39085" "39087" "39089" "39091"
# [27] "39093" "39095" "39097" "39099" "39161" "39163" "39165" "39191" "39193" "39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "90% OTUs"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\n90% OTUs"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Bray-otu90"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 315.06, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) #  "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(60,75,78,80,82,86,85),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "90% OTUs", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\n90% OTUs", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Bray-otu90"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.07138324 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Bray-otu90"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Bray-otu90"]] <- ord


#-------------------------




## Alcoa - with Tree - v. JACCARD - 99% OTUs
#-------------------------

phy_in <- phy_otu_clust[[ "otu99" ]]  #phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "otu99"  
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 18062

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  17584
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 17584 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 17584 taxa by 2 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 17584 tips and 17583 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "99% OTUs"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\n99% OTUs"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-otu99"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 234.17, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) #  2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c        c  
# 6     8     bc       bc  
# 1    14     ab      ab   
# 2    17     ab      ab   
# 4    25     ad      a  d 
# 5    29     de         de
# 7   Ref      e          e

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "3"   "9"   "15"  "18"  "26"  "30"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(25,30,35,37,38,40,45),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "99% OTUs", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\n99% OTUs", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-otu99"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1075423 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-otu99"]] <- p


#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-otu99"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - 97% OTUs
#-------------------------

phy_in <- phy_otu_clust[[ "otu97" ]]  #phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "otu97" 
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 8663

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  8461
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 8461 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 8461 taxa by 2 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 8461 tips and 8460 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "97% OTUs"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\n97% OTUs"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-otu97"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 226.25, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      b         b 
# 6     8     ab        ab 
# 1    14      a        a  
# 2    17      a        a  
# 4    25      a        a  
# 5    29      c          c
# 7   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) #  "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(35,39,43,44,45,49,53),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "97% OTUs", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\n97% OTUs", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-otu97"]] <- anno




## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.09783034 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-otu97"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-otu97"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - 95% OTUs
#-------------------------

phy_in <- phy_otu_clust[[ "otu95" ]]  #phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "otu95"  
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 4868


sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  4772
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 4772 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 4772 taxa by 2 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 4772 tips and 4771 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "95% OTUs"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\n95% OTUs"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-otu95"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 210.4, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      b         b 
# 6     8     ab        ab 
# 1    14      a        a  
# 2    17      a        a  
# 4    25      a        a  
# 5    29      c          c
# 7   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(39,45,51,52,53,55,58),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "95% OTUs", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\n95% OTUs", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-otu95"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.08274257 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)




p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-otu95"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-otu95"]] <- ord


#-------------------------

## Alcoa - with Tree - v. JACCARD - 90% OTUs
#-------------------------

phy_in <- phy_otu_clust[[ "otu90" ]]  #phy.alcoa.withTree
rank_names(phy_in) # "Kingdom" "otu90"
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249
ntaxa(phy_in) # 1512


sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  1489
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 1489 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 1489 taxa by 2 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 1489 tips and 1488 internal nodes ]:
#   taxa are rows

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "90% OTUs"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\n90% OTUs"


# only make this once
#df_results <- list()
length(df_results) # 
names(df_results) # 

df_results[["Alcoa-Jaccard-otu90"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 247.15, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) #  2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)


unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      b         b 
# 6     8     ab        ab 
# 1    14      a        a  
# 2    17     ab        ab 
# 4    25      a        a  
# 5    29      c          c
# 7   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(51,58,61,60,61,65,68),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "90% OTUs", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\n90% OTUs", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) # 

df_anno[["Alcoa-Jaccard-otu90"]] <- anno






## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
# metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.05684303 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 
ord_plots[["Alcoa-Jaccard-otu90"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) # 
ord_obj[["Alcoa-Jaccard-otu90"]] <- ord


#-------------------------





## Plot of 4: All ASV > ">0.001%" > ">0.01%"  > ">0.1%"  
## predict spread in 95% prediction interval for time to reach target?
# length(taxa_rel_abuns) # 31746
# sel <- which(taxa_rel_abuns > 0.001); length(sel) # 17943
# sel <- which(taxa_rel_abuns > 0.01); length(sel) # 1338
# sel <- which(taxa_rel_abuns > 0.1); length(sel) # 72

## ">0.001 %" - with Tree - BRAY-CURTIS - ASV-level 
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
# rel_abun
# taxa_rel_abuns <- taxa_sums(rel_abun)
# hist(taxa_rel_abuns)
# summary(taxa_rel_abuns)
# # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# # 0.00270  0.02025  0.04118  0.11340  0.08373 51.26557 
# sum(taxa_rel_abuns) # 3600

taxa_rel_abuns <- 100*taxa_sums(phy_in)/sum(sample_sums(phy_in))
sum(taxa_rel_abuns) # 100
head(taxa_rel_abuns)
summary(taxa_rel_abuns)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0001129 0.0005643 0.0011849 0.0031500 0.0023699 1.3990698 

length(taxa_rel_abuns) # 31746
#sel <- which(taxa_rel_abuns > 0.0005); length(sel) # 25261
sel <- which(taxa_rel_abuns > 0.001); length(sel) # 17943
# sel <- which(taxa_rel_abuns > 0.005); length(sel) # 3179
# sel <- which(taxa_rel_abuns > 0.01); length(sel) # 1338

keep_taxa <- names(taxa_rel_abuns)[sel]
head(keep_taxa)

phy_in <- prune_taxa(taxa = keep_taxa, x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 17943 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 17943 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 17943 tips and 17942 internal nodes ]

min(taxa_sums(phy_in)) # 18


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 16752

ntaxa(r1.ps) #  17941
sum(sample_sums(r1.ps)) # 603072

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 17941 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 17941 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 17941 tips and 17940 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1"  

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- ">0.001% - ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- ">0.001% - Bray-Curtis\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 0
names(df_results)

df_results[["Alcoa-Bray-ASV-withTree->0.001%"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 359.09, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)


unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14      a       a   
# 2    17      b        b  
# 4    25      b        b  
# 5    29     bd        b d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(22,32,39,48,49,53,59),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( ">0.001% - ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( ">0.001% - Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-Bray-ASV-withTree->0.001%"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1091212 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 0
names(ord_plots)
ord_plots[["Alcoa-Bray-ASV-withTree->0.001%"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj)
ord_obj[["Alcoa-Bray-ASV-withTree->0.001%"]] <- ord

#-------------------------

## ">0.01 %" - with Tree - BRAY-CURTIS - ASV-level 
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
# rel_abun
# taxa_rel_abuns <- taxa_sums(rel_abun)
# hist(taxa_rel_abuns)
# summary(taxa_rel_abuns)
# # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# # 0.00270  0.02025  0.04118  0.11340  0.08373 51.26557 
# sum(taxa_rel_abuns) # 3600

taxa_rel_abuns <- 100*taxa_sums(phy_in)/sum(sample_sums(phy_in))
sum(taxa_rel_abuns) # 100
head(taxa_rel_abuns)
summary(taxa_rel_abuns)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0001129 0.0005643 0.0011849 0.0031500 0.0023699 1.3990698 

length(taxa_rel_abuns) # 31746
#sel <- which(taxa_rel_abuns > 0.0005); length(sel) # 25261
#sel <- which(taxa_rel_abuns > 0.001); length(sel) # 17943
#sel <- which(taxa_rel_abuns > 0.005); length(sel) # 3179
sel <- which(taxa_rel_abuns > 0.01); length(sel) # 1338

keep_taxa <- names(taxa_rel_abuns)[sel]
head(keep_taxa)

phy_in <- prune_taxa(taxa = keep_taxa, x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1338 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 1338 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 1338 tips and 1337 internal nodes ]

min(taxa_sums(phy_in)) # 178


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 49
sample_sums(r1.ps) # all 9691

ntaxa(r1.ps) #  1338
sum(sample_sums(r1.ps)) # 348876

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1338 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 1338 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 1338 tips and 1337 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1"  

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 


df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- ">0.01% - ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- ">0.01% - Bray-Curtis\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 0
names(df_results)

df_results[["Alcoa-Bray-ASV-withTree->0.01%"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 374.3, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)


unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14      a       a   
# 2    17      b        b  
# 4    25      b        b  
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(33,43,55,62,64,65,74),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( ">0.01% - ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( ">0.01% - Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-Bray-ASV-withTree->0.01%"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1061111 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 0
names(ord_plots)
ord_plots[["Alcoa-Bray-ASV-withTree->0.01%"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj)
ord_obj[["Alcoa-Bray-ASV-withTree->0.01%"]] <- ord

#-------------------------

## ">0.1 %" - with Tree - BRAY-CURTIS - ASV-level 
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
# rel_abun
# taxa_rel_abuns <- taxa_sums(rel_abun)
# hist(taxa_rel_abuns)
# summary(taxa_rel_abuns)
# # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# # 0.00270  0.02025  0.04118  0.11340  0.08373 51.26557 
# sum(taxa_rel_abuns) # 3600

taxa_rel_abuns <- 100*taxa_sums(phy_in)/sum(sample_sums(phy_in))
sum(taxa_rel_abuns) # 100
head(taxa_rel_abuns)
summary(taxa_rel_abuns)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0001129 0.0005643 0.0011849 0.0031500 0.0023699 1.3990698 

length(taxa_rel_abuns) # 31746
#sel <- which(taxa_rel_abuns > 0.0005); length(sel) # 25261
#sel <- which(taxa_rel_abuns > 0.001); length(sel) # 17943
#sel <- which(taxa_rel_abuns > 0.005); length(sel) # 3179
#sel <- which(taxa_rel_abuns > 0.01); length(sel) # 1338
sel <- which(taxa_rel_abuns > 0.1); length(sel) # 72

keep_taxa <- names(taxa_rel_abuns)[sel]
head(keep_taxa)

phy_in <- prune_taxa(taxa = keep_taxa, x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 72 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 72 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 72 tips and 71 internal nodes ]

min(taxa_sums(phy_in)) # 1793


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 213
sample_sums(r1.ps) # all 1382

ntaxa(r1.ps) #  72
sum(sample_sums(r1.ps)) # 49752

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 72 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 72 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 72 tips and 71 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1"  

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- ">0.1% - ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- ">0.1% - Bray-Curtis\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 0
names(df_results)

df_results[["Alcoa-Bray-ASV-withTree->0.1%"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 362.03, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     bc        bc 
# 1    14     ab       ab  
# 2    17      a       a   
# 4    25      a       a   
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(45,58,68,75,77,81,85),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( ">0.1% - ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( ">0.1% - Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-Bray-ASV-withTree->0.1%"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1020813 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 0
names(ord_plots)
ord_plots[["Alcoa-Bray-ASV-withTree->0.1%"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj)
ord_obj[["Alcoa-Bray-ASV-withTree->0.1%"]] <- ord

#-------------------------


## ">0.001 %" - with Tree - JACCARD - ASV-level 
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
# rel_abun
# taxa_rel_abuns <- taxa_sums(rel_abun)
# hist(taxa_rel_abuns)
# summary(taxa_rel_abuns)
# # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# # 0.00270  0.02025  0.04118  0.11340  0.08373 51.26557 
# sum(taxa_rel_abuns) # 3600

taxa_rel_abuns <- 100*taxa_sums(phy_in)/sum(sample_sums(phy_in))
sum(taxa_rel_abuns) # 100
head(taxa_rel_abuns)
summary(taxa_rel_abuns)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0001129 0.0005643 0.0011849 0.0031500 0.0023699 1.3990698 

length(taxa_rel_abuns) # 31746

sel <- which(taxa_rel_abuns > 0.001); length(sel) # 17943
# sel <- which(taxa_rel_abuns > 0.01); length(sel) # 1338
# sel <- which(taxa_rel_abuns > 0.1); length(sel) # 1338

keep_taxa <- names(taxa_rel_abuns)[sel]
head(keep_taxa)

phy_in <- prune_taxa(taxa = keep_taxa, x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 17943 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 17943 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 17943 tips and 17942 internal nodes ]

min(taxa_sums(phy_in)) # 18


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 16752

ntaxa(r1.ps) #  17941
sum(sample_sums(r1.ps)) # 603072

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 17941 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 17941 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 17941 tips and 17940 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- ">0.001% - ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- ">0.001% - Jaccard\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 0
names(df_results)

df_results[["Alcoa-Jaccard-ASV-withTree->0.001%"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 247.32, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)


unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c        c  
# 6     8     bc       bc  
# 1    14     ab      ab   
# 2    17      a      a    
# 4    25     ad      a  d 
# 5    29     de         de
# 7   Ref      e          e

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(22,25,30,32,36,38,45),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( ">0.001% - ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( ">0.001% - Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-Jaccard-ASV-withTree->0.001%"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1113814 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 


str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 0
names(ord_plots)
ord_plots[["Alcoa-Jaccard-ASV-withTree->0.001%"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj)
ord_obj[["Alcoa-Jaccard-ASV-withTree->0.001%"]] <- ord

#-------------------------

## ">0.01 %" - with Tree - JACCARD - ASV-level 
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
# rel_abun
# taxa_rel_abuns <- taxa_sums(rel_abun)
# hist(taxa_rel_abuns)
# summary(taxa_rel_abuns)
# # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# # 0.00270  0.02025  0.04118  0.11340  0.08373 51.26557 
# sum(taxa_rel_abuns) # 3600

taxa_rel_abuns <- 100*taxa_sums(phy_in)/sum(sample_sums(phy_in))
sum(taxa_rel_abuns) # 100
head(taxa_rel_abuns)
summary(taxa_rel_abuns)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0001129 0.0005643 0.0011849 0.0031500 0.0023699 1.3990698 

length(taxa_rel_abuns) # 31746
#sel <- which(taxa_rel_abuns > 0.001); length(sel) # 17943
sel <- which(taxa_rel_abuns > 0.01); length(sel) # 1338
#sel <- which(taxa_rel_abuns > 0.1); length(sel) # 1338

keep_taxa <- names(taxa_rel_abuns)[sel]
head(keep_taxa)

phy_in <- prune_taxa(taxa = keep_taxa, x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1338 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 1338 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 1338 tips and 1337 internal nodes ]

min(taxa_sums(phy_in)) # 178


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 49
sample_sums(r1.ps) # all 9691

ntaxa(r1.ps) #  1338
sum(sample_sums(r1.ps)) # 348876

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1338 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 1338 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 1338 tips and 1337 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- ">0.01% - ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- ">0.01% - Jaccard\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 0
names(df_results)

df_results[["Alcoa-Jaccard-ASV-withTree->0.01%"]] <- df_out


p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 304.33, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      d         d 
# 6     8     bd       b d 
# 1    14     ab      ab   
# 2    17     ac      a c  
# 4    25     ac      a c  
# 5    29     ce        c e
# 7   Ref      e          e

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(51,61,69,75,75,78,80),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( ">0.01% - ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( ">0.01% - Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-Jaccard-ASV-withTree->0.01%"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.09820217 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 0
names(ord_plots)
ord_plots[["Alcoa-Jaccard-ASV-withTree->0.01%"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj)
ord_obj[["Alcoa-Jaccard-ASV-withTree->0.01%"]] <- ord

#-------------------------

## ">0.1 %" - with Tree - JACCARD - ASV-level 
#-------------------------

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

sample_sums(phy_in)
min(sample_sums(phy_in)) # 17485

# rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
# rel_abun
# taxa_rel_abuns <- taxa_sums(rel_abun)
# hist(taxa_rel_abuns)
# summary(taxa_rel_abuns)
# # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# # 0.00270  0.02025  0.04118  0.11340  0.08373 51.26557 
# sum(taxa_rel_abuns) # 3600

taxa_rel_abuns <- 100*taxa_sums(phy_in)/sum(sample_sums(phy_in))
sum(taxa_rel_abuns) # 100
head(taxa_rel_abuns)
summary(taxa_rel_abuns)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0001129 0.0005643 0.0011849 0.0031500 0.0023699 1.3990698 

length(taxa_rel_abuns) # 31746
#sel <- which(taxa_rel_abuns > 0.001); length(sel) # 17943
#sel <- which(taxa_rel_abuns > 0.01); length(sel) # 1338
sel <- which(taxa_rel_abuns > 0.1); length(sel) # 72

keep_taxa <- names(taxa_rel_abuns)[sel]
head(keep_taxa)

phy_in <- prune_taxa(taxa = keep_taxa, x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 72 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 72 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 72 tips and 71 internal nodes ]

min(taxa_sums(phy_in)) # 1793


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 213
sample_sums(r1.ps) # all 1382

ntaxa(r1.ps) #  72
sum(sample_sums(r1.ps)) # 49752

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 72 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 72 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 72 tips and 71 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"
identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`Location description`
# [1] "karri 11"     "karri 11"     "mahogany 1"   "mahogany 1"   "coolabah 9"   "coolabah 9"   "karri 6"     
# [8] "karri 6"      "casuarina 1"  "casuarina 1"  "marri 7/9"    "marri 7/9"    "1991-2/G4613" "1991-2/G4613"
# [15] "F4615"        "F4615"        "F4601"        "F4601"        "F4525"        "F4525"        "1991-4"      
# [22] "1991-4"       "E4424"        "E4424"        "2002-2"       "2002-2"       "1999-5"       "1999-5"      
# [29] "1991-1"       "1991-1"       "1999-3"       "1999-3"       "2002-5"       "2002-5"       "1999-1"      
# [36] "1999-1" 

r1.ps@sam_data$Notes
# [1] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [6] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [11] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [16] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [21] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [26] "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural"
# [31] "post mine rehab"  "adjacent natural" "post mine rehab"  "adjacent natural" "post mine rehab" 
# [36] "adjacent natural"

row.names(r1.ps@sam_data)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063"
# [13] "X39065" "X39067" "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087"
# [25] "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18

( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# [1] "X39043" "X39047" "X39051" "X39055" "X39059" "X39063" "X39067" "X39071" "X39075" "X39079" "X39083" "X39087"
# [13] "X39091" "X39095" "X39099" "X39163" "X39191" "X39195"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     site=NA,
                     site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
                                     " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  df_out$similarity[i] <- df_in[row,col]
}


dim(df_out) # 648  6

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 18
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 630  6

unique(df_out$site)
# [1] "karri 11"     "mahogany 1"   "coolabah 9"   "karri 6"      "casuarina 1"  "marri 7/9"    "1991-2/G4613"
# [8] "F4615"        "F4601"        "F4525"        "1991-4"       "E4424"        "2002-2"       "1999-5"      
# [15] "1991-1"       "1999-3"       "2002-5"       "1999-1" 

unique(df_out$site_full_desc)
# [1] "karri 11 (post mine rehab)"      "karri 11 (adjacent natural)"     "mahogany 1 (post mine rehab)"   
# [4] "mahogany 1 (adjacent natural)"   "coolabah 9 (post mine rehab)"    "coolabah 9 (adjacent natural)"  
# [7] "karri 6 (post mine rehab)"       "karri 6 (adjacent natural)"      "casuarina 1 (post mine rehab)"  
# [10] "casuarina 1 (adjacent natural)"  "marri 7/9 (post mine rehab)"     "marri 7/9 (adjacent natural)"   
# [13] "1991-2/G4613 (post mine rehab)"  "1991-2/G4613 (adjacent natural)" "F4615 (post mine rehab)"        
# [16] "F4615 (adjacent natural)"        "F4601 (post mine rehab)"         "F4601 (adjacent natural)"       
# [19] "F4525 (post mine rehab)"         "F4525 (adjacent natural)"        "1991-4 (post mine rehab)"       
# [22] "1991-4 (adjacent natural)"       "E4424 (post mine rehab)"         "E4424 (adjacent natural)"       
# [25] "2002-2 (post mine rehab)"        "2002-2 (adjacent natural)"       "1999-5 (post mine rehab)"       
# [28] "1999-5 (adjacent natural)"       "1991-1 (post mine rehab)"        "1991-1 (adjacent natural)"      
# [31] "1999-3 (post mine rehab)"        "1999-3 (adjacent natural)"       "2002-5 (post mine rehab)"       
# [34] "2002-5 (adjacent natural)"       "1999-1 (post mine rehab)"        "1999-1 (adjacent natural)"

length(df_out$site_full_desc) # 630

sel <- which(df_out$samp %in% ref_sites) # qty 306
unique( df_out$site_full_desc[sel] )
# [1] "karri 11 (adjacent natural)"     "mahogany 1 (adjacent natural)"   "coolabah 9 (adjacent natural)"  
# [4] "karri 6 (adjacent natural)"      "casuarina 1 (adjacent natural)"  "marri 7/9 (adjacent natural)"   
# [7] "1991-2/G4613 (adjacent natural)" "F4615 (adjacent natural)"        "F4601 (adjacent natural)"       
# [10] "F4525 (adjacent natural)"        "1991-4 (adjacent natural)"       "E4424 (adjacent natural)"       
# [13] "2002-2 (adjacent natural)"       "1999-5 (adjacent natural)"       "1991-1 (adjacent natural)"      
# [16] "1999-3 (adjacent natural)"       "2002-5 (adjacent natural)"       "1999-1 (adjacent natural)" 

unique( df_out$site_full_desc[-sel] )
# [1] "karri 11 (post mine rehab)"     "mahogany 1 (post mine rehab)"   "coolabah 9 (post mine rehab)"  
# [4] "karri 6 (post mine rehab)"      "casuarina 1 (post mine rehab)"  "marri 7/9 (post mine rehab)"   
# [7] "1991-2/G4613 (post mine rehab)" "F4615 (post mine rehab)"        "F4601 (post mine rehab)"       
# [10] "F4525 (post mine rehab)"        "1991-4 (post mine rehab)"       "E4424 (post mine rehab)"       
# [13] "2002-2 (post mine rehab)"       "1999-5 (post mine rehab)"       "1991-1 (post mine rehab)"      
# [16] "1999-3 (post mine rehab)"       "2002-5 (post mine rehab)"       "1999-1 (post mine rehab)"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df_out$group[sel] # all N/A
df_out$group[sel] <- "Ref"
unique(df_out$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df_out$group <- factor(df_out$group, 
                       levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                       labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



table(df_out$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 



df_out$dist_measure <- "Jaccard"
df_out$tax_level <- ">0.1% - ASV (with Tree)"
df_out$study <- "Alcoa"
df_out$corrected <- "no"  
df_out$facet_x <- ">0.1% - Jaccard\nASV-level"


# only make this once
#df_results <- list()
length(df_results) # 0
names(df_results)

df_results[["Alcoa-Jaccard-ASV-withTree->0.1%"]] <- df_out



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	630 obs. of  12 variables:


df_out$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 285.01, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_out$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8      a       a   
# 1    14      a       a   
# 2    17      b        b  
# 4    25     bd        b d
# 5    29      d          d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(75,91,93,98,98,98,98),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( ">0.1% - ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( ">0.1% - Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-Jaccard-ASV-withTree->0.1%"]] <- anno






## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.08420512 
# Stress type 1, weak ties
# No convergent solutions - best solution after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 0
names(ord_plots)
ord_plots[["Alcoa-Jaccard-ASV-withTree->0.1%"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj)
ord_obj[["Alcoa-Jaccard-ASV-withTree->0.1%"]] <- ord

#-------------------------






#### 2a. Iluka Eneabba post-mining restoration chronosequence
#    - create phyloseq object from metadata, OTU abundances, taxonomy
#    - data cleaning
#    - build and add phylogenetic tree
#-------------------------

### metadata

context <- read_excel(path= paste0(datadir,"/iluka_eneabba/","AM_ContextualMetadata_iluka.xlsx"),
                      sheet=1, range="A1:EI67", col_names = TRUE)

context <- as.data.frame(context)
str(context)

names(context)
# [1] "Sample_ID"                                             "NCBI_Submission"                                      
# [3] "NCBI_Sample_Accession"                                 "Organism"                                             
# [5] "Tax_ID"                                                "SampleName_depth"                                     
# [7] "NCBI_BioProject"                                       "Date_sampled [YYYY-MM-DD]"                            
# [9] "Time_sampled [hh:mm]"                                  "Latitude [decimal degrees]"                           
# [11] "Longitude [decimal degrees]"                           "Depth [m]"                                            
# [13] "geo_loc [country:subregion]"                           "Sample_Site (location_description)"                   
# [15] "year of rehab or ref"                                  "age of rehab"                                         
# [17] "NOTES"                                                 "Sample_attribution"                                   
# [19] "Sample_descriptor"                                     "Citation"                                             
# [21] "Funding_source"                                        "Sample_collection_device-method"                      
# [23] "Sample_material_processing"                            "Sample_storage_method"                                
# [25] "DNA_extraction_method"                                 "Notes (incl. phys/chem measurement method variation)" 
# [27] "Environment (free living)"                             "Env_material (control vocab 0)"                       
# [29] "Horizon (control vocab 1)"                             "Broad_land_use (control vocab 2)"                     
# [31] "Detailed_land_use (control vocab 2)"                   "General_Env_feature (control vocab 3)"                
# [33] "Env_biome (control vocab 4)"                           "Temperature [deg C]"                                  
# [35] "Vegetation Total cover [%]"                            "Vegetation Dom. Trees [%]"                            
# [37] "Vegetation Dom. Shrubs [%]"                            "Vegetation Dom. Grasses [%]"                          
# [39] "Elevation [m]"                                         "Slope [%]"                                            
# [41] "Slope Aspect [Direction or degrees; e.g., NW or 315?]" "Profile Position (control vocab 5)"                   
# [43] "Australian Soil Classification (control vocab 6)"      "FAO soil classification (control vocab 7)"            
# [45] "Immediate Previous Land Use (control vocab 2)"         "Date since change in Land Use"                        
# [47] "Crop rotation 1yr since present"                       "Crop rotation 2yrs since present"                     
# [49] "Crop rotation 3yrs since present"                      "Crop rotation 4yrs since present"                     
# [51] "Crop rotation 5yrs since present"                      "Agrochemical Additions"                               
# [53] "Tillage (control vocab 9)"                             "Fire"                                                 
# [55] "fire intensity if known"                               "Flooding"                                             
# [57] "Extreme Events"                                        "Soil moisture [%]"                                    
# [59] "Water Holding capacity"                                "bulk density [g/cm3 or kg/m3]"                        
# [61] "Microbial Biomass"                                     "Color (control vocab 10)"                             
# [63] "Gravel [%]- [ >2.0 mm]"                                "Texture []"                                           
# [65] "Course Sand [%] [200-2000 ?m]"                         "Fine Sand [%] [20-200 ?m]"                            
# [67] "Sand [%]"                                              "Silt [%] [2-20 ?m]"                                   
# [69] "Clay [%] [<2 ?m]"                                      "Ammonium Nitrogen [mg/Kg]"                            
# [71] "Nitrate_Nitrogen [mg/Kg]"                              "Phosphorus_Colwell [mg/Kg]"                           
# [73] "Potassium_Colwell [mg/Kg]"                             "Sulphur [mg/Kg]"                                      
# [75] "Organic_Carbon [%]"                                    "Conductivity [dS/m]"                                  
# [77] "Ph_Level_CaCl2 [pH]"                                   "pH_Level_H2O [pH]"                                    
# [79] "DTPA_Copper [mg/Kg]"                                   "DTPA_Iron [mg/Kg]"                                    
# [81] "DTPA_Manganese [mg/Kg]"                                "DTPA_Zinc [mg/Kg]"                                    
# [83] "Exc_Aluminium [meq/100g]"                              "Exc_Calcium [meq/100g]"                               
# [85] "Exc_Magnesium [meq/100g]"                              "Exc_Potassium [meq/100g]"                             
# [87] "Exc_Sodium [meq/100g]"                                 "Boron_Hot_CaCl2 [mg/Kg]"                              
# [89] "Cation Exchange Capacity"                              "Arsenic...90"                                         
# [91] "Scandium"                                              "Vanadium"                                             
# [93] "Chromium"                                              "Cobalt"                                               
# [95] "Nickel"                                                "Gallium"                                              
# [97] "Germanium"                                             "Arsenic...98"                                         
# [99] "Selenium"                                              "Rubidium"                                             
# [101] "Strontium"                                             "Yttrium"                                              
# [103] "Zirconium"                                             "Niobium [Columbium]"                                  
# [105] "Molybdenum"                                            "Ruthenium"                                            
# [107] "Rhodium"                                               "Palladium"                                            
# [109] "Silver"                                                "Cadmium"                                              
# [111] "Tin"                                                   "Antimony"                                             
# [113] "Cesium"                                                "Barium"                                               
# [115] "Lanthanum"                                             "Cerium"                                               
# [117] "Praseodymium"                                          "Neodymium"                                            
# [119] "Samarium"                                              "Europium"                                             
# [121] "Gadolinium"                                            "Terbium"                                              
# [123] "Dysprosium"                                            "Holmium"                                              
# [125] "Erbium"                                                "Thulium"                                              
# [127] "Ytterbium"                                             "Lutetium"                                             
# [129] "Hafnium"                                               "Tantalum"                                             
# [131] "Tungsten or Wolfram"                                   "Osmium"                                               
# [133] "Iridium"                                               "Platinum"                                             
# [135] "Gold"                                                  "Lead"                                                 
# [137] "Thorium"                                               "Uranium"                                              
# [139] "site"

# Note field: "year of rehab or ref"
row.names(context) <- context$Sample_ID
str(context)

gsub(pattern = "102.100.100/", replacement = "", x = row.names(context))
row.names(context) <- gsub(pattern = "102.100.100/", replacement = "", x = row.names(context))

row.names(context)
# [1] "138408" "138409" "138410" "138411" "138412" "138413" "138414" "138415" "138416" "138417" "138418" "138419" "138420"
# [14] "138421" "138422" "138423" "138424" "138425" "138426" "138427" "138428" "138429" "138430" "138431" "138432" "138433"
# [27] "138434" "138435" "138436" "138437" "138438" "138439" "138440" "138441" "138444" "138445" "138446" "138447" "138448"
# [40] "138449" "138450" "138451" "138452" "138453" "138454" "138455" "138456" "138457" "138458" "138459" "138460" "138461"
# [53] "138462" "138463" "138464" "138465" "138466" "138467" "138468" "138469" "138470" "138471" "138472" "138473" "138474"
# [66] "138475"

row.names(context) <- paste0("X",row.names(context))

table(context$`Depth [m]`)
# 0.01 0.02 
# 33   33 

sel <- which(context$`Depth [m]` == "0.01") # qty 33

context <- context[sel, ]


samp.16s <- context

write.csv(context, file = "iluka_eneabba-metadata.csv")


### OTU raw data matrix 

# raw.otu.16s <- read_excel(path= paste0(datadir,"/iluka_eneabba/","Iluka_otu_cluss_tax.xlsx"),
#                           sheet=1, range="A1:BU19427", col_names = TRUE)
# raw.otu.16s <- as.data.frame(raw.otu.16s)

raw.otu.16s <- read.delim(file = paste0(datadir,"/iluka_eneabba/","Iluka_otu_cluss_tax.txt"), header = TRUE, sep = "\t")
str(raw.otu.16s)
# 'data.frame':	133532 obs. of  75 variables:
# $ X.OTU.ID: chr  "AAAGAACGCTAGCAACGGATATAACACATGCAAGTCAATGTCAAACATGGCGTACGAGTGAGTAATAGTTCGGAATTTGCTTGAAAATTGGGAATAGAAAGTAATACCCAA"| __truncated__ "AAAGAACGCTAGCACGATACATTACACATGCAAGTGGAATGGCTAAAGTAAGAACAGGCCATCACGTACGGGTGAGTAAAATTTCTAAACATACCCTTTAGACCGAAGGAA"| __truncated__ "AAAGAACGCTGGCGGCATGCCTAACACATGCAAGTTGAACGAGTAGTAGCAATACTATTAGTAGCAAACGGGTGAGTAATATGTAGGAATCTACCCTATAGTATGGAACAA"| __truncated__ "AAAGAACGCTGGCGGCATGCTTAACACATGCAAGTCGAACGAGTAGTAGCAATACTATTAGTGGCAGACGGGTGAGTAATACGTAGGAATCTCCCCTATAGTACGGAACAA"| __truncated__ ...
# $ X138408 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138409 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138410 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138411 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138412 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138413 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138414 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138415 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138416 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138417 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138418 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138419 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138420 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138421 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138422 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138423 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138424 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138425 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138426 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138427 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138428 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138429 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138430 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138431 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138432 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138433 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138434 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138435 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138436 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138437 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138438 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138439 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138440 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138441 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138444 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138445 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138446 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138447 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138448 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138449 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138450 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138451 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138452 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138453 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138454 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138455 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138456 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138457 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138458 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138459 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138460 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138461 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138462 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138463 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138464 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138465 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138466 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138467 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138468 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138469 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138470 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138471 : num  0 0 0 0 0 10 0 0 0 0 ...
# $ X138472 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138473 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138474 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138475 : num  0 0 0 0 0 0 0 0 0 0 ...
# $ class   : chr  "Proteobacteria_unclassified" "Proteobacteria_unclassified" "Alphaproteobacteria" "Alphaproteobacteria" ...
# $ clus97ID: chr  "OTU97_12719" "OTU97_21472" "OTU97_21919" "OTU97_3248" ...
# $ clus99ID: chr  "OTU99_23592" "OTU99_45338" "OTU99_46371" "OTU99_5133" ...
# $ family  : chr  "Proteobacteria_unclassified" "Proteobacteria_unclassified" "Rickettsiaceae" "Rickettsiaceae" ...
# $ genus   : chr  "0" "0" "uncultured" "uncultured" ...
# $ kingdom : chr  "Bacteria" "Bacteria" "Bacteria" "Bacteria" ...
# $ order   : chr  "Proteobacteria_unclassified" "Proteobacteria_unclassified" "Rickettsiales" "Rickettsiales" ...
# $ phylum  : chr  "Proteobacteria" "Proteobacteria" "Proteobacteria" "Proteobacteria" ...


# store rep sequences
rep_seq.iluka_eneabba <- as.character( raw.otu.16s[, 1] )
head(rep_seq.iluka_eneabba)
names(rep_seq.iluka_eneabba) <- paste("OTU_",1:length(rep_seq.iluka_eneabba))
head(rep_seq.iluka_eneabba)


row.names(raw.otu.16s) <- names(rep_seq.iluka_eneabba)
# remove first column, it is saved in rep_seq
raw.otu.16s <- raw.otu.16s[ , -1]

temp.otu.iluka_eneabba <- raw.otu.16s

length(names(raw.otu.16s)) # 74

sel <- which(names(raw.otu.16s) %in% row.names(samp.16s)) # qty 33
otu.16s <- raw.otu.16s[ ,sel]

## align available OTU abundance data with samples?
length(row.names(samp.16s)) # 33
length(names(otu.16s)) # 33
class(row.names(samp.16s)) # "character"
class(names(otu.16s)) # "character"
identical( row.names(samp.16s), names(otu.16s) ) # TRUE

dim(samp.16s)
#  33  139

dim(otu.16s)
# 133532    33


# trim? Are there any OTUs with zero counts after limiting to surface samples only?

sel <- which( rowSums(otu.16s) < 1 )
otu.16s[sel[1:10], ] # all zeros

otu.16s <- otu.16s[-sel, ]
dim(otu.16s)
# 50845    33



### derive Taxonomy table

## read in representative sequences
length(rep_seq.iluka_eneabba) # 133532

keep_otus <- row.names(otu.16s)
rep_seq.iluka_eneabba <- rep_seq.iluka_eneabba[keep_otus]
length(rep_seq.iluka_eneabba) # 50845

identical(names(rep_seq.iluka_eneabba),row.names(otu.16s)) # TRUE

### Assign taxonomy using dada2
# https://benjjneb.github.io/dada2/tutorial_1_8.html

# SILVA taxonomy training databases from here 
# https://zenodo.org/record/3986799#.X5kj-aozaUl

# assignTaxonomy(seqs, refFasta, minBoot = 50, tryRC = FALSE,
#                outputBootstraps = FALSE, taxLevels = c("Kingdom", "Phylum", "Class",
#                                                        "Order", "Family", "Genus", "Species"), multithread = FALSE,
#                verbose = FALSE)



head(as.character(rep_seq.iluka_eneabba))

taxa <- assignTaxonomy(seqs = as.character(rep_seq.iluka_eneabba), 
                       refFasta = "/Users/lidd0026/WORKSPACE/DATA/SILVA_DB/silva_nr99_v138_train_set.fa.gz",
                       tryRC = TRUE,
                       multithread=TRUE, verbose = TRUE)

temp.taxa.iluka_eneabba <- taxa
#taxa <- temp.taxa.iluka_eneabba

str(taxa)
# Large matrix

tax.16s <- as.data.frame(taxa)
head(row.names(tax.16s)) # these are sequences ... swap for OTU names

identical(names(rep_seq.iluka_eneabba),row.names(otu.16s)) # TRUE
identical(toupper(as.character(rep_seq.iluka_eneabba)),row.names(tax.16s)) # TRUE

# swap labels
row.names(tax.16s) <- names(rep_seq.iluka_eneabba)





# tidy  up taxa table

tax <- tax.16s
names(tax)
# "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

dim(tax) # 50845     6

temp <- tax
#tax <- temp

#taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__", "s__")
taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__")

for (i in 1:length(names(tax))) {
  #i<-1
  tax[ , names(tax)[i] ] <- paste0(taxa_prefix[i],tax[ , names(tax)[i] ])
  
  sel <- which( is.na(tax[, i]) | tax[, i] == "" | tax[, i] == taxa_prefix[i] | tax[, i] == paste0(taxa_prefix[i],NA) )
  if (length(sel)>=1) {
    tax[ sel, names(tax)[i] ] <- paste0(taxa_prefix[i],"Unclassified")
  }
  print(paste0("completed ",i))
}

write.csv(tax, file = "iluka_eneabba-surf-16s-Taxonomy-Genus-level.csv", row.names = TRUE )


# confirm matching row names between Taxonomy and OTUs
identical(row.names(tax),row.names(otu.16s)) # TRUE


## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax )
TAX.16s <- tax_table( tax.m )





## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( otu.16s )
dim(otu.m)
# 50845    33

OTU.16s <- otu_table(otu.16s, taxa_are_rows = TRUE)





## Create a phyloseq object, merging OTU & TAX tables
phy.16s = phyloseq(OTU.16s, TAX.16s)
phy.16s
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 50845 taxa and 33 samples ]
# tax_table()   Taxonomy Table:    [ 50845 taxa by 6 taxonomic ranks ]

sample_names(phy.16s)
# [1] "X138408" "X138410" "X138412" "X138414" "X138416" "X138418" "X138420" "X138422" "X138424" "X138426" "X138428" "X138430"
# [13] "X138432" "X138434" "X138436" "X138438" "X138440" "X138444" "X138446" "X138448" "X138450" "X138452" "X138454" "X138456"
# [25] "X138458" "X138460" "X138462" "X138464" "X138466" "X138468" "X138470" "X138472" "X138474"


### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(samp.16s) #  33 139
identical( row.names(samp.16s), sample_names(phy.16s) ) # TRUE

# # re-order names/rows to match phyloseq object
# sel <- which( row.names(samp.16s) %in% sample_names(phy.16s)) # qty 50
# 
# samp2.16s <- samp.16s[ sample_names(phy.16s), ]
# identical( row.names(samp2.16s), sample_names(phy.16s) ) # TRUE

SAMP.16s <- sample_data(samp.16s)
#SAMP.16s <- sample_data(samp2.16s)


### Combine SAMPDATA into phyloseq object
phy.16s <- merge_phyloseq(phy.16s, SAMP.16s)
phy.16s
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 50845 taxa and 33 samples ]
# sample_data() Sample Data:       [ 33 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 50845 taxa by 6 taxonomic ranks ]




### data cleaning

## remove taxa not assigned as Bacteria

phy_in <- phy.16s

sum(sample_sums(phy_in)) # 2271627 = 2,271,627

rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

levels(factor(tax_table(phy_in)[, "Kingdom"])) # "k__Bacteria"
# rem_taxa <- which(tax_table(phy_in)[, "Kingdom"] %in%  c("Archaea", "Eukaryota") ) #
# phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-rem_taxa, ]) )
# phy_in


## remove taxa not assigned at the phylum level
sort( unique( as.character(tax_table(phy_in)[, "Phylum"]) ) )
# [1] "p__Abditibacteriota"             "p__Acidobacteriota"              "p__Actinobacteriota"            
# [4] "p__Armatimonadota"               "p__Bacteroidota"                 "p__Bdellovibrionota"            
# [7] "p__Chloroflexi"                  "p__Cyanobacteria"                "p__Dadabacteria"                
# [10] "p__Deinococcota"                 "p__Dependentiae"                 "p__Desulfobacterota"            
# [13] "p__Elusimicrobiota"              "p__Entotheonellaeota"            "p__FCPU426"                     
# [16] "p__Fibrobacterota"               "p__Firmicutes"                   "p__GAL15"                       
# [19] "p__Gemmatimonadota"              "p__Hydrogenedentes"              "p__Latescibacterota"            
# [22] "p__MBNT15"                       "p__Methylomirabilota"            "p__Myxococcota"                 
# [25] "p__NB1-j"                        "p__Nitrospirota"                 "p__Patescibacteria"             
# [28] "p__Planctomycetota"              "p__Proteobacteria"               "p__RCP2-54"                     
# [31] "p__SAR324 clade(Marine group B)" "p__Spirochaetota"                "p__Sumerlaeota"                 
# [34] "p__Unclassified"                 "p__Verrucomicrobiota"            "p__WPS-2"                       
# [37] "p__WS2"                          "p__WS4" 


rem_taxa <- which(tax_table(phy_in)[, "Phylum"] == "p__Unclassified") # qty 9
phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-rem_taxa, ]) )
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 50836 taxa and 33 samples ]
# sample_data() Sample Data:       [ 33 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 50836 taxa by 6 taxonomic ranks ]

sum(sample_sums(phy_in)) # 2271552

rank_names(phy_in) #  "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"
sort( as.character( unique( tax_table(phy_in)[, "Phylum"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Class"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Order"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Family"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Genus"] ) ))


## remove taxa associated with chloroplast and mitochondria

rem_taxa1 <- which(tax_table(phy_in)[, "Order"] == "o__Chloroplast") # qty 4... ASV
rem_taxa2 <- which(tax_table(phy_in)[, "Family"] == "f__Mitochondria") # qty 65... ASV

length( c(rem_taxa1,rem_taxa2) ) # 69
length( unique(c(rem_taxa1,rem_taxa2)) ) # 69

# remove chloroplast and mitochondria
phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-unique(c(rem_taxa1,rem_taxa2)), ]) )
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 50767 taxa and 33 samples ]
# sample_data() Sample Data:       [ 33 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 50767 taxa by 6 taxonomic ranks ]

sum(sample_sums(phy_in)) # 2270185


## remove taxa that do not occur in at least two samples (e.g avoiding unrepresentative taxa only present in a single sample)
temp.phy_in <- phy_in
phy_in <- prune_taxa( taxa = apply( otu_table(phy_in), 1, function(x) {sum(x > 0) >= 2 }), x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 33641 taxa and 33 samples ]
# sample_data() Sample Data:       [ 33 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 33641 taxa by 6 taxonomic ranks ]


sum(sample_sums(phy_in)) # 2173342


## use only the surface samples

table(phy_in@sam_data$Depth)
# 0.01 
# 33 

# phy.surf <- prune_samples( samples = phy_in@sam_data$depth == "0-10 cm" , x = phy_in )
# phy.surf
# # phyloseq-class experiment-level object
# # otu_table()   OTU Table:         [ 14132 taxa and 25 samples ]
# # sample_data() Sample Data:       [ 25 samples by 54 sample variables ]
# # tax_table()   Taxonomy Table:    [ 14132 taxa by 7 taxonomic ranks ]

min(taxa_sums(phy_in)) # 2

# # prune taxa that have zero sequence reads
# phy.surf <- prune_taxa(taxa = taxa_sums(phy.surf) > 0, x = phy.surf)
# phy.surf
# # phyloseq-class experiment-level object
# # otu_table()   OTU Table:         [ 12688 taxa and 25 samples ]
# # sample_data() Sample Data:       [ 25 samples by 54 sample variables ]
# # tax_table()   Taxonomy Table:    [ 12688 taxa by 7 taxonomic ranks ]

min(taxa_sums(phy_in)) # 2
min(sample_sums(phy_in)) # 87
sum(sample_sums(phy_in)) # 2173342

sort(sample_sums(phy_in))
# X138414 X138422 X138474 X138446 X138420 X138416 X138440 X138444 X138418 X138450 X138428 X138424 X138426 X138436 X138452 
#      87     626    1299    2775    3545    3902    5897   10142   10338   26393   32464   43173   46582   49293   53219 
# X138448 X138410 X138408 X138432 X138438 X138430 X138412 X138434 X138456 X138460 X138454 X138470 X138472 X138466 X138468 
#   54430   54772   55139   59488   63018   63169   66568   75842   79511  117629  119783  121369  140083  140472  157241 
# X138464 X138458 X138462 
#  163470  169761  181862 


## prune samples that have low read counts
rem_samps <- c("X138414", "X138422", "X138474", "X138446", "X138420", "X138416", "X138440") # removed 7 samples
keep_samps <- which(!sample_names(phy_in) %in% rem_samps)
keep_samps <- sample_names(phy_in)[keep_samps]
keep_samps
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

phy_in <- prune_samples(samples = keep_samps, x = phy_in)
min(taxa_sums(phy_in)) # 0

# prune taxa that have zero sequence reads
phy_in <- prune_taxa(taxa = taxa_sums(phy_in) > 0, x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 33636 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 33636 taxa by 6 taxonomic ranks ]

min(taxa_sums(phy_in)) # 1
min(sample_sums(phy_in)) #  10142
sum(sample_sums(phy_in)) # 2155211

phy.iluka_eneabba <- phy_in




### build and add phylogenetic tree to the phyloseq object

ps <- phy.iluka_eneabba

head( taxa_names(ps) )
# "OTU_ 13" "OTU_ 14" "OTU_ 15" "OTU_ 16" "OTU_ 30" "OTU_ 32"

head( gsub(pattern = " ", replacement = "", x = taxa_names(ps) ) )

taxa_names(ps) <- gsub(pattern = " ", replacement = "", x = taxa_names(ps) )

taxa_names_ps <- taxa_names(ps)
length(taxa_names_ps) # 33636


## read rep seq fasta file
#rep_seq <- read.fasta(file = paste0(datadir,"/mtbold/","Mt-Bold-AMD-16S-table-otus.fasta"), seqtype = "DNA", as.string = TRUE)
length( rep_seq.iluka_eneabba ) # 50845
rep_seq <- rep_seq.iluka_eneabba

head(rep_seq)
names(rep_seq) <- gsub(pattern = " ", replacement = "", x = names(rep_seq) )
head(rep_seq)

## subselect rep seqs
rep_seq <- rep_seq[taxa_names_ps]
length(rep_seq) # 33636
head(names(rep_seq))
# "OTU_13" "OTU_14" "OTU_15" "OTU_16" "OTU_30" "OTU_32"

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"
#write.fasta(sequences = rep_seq, names = names(rep_seq), file.out = "iluka_eneabba_clean_surf_16s_rep_seq.fasta", open = "w",nbchar = 80 )
write.fasta(sequences = rep_seq[1], names = names(rep_seq)[1], file.out = "iluka_eneabba_clean_surf_16s_rep_seq.fasta", open = "w",nbchar = 80 )
for (i in 2:length(rep_seq)) {
  write.fasta(sequences = rep_seq[i], names = names(rep_seq)[i], file.out = "iluka_eneabba_clean_surf_16s_rep_seq.fasta", open = "a",nbchar = 80 )
}


## run MAFFT multiple sequence alignment on DeepThought HPC ...

# # # # # # # # # # #
# # # # MAFFT # # # #
# # # # # # # # # # # 
# # https://mafft.cbrc.jp/alignment/software/
# # A fast option (FFT-NS-2) for a larger sequence alignment:
# # mafft input > output
# 
# $ whereis mafft   #:mafft: /home/lidd0026/miniconda3/bin/mafft
# 
# # run mafft using slurm script called 'run_mafft_iluka_eneabba_slurm.sh'
# - - - - - - - - - - - - - - - 
# #!/bin/bash
#   
# # How many tasks and processes
# #SBATCH --ntasks=1
#   
# #SBATCH --cpus-per-task=32
#   
# # How much memory
# #SBATCH --mem=64G
#   
# # Put your code to run here:
# mafft --thread -1 iluka_eneabba_clean_surf_16s_rep_seq.fasta > iluka_eneabba_clean_surf_16s_mafft_aln.fasta
# 
# - - - - - - - - - - - - - - - 
#   
# # RUN USING THIS COMMAND:
# $ cd /scratch/user/lidd0026/resto_traj
# $ ls
# 
# # more detailed view of the cluster with sinfo:
# $ sinfo -N -o "%N %C %e %m"
# 
# $ sbatch run_mafft_iluka_eneabba_slurm.sh

# Strategy:
#   FFT-NS-2 (Fast but rough)
# Progressive method (guide trees were built 2 times.)

## output of MAFFT is: 'iluka_eneabba_clean_surf_16s_mafft_aln.fasta'




## run Gblocks on Mac to clean noisy alignments
# set -b2 to 60% of sequences = 0.6*33636 = 20181.6 = 20182
## run 'relaxed' cleaning with Gblocks
# cd /Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling/deepthoughtHPC/
# /Users/lidd0026/opt/Gblocks_0.91b/Gblocks iluka_eneabba_clean_surf_16s_mafft_aln.fasta -t=d -b2=20182 -b3=10 -b4=5 -b5=h -e=gbrel 
# 33636 sequences and 1749 positions in the first alignment file:
# iluka_eneabba_clean_surf_16s_mafft_aln.fasta
# 
# iluka_eneabba_clean_surf_16s_mafft_aln.fasta
# Original alignment: 1749 positions
# Gblocks alignment:  372 positions (21 %) in 20 selected block(s)

## output of Gblocks is: 'iluka_eneabba_clean_surf_16s_mafft_aln_Gblocks_relaxed.fasta'



gblocks_aln <- read.fasta( paste0(getwd(),"/deepthoughtHPC/","iluka_eneabba_clean_surf_16s_mafft_aln_Gblocks_relaxed.fasta"),
                           as.string = FALSE)
head(gblocks_aln)
unique(unlist(gblocks_aln)) # "a" "g" "c" "t" " " "-"
gblocks_aln <- read.fasta( paste0(getwd(),"/deepthoughtHPC/","iluka_eneabba_clean_surf_16s_mafft_aln_Gblocks_relaxed.fasta"),
                           as.string = TRUE)
names_gblocks_aln <- names(gblocks_aln)
head(names_gblocks_aln) # "OTU_13" "OTU_14" "OTU_15" "OTU_16" "OTU_30" "OTU_32"
gblocks_aln <- toupper(gblocks_aln)
gblocks_aln <- gsub(pattern = " ",replacement = "-", x = gblocks_aln)
names(gblocks_aln) <- names_gblocks_aln
head(gblocks_aln)


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"
write.fasta(sequences = gblocks_aln[1], names = names(gblocks_aln)[1], file.out = "iluka_eneabba_clean_16s_mafft_Gblocks_rel_tidy.fasta", open = "w",nbchar = 80 )
for (i in 2:length(gblocks_aln)) {
  write.fasta(sequences = gblocks_aln[i], names = names(gblocks_aln)[i], file.out = "iluka_eneabba_clean_16s_mafft_Gblocks_rel_tidy.fasta", open = "a",nbchar = 80 )
  print(paste0("completed ",i))
}



## run IQTREE2 to build phylogenetic tree on DeepThought HPC

# # # # # # # # # # #
# # # # IQTREE2 # # #
# # # # # # # # # # # 
# 
# # run iqtree2 using slurm script called 'run_iqtree2_iluka_eneabba_slurm.sh'
# 
# - - - - - - - - - - - - - - - 
# #!/bin/bash
# # How many tasks and processes
# #SBATCH --ntasks=1
# #SBATCH --cpus-per-task=32
# # How much memory
# #SBATCH --mem=64G
# # Put your code to run here:
# /home/lidd0026/opt/iqtree-2.1.2-Linux/bin/iqtree2 -s /scratch/user/lidd0026/resto_traj/iluka_eneabba_clean_16s_mafft_Gblocks_rel_tidy.fasta --prefix iqtree2_iluka_gblocks -m GTR+I+G -T AUTO -n 200
# - - - - - - - - - - - - - - - - - - - - 
#   
# # RUN ON HPC USING THIS COMMAND:
# $ cd /scratch/user/lidd0026/resto_traj
# $ ls
# 
# $ sinfo -N -o "%N %C %e %m"
# 
# $ sbatch run_iqtree2_iluka_eneabba_slurm.sh




### add phylo tree

phy_in <- phy.iluka_eneabba

phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 33636 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 33636 taxa by 6 taxonomic ranks ]

tree <- read_tree("iqtree2_iluka_gblocks.treefile")

# need to match taxa names ...
head( taxa_names(phy_in) ) # "OTU_ 13" "OTU_ 14" "OTU_ 15" "OTU_ 16" "OTU_ 30" "OTU_ 32"
head( taxa_names(tree)) # "OTU_13"    "OTU_14"    "OTU_48062" "OTU_16"    "OTU_18566" "OTU_18570"

# fix issue with phy taxa names
taxa_names(phy_in) <- gsub(pattern = " ", replacement = "", x = taxa_names(phy_in))

length(taxa_names(phy_in)) # 33636
length(taxa_names(tree)) # 33636

sel <- which(taxa_names(tree) %in% taxa_names(phy_in))
length(sel) # 33636
identical(sort(taxa_names(tree)),sort(taxa_names(phy_in))) # TRUE

phy_in <- merge_phyloseq(phy_in, tree)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 33636 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 33636 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 33636 tips and 33634 internal nodes ]

phy.iluka_eneabba.withTree <- phy_in

#-------------------------

#### 2b. Iluka Eneabba distance plots

## Iluka-Eneabba - with Tree - i. Bray-Curtis - ASV-level
#-------------------------

phy_in <- phy.iluka_eneabba.withTree
min(taxa_sums(phy_in)) # 1
sum(sample_sums(phy_in)) # 2155211

sample_sums(phy_in)
min(sample_sums(phy_in)) # 10142

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142

ntaxa(r1.ps) #  27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 27115 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 27115 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 27115 tips and 27114 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$year.of.rehab.or.ref
# [1] "REF"  "REF"  "REF"  "1989" "1995" "1995" "1981" "REF"  "REF"  "1981" "1981" "REF"  "REF"  "2012" "2012" "REF"  "REF" 
# [18] "2009" "2004" "2009" "2004" "2000" "2000" "1995" "2004" "1989"

table(r1.ps@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 

row.names(r1.ps@sam_data)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.bray), row.names(r1.ps@sam_data)) # TRUE

sel <- which(r1.ps@sam_data$year.of.rehab.or.ref == "REF") # qty 9

( ref_sites <- row.names(r1.ps@sam_data)[sel] )

# "X138408" "X138410" "X138412" "X138430" "X138432" "X138438" "X138444" "X138452" "X138454"

df_in <- sim.bray

#df_out <- data.frame(samp=colnames(ps.sim.bray))
df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  #df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  #df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
  #                                   " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  #df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$year.of.rehab.or.ref[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 234   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 9
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 225   4


unique(df_out$year_rehab)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

2019 - c(2012,2009,2004,2000,1995,1989,1981)
# 7 10 15 19 24 30 38

df_out$group <- factor(df_out$group, 
                       levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                       labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)

table(df_out$group)
#  7  10  15  19  24  30  38 Ref 
# 18  18  27  18  27  18  27  72 

df_out$group <- factor(df_out$group, levels=as.character(c(0:40,"Ref")),ordered=TRUE)
df_out$group

dim(df_out) # 225 4


df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "Iluka-Eneabba"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["Iluka-Eneabba-Bray-ASV-withTree"]] <- df_out
#df_out <- df_results[[1]]



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	225 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 141.02, df = 7, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# Ref 30  24  38  7   10  15  19

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("7","10","15","19","24","30","38","Ref"),
                     labels = c("7","ten","15","19","24","thirty","38","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	8 obs. of  3 variables:
#   $ Group     : chr  "15" "19" "24" "38" ...
# $ Letter    : chr  "ab" "ac" "ab" "cd" ...
# $ MonoLetter: chr  "ab  " "a c " "ab  " "  cd" ...

unique(sig$Group) # "15"     "19"     "24"     "38"     "7"      "Ref"    "ten"    "thirty"

sig$Group <- factor( sig$Group, levels=c("7","ten","15","19","24","thirty","38","Ref"),
                     labels = c("7","10","15","19","24","30","38","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 5     7     ab       ab  
# 7    10      b        b  
# 1    15     ab       ab  
# 2    19     ac       a c 
# 3    24     ab       ab  
# 8    30    acd       a cd
# 4    38     cd         cd
# 6   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "7"   "10"  "15"  "19"  "24"  "30"  "38"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(28,25,28,30,32,34,35,41.5),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with tree)", times = dim(sig)[1]),
                   study  = rep( "Iluka-Eneabba", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) 

df_anno[["Iluka-Eneabba-Bray-ASV-withTree"]] <- anno




## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1468231 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$year.of.rehab.or.ref
unique(r1.ps@sam_data$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                               labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) #
names(ord_plots) #
ord_plots[["Iluka-Eneabba-Bray-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) #
ord_obj[["Iluka-Eneabba-Bray-ASV-withTree"]] <- ord

#-------------------------

## Iluka-Eneabba - with Tree - ii. Jaccard - ASV-level
#-------------------------
phy_in <- phy.iluka_eneabba.withTree
min(taxa_sums(phy_in)) # 1
sum(sample_sums(phy_in)) # 2155211

sample_sums(phy_in)
min(sample_sums(phy_in)) # 10142

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142

ntaxa(r1.ps) #  27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 27115 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 27115 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 27115 tips and 27114 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE) 
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$year.of.rehab.or.ref
# [1] "REF"  "REF"  "REF"  "1989" "1995" "1995" "1981" "REF"  "REF"  "1981" "1981" "REF"  "REF"  "2012" "2012" "REF"  "REF" 
# [18] "2009" "2004" "2009" "2004" "2000" "2000" "1995" "2004" "1989"

table(r1.ps@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 

row.names(r1.ps@sam_data)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.jacc), row.names(r1.ps@sam_data)) # TRUE

sel <- which(r1.ps@sam_data$year.of.rehab.or.ref == "REF") # qty 9

( ref_sites <- row.names(r1.ps@sam_data)[sel] )

# "X138408" "X138410" "X138412" "X138430" "X138432" "X138438" "X138444" "X138452" "X138454"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  #df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  #df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
  #                                   " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  #df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$year.of.rehab.or.ref[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 234   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 9
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 225   4


unique(df_out$year_rehab)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

2019 - c(2012,2009,2004,2000,1995,1989,1981)
# 7 10 15 19 24 30 38

df_out$group <- factor(df_out$group, 
                       levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                       labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)

table(df_out$group)
#  7  10  15  19  24  30  38 Ref 
# 18  18  27  18  27  18  27  72 


# df_out$group <- factor(df_out$group, levels=as.character(c(0:40,"Ref")),ordered=TRUE)
# df_out$group

dim(df_out) # 225 4


df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "ASV (with tree)"
df_out$study <- "Iluka-Eneabba"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["Iluka-Eneabba-Jaccard-ASV-withTree"]] <- df_out




p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	225 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 153.81, df = 7, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# Ref 30  24  38  7   10  15  19 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("7","10","15","19","24","30","38","Ref"),
                     labels = c("7","ten","15","19","24","thirty","38","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	8 obs. of  3 variables:
#   $ Group     : chr  "15" "19" "24" "38" ...
# $ Letter    : chr  "ab" "a" "a" "cd" ...
# $ MonoLetter: chr  "ab  " "a   " "a   " "  cd" ...

unique(sig$Group) # "15"     "19"     "24"     "38"     "7"      "Ref"    "ten"    "thirty"

sig$Group <- factor( sig$Group, levels=c("7","ten","15","19","24","thirty","38","Ref"),
                     labels = c("7","10","15","19","24","30","38","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 5     7     ab       ab  
# 7    10      b        b  
# 1    15     ab       ab  
# 2    19      a       a   
# 3    24      a       a   
# 8    30     ad       a  d
# 4    38     cd         cd
# 6   Ref      c         c 

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "7"   "10"  "15"  "19"  "24"  "30"  "38"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(17.5,14,17.5,18,18,20,24,25),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with tree)", times = dim(sig)[1]),
                   study  = rep( "Iluka-Eneabba", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 
names(df_anno) 

df_anno[["Iluka-Eneabba-Jaccard-ASV-withTree"]] <- anno




## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.152997 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’ 

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$year.of.rehab.or.ref
unique(r1.ps@sam_data$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                               labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 

ord_plots[["Iluka-Eneabba-Jaccard-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) #

ord_obj[["Iluka-Eneabba-Jaccard-ASV-withTree"]] <- ord


#-------------------------

## Iluka-Eneabba - with Tree - iii. Weighted UniFrac - ASV-level
#-------------------------

phy_in <- phy.iluka_eneabba.withTree
min(taxa_sums(phy_in)) # 1
sum(sample_sums(phy_in)) # 2155211

sample_sums(phy_in)
min(sample_sums(phy_in)) # 10142

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142

ntaxa(r1.ps) #  27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 27115 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 27115 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 27115 tips and 27114 internal nodes ]

dist.wuni <- phyloseq::distance(r1.ps, "wunifrac") # weighted UniFrac
class(dist.wuni) # "dist"
dist.wuni <- as(dist.wuni, "matrix")
# express as similarity
sim.wuni <- 100*(1 - dist.wuni)
# list all sample names
colnames(sim.wuni)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"
identical(colnames(sim.wuni),rownames(sim.wuni)) # TRUE

## which are the reference sites?
r1.ps@sam_data$year.of.rehab.or.ref
# [1] "REF"  "REF"  "REF"  "1989" "1995" "1995" "1981" "REF"  "REF"  "1981" "1981" "REF"  "REF"  "2012" "2012" "REF"  "REF" 
# [18] "2009" "2004" "2009" "2004" "2000" "2000" "1995" "2004" "1989"

table(r1.ps@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 

row.names(r1.ps@sam_data)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.wuni), row.names(r1.ps@sam_data)) # TRUE

sel <- which(r1.ps@sam_data$year.of.rehab.or.ref == "REF") # qty 9

( ref_sites <- row.names(r1.ps@sam_data)[sel] )

# "X138408" "X138410" "X138412" "X138430" "X138432" "X138438" "X138444" "X138452" "X138454"

df_in <- sim.wuni

df_out <- data.frame(samp= rep( x = colnames(sim.wuni), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.wuni))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)


for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  #df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  #df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
  #                                   " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  #df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$year.of.rehab.or.ref[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 234   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 9
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 225   4


unique(df_out$year_rehab)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

2019 - c(2012,2009,2004,2000,1995,1989,1981)
# 7 10 15 19 24 30 38

df_out$group <- factor(df_out$group, 
                       levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                       labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)

table(df_out$group)
#  7  10  15  19  24  30  38 Ref 
# 18  18  27  18  27  18  27  72 


dim(df_out) # 225 4


df_out$dist_measure <- "Weighted UniFrac"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "Iluka-Eneabba"
df_out$corrected <- "no"  
df_out$facet_x <- "Weighted UniFrac\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["Iluka-Eneabba-WtUniFrac-ASV-withTree"]] <- df_out
#df_out <- df_results[[1]]



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Weighted Unifrac Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	225 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 54.608, df = 7, p-value = 1.783e-09


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# Ref 30  24  38  7   10  15  19

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("7","10","15","19","24","30","38","Ref"),
                     labels = c("7","ten","15","19","24","thirty","38","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	8 obs. of  3 variables:
#   $ Group     : chr  "15" "19" "24" "38" ...
# $ Letter    : chr  "a" "ab" "ab" "bc" ...
# $ MonoLetter: chr  "a  " "ab " "ab " " bc" ...

unique(sig$Group) # "15"     "19"     "24"     "38"     "7"      "Ref"    "ten"    "thirty"

sig$Group <- factor( sig$Group, levels=c("7","ten","15","19","24","thirty","38","Ref"),
                     labels = c("7","10","15","19","24","30","38","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
#   Group Letter MonoLetter
# 5     7    abc        abc
# 7    10      a        a  
# 1    15      a        a  
# 2    19     ab        ab 
# 3    24     ab        ab 
# 8    30     bc         bc
# 4    38     bc         bc
# 6   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "7"   "10"  "15"  "19"  "24"  "30"  "38"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(93,92,93,93,93,94,94,95),
                   dist_measure = rep( "Weighted UniFrac", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with tree)", times = dim(sig)[1]),
                   study  = rep( "Iluka-Eneabba", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Weighted UniFrac\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) 

df_anno[["Iluka-Eneabba-WtUniFrac-ASV-withTree"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "unifrac", weighted = TRUE)

ord
# Call:
#   metaMDS(comm = ps.dist) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     ps.dist 
# Distance: user supplied 
# 
# Dimensions: 2 
# Stress:     0.1067698 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation 
# Species: scores missing

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$year.of.rehab.or.ref
unique(r1.ps@sam_data$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                               labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) #
names(ord_plots) #
ord_plots[["Iluka-Eneabba-WtUniFrac-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) #
ord_obj[["Iluka-Eneabba-WtUniFrac-ASV-withTree"]] <- ord

#-------------------------

## Iluka-Eneabba - with Tree - iv. Unweighted UniFrac - ASV-level
#-------------------------
phy_in <- phy.iluka_eneabba.withTree
min(taxa_sums(phy_in)) # 1
sum(sample_sums(phy_in)) # 2155211

sample_sums(phy_in)
min(sample_sums(phy_in)) # 10142
ntaxa(phy_in) # 33636

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142

ntaxa(r1.ps) #  27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 27115 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 27115 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 27115 tips and 27114 internal nodes ]


dist.uuni <- phyloseq::distance(r1.ps, "uunifrac") # Unweighted UniFrac
class(dist.uuni) # "dist"
dist.uuni <- as(dist.uuni, "matrix")
# express as similarity
sim.uuni <- 100*(1 - dist.uuni)
# list all sample names
colnames(sim.uuni)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"
identical(colnames(sim.uuni),rownames(sim.uuni)) # TRUE



## which are the reference sites?
r1.ps@sam_data$year.of.rehab.or.ref
# [1] "REF"  "REF"  "REF"  "1989" "1995" "1995" "1981" "REF"  "REF"  "1981" "1981" "REF"  "REF"  "2012" "2012" "REF"  "REF" 
# [18] "2009" "2004" "2009" "2004" "2000" "2000" "1995" "2004" "1989"

table(r1.ps@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 

row.names(r1.ps@sam_data)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.uuni), row.names(r1.ps@sam_data)) # TRUE

sel <- which(r1.ps@sam_data$year.of.rehab.or.ref == "REF") # qty 9

( ref_sites <- row.names(r1.ps@sam_data)[sel] )

# "X138408" "X138410" "X138412" "X138430" "X138432" "X138438" "X138444" "X138452" "X138454"

df_in <- sim.uuni

df_out <- data.frame(samp= rep( x = colnames(sim.uuni), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.uuni))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  #df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  #df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
  #                                   " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  #df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$year.of.rehab.or.ref[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 234   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 9
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 225   4


unique(df_out$year_rehab)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

2019 - c(2012,2009,2004,2000,1995,1989,1981)
# 7 10 15 19 24 30 38

df_out$group <- factor(df_out$group, 
                       levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                       labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)

table(df_out$group)
#  7  10  15  19  24  30  38 Ref 
# 18  18  27  18  27  18  27  72 


# df_out$group <- factor(df_out$group, levels=as.character(c(0:40,"Ref")),ordered=TRUE)
# df_out$group

dim(df_out) # 225 4


df_out$dist_measure <- "Unweighted UniFrac"
df_out$tax_level <- "ASV (with tree)"
df_out$study <- "Iluka-Eneabba"
df_out$corrected <- "no"  
df_out$facet_x <- "Unweighted UniFrac\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["Iluka-Eneabba-UnWtUniFrac-ASV-withTree"]] <- df_out




p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Unweighted UniFrac Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	225 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 83.072, df = 7, p-value = 3.25e-15


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# Ref 30  24  38  7   10  15  19 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("7","10","15","19","24","30","38","Ref"),
                     labels = c("7","ten","15","19","24","thirty","38","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	8 obs. of  3 variables:
#   $ Group     : chr  "15" "19" "24" "38" ...
# $ Letter    : chr  "ab" "acd" "abc" "cd" ...
# $ MonoLetter: chr  "ab  " "a cd" "abc " "  cd" ...

unique(sig$Group) # "15"     "19"     "24"     "38"     "7"      "Ref"    "ten"    "thirty"

sig$Group <- factor( sig$Group, levels=c("7","ten","15","19","24","thirty","38","Ref"),
                     labels = c("7","10","15","19","24","30","38","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
#   Group Letter MonoLetter
# 5     7     ab       ab  
# 7    10      b        b  
# 1    15     ab       ab  
# 2    19    acd       a cd
# 3    24    abc       abc 
# 8    30    acd       a cd
# 4    38     cd         cd
# 6   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "7"   "10"  "15"  "19"  "24"  "30"  "38"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(54,50,54,55,55,55,57,58),
                   dist_measure = rep( "Unweighted UniFrac", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with tree)", times = dim(sig)[1]),
                   study  = rep( "Iluka-Eneabba", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Unweighted UniFrac\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 
names(df_anno) 

df_anno[["Iluka-Eneabba-UnWtUniFrac-ASV-withTree"]] <- anno





## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "unifrac", weighted = FALSE)

ord
# Call:
#   metaMDS(comm = ps.dist) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     ps.dist 
# Distance: user supplied 
# 
# Dimensions: 2 
# Stress:     0.1270845 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation 
# Species: scores missing

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$year.of.rehab.or.ref
unique(r1.ps@sam_data$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                               labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) # 

ord_plots[["Iluka-Eneabba-UnWtUniFrac-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) #

ord_obj[["Iluka-Eneabba-UnWtUniFrac-ASV-withTree"]] <- ord


#-------------------------


## Iluka-Eneabba - with Tree - Bray-Curtis - ASV-level - EXCLUDE youngest rehab_age
#-------------------------

phy_in <- phy.iluka_eneabba.withTree
min(taxa_sums(phy_in)) # 1
sum(sample_sums(phy_in)) # 2155211

sample_sums(phy_in)
min(sample_sums(phy_in)) # 10142

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142

ntaxa(r1.ps) #  27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 27115 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 27115 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 27115 tips and 27114 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?
r1.ps@sam_data$year.of.rehab.or.ref
# [1] "REF"  "REF"  "REF"  "1989" "1995" "1995" "1981" "REF"  "REF"  "1981" "1981" "REF"  "REF"  "2012" "2012" "REF"  "REF" 
# [18] "2009" "2004" "2009" "2004" "2000" "2000" "1995" "2004" "1989"

table(r1.ps@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 

row.names(r1.ps@sam_data)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.bray), row.names(r1.ps@sam_data)) # TRUE

sel <- which(r1.ps@sam_data$year.of.rehab.or.ref == "REF") # qty 9

( ref_sites <- row.names(r1.ps@sam_data)[sel] )

# "X138408" "X138410" "X138412" "X138430" "X138432" "X138438" "X138444" "X138452" "X138454"

df_in <- sim.bray

#df_out <- data.frame(samp=colnames(ps.sim.bray))
df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  #df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  #df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
  #                                   " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  #df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$year.of.rehab.or.ref[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 234   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 9
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 225   4


unique(df_out$year_rehab)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

2019 - c(2012,2009,2004,2000,1995,1989,1981)
# 7 10 15 19 24 30 38

class(df_out$group) # "character"

dim(df_out) # 225   5

## exclude youngest rehab_age group
sel <- which(df_out$group == "2012") # qty 18
df_out <- df_out[-sel, ]
dim(df_out) # 207   5

unique(df_out$group) # "REF"  "1989" "1995" "1981" "2009" "2004" "2000"

df_out$group <- factor(df_out$group, 
                       levels = c("2009","2004","2000","1995","1989","1981","REF"),
                       labels = c("10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 10-38 yr old) - if exclude 7-year rehab

table(df_out$group)
# 10  15  19  24  30  38 Ref 
# 18  27  18  27  18  27  72 


df_out$group <- factor(df_out$group, levels=as.character(c(0:40,"Ref")),ordered=TRUE)
df_out$group



df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "Iluka-Eneabba\n(exclude 7-year rehab)"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["Iluka-Eneabba-Bray-ASV-withTree-exclude7yr"]] <- df_out
#df_out <- df_results[[1]]



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	207 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 125.71, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# Ref 30  24  38  10  15  19 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("10","15","19","24","30","38","Ref"),
                     labels = c("ten","15","19","24","thirty","38","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "15"     "19"     "24"     "38"       "Ref"    "ten"    "thirty"

sig$Group <- factor( sig$Group, levels=c("ten","15","19","24","thirty","38","Ref"),
                     labels = c("10","15","19","24","30","38","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 6    10      a       a   
# 1    15     ab       ab  
# 2    19    abc       abc 
# 3    24     ab       ab  
# 7    30    bcd        bcd
# 4    38     cd         cd
# 5   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) #  "10"  "15"  "19"  "24"  "30"  "38"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(25,28,30,32,34,35,41.5),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with tree)", times = dim(sig)[1]),
                   study  = rep( "Iluka-Eneabba\n(exclude 7-year rehab)", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) 

df_anno[["Iluka-Eneabba-Bray-ASV-withTree-exclude7yr"]] <- anno


#-------------------------

## Iluka-Eneabba - with Tree - Jaccard - ASV-level - EXCLUDE youngest rehab_age
#-------------------------
phy_in <- phy.iluka_eneabba.withTree
min(taxa_sums(phy_in)) # 1
sum(sample_sums(phy_in)) # 2155211

sample_sums(phy_in)
min(sample_sums(phy_in)) # 10142

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142

ntaxa(r1.ps) #  27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 27115 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 27115 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 27115 tips and 27114 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$year.of.rehab.or.ref
# [1] "REF"  "REF"  "REF"  "1989" "1995" "1995" "1981" "REF"  "REF"  "1981" "1981" "REF"  "REF"  "2012" "2012" "REF"  "REF" 
# [18] "2009" "2004" "2009" "2004" "2000" "2000" "1995" "2004" "1989"

table(r1.ps@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 

row.names(r1.ps@sam_data)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

identical(colnames(sim.jacc), row.names(r1.ps@sam_data)) # TRUE

sel <- which(r1.ps@sam_data$year.of.rehab.or.ref == "REF") # qty 9

( ref_sites <- row.names(r1.ps@sam_data)[sel] )

# "X138408" "X138410" "X138412" "X138430" "X138432" "X138438" "X138444" "X138452" "X138454"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  #df_out$site[i] <- as.character(r1.ps@sam_data$`Location description`[sel])
  #df_out$site_full_desc[i] <- paste0(as.character(r1.ps@sam_data$`Location description`[sel]),
  #                                   " (",as.character(r1.ps@sam_data$Notes[sel]),")")
  #df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`Date since change in Land Use`[sel])
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$year.of.rehab.or.ref[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 234   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 9
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 225   4


unique(df_out$year_rehab)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"


df_out$group <- df_out$year_rehab
unique(df_out$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

2019 - c(2012,2009,2004,2000,1995,1989,1981)
# 7 10 15 19 24 30 38

## exclude youngest rehab_age 7 year group
sel <- which(df_out$group == "2012") # qty 18
df_out <- df_out[-sel, ]
dim(df_out) # 207   5

unique(df_out$group) # "REF"  "1989" "1995" "1981" "2009" "2004" "2000"

df_out$group <- factor(df_out$group, 
                       levels = c("2009","2004","2000","1995","1989","1981","REF"),
                       labels = c("10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 10-38 yr old) - with exclusion of 7 yr rehab_age

table(df_out$group)
# 10  15  19  24  30  38 Ref 
# 18  27  18  27  18  27  72 


# df_out$group <- factor(df_out$group, levels=as.character(c(0:40,"Ref")),ordered=TRUE)
# df_out$group

dim(df_out) # 207 4


df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "ASV (with tree)"
df_out$study <- "Iluka-Eneabba\n(exclude 7-year rehab)"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["Iluka-Eneabba-Jaccard-ASV-withTree-exclude7yr"]] <- df_out




p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	207 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 141.17, df = 6, p-value < 2.2e-16


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# Ref 30  24  38  10  15  19 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("10","15","19","24","30","38","Ref"),
                     labels = c("ten","15","19","24","thirty","38","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "15"     "19"     "24"     "38"      "Ref"    "ten"    "thirty"

sig$Group <- factor( sig$Group, levels=c("ten","15","19","24","thirty","38","Ref"),
                     labels = c("10","15","19","24","30","38","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 6    10      b        b  
# 1    15     ab       ab  
# 2    19      a       a   
# 3    24     ab       ab  
# 7    30     ad       a  d
# 4    38     cd         cd
# 5   Ref      c         c 

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "10"  "15"  "19"  "24"  "30"  "38"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(14,17.5,18,18,20,24,25),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with tree)", times = dim(sig)[1]),
                   study  = rep( "Iluka-Eneabba\n(exclude 7-year rehab)", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 
names(df_anno) 

df_anno[["Iluka-Eneabba-Jaccard-ASV-withTree-exclude7yr"]] <- anno


#-------------------------






#### 4a. South32 post-mining restoration chronosequence
#    - create phyloseq object from metadata, OTU abundances, taxonomy
#    - data cleaning
#    - build and add phylogenetic tree
#-------------------------

### metadata

context <- read_excel(path= paste0(datadir,"/south32/","AM_ContextualMetadata_south32.xlsx"),
                      sheet=1, range="A1:EI51", col_names = TRUE)

context <- as.data.frame(context)
str(context)

names(context)
# [1] "Sample_ID"                                             "NCBI_Submission"                                      
# [3] "NCBI_Sample_Accession"                                 "Organism"                                             
# [5] "Tax_ID"                                                "SampleName_depth"                                     
# [7] "NCBI_BioProject"                                       "Date_sampled [YYYY-MM-DD]"                            
# [9] "Time_sampled [hh:mm]"                                  "Latitude [decimal degrees]"                           
# [11] "Longitude [decimal degrees]"                           "Depth [m]"                                            
# [13] "geo_loc [country:subregion]"                           "Sample_Site (location_description)"                   
# [15] "year of rehab or ref"                                  "age of rehab"                                         
# [17] "NOTES"                                                 "Sample_attribution"                                   
# [19] "Sample_descriptor"                                     "Citation"                                             
# [21] "Funding_source"                                        "Sample_collection_device-method"                      
# [23] "Sample_material_processing"                            "Sample_storage_method"                                
# [25] "DNA_extraction_method"                                 "Notes (incl. phys/chem measurement method variation)" 
# [27] "Environment (free living)"                             "Env_material (control vocab 0)"                       
# [29] "Horizon (control vocab 1)"                             "Broad_land_use (control vocab 2)"                     
# [31] "Detailed_land_use (control vocab 2)"                   "General_Env_feature (control vocab 3)"                
# [33] "Env_biome (control vocab 4)"                           "Temperature [deg C]"                                  
# [35] "Vegetation Total cover [%]"                            "Vegetation Dom. Trees [%]"                            
# [37] "Vegetation Dom. Shrubs [%]"                            "Vegetation Dom. Grasses [%]"                          
# [39] "Elevation [m]"                                         "Slope [%]"                                            
# [41] "Slope Aspect [Direction or degrees; e.g., NW or 315?]" "Profile Position (control vocab 5)"                   
# [43] "Australian Soil Classification (control vocab 6)"      "FAO soil classification (control vocab 7)"            
# [45] "Immediate Previous Land Use (control vocab 2)"         "Date since change in Land Use"                        
# [47] "Crop rotation 1yr since present"                       "Crop rotation 2yrs since present"                     
# [49] "Crop rotation 3yrs since present"                      "Crop rotation 4yrs since present"                     
# [51] "Crop rotation 5yrs since present"                      "Agrochemical Additions"                               
# [53] "Tillage (control vocab 9)"                             "Fire"                                                 
# [55] "fire intensity if known"                               "Flooding"                                             
# [57] "Extreme Events"                                        "Soil moisture [%]"                                    
# [59] "Water Holding capacity"                                "bulk density [g/cm3 or kg/m3]"                        
# [61] "Microbial Biomass"                                     "Color (control vocab 10)"                             
# [63] "Gravel [%]- [ >2.0 mm]"                                "Texture []"                                           
# [65] "Course Sand [%] [200-2000 ?m]"                         "Fine Sand [%] [20-200 ?m]"                            
# [67] "Sand [%]"                                              "Silt [%] [2-20 ?m]"                                   
# [69] "Clay [%] [<2 ?m]"                                      "Ammonium Nitrogen [mg/Kg]"                            
# [71] "Nitrate_Nitrogen [mg/Kg]"                              "Phosphorus_Colwell [mg/Kg]"                           
# [73] "Potassium_Colwell [mg/Kg]"                             "Sulphur [mg/Kg]"                                      
# [75] "Organic_Carbon [%]"                                    "Conductivity [dS/m]"                                  
# [77] "Ph_Level_CaCl2 [pH]"                                   "pH_Level_H2O [pH]"                                    
# [79] "DTPA_Copper [mg/Kg]"                                   "DTPA_Iron [mg/Kg]"                                    
# [81] "DTPA_Manganese [mg/Kg]"                                "DTPA_Zinc [mg/Kg]"                                    
# [83] "Exc_Aluminium [meq/100g]"                              "Exc_Calcium [meq/100g]"                               
# [85] "Exc_Magnesium [meq/100g]"                              "Exc_Potassium [meq/100g]"                             
# [87] "Exc_Sodium [meq/100g]"                                 "Boron_Hot_CaCl2 [mg/Kg]"                              
# [89] "Cation Exchange Capacity"                              "Arsenic...90"                                         
# [91] "Scandium"                                              "Vanadium"                                             
# [93] "Chromium"                                              "Cobalt"                                               
# [95] "Nickel"                                                "Gallium"                                              
# [97] "Germanium"                                             "Arsenic...98"                                         
# [99] "Selenium"                                              "Rubidium"                                             
# [101] "Strontium"                                             "Yttrium"                                              
# [103] "Zirconium"                                             "Niobium [Columbium]"                                  
# [105] "Molybdenum"                                            "Ruthenium"                                            
# [107] "Rhodium"                                               "Palladium"                                            
# [109] "Silver"                                                "Cadmium"                                              
# [111] "Tin"                                                   "Antimony"                                             
# [113] "Cesium"                                                "Barium"                                               
# [115] "Lanthanum"                                             "Cerium"                                               
# [117] "Praseodymium"                                          "Neodymium"                                            
# [119] "Samarium"                                              "Europium"                                             
# [121] "Gadolinium"                                            "Terbium"                                              
# [123] "Dysprosium"                                            "Holmium"                                              
# [125] "Erbium"                                                "Thulium"                                              
# [127] "Ytterbium"                                             "Lutetium"                                             
# [129] "Hafnium"                                               "Tantalum"                                             
# [131] "Tungsten or Wolfram"                                   "Osmium"                                               
# [133] "Iridium"                                               "Platinum"                                             
# [135] "Gold"                                                  "Lead"                                                 
# [137] "Thorium"                                               "Uranium"                                              
# [139] "site"  

# Note field: "year of rehab or ref"
row.names(context) <- context$Sample_ID
str(context)

gsub(pattern = "102.100.100/", replacement = "", x = row.names(context))
row.names(context) <- gsub(pattern = "102.100.100/", replacement = "", x = row.names(context))

row.names(context)
# [1] "138358" "138359" "138360" "138361" "138362" "138363" "138364" "138365" "138366" "138367" "138368" "138369" "138370"
# [14] "138371" "138372" "138373" "138374" "138375" "138376" "138377" "138378" "138379" "138380" "138381" "138382" "138383"
# [27] "138384" "138385" "138386" "138387" "138388" "138389" "138390" "138391" "138392" "138393" "138394" "138395" "138396"
# [40] "138397" "138398" "138399" "138400" "138401" "138402" "138403" "138404" "138405" "138406" "138407"

row.names(context) <- paste0("X",row.names(context))

table(context$`Depth [m]`)
# 0.01 0.02 
#   25   25

sel <- which(context$`Depth [m]` == "0.01") # qty 25

context <- context[sel, ]


samp.16s <- context

write.csv(context, file = "south32-metadata.csv")


### OTU raw data matrix 

raw.otu.16s <- read_excel(path= paste0(datadir,"/south32/","south32_otu_cluss_tax.xlsx"),
                          #sheet=1, range="A1:BG133533", col_names = TRUE) # ERROR due to large cell range - "Error: Cell references aren't uniformly A1 or R1C1 format:"
                          # https://community.rstudio.com/t/read-excel-range-limits/6894
                          sheet=1, range = cell_limits(c(1, 1), c(NA, 59)), col_names = TRUE)

raw.otu.16s <- as.data.frame(raw.otu.16s)

#raw.otu.16s <- read.delim(file = paste0(datadir,"/iluka_eneabba/","Iluka_otu_cluss_tax.txt"), header = TRUE, sep = "\t")

names(raw.otu.16s)
names(raw.otu.16s) <- paste0("X",names(raw.otu.16s))

str(raw.otu.16s)
# 'data.frame':	133532 obs. of  59 variables:
#   $ X#OTU ID : chr  "AAAGAACGCTAGCAACGGATATAACACATGCAAGTCAATGTCAAACATGGCGTACGAGTGAGTAATAGTTCGGAATTTGCTTGAAAATTGGGAATAGAAAGTAATACCCAA"| __truncated__ "AAAGAACGCTAGCACGATACATTACACATGCAAGTGGAATGGCTAAAGTAAGAACAGGCCATCACGTACGGGTGAGTAAAATTTCTAAACATACCCTTTAGACCGAAGGAA"| __truncated__ "AAAGAACGCTGGCGGCATGCCTAACACATGCAAGTTGAACGAGTAGTAGCAATACTATTAGTAGCAAACGGGTGAGTAATATGTAGGAATCTACCCTATAGTATGGAACAA"| __truncated__ "AAAGAACGCTGGCGGCATGCTTAACACATGCAAGTCGAACGAGTAGTAGCAATACTATTAGTGGCAGACGGGTGAGTAATACGTAGGAATCTCCCCTATAGTACGGAACAA"| __truncated__ ...
# $ X138358  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138359  : num  0 0 0 1 0 0 0 0 0 0 ...
# $ X138360  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138361  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138362  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138363  : num  0 0 0 0 1 0 0 0 0 0 ...
# $ X138364  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138365  : num  0 0 0 0 3 0 0 0 0 0 ...
# $ X138366  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138367  : num  0 0 0 0 1 0 0 0 0 0 ...
# $ X138368  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138369  : num  0 0 0 0 1 0 0 0 0 0 ...
# $ X138370  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138371  : num  0 0 0 2 0 0 0 0 0 0 ...
# $ X138372  : num  0 0 0 0 1 0 0 0 0 0 ...
# $ X138373  : num  0 0 0 0 1 0 0 0 0 0 ...
# $ X138374  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138375  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138376  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138377  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138378  : num  0 0 0 0 1 0 0 0 0 0 ...
# $ X138379  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138380  : num  0 0 0 0 2 0 0 0 0 0 ...
# $ X138381  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138382  : num  0 0 0 5 0 0 0 0 0 0 ...
# $ X138383  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138384  : num  0 0 0 4 0 0 0 0 0 0 ...
# $ X138385  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138386  : num  0 0 0 1 0 0 0 0 0 0 ...
# $ X138387  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138388  : num  0 0 0 0 3 0 0 0 0 0 ...
# $ X138389  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138390  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138391  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138392  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138393  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138394  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138395  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138396  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138397  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138398  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138399  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138400  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138401  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138402  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138403  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138404  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138405  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138406  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ X138407  : num  0 0 0 0 0 0 0 0 0 0 ...
# $ Xclus97ID: chr  "OTU97_12719" "OTU97_21472" "OTU97_21919" "OTU97_3248" ...
# $ Xclus99ID: chr  "OTU99_23592" "OTU99_45338" "OTU99_46371" "OTU99_5133" ...
# $ Xkingdom : chr  "Bacteria" "Bacteria" "Bacteria" "Bacteria" ...
# $ Xphylum  : chr  "Proteobacteria" "Proteobacteria" "Proteobacteria" "Proteobacteria" ...
# $ Xclass   : chr  "Proteobacteria_unclassified" "Proteobacteria_unclassified" "Alphaproteobacteria" "Alphaproteobacteria" ...
# $ Xorder   : chr  "Proteobacteria_unclassified" "Proteobacteria_unclassified" "Rickettsiales" "Rickettsiales" ...
# $ Xfamily  : chr  "Proteobacteria_unclassified" "Proteobacteria_unclassified" "Rickettsiaceae" "Rickettsiaceae" ...
# $ Xgenus   : chr  "0" "0" "uncultured" "uncultured" ...

# store rep sequences

rep_seq.south32 <- as.character( raw.otu.16s[, 1] )
head(rep_seq.south32)
names(rep_seq.south32) <- paste("OTU_",1:length(rep_seq.south32))
head(rep_seq.south32)


row.names(raw.otu.16s) <- names(rep_seq.south32)
# remove first column, it is saved in rep_seq
raw.otu.16s <- raw.otu.16s[ , -1]

temp.otu.south32 <- raw.otu.16s

length(names(raw.otu.16s)) # 58

sel <- which(names(raw.otu.16s) %in% row.names(samp.16s)) # qty 25
otu.16s <- raw.otu.16s[ ,sel]

## align available OTU abundance data with samples?
length(row.names(samp.16s)) # 25
length(names(otu.16s)) # 25
class(row.names(samp.16s)) # "character"
class(names(otu.16s)) # "character"
identical( row.names(samp.16s), names(otu.16s) ) # TRUE

dim(samp.16s)
#  25  139

dim(otu.16s)
# 133532    25


# trim? Are there any OTUs with zero counts after limiting to surface samples only?

sel <- which( rowSums(otu.16s) < 1 )
otu.16s[sel[1:10], ] # all zeros

otu.16s <- otu.16s[-sel, ]
dim(otu.16s)
# 64026    25



### derive Taxonomy table

## read in representative sequences

length(rep_seq.south32) # 133532

keep_otus <- row.names(otu.16s)
rep_seq.south32 <- rep_seq.south32[keep_otus]
length(rep_seq.south32) # 64026

identical(names(rep_seq.south32),row.names(otu.16s)) # TRUE

### Assign taxonomy using dada2
# https://benjjneb.github.io/dada2/tutorial_1_8.html

# SILVA taxonomy training databases from here 
# https://zenodo.org/record/3986799#.X5kj-aozaUl

# assignTaxonomy(seqs, refFasta, minBoot = 50, tryRC = FALSE,
#                outputBootstraps = FALSE, taxLevels = c("Kingdom", "Phylum", "Class",
#                                                        "Order", "Family", "Genus", "Species"), multithread = FALSE,
#                verbose = FALSE)


head(as.character(rep_seq.south32))

taxa <- assignTaxonomy(seqs = as.character(rep_seq.south32), 
                       refFasta = "/Users/lidd0026/WORKSPACE/DATA/SILVA_DB/silva_nr99_v138_train_set.fa.gz",
                       tryRC = TRUE,
                       multithread=TRUE, verbose = TRUE)

temp.taxa.south32 <- taxa
#taxa <- temp.taxa.south32

str(taxa)
# Large matrix

tax.16s <- as.data.frame(taxa)
head(row.names(tax.16s)) # these are sequences ... swap for OTU names

identical(names(rep_seq.south32),row.names(otu.16s)) # TRUE
identical(toupper(as.character(rep_seq.south32)),row.names(tax.16s)) # TRUE

# swap labels
row.names(tax.16s) <- names(rep_seq.south32)




# tidy  up taxa table

tax <- tax.16s
names(tax)
# "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

dim(tax) # 64026     6

temp <- tax
#tax <- temp

#taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__", "s__")
taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__")

for (i in 1:length(names(tax))) {
  #i<-1
  tax[ , names(tax)[i] ] <- paste0(taxa_prefix[i],tax[ , names(tax)[i] ])
  
  sel <- which( is.na(tax[, i]) | tax[, i] == "" | tax[, i] == taxa_prefix[i] | tax[, i] == paste0(taxa_prefix[i],NA) )
  if (length(sel)>=1) {
    tax[ sel, names(tax)[i] ] <- paste0(taxa_prefix[i],"Unclassified")
  }
  print(paste0("completed ",i))
}

write.csv(tax, file = "south32-surf-16s-Taxonomy-Genus-level.csv", row.names = TRUE )


# confirm matching row names between Taxonomy and OTUs
identical(row.names(tax),row.names(otu.16s)) # TRUE


## Create 'taxonomyTable'
#  tax_table - Works on any character matrix. 
#  The rownames must match the OTU names (taxa_names) of the otu_table if you plan to combine it with a phyloseq-object.
tax.m <- as.matrix( tax )
TAX.16s <- tax_table( tax.m )





## Create 'otuTable'
#  otu_table - Works on any numeric matrix. 
#  You must also specify if the species are rows or columns
otu.m <- as.matrix( otu.16s )
dim(otu.m)
# 64026    25

OTU.16s <- otu_table(otu.16s, taxa_are_rows = TRUE)





## Create a phyloseq object, merging OTU & TAX tables
phy.16s = phyloseq(OTU.16s, TAX.16s)
phy.16s
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 64026 taxa and 25 samples ]
# tax_table()   Taxonomy Table:    [ 64026 taxa by 6 taxonomic ranks ]

sample_names(phy.16s)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374" "X138376" "X138378" "X138380"
# [13] "X138382" "X138384" "X138386" "X138388" "X138390" "X138392" "X138394" "X138396" "X138398" "X138400" "X138402" "X138404"
# [25] "X138406"

### Now Add sample data to phyloseq object
# sample_data - Works on any data.frame. The rownames must match the sample names in
# the otu_table if you plan to combine them as a phyloseq-object

dim(samp.16s) #  25 139
identical( row.names(samp.16s), sample_names(phy.16s) ) # TRUE

# # re-order names/rows to match phyloseq object
# sel <- which( row.names(samp.16s) %in% sample_names(phy.16s)) # qty 50
# 
# samp2.16s <- samp.16s[ sample_names(phy.16s), ]
# identical( row.names(samp2.16s), sample_names(phy.16s) ) # TRUE

SAMP.16s <- sample_data(samp.16s)
#SAMP.16s <- sample_data(samp2.16s)


### Combine SAMPDATA into phyloseq object
phy.16s <- merge_phyloseq(phy.16s, SAMP.16s)
phy.16s
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 64026 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 64026 taxa by 6 taxonomic ranks ]




### data cleaning

## remove taxa not assigned as Bacteria

phy_in <- phy.16s

sum(sample_sums(phy_in)) # 2097206 = 2,097,206

rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

levels(factor(tax_table(phy_in)[, "Kingdom"])) # "k__Bacteria"
# rem_taxa <- which(tax_table(phy_in)[, "Kingdom"] %in%  c("Archaea", "Eukaryota") ) #
# phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-rem_taxa, ]) )
# phy_in


## remove taxa not assigned at the phylum level
sort( unique( as.character(tax_table(phy_in)[, "Phylum"]) ) )
# [1] "p__Abditibacteriota"              "p__Acidobacteriota"               "p__Actinobacteriota"             
# [4] "p__Armatimonadota"                "p__Bacteroidota"                  "p__Bdellovibrionota"             
# [7] "p__Chloroflexi"                   "p__Cyanobacteria"                 "p__Dadabacteria"                 
# [10] "p__Deinococcota"                  "p__Dependentiae"                  "p__Desulfobacterota"             
# [13] "p__Elusimicrobiota"               "p__Entotheonellaeota"             "p__FCPU426"                      
# [16] "p__Fibrobacterota"                "p__Firmicutes"                    "p__Gemmatimonadota"              
# [19] "p__Latescibacterota"              "p__Margulisbacteria"              "p__Marinimicrobia (SAR406 clade)"
# [22] "p__MBNT15"                        "p__Methylomirabilota"             "p__Myxococcota"                  
# [25] "p__Nitrospirota"                  "p__Patescibacteria"               "p__PAUC34f"                      
# [28] "p__Planctomycetota"               "p__Proteobacteria"                "p__RCP2-54"                      
# [31] "p__SAR324 clade(Marine group B)"  "p__Sumerlaeota"                   "p__Unclassified"                 
# [34] "p__Verrucomicrobiota"             "p__WPS-2"                         "p__WS2"                          
# [37] "p__WS4"                             "p__WS4" 


rem_taxa <- which(tax_table(phy_in)[, "Phylum"] == "p__Unclassified") # qty 20
phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-rem_taxa, ]) )
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 64006 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 64006 taxa by 6 taxonomic ranks ]

sum(sample_sums(phy_in)) #  2097021

rank_names(phy_in) #  "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"
sort( as.character( unique( tax_table(phy_in)[, "Phylum"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Class"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Order"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Family"] ) ))
sort( as.character( unique( tax_table(phy_in)[, "Genus"] ) ))


## remove taxa associated with chloroplast and mitochondria

# rem_taxa1 <- which(tax_table(phy_in)[, "Class"] == "c__Chloroplast") # qty  OTUs
# rem_taxa2 <- which(tax_table(phy_in)[, "Family"] == "f__mitochondria") # qty  OTUs
rem_taxa1 <- which(tax_table(phy_in)[, "Order"] == "o__Chloroplast") # qty 2... ASV
rem_taxa2 <- which(tax_table(phy_in)[, "Family"] == "f__Mitochondria") # qty 19... ASV

length( c(rem_taxa1,rem_taxa2) ) # 21
length( unique(c(rem_taxa1,rem_taxa2)) ) # 21

# remove chloroplast and mitochondria
phy_in <- prune_taxa(phy_in, taxa = row.names(tax_table(phy_in)[-unique(c(rem_taxa1,rem_taxa2)), ]) )
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 63985 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 63985 taxa by 6 taxonomic ranks ]

sum(sample_sums(phy_in)) # 2096733


## remove taxa that do not occur in at least two samples (e.g avoiding unrepresentative taxa only present in a single sample)
temp.phy_in <- phy_in
phy_in <- prune_taxa( taxa = apply( otu_table(phy_in), 1, function(x) {sum(x > 0) >= 2 }), x = phy_in)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54671 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54671 taxa by 6 taxonomic ranks ]


sum(sample_sums(phy_in)) # 2049625


## use only the surface samples

table(phy_in@sam_data$Depth)
# 0.01 
# 25 

min(taxa_sums(phy_in)) # 2


min(taxa_sums(phy_in)) # 2
min(sample_sums(phy_in)) # 54122
sum(sample_sums(phy_in)) # 2049625

sort(sample_sums(phy_in))
# X138400 X138404 X138398 X138380 X138396 X138392 X138360 X138402 X138376 X138394 X138406 X138374 X138382 X138368 
# 54122   59216   63268   71359   71682   72227   74378   74677   74763   74906   76559   79543   81475   83191 
# X138372 X138366 X138364 X138384 X138388 X138390 X138362 X138386 X138378 X138358 X138370 
# 83599   84782   85503   87939   89314   95344   98235   98769   99395  100138  115241 




phy.south32 <- phy_in




### build and add phylogenetic tree to the phyloseq object

ps <- phy.south32

head( taxa_names(ps) )
# "OTU_ 4"  "OTU_ 5"  "OTU_ 14" "OTU_ 15" "OTU_ 21" "OTU_ 23"

taxa_names(ps) <- gsub(pattern = " ", replacement = "", x = taxa_names(ps) )

taxa_names_ps <- taxa_names(ps)
length(taxa_names_ps) # 54671


## read rep seq fasta file
#rep_seq <- read.fasta(file = paste0(datadir,"/mtbold/","Mt-Bold-AMD-16S-table-otus.fasta"), seqtype = "DNA", as.string = TRUE)
length( rep_seq.south32 ) # 64026
rep_seq <- rep_seq.south32
names(rep_seq) <- gsub(pattern = " ", replacement = "", x = names(rep_seq) )
head(rep_seq)

## subselect rep seqs
rep_seq <- rep_seq[taxa_names_ps]
length(rep_seq) # 54671
head(names(rep_seq))
# "OTU_4"  "OTU_5"  "OTU_14" "OTU_15" "OTU_21" "OTU_23"

getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"
#write.fasta(sequences = rep_seq, names = names(rep_seq), file.out = "south32_clean_surf_16s_rep_seq.fasta", open = "w",nbchar = 80 )
write.fasta(sequences = rep_seq[1], names = names(rep_seq)[1], file.out = "south32_clean_surf_16s_rep_seq.fasta", open = "w",nbchar = 80 )
for (i in 2:length(rep_seq)) {
  write.fasta(sequences = rep_seq[i], names = names(rep_seq)[i], file.out = "south32_clean_surf_16s_rep_seq.fasta", open = "a",nbchar = 80 )
  print(paste0("completed ",i))
}


## run MAFFT multiple sequence alignment on DeepThought HPC ...

# # # # # # # # # # # #
# # # # # MAFFT # # # #
# # # # # # # # # # # # 
# # https://mafft.cbrc.jp/alignment/software/
# # A fast option (FFT-NS-2) for a larger sequence alignment:
# # mafft input > output
# 
# $ whereis mafft   #:mafft: /home/lidd0026/miniconda3/bin/mafft
# 
# # run mafft using slurm script called 'run_mafft_south32_slurm.sh'
# - - - - - - - - - - - - - - - 
# #!/bin/bash
#   
# # How many tasks and processes
# #SBATCH --ntasks=1
#   
# #SBATCH --cpus-per-task=32
#   
# # How much memory
# #SBATCH --mem=64G
#   
# # Put your code to run here:
# mafft --thread -1 south32_clean_surf_16s_rep_seq.fasta > south32_clean_surf_16s_mafft_aln.fasta
# 
# - - - - - - - - - - - - - - - 
#   
# # RUN USING THIS COMMAND:
# $ cd /scratch/user/lidd0026/resto_traj
# $ ls
# 
# # more detailed view of the cluster with sinfo:
# $ sinfo -N -o "%N %C %e %m"
# 
# $ sbatch run_mafft_south32_slurm.sh

# Strategy:
#   FFT-NS-2 (Fast but rough)
# Progressive method (guide trees were built 2 times.)

## output of MAFFT is: 'south32_clean_surf_16s_mafft_aln.fasta'


## run Gblocks on Mac to clean noisy alignments
# set -b2 to 60% of sequences = 0.6*54671 = 32802.6 = 32803
## run 'relaxed' cleaning with Gblocks
# /Users/lidd0026/opt/Gblocks_0.91b/Gblocks south32_clean_surf_16s_mafft_aln.fasta -t=d -b2=32803 -b3=10 -b4=5 -b5=h -e=gbrel 

# 54671 sequences and 2610 positions in the first alignment file:
#   south32_clean_surf_16s_mafft_aln.fasta
# 
# south32_clean_surf_16s_mafft_aln.fasta
# Original alignment: 2610 positions
# Gblocks alignment:  346 positions (13 %) in 25 selected block(s)

## output of Gblocks is: 'south32_clean_surf_16s_mafft_aln_Gblocks_relaxed.fasta'

gblocks_aln <- read.fasta( paste0(getwd(),"/deepthoughtHPC/","south32_clean_surf_16s_mafft_aln_Gblocks_relaxed.fasta"),
                           as.string = FALSE)
head(gblocks_aln)
unique(unlist(gblocks_aln)) # "a" "g" "c" "t" " " "-"
gblocks_aln <- read.fasta( paste0(getwd(),"/deepthoughtHPC/","south32_clean_surf_16s_mafft_aln_Gblocks_relaxed.fasta"),
                           as.string = TRUE)
names_gblocks_aln <- names(gblocks_aln)
head(names_gblocks_aln) # "OTU_4"  "OTU_5"  "OTU_14" "OTU_15" "OTU_21" "OTU_23"
gblocks_aln <- toupper(gblocks_aln)
gblocks_aln <- gsub(pattern = " ",replacement = "-", x = gblocks_aln)
names(gblocks_aln) <- names_gblocks_aln
head(gblocks_aln)


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"
write.fasta(sequences = gblocks_aln[1], names = names(gblocks_aln)[1], file.out = "south32_clean_16s_mafft_Gblocks_rel_tidy.fasta", open = "w",nbchar = 80 )
for (i in 2:length(gblocks_aln)) {
  write.fasta(sequences = gblocks_aln[i], names = names(gblocks_aln)[i], file.out = "south32_clean_16s_mafft_Gblocks_rel_tidy.fasta", open = "a",nbchar = 80 )
  print(paste0("completed ",i))
}



## run IQTREE2 to build phylogenetic tree on DeepThought HPC

# # # # # # # # # # #
# # # # IQTREE2 # # #
# # # # # # # # # # # 
# 
# # run iqtree2 using slurm script called 'run_iqtree2_south32_slurm.sh'
# 
# - - - - - - - - - - - - - - - 
# #!/bin/bash
#   
# # How many tasks and processes
# #SBATCH --ntasks=1
#   
# #SBATCH --cpus-per-task=32
#   
# # How much memory
# #SBATCH --mem=64G
#   
# # Put your code to run here:
# /home/lidd0026/opt/iqtree-2.1.2-Linux/bin/iqtree2 -s /scratch/user/lidd0026/resto_traj/south32_clean_16s_mafft_Gblocks_rel_tidy.fasta --prefix iqtree2_south32_gblocks -m GTR+I+G -T 32 -n 200
# 
# - - - - - - - - - - - - - - - 
#   
# # RUN USING THIS COMMAND:
# $ cd /scratch/user/lidd0026/resto_traj
# $ ls
# 
# $ sinfo -N -o "%N %C %e %m"
# 
# $ sbatch run_iqtree2_south32_slurm.sh



### add phylo tree

phy_in <- phy.south32

phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54671 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54671 taxa by 6 taxonomic ranks ]

tree <- read_tree("iqtree2_south32_gblocks.treefile")

# need to match taxa names ...
head( taxa_names(phy_in) ) # "OTU_ 4"  "OTU_ 5"  "OTU_ 14" "OTU_ 15" "OTU_ 21" "OTU_ 23"
head( taxa_names(tree)) # "OTU_4"     "OTU_5"     "OTU_14"    "OTU_15"    "OTU_18569" "OTU_18572"

# fix issue with phy_in taxa names
taxa_names(phy_in) <- gsub(pattern = " ", replacement = "", x = taxa_names(phy_in))

length(taxa_names(phy_in)) # 54671
length(taxa_names(tree)) # 54671

sel <- which(taxa_names(tree) %in% taxa_names(phy_in))
length(sel) # 54671
identical(sort(taxa_names(tree)),sort(taxa_names(phy_in))) # TRUE

phy_in <- merge_phyloseq(phy_in, tree)
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54671 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54671 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54671 tips and 54669 internal nodes ]

phy.south32.withTree <- phy_in


#
#-------------------------

#### 4b. South32 distance plots

## South32 - with Tree - i. Bray-Curtis - ASV-level
#-------------------------

phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

sample_sums(phy_in)
min(sample_sums(phy_in)) # 54122

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122

ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374"
# [10] "X138376" "X138378" "X138380" "X138382" "X138384" "X138386" "X138388" "X138390" "X138392"
# [19] "X138394" "X138396" "X138398" "X138400" "X138402" "X138404" "X138406"

identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?

sel <- which(r1.ps@sam_data$`year of rehab or ref` == "REF") # qty 6
( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# "X138362" "X138366" "X138374" "X138376" "X138378" "X138404"

table(r1.ps@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6 

row.names(r1.ps@sam_data)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374"
# [10] "X138376" "X138378" "X138380" "X138382" "X138384" "X138386" "X138388" "X138390" "X138392"
# [19] "X138394" "X138396" "X138398" "X138400" "X138402" "X138404" "X138406"

identical(colnames(sim.bray), row.names(r1.ps@sam_data)) # TRUE


df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`year of rehab or ref`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 150   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 6
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 144   4

unique(df_out$year_rehab)
# "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

df_out$group <- df_out$year_rehab
sort( unique(df_out$group) , decreasing = TRUE)
#"REF"  "2017" "2011" "2007" "2005" "2002" "1999" "1996" "1991"

2019 - c(2017,2011,2007,2005,2002,1999,1996,1991)
# 2  8 12 14 17 20 23 28

df_out$group <- factor(df_out$group, 
                       levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                       labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)

table(df_out$group)
#  2   8  12  14  17  20  23  28 Ref 
# 18  18   6  12  12  12  24  12  30 

# df_out$group <- factor(df_out$group, levels=as.character(c(0:40,"Ref")),ordered=TRUE)
# df_out$group

dim(df_out) # 144  5


df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "South32"
df_out$corrected <- "no"  
df_out$facet_x <- "Bray-Curtis\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["South32-Bray-ASV-withTree"]] <- df_out
#df_out <- df_results[[1]]



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	144 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 80.84, df = 8, p-value = 3.312e-14


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# 23  Ref 2   17  12  20  8   14  28 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("2","8","12","14","17","20","23","28","Ref"),
                     labels = c("2","8","12","14","17","twenty","23","28","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

#    Group Letter MonoLetter
# 1     12    abc        abc
# 2     14     ab        ab 
# 3     17    abc        abc
# 4      2      a        a  
# 5     23     bc         bc
# 6     28     bc         bc
# 7      8      a        a  
# 8    Ref      c          c
# 9 twenty     bc         bc

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	9 obs. of  3 variables:
# $ Group     : chr  "12" "14" "17" "2" ...
# $ Letter    : chr  "abc" "ab" "abc" "a" ...
# $ MonoLetter: chr  "abc" "ab " "abc" "a  " ...

unique(sig$Group) #  "12"     "14"     "17"     "2"      "23"     "28"     "8"      "Ref"    "twenty"

sig$Group <- factor( sig$Group, levels=c("2","8","12","14","17","twenty","23","28","Ref"),
                     labels = c("2","8","12","14","17","20","23","28","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
#  Group Letter MonoLetter
# 4     2      a        a  
# 7     8      a        a  
# 1    12    abc        abc
# 2    14     ab        ab 
# 3    17    abc        abc
# 9    20     bc         bc
# 5    23     bc         bc
# 6    28     bc         bc
# 8   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "12"  "14"  "17"  "20"  "23"  "28"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(31,32,40,35,40,45,46,45,48),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "South32", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) 

df_anno[["South32-Bray-ASV-withTree"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "bray")

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: bray 
# 
# Dimensions: 2 
# Stress:     0.1617845 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`year of rehab or ref`
unique(r1.ps@sam_data$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                               labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) #
names(ord_plots) #

ord_plots[["South32-Bray-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) #

ord_obj[["South32-Bray-ASV-withTree"]] <- ord

#-------------------------

## South32 - with Tree - ii. Jaccard - ASV-level
#-------------------------
phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

sample_sums(phy_in)
min(sample_sums(phy_in)) # 54122

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122

ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374" "X138376" "X138378" "X138380"
# [13] "X138382" "X138384" "X138386" "X138388" "X138390" "X138392" "X138394" "X138396" "X138398" "X138400" "X138402" "X138404"
# [25] "X138406"

identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`year of rehab or ref`
# [1] "1996" "1996" "REF"  "2017" "REF"  "2002" "2007" "1996" "REF"  "REF"  "REF"  "1999" "2011" "2011" "2005" "1999" "1996"
# [18] "2002" "2011" "2005" "1991" "1991" "2017" "REF"  "2017"

table(r1.ps@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6 

row.names(r1.ps@sam_data)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374" "X138376" "X138378" "X138380"
# [13] "X138382" "X138384" "X138386" "X138388" "X138390" "X138392" "X138394" "X138396" "X138398" "X138400" "X138402" "X138404"
# [25] "X138406"

identical(colnames(sim.jacc), row.names(r1.ps@sam_data)) # TRUE

sel <- which(r1.ps@sam_data$`year of rehab or ref` == "REF") # qty 6

( ref_sites <- row.names(r1.ps@sam_data)[sel] )

# "X138362" "X138366" "X138374" "X138376" "X138378" "X138404"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`year of rehab or ref`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 150   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 6
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 144  4


sort( unique(df_out$year_rehab) , decreasing = TRUE )
#  "REF"  "2017" "2011" "2007" "2005" "2002" "1999" "1996" "1991"

df_out$group <- df_out$year_rehab
sort( unique(df_out$group) , decreasing = TRUE )
# "REF"  "2017" "2011" "2007" "2005" "2002" "1999" "1996" "1991"

2019 - c(2017,2011,2007,2005,2002,1999,1996,1991)
# 2  8 12 14 17 20 23 28

df_out$group <- factor(df_out$group, 
                       levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                       labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)

table(df_out$group)
# 2   8  12  14  17  20  23  28 Ref 
# 18  18   6  12  12  12  24  12  30 


dim(df_out) # 144 4


df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "South32"
df_out$corrected <- "no"  
df_out$facet_x <- "Jaccard\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["South32-Jaccard-ASV-withTree"]] <- df_out




p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	144 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 70.968, df = 8, p-value = 3.153e-12


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# 23  Ref 2   17  12  20  8   14  28 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("2","8","12","14","17","20","23","28","Ref"),
                     labels = c("2","8","12","14","17","twenty","23","28","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)


sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	9 obs. of  3 variables:
#   $ Group     : chr  "12" "14" "17" "2" ...
# $ Letter    : chr  "abc" "ab" "abc" "ab" ...
# $ MonoLetter: chr  "abc" "ab " "abc" "ab " ...

unique(sig$Group) #  "12"     "14"     "17"     "2"      "23"     "28"     "8"      "Ref"    "twenty"

sig$Group <- factor( sig$Group, levels=c("2","8","12","14","17","twenty","23","28","Ref"),
                     labels = c("2","8","12","14","17","20","23","28","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 4     2     ab        ab 
# 7     8      b         b 
# 1    12    abc        abc
# 2    14     ab        ab 
# 3    17    abc        abc
# 9    20      c          c
# 5    23     ac        a c
# 6    28      c          c
# 8   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "12"  "14"  "17"  "20"  "23"  "28"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(26,24,28,26,31,35,33,35,35),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "South32", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) 

df_anno[["South32-Jaccard-ASV-withTree"]] <- anno





## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "jaccard", binary=TRUE)

ord
# Call:
#   metaMDS(comm = veganifyOTU(physeq), distance = distance, binary = TRUE) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     wisconsin(sqrt(veganifyOTU(physeq))) 
# Distance: binary jaccard 
# 
# Dimensions: 2 
# Stress:     0.1589099 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation, halfchange scaling 
# Species: expanded scores based on ‘wisconsin(sqrt(veganifyOTU(physeq)))’

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`year of rehab or ref`
unique(r1.ps@sam_data$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"


r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                               labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) #
names(ord_plots) #

ord_plots[["South32-Jaccard-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) #

ord_obj[["South32-Jaccard-ASV-withTree"]] <- ord



#-------------------------

## South32 - with Tree - iii. Weighted UniFrac - ASV-level
#-------------------------

phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

sample_sums(phy_in)
min(sample_sums(phy_in)) # 54122

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122

ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]

head(r1.ps@otu_table)

dist.wuni <- phyloseq::distance(r1.ps, "wunifrac") # weighted UniFrac
class(dist.wuni) # "dist"
dist.wuni <- as(dist.wuni, "matrix")
# express as similarity
sim.wuni <- 100*(1 - dist.wuni)
# list all sample names
colnames(sim.wuni)

identical(colnames(sim.wuni),rownames(sim.wuni)) # TRUE

## which are the reference sites?

sel <- which(r1.ps@sam_data$`year of rehab or ref` == "REF") # qty 6
( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# "X138362" "X138366" "X138374" "X138376" "X138378" "X138404"

table(r1.ps@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6 

row.names(r1.ps@sam_data)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374"
# [10] "X138376" "X138378" "X138380" "X138382" "X138384" "X138386" "X138388" "X138390" "X138392"
# [19] "X138394" "X138396" "X138398" "X138400" "X138402" "X138404" "X138406"

identical(colnames(sim.wuni), row.names(r1.ps@sam_data)) # TRUE


df_in <- sim.wuni

df_out <- data.frame(samp= rep( x = colnames(sim.wuni), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.wuni))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`year of rehab or ref`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 150   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 6
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 144   4

unique(df_out$year_rehab)
# "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

df_out$group <- df_out$year_rehab
sort( unique(df_out$group) , decreasing = TRUE)
#"REF"  "2017" "2011" "2007" "2005" "2002" "1999" "1996" "1991"

2019 - c(2017,2011,2007,2005,2002,1999,1996,1991)
# 2  8 12 14 17 20 23 28

df_out$group <- factor(df_out$group, 
                       levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                       labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)

table(df_out$group)
#  2   8  12  14  17  20  23  28 Ref 
# 18  18   6  12  12  12  24  12  30 


dim(df_out) # 144  5


df_out$dist_measure <- "Weighted UniFrac"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "South32"
df_out$corrected <- "no"   
df_out$facet_x <- "Weighted UniFrac\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["South32-WtUniFrac-ASV-withTree"]] <- df_out
#df_out <- df_results[[1]]



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Weighted Unifrac Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Weighted-Unifrac-Similarity--South32.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw")



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	144 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 57.543, df = 8, p-value = 1.411e-09


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# 23  Ref 2   17  12  20  8   14  28 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("2","8","12","14","17","20","23","28","Ref"),
                     labels = c("2","8","12","14","17","twenty","23","28","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)


sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	9 obs. of  3 variables:
#   $ Group     : chr  "12" "14" "17" "2" ...
# $ Letter    : chr  "ab" "ab" "ab" "a" ...
# $ MonoLetter: chr  "ab" "ab" "ab" "a " ...

unique(sig$Group) #  "12"     "14"     "17"     "2"      "23"     "28"     "8"      "Ref"    "twenty"

sig$Group <- factor( sig$Group, levels=c("2","8","12","14","17","twenty","23","28","Ref"),
                     labels = c("2","8","12","14","17","20","23","28","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 4     2      a         a 
# 7     8      a         a 
# 1    12     ab         ab
# 2    14     ab         ab
# 3    17     ab         ab
# 9    20      b          b
# 5    23      b          b
# 6    28      b          b
# 8   Ref      b          b

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "12"  "14"  "17"  "20"  "23"  "28"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(95,95.5,96.5,96,96.5,97.5,98,97.5,98),
                   dist_measure = rep( "Weighted UniFrac", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "South32", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Weighted UniFrac\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) 

df_anno[["South32-WtUniFrac-ASV-withTree"]] <- anno





## ordination plot
## NMDS + Bray-Curtis

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "unifrac", weighted = TRUE)

ord
# Call:
#   metaMDS(comm = ps.dist) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     ps.dist 
# Distance: user supplied 
# 
# Dimensions: 2 
# Stress:     0.1399623 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation 
# Species: scores missing

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`year of rehab or ref`
unique(r1.ps@sam_data$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                               labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) # 
names(ord_plots) #

ord_plots[["South32-WtUniFrac-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) #


ord_obj[["South32-WtUniFrac-ASV-withTree"]] <- ord

#-------------------------

## South32 - with Tree - iv. Unweighted UniFrac - ASV-level
#-------------------------
phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

sample_sums(phy_in)
min(sample_sums(phy_in)) # 54122

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122

ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]

head(r1.ps@otu_table)

dist.uuni <- phyloseq::distance(r1.ps, "uunifrac") # Unweighted UniFrac
class(dist.uuni) # "dist"
dist.uuni <- as(dist.uuni, "matrix")
# express as similarity
sim.uuni <- 100*(1 - dist.uuni)
# list all sample names
colnames(sim.uuni)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374" "X138376" "X138378" "X138380" "X138382" "X138384" "X138386"
# [16] "X138388" "X138390" "X138392" "X138394" "X138396" "X138398" "X138400" "X138402" "X138404" "X138406"

identical(colnames(sim.uuni),rownames(sim.uuni)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`year of rehab or ref`
# [1] "1996" "1996" "REF"  "2017" "REF"  "2002" "2007" "1996" "REF"  "REF"  "REF"  "1999" "2011" "2011" "2005" "1999" "1996"
# [18] "2002" "2011" "2005" "1991" "1991" "2017" "REF"  "2017"

table(r1.ps@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6 

row.names(r1.ps@sam_data)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374" "X138376" "X138378" "X138380"
# [13] "X138382" "X138384" "X138386" "X138388" "X138390" "X138392" "X138394" "X138396" "X138398" "X138400" "X138402" "X138404"
# [25] "X138406"

identical(colnames(sim.uuni), row.names(r1.ps@sam_data)) # TRUE

sel <- which(r1.ps@sam_data$`year of rehab or ref` == "REF") # qty 6

( ref_sites <- row.names(r1.ps@sam_data)[sel] )

# "X138362" "X138366" "X138374" "X138376" "X138378" "X138404"

df_in <- sim.uuni

df_out <- data.frame(samp= rep( x = colnames(sim.uuni), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.uuni))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`year of rehab or ref`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 150   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 6
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 144  4


sort( unique(df_out$year_rehab) , decreasing = TRUE )
#  "REF"  "2017" "2011" "2007" "2005" "2002" "1999" "1996" "1991"

df_out$group <- df_out$year_rehab
sort( unique(df_out$group) , decreasing = TRUE )
# "REF"  "2017" "2011" "2007" "2005" "2002" "1999" "1996" "1991"

2019 - c(2017,2011,2007,2005,2002,1999,1996,1991)
# 2  8 12 14 17 20 23 28

df_out$group <- factor(df_out$group, 
                       levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                       labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)

table(df_out$group)
# 2   8  12  14  17  20  23  28 Ref 
# 18  18   6  12  12  12  24  12  30 


dim(df_out) # 144 4


df_out$dist_measure <- "Unweighted UniFrac"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "South32"
df_out$corrected <- "no"   
df_out$facet_x <- "Unweighted UniFrac\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["South32-UnWtUniFrac-ASV-withTree"]] <- df_out




p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Unweighted UniFrac Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Unweighted-UniFrac-Similarity--South32.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw")



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	144 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 77.136, df = 8, p-value = 1.84e-13


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# 23  Ref 2   17  12  20  8   14  28 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("2","8","12","14","17","20","23","28","Ref"),
                     labels = c("2","8","12","14","17","twenty","23","28","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)


sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	9 obs. of  3 variables:
#   $ Group     : chr  "12" "14" "17" "2" ...
# $ Letter    : chr  "abc" "ab" "abc" "a" ...
# $ MonoLetter: chr  "abc" "ab " "abc" "a  " ...

unique(sig$Group) #  "12"     "14"     "17"     "2"      "23"     "28"     "8"      "Ref"    "twenty"

sig$Group <- factor( sig$Group, levels=c("2","8","12","14","17","twenty","23","28","Ref"),
                     labels = c("2","8","12","14","17","20","23","28","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
#    Group Letter MonoLetter
# 4     2      a        a  
# 7     8      a        a  
# 1    12    abc        abc
# 2    14     ab        ab 
# 3    17    abc        abc
# 9    20      c          c
# 5    23     bc         bc
# 6    28     bc         bc
# 8   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "12"  "14"  "17"  "20"  "23"  "28"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(56,55.5,60,58,61,63,61,61.5,63),
                   dist_measure = rep( "Unweighted UniFrac", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "South32", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Unweighted UniFrac\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) 


df_anno[["South32-UnWtUniFrac-ASV-withTree"]] <- anno





## ordination plot
## NMDS + Jaccard

set.seed(123)
ord <- ordinate(r1.ps, "NMDS", "unifrac", weighted = FALSE)

ord

# Call:
#   metaMDS(comm = ps.dist) 
# 
# global Multidimensional Scaling using monoMDS
# 
# Data:     ps.dist 
# Distance: user supplied 
# 
# Dimensions: 2 
# Stress:     0.1452691 
# Stress type 1, weak ties
# Two convergent solutions found after 20 tries
# Scaling: centring, PC rotation 
# Species: scores missing

str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`year of rehab or ref`
unique(r1.ps@sam_data$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"


r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                               labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


p <- plot_ordination(r1.ps, ord, type="samples", color="group")
p

str(p$data)

# only make this once
#ord_plots <- list()
length(ord_plots) #
names(ord_plots) #

ord_plots[["South32-UnWtUniFrac-ASV-withTree"]] <- p

#ord_obj <- list()
length(ord_obj) #
names(ord_obj) #

ord_obj[["South32-UnWtUniFrac-ASV-withTree"]] <- ord


#-------------------------



## South32 - with Tree - Bray-Curtis - ASV-level - EXCLUDE youngest rehab_age & associated Ref
#-------------------------

phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

sample_sums(phy_in)
min(sample_sums(phy_in)) # 54122

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122

ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")
# express as similarity
sim.bray <- 100*(1 - dist.bray)
# list all sample names
colnames(sim.bray)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374"
# [10] "X138376" "X138378" "X138380" "X138382" "X138384" "X138386" "X138388" "X138390" "X138392"
# [19] "X138394" "X138396" "X138398" "X138400" "X138402" "X138404" "X138406"

identical(colnames(sim.bray),rownames(sim.bray)) # TRUE

## which are the reference sites?

sel <- which(r1.ps@sam_data$`year of rehab or ref` == "REF") # qty 6
( ref_sites <- row.names(r1.ps@sam_data)[sel] )
# "X138362" "X138366" "X138374" "X138376" "X138378" "X138404"

table(r1.ps@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6 

row.names(r1.ps@sam_data)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374"
# [10] "X138376" "X138378" "X138380" "X138382" "X138384" "X138386" "X138388" "X138390" "X138392"
# [19] "X138394" "X138396" "X138398" "X138400" "X138402" "X138404" "X138406"

identical(colnames(sim.bray), row.names(r1.ps@sam_data)) # TRUE


df_in <- sim.bray

df_out <- data.frame(samp= rep( x = colnames(sim.bray), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.bray))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`year of rehab or ref`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 150   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 6
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 144   4

unique(df_out$year_rehab)
# "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

df_out$group <- df_out$year_rehab
sort( unique(df_out$group) , decreasing = TRUE)
#"REF"  "2017" "2011" "2007" "2005" "2002" "1999" "1996" "1991"

2019 - c(2017,2011,2007,2005,2002,1999,1996,1991)
# 2  8 12 14 17 20 23 28


class(df_out$group) # "character"

dim(df_out) # 144   5

## exclude youngest rehab_age group
sel <- which(df_out$group == "2017") # qty 18
df_out <- df_out[-sel, ]
dim(df_out) # 126   5

unique(df_out$group) # "1996" "REF"  "2002" "2007" "1999" "2011" "2005" "1991"

## also exclude the Ref site that was associated with the southernmost youngest sites

hist(r1.ps@sam_data$`Latitude [decimal degrees]`)
sel <- which(r1.ps@sam_data$`Latitude [decimal degrees]` > 32.95)
as(r1.ps@sam_data, "data.frame")[sel, c("Sample_ID","year.of.rehab.or.ref")  ] #
#                  Sample_ID year.of.rehab.or.ref
# X138402 102.100.100/138402                 2017
# X138404 102.100.100/138404                  REF
# X138406 102.100.100/138406                 2017

sel <- which(df_out$samp %in% c("X138402","X138404","X138406")) # qty 5
df_out[sel, ]
#        samp compare_with year_rehab similarity group
# 139 X138404      X138362        REF   28.89398   REF
# 140 X138404      X138366        REF   21.39426   REF
# 141 X138404      X138374        REF   32.50249   REF
# 142 X138404      X138376        REF   24.15654   REF
# 143 X138404      X138378        REF   25.63098   REF

df_out <- df_out[-sel, ]
dim(df_out) # 121   5

table(df_out$group)
# 1991 1996 1999 2002 2005 2007 2011  REF 
#   12   24   12   12   12    6   18   25



df_out$group <- factor(df_out$group, 
                       levels = c("2011","2007","2005","2002","1999","1996","1991","REF"),
                       labels = c("8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 8-28 yr old)

table(df_out$group)
#  8  12  14  17  20  23  28 Ref 
# 18   6  12  12  12  24  12  25 

# df_out$group <- factor(df_out$group, levels=as.character(c(0:40,"Ref")),ordered=TRUE)
# df_out$group

dim(df_out) # 121  5


df_out$dist_measure <- "Bray-Curtis"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "South32\n(exclude 2-year rehab)"
df_out$corrected <- "no"   
df_out$facet_x <- "Bray-Curtis\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["South32-Bray-ASV-withTree-exclude2yr"]] <- df_out
#df_out <- df_results[[1]]



p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	121 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 67.727, df = 7, p-value = 4.249e-12


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# 23  Ref 17  12  20  8   14  28 



# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("8","12","14","17","20","23","28","Ref"),
                     labels = c("8","12","14","17","twenty","23","28","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)


sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	8 obs. of  3 variables:

unique(sig$Group) #  "12"     "14"     "17"     "23"     "28"     "8"      "Ref"    "twenty"

sig$Group <- factor( sig$Group, levels=c("8","12","14","17","twenty","23","28","Ref"),
                     labels = c("8","12","14","17","20","23","28","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 6     8      b         b 
# 1    12     ab        ab 
# 2    14     ab        ab 
# 3    17     ab        ab 
# 8    20     ac        a c
# 4    23     ac        a c
# 5    28     ac        a c
# 7   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "8"   "12"  "14"  "17"  "20"  "23"  "28"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(32,40,35,40,47,45,46,48),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "South32\n(exclude 2-year rehab)", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) 

df_anno[["South32-Bray-ASV-withTree-exclude2yr"]] <- anno



#-------------------------

## South32 - with Tree - Jaccard - ASV-level - EXCLUDE youngest rehab_age & associated Ref
#-------------------------
phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

sample_sums(phy_in)
min(sample_sums(phy_in)) # 54122

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122

ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")
# express as similarity
sim.jacc <- 100*(1 - dist.jacc)
# list all sample names
colnames(sim.jacc)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374" "X138376" "X138378" "X138380"
# [13] "X138382" "X138384" "X138386" "X138388" "X138390" "X138392" "X138394" "X138396" "X138398" "X138400" "X138402" "X138404"
# [25] "X138406"

identical(colnames(sim.jacc),rownames(sim.jacc)) # TRUE

## which are the reference sites?
r1.ps@sam_data$`year of rehab or ref`
# [1] "1996" "1996" "REF"  "2017" "REF"  "2002" "2007" "1996" "REF"  "REF"  "REF"  "1999" "2011" "2011" "2005" "1999" "1996"
# [18] "2002" "2011" "2005" "1991" "1991" "2017" "REF"  "2017"

table(r1.ps@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6 

row.names(r1.ps@sam_data)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374" "X138376" "X138378" "X138380"
# [13] "X138382" "X138384" "X138386" "X138388" "X138390" "X138392" "X138394" "X138396" "X138398" "X138400" "X138402" "X138404"
# [25] "X138406"

identical(colnames(sim.jacc), row.names(r1.ps@sam_data)) # TRUE

sel <- which(r1.ps@sam_data$`year of rehab or ref` == "REF") # qty 6

( ref_sites <- row.names(r1.ps@sam_data)[sel] )

# "X138362" "X138366" "X138374" "X138376" "X138378" "X138404"

df_in <- sim.jacc

df_out <- data.frame(samp= rep( x = colnames(sim.jacc), each=length(ref_sites) ),
                     compare_with = rep( x = ref_sites, times = length(colnames(sim.jacc))),
                     #site=NA,
                     #site_full_desc=NA,
                     year_rehab=NA,
                     similarity=NA
)



for (i in 1:dim(df_out)[1]) {
  #i<-1
  this_samp <- df_out$samp[i]
  this_compare <- df_out$compare_with[i]
  sel <- which( row.names(r1.ps@sam_data) == this_samp)
  
  df_out$year_rehab[i] <- as.character(r1.ps@sam_data$`year of rehab or ref`[sel])
  
  row <- which(rownames(df_in)==this_samp)
  col <- which(colnames(df_in)==this_compare)
  
  df_out$similarity[i] <- df_in[row,col]
  
}

dim(df_out) # 150   4

# remove remnant self-comparisons
sel <- which(df_out$samp==df_out$compare_with) # qty 6
identical( which(df_out$samp==df_out$compare_with), which(df_out$similarity==100) ) # TRUE
df_out <- df_out[-sel, ]
dim(df_out) # 144  4


sort( unique(df_out$year_rehab) , decreasing = TRUE )
#  "REF"  "2017" "2011" "2007" "2005" "2002" "1999" "1996" "1991"

df_out$group <- df_out$year_rehab
sort( unique(df_out$group) , decreasing = TRUE )
# "REF"  "2017" "2011" "2007" "2005" "2002" "1999" "1996" "1991"

2019 - c(2017,2011,2007,2005,2002,1999,1996,1991)
# 2  8 12 14 17 20 23 28


## exclude youngest rehab_age group
sel <- which(df_out$group == "2017") # qty 18
df_out <- df_out[-sel, ]
dim(df_out) # 126   5

unique(df_out$group) # "1996" "REF"  "2002" "2007" "1999" "2011" "2005" "1991"

## also exclude the Ref site that was associated with the southernmost youngest sites

hist(r1.ps@sam_data$`Latitude [decimal degrees]`)
sel <- which(r1.ps@sam_data$`Latitude [decimal degrees]` > 32.95)
as(r1.ps@sam_data, "data.frame")[sel, c("Sample_ID","year.of.rehab.or.ref")  ] #
#                  Sample_ID year.of.rehab.or.ref
# X138402 102.100.100/138402                 2017
# X138404 102.100.100/138404                  REF
# X138406 102.100.100/138406                 2017

sel <- which(df_out$samp %in% c("X138402","X138404","X138406")) # qty 5
df_out[sel, ]
#        samp compare_with year_rehab similarity group
# 139 X138404      X138362        REF   28.89398   REF
# 140 X138404      X138366        REF   21.39426   REF
# 141 X138404      X138374        REF   32.50249   REF
# 142 X138404      X138376        REF   24.15654   REF
# 143 X138404      X138378        REF   25.63098   REF

df_out <- df_out[-sel, ]
dim(df_out) # 121   5

table(df_out$group)
# 1991 1996 1999 2002 2005 2007 2011  REF 
#   12   24   12   12   12    6   18   25




df_out$group <- factor(df_out$group, 
                       levels = c("2011","2007","2005","2002","1999","1996","1991","REF"),
                       labels = c("8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 8-28 yr old)

table(df_out$group)
# 8  12  14  17  20  23  28 Ref 
# 18   6  12  12  12  24  12  25


dim(df_out) # 121 4


df_out$dist_measure <- "Jaccard"
df_out$tax_level <- "ASV (with Tree)"
df_out$study <- "South32\n(exclude 2-year rehab)"
df_out$corrected <- "no"   
df_out$facet_x <- "Jaccard\nASV-level"


# only make this once
#df_results <- list()
length(df_results) #
names(df_results) #

df_results[["South32-Jaccard-ASV-withTree-exclude2yr"]] <- df_out




p <- ggplot(data=df_out, aes(x=group, similarity)) + ggtitle("Jaccard Similarity") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_out)
# 'data.frame':	121 obs. of  10 variables:


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_out) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  similarity by group
# Kruskal-Wallis chi-squared = 62.957, df = 7, p-value = 3.867e-11


## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique(df_out$group)
# 23  Ref 2   17  12  20  8   14  28 

# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_out
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("8","12","14","17","20","23","28","Ref"),
                     labels = c("8","12","14","17","twenty","23","28","Ref"),ordered = TRUE)

 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt



pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)


sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	8 obs. of  3 variables:

unique(sig$Group) #  "12"     "14"     "17"     "23"     "28"     "8"      "Ref"    "twenty"

sig$Group <- factor( sig$Group, levels=c("8","12","14","17","twenty","23","28","Ref"),
                     labels = c("8","12","14","17","20","23","28","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 6     8      b         b 
# 1    12    abc        abc
# 2    14     ab        ab 
# 3    17    abc        abc
# 8    20      c          c
# 4    23     ac        a c
# 5    28      c          c
# 7   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "8"   "12"  "14"  "17"  "20"  "23"  "28"  "Ref"

## store annotations for later facet_grid ggplot
names(df_out)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(24,28,26,32,35,34,33,35),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "South32\n(exclude 2-year rehab)", times = dim(sig)[1]),
                   corrected = rep( "no" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) #
names(df_anno) 

df_anno[["South32-Jaccard-ASV-withTree-exclude2yr"]] <- anno

## UP TO HERE !!!

#-------------------------



#### Rarefaction curves: Alcoa-Huntly, Iluka-Eneabba, South32-Worsley
#-------------------------

### Rarefaction curve


## Code for calculating rarefaction curves using phyloseq are here (provided by Bela Hausmann)
# https://github.com/joey711/phyloseq/issues/143



calculate_rarefaction_curves <- function(psdata, measures, depths) {
  require('plyr') # ldply
  require('reshape2') # melt
  
  estimate_rarified_richness <- function(psdata, measures, depth) {
    if(max(sample_sums(psdata)) < depth) return()
    psdata <- prune_samples(sample_sums(psdata) >= depth, psdata)
    
    rarified_psdata <- rarefy_even_depth(psdata, depth, replace = FALSE, verbose = FALSE) # default: replace = TRUE
    
    alpha_diversity <- estimate_richness(rarified_psdata, measures = measures)
    
    # as.matrix forces the use of melt.array, which includes the Sample names (rownames)
    molten_alpha_diversity <- melt(as.matrix(alpha_diversity), varnames = c('Sample', 'Measure'), value.name = 'Alpha_diversity')
    
    molten_alpha_diversity
  }
  
  names(depths) <- depths # this enables automatic addition of the Depth to the output by ldply
  rarefaction_curve_data <- ldply(depths, estimate_rarified_richness, psdata = psdata, measures = measures, .id = 'Depth', .progress = ifelse(interactive(), 'text', 'none'))
  
  # convert Depth from factor to numeric
  rarefaction_curve_data$Depth <- as.numeric(levels(rarefaction_curve_data$Depth))[rarefaction_curve_data$Depth]
  
  rarefaction_curve_data
}


### Alcoa-Huntly

phy_in <- phy.alcoa.withTree


sample_sums(phy_in)
# X39041 X39043 X39045 X39047 X39049 X39051 X39053 X39055 X39057 X39059 X39061 X39063 X39065 X39067 X39069 X39071 X39073 
# 35628  48218  31768  61698  58221  39709  42476  79269  58093  31804  32412  43924  67413  43797  61056  42302  60439 
# X39075 X39077 X39079 X39081 X39083 X39085 X39087 X39089 X39091 X39093 X39095 X39097 X39099 X39161 X39163 X39165 X39191 
# 45554  45587  51794  59390  52427  58043  51424  53532  69298  57338  42579  57002  34570  42102  46572  62357  67370 
# X39193 X39195 
# 19598  17485 

summary(sample_sums(phy_in))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 17485   42252   49821   49229   58513   79269


## Examine if species accumulation (rarefaction curves) are converging
## (i.e. reaching horizontal asymptote) or not?

psdata <- phy_in
set.seed(123)
rarefaction_curve_data <- calculate_rarefaction_curves(psdata, c('Observed'), sort(c(min(sample_sums(psdata)), 1, 10, 100, 1000, 1:80 * 1000)) )
# cover sequence read depths up to max value: 79269
rarefaction_curve_data

str( data.frame(sample_data(psdata)) )
row.names(data.frame(sample_data(psdata)))

## Add sample data
df <- merge(rarefaction_curve_data, data.frame(sample_data(psdata)), by.x = 'Sample', by.y = 'row.names')

sample_data(psdata)

# $Date.since.change.in.Land.Use

df$group <- df$Date.since.change.in.Land.Use
unique(df$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
sel <- which(df$group == "N/A")
df$group[sel] # all N/A
df$group[sel] <- "Ref"
unique(df$group) # "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df$group <- factor(df$group, 
                   levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                   labels = c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



cols.group.alcoa <- c("2 yr" = "#9e0142","8 yr" = "#d53e4f","14 yr" = "#fdae61","17 yr" = "#e6f598",
                      "25 yr" = "#abdda4","29 yr" = "#66c2a5","Ref" = "#5e4fa2")

cols <- cols.group.alcoa

str(df)

summary(df$Alpha_diversity)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#    1    3648    5928    5607    7608   12142 

p <- ggplot(
  data = df,
  # note the table merge has Depth.x from sequence depth; and Depth.y from soil horizon depth
  mapping = aes(x = Depth.x, y = Alpha_diversity, colour = group, group = Sample)) +
  theme_bw() +
  scale_colour_manual(values = cols, name  ="Rehab\nage") +
  ggtitle("(A) Huntly") +
  labs(x = "ASV sequence depth", y = "Observed ASVs (count)") +
  geom_line() +
  geom_vline(xintercept = min(sample_sums(psdata)), linetype="dotted") +
  ylim(0,12200) +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )

p

ggsave(plot=p, filename = paste0("plots/","Rarefaction-curve-A-Alcoa-Huntly.tiff"), width = 12, height = 8, units = "cm", dpi = 600, compression = "lzw", type = "cairo")



### Iluka-Eneabba

phy_in <- phy.iluka_eneabba.withTree


sample_sums(phy_in)
# X138408 X138410 X138412 X138418 X138424 X138426 X138428 X138430 X138432 X138434 X138436 X138438 X138444 X138448 X138450 
# 55139   54772   66568   10338   43173   46582   32464   63169   59488   75842   49293   63018   10142   54430   26393 
# X138452 X138454 X138456 X138458 X138460 X138462 X138464 X138466 X138468 X138470 X138472 
# 53219  119783   79511  169761  117629  181862  163470  140472  157241  121369  140083 

summary(sample_sums(phy_in))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 10142   50274   63094   82893  120972  181862 


## Examine if species accumulation (rarefaction curves) are converging
## (i.e. reaching horizontal asymptote) or not?

psdata <- phy_in
set.seed(123)
rarefaction_curve_data <- calculate_rarefaction_curves(psdata, c('Observed'), sort(c(min(sample_sums(psdata)), 1, 10, 100, 1000, 1:182 * 1000)) )
# cover sequence read depths up to max value: 181862
rarefaction_curve_data


str( data.frame(sample_data(psdata)) )
row.names(psdata@sam_data)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438"
# [13] "X138444" "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468"
# [25] "X138470" "X138472"

## Add sample data
df <- merge(rarefaction_curve_data, data.frame(sample_data(psdata)), by.x = 'Sample', by.y = 'row.names')

sample_data(psdata)

# $Date.since.change.in.Land.Use

df$group <- df$year.of.rehab.or.ref
unique(df$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"
# sel <- which(df$group == "N/A")
# df$group[sel] # all N/A
# df$group[sel] <- "Ref"
# unique(df$group) # 
df$group <- factor(df$group, 
                   levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                   labels = c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


cols.group.iluka <- c("7 yr" = "#d53e4f","10 yr" = "#f46d43", "15 yr" = "#fdae61","19 yr" = "#e6f598",
                      "24 yr" = "#abdda4","30 yr" = "#66c2a5", "38 yr" = "#3288bd", "Ref" = "#5e4fa2")

cols <- cols.group.iluka

str(df)

summary(df$Alpha_diversity)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1    3856    5760    6176    8631   14018 

p <- ggplot(
  data = df,
  mapping = aes(x = Depth, y = Alpha_diversity, colour = group, group = Sample)) +
  theme_bw() +
  scale_colour_manual(values = cols, name  ="Rehab\nage") +
  ggtitle("(B) Eneabba") +
  labs(x = "ASV sequence depth", y = "Observed ASVs (count)") +
  geom_line() +
  geom_vline(xintercept = min(sample_sums(psdata)), linetype="dotted") +
  ylim(0,14100) +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )

p

ggsave(plot=p, filename = paste0("plots/","Rarefaction-curve-B-Iluka-Eneabba.tiff"), width = 12, height = 8, units = "cm", dpi = 600, compression = "lzw", type = "cairo")




### South32


phy_in <- phy.south32.withTree


sample_sums(phy_in)
# X138358 X138360 X138362 X138364 X138366 X138368 X138370 X138372 X138374 X138376 X138378 X138380 X138382 X138384 X138386 
# 100138   74378   98235   85503   84782   83191  115241   83599   79543   74763   99395   71359   81475   87939   98769 
# X138388 X138390 X138392 X138394 X138396 X138398 X138400 X138402 X138404 X138406 
# 89314   95344   72227   74906   71682   63268   54122   74677   59216   76559 

summary(sample_sums(phy_in))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 54122   74378   81475   81985   89314  115241 


## Examine if species accumulation (rarefaction curves) are converging
## (i.e. reaching horizontal asymptote) or not?

psdata <- phy_in
set.seed(123)
rarefaction_curve_data <- calculate_rarefaction_curves(psdata, c('Observed'), sort(c(min(sample_sums(psdata)), 1, 10, 100, 1000, 1:116 * 1000)) )
# cover sequence read depths up to max value: 115241 
rarefaction_curve_data


str( data.frame(sample_data(psdata)) )
row.names(psdata@sam_data)

## Add sample data
df <- merge(rarefaction_curve_data, data.frame(sample_data(psdata)), by.x = 'Sample', by.y = 'row.names')

sample_data(psdata)

# $Date.since.change.in.Land.Use

df$group <- df$year.of.rehab.or.ref
unique(df$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"
# sel <- which(df$group == "N/A")
# df$group[sel] # all N/A
# df$group[sel] <- "Ref"
# unique(df$group) # 
df$group <- factor(df$group, 
                   levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                   labels = c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


cols.group.south32 <- c("2 yr" = "#9e0142","8 yr" = "#d53e4f","12 yr" = "#f46d43", "14 yr" = "#fdae61",
                        "17 yr" = "#fee08b", "20 yr" = "#e6f598", "23 yr" = "#abdda4","28 yr" = "#66c2a5",
                        "Ref" = "#5e4fa2")

cols <- cols.group.south32

str(df)

summary(df$Alpha_diversity)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 1    6892   10688   10118   13786   20072 

p <- ggplot(
  data = df,
  mapping = aes(x = Depth, y = Alpha_diversity, colour = group, group = Sample)) +
  theme_bw() +
  scale_colour_manual(values = cols, name  ="Rehab\nage") +
  ggtitle("(C) Worsley") +
  labs(x = "ASV sequence depth", y = "Observed ASVs (count)") +
  geom_line() +
  geom_vline(xintercept = min(sample_sums(psdata)), linetype="dotted") +
  ylim(0,20100) +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )

p

ggsave(plot=p, filename = paste0("plots/","Rarefaction-curve-C-South32-Worsley.tiff"), width = 12, height = 8, units = "cm", dpi = 600, compression = "lzw", type = "cairo")

#-------------------------



#### facet_grid trajectory plots + predict time to reach ref:
##   ASV -- Jaccard + Bray-Curtis + Unweighted UniFrac + Weighted UniFrac -- Alcoa + Iluka-Eneabba + South32
#-------------------------

names(df_results)


names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"    
# [7] "group"          "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"  


keep_results <- c("Alcoa-Jaccard-ASV-withTree"     , "Alcoa-Bray-ASV-withTree" ,
                  "Alcoa-UnWtUniFrac-ASV-withTree"   , "Alcoa-WtUniFrac-ASV-withTree" ,
  
                  "Iluka-Eneabba-Jaccard-ASV-withTree" , "Iluka-Eneabba-Bray-ASV-withTree" ,
                  "Iluka-Eneabba-UnWtUniFrac-ASV-withTree" , "Iluka-Eneabba-WtUniFrac-ASV-withTree" ,
                  
                  "South32-Jaccard-ASV-withTree"    , "South32-Bray-ASV-withTree" ,      
                  "South32-UnWtUniFrac-ASV-withTree" , "South32-WtUniFrac-ASV-withTree" 
                  )


keep_cols <- c("samp","compare_with","similarity","group","dist_measure","tax_level","study","corrected","facet_x")
traj_data <- df_results[[ keep_results[1] ]][ ,keep_cols]
for (i in 2:length(keep_results)) {
  traj_data <- rbind(traj_data,df_results[[ keep_results[i] ]][ ,keep_cols])
  print(paste0("completed ",i))
}
dim(traj_data) # 3996    9
unique(traj_data$facet_x) # "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"        "Unweighted UniFrac\nASV-level" "Weighted UniFrac\nASV-level" 

keep_cols_anno <- c("group","sig_letter","similarity","dist_measure","tax_level","study","corrected","facet_x")
anno_data <- df_anno[[ keep_results[1] ]][ ,keep_cols_anno]
for (i in 2:length(keep_results)) {
  anno_data <- rbind(anno_data,df_anno[[ keep_results[i] ]][ ,keep_cols_anno])
  print(paste0("completed ",i))
}
dim(anno_data) # 96  8
class(anno_data$facet_x) # "character"
unique(anno_data$facet_x) # "character"
# [1] "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"        "Unweighted UniFrac\nASV-level" "Weighted UniFrac\nASV-level"  
anno_data$facet_x <- factor(anno_data$facet_x,
                            levels = c("Jaccard\nASV-level","Bray-Curtis\nASV-level",
                                       "Unweighted UniFrac\nASV-level","Weighted UniFrac\nASV-level"),
                            ordered = TRUE)

class(traj_data$group) # "ordered" "factor" 
unique(traj_data$facet_x)
# [1] "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"        "Unweighted UniFrac\nASV-level" "Weighted UniFrac\nASV-level"  
# [5] "Weighted Unifrac\nASV-level" 
class(traj_data$facet_x) # "character"
traj_data$facet_x <- factor(traj_data$facet_x,
                            levels = c("Jaccard\nASV-level","Bray-Curtis\nASV-level",
                                       "Unweighted UniFrac\nASV-level","Weighted UniFrac\nASV-level"),
                            ordered = TRUE)
class(traj_data$group)
# "ordered" "factor" 

traj_data$group <- factor(traj_data$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)

levels(traj_data$group)
# [1] "0"   "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14" 
# [16] "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29" 
# [31] "30"  "31"  "32"  "33"  "34"  "35"  "36"  "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44" 
# [46] "45"  "Ref"

length(levels(traj_data$group)) #47

x_labels <- c("0","","","","","","","","","","10",
              "","","","","","","","","","20",
              "","","","","","","","","","30",
              "","","","","","","","","","40",
              "","","","","","Ref")
length(x_labels)
# 47
x_ticks <- c(1,0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,1)
length(x_ticks)
#47


unique(traj_data$study) # "Alcoa"         "Iluka-Eneabba" "South32" 

unique(anno_data$study) # "Alcoa"         "Iluka-Eneabba" "South32"





## median reference horizontal lines ...

href_lines <- data.frame(
  study  = rep( c("Alcoa", "Iluka-Eneabba", "South32"), times = 4 ),
  facet_x  = rep( c("Jaccard\nASV-level","Bray-Curtis\nASV-level",
                    "Unweighted UniFrac\nASV-level","Weighted UniFrac\nASV-level"), each = 3),
  similarity= NA
)

x <- traj_data

for (i in 1:dim(href_lines)[1]) {
  #i<-1
  sel <- which(x$study == href_lines$study[i] & x$facet_x == href_lines$facet_x[i] & x$group == "Ref")
  href_lines$similarity[i] <- median( x$similarity[sel] )
}
href_lines$facet_x <- factor(href_lines$facet_x,
                             levels = c("Jaccard\nASV-level","Bray-Curtis\nASV-level",
                                        "Unweighted UniFrac\nASV-level","Weighted UniFrac\nASV-level"),
                             ordered = TRUE)


## fit logarithmic models for each data type (facet_x)



## BOOTSTRAP MODELS ...

pred_models_boot <- data.frame(study=NA, # one of: "Alcoa", "Iluka-Eneabba", "South32"
                                  facet_x=NA, # one of: c("Jaccard\nASV-level","Bray-Curtis\nASV-level", "Jaccard\nGenus-level","Bray-Curtis\nGenus-level")
                                  study_and_metric=NA, # combine the two values above
                                  group=NA,
                                  group_num=NA,
                                  similarity=NA,
                                  boot_no=NA,                 # EXTRA
                                  ref_median=NA,              # EXTRA
                                  pred_time_to_ref_median=NA, # EXTRA
                                  pred_type=NA)     # c("rectangular hyperbola", "logarithmic")
pred_models_boot.temp <- pred_models_boot

studies <- c("Alcoa", "Iluka-Eneabba", "South32")
metrics <- c("Jaccard\nASV-level","Bray-Curtis\nASV-level",
             "Unweighted UniFrac\nASV-level","Weighted UniFrac\nASV-level")
pred_types <- c("rectangular hyperbola", "negative exponential", "logarithmic")
k<-3

## run logarithmic model only !!!

x <- traj_data
dim(x) # 3996    9

for (i in 1:length(studies)) {
  #i<-1
  this_study <- studies[i]
  for (j in 1:length(metrics)) {
    #j<-3
    this_metric <- metrics[j]
    
    # prepare data for modelling
    subsel <- which(x$facet_x == this_metric & x$study==this_study)
    newdat <- x[subsel, c("similarity","group") ]
    newdat$group_num <- as.numeric(as.character(newdat$group))
    # remove 'Ref' similarities which have now been converted to NA
    sel.na <- which(is.na(newdat$group_num))
    
    this_ref_median <- median(newdat$similarity[sel.na])
    
    newdat <- newdat[-sel.na, ]
    
    # prep output dataframe for prediction model to populate
    sel.dup <- which(duplicated(newdat$group)==TRUE)
    new_preds <- data.frame(matrix( nrow=dim(newdat[-sel.dup, ])[1], ncol = length(names(pred_models_boot.temp)) ))
    names(new_preds) <- names(pred_models_boot.temp)
    new_preds$group_num <- sort( newdat$group_num[-sel.dup] )
    new_preds$study <- this_study
    new_preds$facet_x <- this_metric
    new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$facet_x)
    new_preds$group <- as.character(new_preds$group_num)
    new_preds$ref_median <- this_ref_median
    
    #for (k in 1:length(pred_types)) {
    
    this_pred_model_type <- pred_types[k]
    
    ## NOTE:
    #  rectangular hyperbola ( fct = MM.2() from ‘drc' package; https://www.statforbiology.com/nonlinearregression/usefulequations#michaelis-menten_equation), and 
    #  negative exponential ( fct = DRC.negExp() from 'aomisc' package; https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function)
    #  models were trialled however these regularly failed to achieve model fits.
    
    
    # NOW using "logarithmic" model only !!!!!
    
    # negative exponential model, using fct = DRC.negExp() from 'aomisc' package
    # https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function
    
    for (b in 1:100) {
      #b<-1
      new_preds$boot_no <- b
      set.seed(b + 1234)
      sel.boot <- sample(x = c(1:dim(newdat)[1]), size = dim(newdat)[1], replace = TRUE)
      
      bootdat <- newdat[sel.boot, ]
      
      # exclude outlying data points from trajectory data-cloud ... 
      # as outliers are likely to throw out model-fitting
      bp <- boxplot(bootdat$similarity, print=FALSE)
      sel.outlier <- which(bootdat$similarity %in% bp$out)
      if (length(sel.outlier)>0) {
        bootdat <- bootdat[-sel.outlier, ]
      }
      
      #plot(bootdat$group_num, bootdat$similarity)
      
      model <- drm(similarity ~ group_num , fct = DRC.logCurve(), # DRC.negExp()
                   data = bootdat)
      
      new_preds$similarity <- predict(model, new_preds )
      new_preds$pred_type <- this_pred_model_type
      
      y <- data.frame(group_num=1:500, sim_pred=NA)
      y$sim_pred <- predict(model, y )
      y$reached_target <- y$sim_pred >= this_ref_median
      #plot(y$group_num, y$sim_pred)
      row.names(y)
      
      sel <- which(y$reached_target == TRUE)
      
      if (length(sel) > 0) {
        years_to_ref <- min(sel)
        } else {
          years_to_ref <- NA
        }
      new_preds$pred_time_to_ref_median <- years_to_ref
        
      # add predictions from end of last year to 40 years ...
      max_year <- max(new_preds$group_num)
      year <- max_year + 1
      while (year < 40) {
        new_preds <- rbind(new_preds, new_preds[ dim(new_preds)[1], ])
        year <- year + 1
        new_preds$group_num[dim(new_preds)[1]] <- year
        new_preds$group[dim(new_preds)[1]] <- as.character( new_preds$group_num[dim(new_preds)[1]] )
        new_preds$similarity[dim(new_preds)[1]] <- y[ year, "sim_pred"]
      }
      
      pred_models_boot <- rbind(pred_models_boot, new_preds)
          
      }# END bootstrap
    
    print(paste0("completed study: ",studies[i],"; metric: ",metrics[j],"; pred_type: ",pred_types[k]))
    
    #} # END pred_types
  } # END metrics
} # END studies


# remove NA first row
pred_models_boot <- pred_models_boot[-1, ]

unique(pred_models_boot$facet_x)
pred_models_boot$facet_x <- factor(pred_models_boot$facet_x,
                                      levels = c("Jaccard\nASV-level", "Bray-Curtis\nASV-level", 
                                                 "Unweighted UniFrac\nASV-level", "Weighted UniFrac\nASV-level"),
                                      ordered = TRUE)

dim(pred_models_boot) # 17200    10


# correct time prediction to ref; update NA to (>) 500 yrs
sel <- which(is.na(pred_models_boot$pred_time_to_ref_median))
pred_models_boot$pred_time_to_ref_median[sel] <- 500


pred_models_boot$group <- factor(pred_models_boot$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)
str(pred_models_boot)
# 'data.frame':	17200 obs. of  10 variables:
#   $ study                  : chr  "Alcoa" "Alcoa" "Alcoa" "Alcoa" ...
# $ facet_x                : Ord.factor w/ 4 levels "Jaccard\nASV-level"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ study_and_metric       : chr  "Alcoa__Jaccard\nASV-level" "Alcoa__Jaccard\nASV-level" "Alcoa__Jaccard\nASV-level" "Alcoa__Jaccard\nASV-level" ...
# $ group                  : Ord.factor w/ 47 levels "0"<"1"<"2"<"3"<..: 3 9 15 18 26 30 32 33 34 35 ...
# $ group_num              : num  2 8 14 17 25 29 31 32 33 34 ...
# $ similarity             : num  8.82 13.39 15.24 15.88 17.15 ...
# $ boot_no                : int  1 1 1 1 1 1 1 1 1 1 ...
# $ ref_median             : num  20.3 20.3 20.3 20.3 20.3 ...
# $ pred_time_to_ref_median: num  66 66 66 66 66 66 66 66 66 66 ...
# $ pred_type              : chr  "logarithmic" "logarithmic" "logarithmic" "logarithmic" ...

str(traj_data)
# 'data.frame':	3996 obs. of  9 variables:
#   $ samp        : chr  "X39041" "X39041" "X39041" "X39041" ...
# $ compare_with: chr  "X39043" "X39047" "X39051" "X39055" ...
# $ similarity  : num  11.2 11.9 11.6 10.2 12.1 ...
# $ group       : Ord.factor w/ 47 levels "0"<"1"<"2"<"3"<..: 3 3 3 3 3 3 3 3 3 3 ...
# $ dist_measure: chr  "Jaccard" "Jaccard" "Jaccard" "Jaccard" ...
# $ tax_level   : chr  "ASV (with Tree)" "ASV (with Tree)" "ASV (with Tree)" "ASV (with Tree)" ...
# $ study       : chr  "Alcoa" "Alcoa" "Alcoa" "Alcoa" ...
# $ corrected   : chr  "no" "no" "no" "no" ...
# $ facet_x     : Ord.factor w/ 4 levels "Jaccard\nASV-level"<..: 1 1 1 1 1 1 1 1 1 1 ...

temp.traj_data <- traj_data
temp.pred_models_boot <- pred_models_boot

unique(traj_data$study) # "Alcoa"         "Iluka-Eneabba" "South32" 


study_names <- c(
  Alcoa="Huntly",
  `Iluka-Eneabba`="Eneabba",
  South32="Worsley"
)



p <- ggplot(data=traj_data, aes(x=group, y=similarity)) +
  geom_line(data=pred_models_boot, aes(  x=group, y=similarity, group = boot_no ), colour = "#e34a33", alpha = 0.02) + # "red"
  geom_hline(data = href_lines, aes(yintercept = similarity),  color= "#2c7fb8", linetype="longdash") + # "blue"
  
  geom_boxplot(outlier.shape = NA, width = 2) +
  #geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  
  theme_bw() +
  geom_text(data=anno_data, aes(x=group, y=similarity, label=sig_letter),nudge_y = 2, size=2.75) +
  scale_x_discrete(drop=FALSE, expand = c(0.05,0.05),
                   labels = x_labels) +
  theme(axis.text.x  = element_text(angle=90,hjust=1, vjust=0.5), # 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_line(linetype = 0)
        axis.ticks.x = element_line(linetype = x_ticks),
        strip.text.x = element_text(size = rel(1.1)),
        strip.text.y = element_text(size = rel(1.1))
  ) +
  labs(x = "Rehabilitation age (years)", y = "Similarity to Reference (%)") +
  facet_grid(rows = vars(study), cols = vars(facet_x), scales = "fixed", labeller = labeller(study = study_names) ) # study_labeller   as_labeller(study_names)

p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Restoration-trajectories--AlcoaHuntly-IlukaEneabba-South32Worsley--ASV-Jaccard-Bray-UnWt-Wt-UniFrac-with-Preds.tiff"), width = 18, height = 16, units = "cm", dpi = 600, compression = "lzw",type="cairo")



str(traj_data)
unique(traj_data$study) # "Alcoa"         "Iluka-Eneabba" "South32"
unique(traj_data$group) # use "ref" only
unique(traj_data$facet_x)
# [1] Jaccard\nASV-level            Bray-Curtis\nASV-level        Unweighted UniFrac\nASV-level Weighted UniFrac\nASV-level 

## Huntly-Alcoa
sel <- which(traj_data$study == "Alcoa")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Jaccard\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 9.503  17.824  20.345  20.092  23.213  35.809
sel <- which(traj_data$study == "Alcoa")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Bray-Curtis\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 22.38   31.28   35.55   35.79   39.93   60.91 
sel <- which(traj_data$study == "Alcoa")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Unweighted UniFrac\nASV-level")
summary(traj_data$similarity[sel[subsel]])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 30.86   51.60   54.09   52.11   56.19   65.95 
sel <- which(traj_data$study == "Alcoa")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Weighted UniFrac\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 90.62   94.60   95.37   95.18   96.08   97.92 

## "Iluka-Eneabba"
sel <- which(traj_data$study == "Iluka-Eneabba")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Jaccard\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 12.14   16.43   18.59   18.42   20.54   24.26 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Bray-Curtis\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 21.08   26.33   29.82   29.45   32.40   39.76 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Unweighted UniFrac\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 45.03   48.69   50.91   50.75   52.73   55.81 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Weighted UniFrac\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 83.21   88.85   89.89   89.94   91.84   94.25 

## "South32"
sel <- which(traj_data$study == "South32")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Jaccard\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 17.25   20.16   25.72   24.56   28.41   31.25 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Bray-Curtis\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 21.39   26.45   36.02   34.32   39.96   46.11 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Unweighted UniFrac\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 51.53   54.41   56.62   56.52   58.67   60.32 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Weighted UniFrac\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 91.95   93.63   94.53   94.58   95.91   96.85 


str(pred_models_boot)

unique(pred_models_boot$study_and_metric)
# [1] "Alcoa__Jaccard\nASV-level"                    "Alcoa__Bray-Curtis\nASV-level"                "Alcoa__Unweighted UniFrac\nASV-level"        
# [4] "Alcoa__Weighted UniFrac\nASV-level"           "Iluka-Eneabba__Jaccard\nASV-level"            "Iluka-Eneabba__Bray-Curtis\nASV-level"       
# [7] "Iluka-Eneabba__Unweighted UniFrac\nASV-level" "Iluka-Eneabba__Weighted UniFrac\nASV-level"   "South32__Jaccard\nASV-level"                 
# [10] "South32__Bray-Curtis\nASV-level"              "South32__Unweighted UniFrac\nASV-level"       "South32__Weighted UniFrac\nASV-level"  

pred_models_boot$study_and_metric <- factor(pred_models_boot$study_and_metric,
                                            levels = c(
                                              "Alcoa__Jaccard\nASV-level"                   , "Alcoa__Bray-Curtis\nASV-level"               , "Alcoa__Unweighted UniFrac\nASV-level"  ,      
                                              "Alcoa__Weighted UniFrac\nASV-level"          , "Iluka-Eneabba__Jaccard\nASV-level"           , "Iluka-Eneabba__Bray-Curtis\nASV-level"  ,     
                                              "Iluka-Eneabba__Unweighted UniFrac\nASV-level", "Iluka-Eneabba__Weighted UniFrac\nASV-level"  , "South32__Jaccard\nASV-level"             ,    
                                              "South32__Bray-Curtis\nASV-level"             , "South32__Unweighted UniFrac\nASV-level"      , "South32__Weighted UniFrac\nASV-level"  
                                            ))

dim(pred_models_boot) # 17200    10

dat.sum <- pred_models_boot
head(dat.sum)
# study            facet_x          study_and_metric group group_num similarity boot_no ref_median pred_time_to_ref_median
# 11 Alcoa Jaccard\nASV-level Alcoa__Jaccard\nASV-level     2         2   8.819282       1   20.34483                      66
# 2  Alcoa Jaccard\nASV-level Alcoa__Jaccard\nASV-level     8         8  13.392202       1   20.34483                      66
# 3  Alcoa Jaccard\nASV-level Alcoa__Jaccard\nASV-level    14        14  15.238187       1   20.34483                      66
# 4  Alcoa Jaccard\nASV-level Alcoa__Jaccard\nASV-level    17        17  15.878643       1   20.34483                      66
# 5  Alcoa Jaccard\nASV-level Alcoa__Jaccard\nASV-level    25        25  17.150814       1   20.34483                      66
# 6  Alcoa Jaccard\nASV-level Alcoa__Jaccard\nASV-level    29        29  17.640402       1   20.34483                      66
# pred_type
# 11 logarithmic
# 2  logarithmic
# 3  logarithmic
# 4  logarithmic
# 5  logarithmic
# 6  logarithmic

for (i in 1:length(levels(dat.sum$study_and_metric))) {
  #i<-1
  this_study_and_metric <- levels(dat.sum$study_and_metric)[i]
  sel <- which(dat.sum$study_and_metric == this_study_and_metric)
  #unique(dat.sum$ref_median[sel])
  temp <- dat.sum$pred_time_to_ref_median[sel]
  #quantile(temp, probs = c(0.05, 0.25, 0.5, 0.75, 0.95), na.rm = TRUE)
  #hist(temp)
  res <- quantile(temp, probs = c(0.05, 0.5, 0.95), na.rm = TRUE)
  names(res)
  #"5%"  "50%" "95%"
  #print(paste0(this_study_and_metric))
  print(paste0(this_study_and_metric,": ",res["50%"]," (",res["5%"],", ",res["95%"],")"))
}
# [1] "Alcoa__Jaccard\nASV-level: 60.5 (51, 79)"
# [1] "Alcoa__Bray-Curtis\nASV-level: 43 (39, 49)"
# [1] "Alcoa__Unweighted UniFrac\nASV-level: 121.5 (81.9, 237.549999999999)"
# [1] "Alcoa__Weighted UniFrac\nASV-level: 42 (38.95, 49)"
# [1] "Iluka-Eneabba__Jaccard\nASV-level: 97 (73.9, 133.45)"
# [1] "Iluka-Eneabba__Bray-Curtis\nASV-level: 93.5 (69, 131.35)"
# [1] "Iluka-Eneabba__Unweighted UniFrac\nASV-level: 65 (49.95, 118.05)"
# [1] "Iluka-Eneabba__Weighted UniFrac\nASV-level: 69.5 (47, 153.299999999999)"
# [1] "South32__Jaccard\nASV-level: 131.5 (68.8, 313.25)"
# [1] "South32__Bray-Curtis\nASV-level: 103.5 (68.9, 173.549999999999)"
# [1] "South32__Unweighted UniFrac\nASV-level: 60 (44, 91.3499999999997)"
# [1] "South32__Weighted UniFrac\nASV-level: 29 (24, 38)"


levels( pred_models_boot$facet_x )
# "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"        "Unweighted UniFrac\nASV-level" "Weighted UniFrac\nASV-level"  




## plot time to target

names(pred_models_boot)
# [1] "study"                   "facet_x"                 "study_and_metric"        "group"                   "group_num"              
# [6] "similarity"              "boot_no"                 "ref_median"              "pred_time_to_ref_median" "pred_type"



df <- pred_models_boot
unique(df$study) # "Alcoa"         "Iluka-Eneabba" "South32"  
df$study <- factor(df$study,
                   levels = c("Alcoa", "Iluka-Eneabba", "South32" ),
                   labels = c("Huntly", "Eneabba", "Worsley"), ordered = TRUE)
df$facet_x
# Levels: Jaccard\nASV-level < Bray-Curtis\nASV-level < Unweighted UniFrac\nASV-level < Weighted UniFrac\nASV-level
df$facet_x <- factor(df$facet_x, 
                     levels = c("Jaccard\nASV-level",
                                "Bray-Curtis\nASV-level",
                                "Unweighted UniFrac\nASV-level",
                                "Weighted UniFrac\nASV-level"),
                     labels = c("Jaccard",
                                "Bray-Curtis",
                                "Unweighted\nUniFrac",
                                "Weighted\nUniFrac"), ordered = TRUE)

p <- ggplot(data=df, aes(x=facet_x, y=pred_time_to_ref_median)) +  # , color = facet_x
  geom_boxplot() +
  theme_bw() +
  facet_wrap(vars(study) , nrow=1) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = rel(1.1))
  ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Predicted-time-to-target--Alcoa-Iluka-South32--ASV-Jaccard-Bray-UnWtUniFrac-WtUniFrac-facet-study.tiff"), width = 16, height = 8, units = "cm", dpi = 600, compression = "lzw",type="cairo")


#-------------------------


#### facet_grid trajectory plots + predict time to reach ref:
##   ASV -- Jaccard + Bray-Curtis -- Iluka-Eneabba only - with & without youngest rehab_age group
#-------------------------

names(df_results)


names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"    
# [7] "group"          "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"  


keep_results <- c("Iluka-Eneabba-Jaccard-ASV-withTree" ,           "Iluka-Eneabba-Bray-ASV-withTree" ,
                  "Iluka-Eneabba-Jaccard-ASV-withTree-exclude7yr", "Iluka-Eneabba-Bray-ASV-withTree-exclude7yr"
                  )


keep_cols <- c("samp","compare_with","similarity","group","dist_measure","tax_level","study","corrected","facet_x")
traj_data <- df_results[[ keep_results[1] ]][ ,keep_cols]
for (i in 2:length(keep_results)) {
  traj_data <- rbind(traj_data,df_results[[ keep_results[i] ]][ ,keep_cols])
  print(paste0("completed ",i))
}
dim(traj_data) # 864    9
unique(traj_data$facet_x) # "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"        "Unweighted UniFrac\nASV-level" "Weighted UniFrac\nASV-level" 

keep_cols_anno <- c("group","sig_letter","similarity","dist_measure","tax_level","study","corrected","facet_x")
anno_data <- df_anno[[ keep_results[1] ]][ ,keep_cols_anno]
for (i in 2:length(keep_results)) {
  anno_data <- rbind(anno_data,df_anno[[ keep_results[i] ]][ ,keep_cols_anno])
  print(paste0("completed ",i))
}
dim(anno_data) # 30  8
class(anno_data$facet_x) # "character"
unique(anno_data$facet_x) # "character"
# [1] "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"
anno_data$facet_x <- factor(anno_data$facet_x,
                            levels = c("Jaccard\nASV-level","Bray-Curtis\nASV-level"),
                            ordered = TRUE)

unique(traj_data$facet_x)
# [1] "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"
class(traj_data$facet_x) # "character"
traj_data$facet_x <- factor(traj_data$facet_x,
                            levels = c("Jaccard\nASV-level","Bray-Curtis\nASV-level"),
                            ordered = TRUE)

unique(traj_data$study)
# [1] "Iluka-Eneabba"                         "Iluka-Eneabba\n(exclude 7-year rehab)"
class(traj_data$study) # "character"
traj_data$study <- factor(traj_data$study,
                            levels = c("Iluka-Eneabba", "Iluka-Eneabba\n(exclude 7-year rehab)"),
                            ordered = TRUE)

unique(anno_data$study)
# [1] "Iluka-Eneabba"                         "Iluka-Eneabba\n(exclude 7-year rehab)"
class(anno_data$study) # "character"
anno_data$study <- factor(anno_data$study,
                          levels = c("Iluka-Eneabba", "Iluka-Eneabba\n(exclude 7-year rehab)"),
                          ordered = TRUE)

class(traj_data$group)
# "ordered" "factor" 

traj_data$group <- factor(traj_data$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)

levels(traj_data$group)
# [1] "0"   "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14" 
# [16] "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29" 
# [31] "30"  "31"  "32"  "33"  "34"  "35"  "36"  "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44" 
# [46] "45"  "Ref"

length(levels(traj_data$group)) #47

x_labels <- c("0","","","","","","","","","","10",
              "","","","","","","","","","20",
              "","","","","","","","","","30",
              "","","","","","","","","","40",
              "","","","","","Ref")
length(x_labels)
# 47
x_ticks <- c(1,0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,1)
length(x_ticks)
#47


unique(traj_data$study) # Iluka-Eneabba                         Iluka-Eneabba\n(exclude 7-year rehab)

unique(anno_data$study) # "Iluka-Eneabba"                         "Iluka-Eneabba\n(exclude 7-year rehab)"





## median reference horizontal lines ...

href_lines <- data.frame(
  study  = rep( c("Iluka-Eneabba", "Iluka-Eneabba\n(exclude 7-year rehab)"), times = 2 ),
  facet_x  = rep( c("Jaccard\nASV-level","Bray-Curtis\nASV-level"), each = 2),
  similarity= NA
)

x <- traj_data

for (i in 1:dim(href_lines)[1]) {
  #i<-1
  sel <- which(x$study == href_lines$study[i] & x$facet_x == href_lines$facet_x[i] & x$group == "Ref")
  href_lines$similarity[i] <- median( x$similarity[sel] )
}
href_lines$facet_x <- factor(href_lines$facet_x,
                             levels = c("Jaccard\nASV-level","Bray-Curtis\nASV-level"),
                             ordered = TRUE)


## fit logarithmic models for each data type (facet_x)



## BOOTSTRAP MODELS ...

pred_models_boot <- data.frame(study=NA, # one of: "Alcoa", "Iluka-Eneabba", "South32"
                               facet_x=NA, # one of: c("Jaccard\nASV-level","Bray-Curtis\nASV-level", "Jaccard\nGenus-level","Bray-Curtis\nGenus-level")
                               study_and_metric=NA, # combine the two values above
                               group=NA,
                               group_num=NA,
                               similarity=NA,
                               boot_no=NA,                 # EXTRA
                               ref_median=NA,              # EXTRA
                               pred_time_to_ref_median=NA, # EXTRA
                               pred_type=NA)     # c("rectangular hyperbola", "logarithmic")
pred_models_boot.temp <- pred_models_boot

studies <- c("Iluka-Eneabba", "Iluka-Eneabba\n(exclude 7-year rehab)")
metrics <- c("Jaccard\nASV-level","Bray-Curtis\nASV-level")
pred_types <- c("rectangular hyperbola", "negative exponential", "logarithmic")
k<-3

## run logarithmic model only !!!

x <- traj_data
dim(x) # 864    9

for (i in 1:length(studies)) {
  #i<-1
  this_study <- studies[i]
  for (j in 1:length(metrics)) {
    #j<-3
    this_metric <- metrics[j]
    
    # prepare data for modelling
    subsel <- which(x$facet_x == this_metric & x$study==this_study)
    newdat <- x[subsel, c("similarity","group") ]
    newdat$group_num <- as.numeric(as.character(newdat$group))
    # remove 'Ref' similarities which have now been converted to NA
    sel.na <- which(is.na(newdat$group_num))
    
    this_ref_median <- median(newdat$similarity[sel.na])
    
    newdat <- newdat[-sel.na, ]
    
    # prep output dataframe for prediction model to populate
    sel.dup <- which(duplicated(newdat$group)==TRUE)
    new_preds <- data.frame(matrix( nrow=dim(newdat[-sel.dup, ])[1], ncol = length(names(pred_models_boot.temp)) ))
    names(new_preds) <- names(pred_models_boot.temp)
    new_preds$group_num <- sort( newdat$group_num[-sel.dup] )
    new_preds$study <- this_study
    new_preds$facet_x <- this_metric
    new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$facet_x)
    new_preds$group <- as.character(new_preds$group_num)
    new_preds$ref_median <- this_ref_median
    
    #for (k in 1:length(pred_types)) {
    
    
    this_pred_model_type <- pred_types[k]
    
    ## NOTE:
    #  rectangular hyperbola ( fct = MM.2() from ‘drc' package; https://www.statforbiology.com/nonlinearregression/usefulequations#michaelis-menten_equation), and 
    #  negative exponential ( fct = DRC.negExp() from 'aomisc' package; https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function)
    #  models were trialled however these regularly failed to achieve model fits.
    
    
    # NOW using "logarithmic" model only !!!!!
    
    # negative exponential model, using fct = DRC.negExp() from 'aomisc' package
    # https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function
    
    
    for (b in 1:100) {
      #b<-1
      new_preds$boot_no <- b
      set.seed(b + 1234)
      sel.boot <- sample(x = c(1:dim(newdat)[1]), size = dim(newdat)[1], replace = TRUE)
      
      bootdat <- newdat[sel.boot, ]
      
      # exclude outlying data points from trajectory data-cloud ... 
      # as outliers are likely to throw out model-fitting
      bp <- boxplot(bootdat$similarity, print=FALSE)
      sel.outlier <- which(bootdat$similarity %in% bp$out)
      if (length(sel.outlier)>0) {
        bootdat <- bootdat[-sel.outlier, ]
      }
      
      #plot(bootdat$group_num, bootdat$similarity)
      
      model <- drm(similarity ~ group_num , fct = DRC.logCurve(), # DRC.negExp()
                   data = bootdat)
      
      new_preds$similarity <- predict(model, new_preds )
      new_preds$pred_type <- this_pred_model_type
      
      y <- data.frame(group_num=1:500, sim_pred=NA)
      y$sim_pred <- predict(model, y )
      y$reached_target <- y$sim_pred >= this_ref_median
      #plot(y$group_num, y$sim_pred)
      row.names(y)
      
      sel <- which(y$reached_target == TRUE)
      
      if (length(sel) > 0) {
        years_to_ref <- min(sel)
      } else {
        years_to_ref <- NA
      }
      new_preds$pred_time_to_ref_median <- years_to_ref
      
      # add predictions from end of last year to 40 years ...
      max_year <- max(new_preds$group_num)
      year <- max_year + 1
      while (year < 40) {
        new_preds <- rbind(new_preds, new_preds[ dim(new_preds)[1], ])
        year <- year + 1
        new_preds$group_num[dim(new_preds)[1]] <- year
        new_preds$group[dim(new_preds)[1]] <- as.character( new_preds$group_num[dim(new_preds)[1]] )
        new_preds$similarity[dim(new_preds)[1]] <- y[ year, "sim_pred"]
      }
      
      pred_models_boot <- rbind(pred_models_boot, new_preds)
      
    } # END bootstrap
    
    print(paste0("completed study: ",studies[i],"; metric: ",metrics[j],"; pred_type: ",pred_types[k]))  
    #} # END pred_types
  } # END metrics
} # END studies


# remove NA first row
pred_models_boot <- pred_models_boot[-1, ]

unique(pred_models_boot$facet_x)
pred_models_boot$facet_x <- factor(pred_models_boot$facet_x,
                                   levels = c("Jaccard\nASV-level", "Bray-Curtis\nASV-level"),
                                   ordered = TRUE)

unique(pred_models_boot$study)
pred_models_boot$study <- factor(pred_models_boot$study,
                                   levels = c("Iluka-Eneabba", "Iluka-Eneabba\n(exclude 7-year rehab)"),
                                   ordered = TRUE)

dim(pred_models_boot) # 3000    10


# correct time prediction to ref; update NA to (>) 500 yrs
sel <- which(is.na(pred_models_boot$pred_time_to_ref_median))
pred_models_boot$pred_time_to_ref_median[sel] <- 500


pred_models_boot$group <- factor(pred_models_boot$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)
str(pred_models_boot)


str(traj_data)


temp.traj_data <- traj_data
temp.pred_models_boot <- pred_models_boot

unique(traj_data$study) #  Iluka-Eneabba                         Iluka-Eneabba\n(exclude 7-year rehab)

study_names <- c(
  `Iluka-Eneabba`="Eneabba",
  `Iluka-Eneabba\n(exclude 7-year rehab)`="Eneabba\n(exclude 7-year)"
)



p <- ggplot(data=traj_data, aes(x=group, y=similarity)) +
  geom_line(data=pred_models_boot, aes(  x=group, y=similarity, group = boot_no ), colour = "#e34a33", alpha = 0.02) + # "red"
  geom_hline(data = href_lines, aes(yintercept = similarity),  color= "#2c7fb8", linetype="longdash") + # "blue"
  #geom_line(data= pred_models_alldata, aes(  x=group, y=similarity, group = 1), colour = "#e34a33" ) + # "red"
  
  geom_boxplot(outlier.shape = NA, width = 2) +
  #geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  
  theme_bw() +
  geom_text(data=anno_data, aes(x=group, y=similarity, label=sig_letter),nudge_y = 2, size=2.75) +
  scale_x_discrete(drop=FALSE, expand = c(0.05,0.05),
                   labels = x_labels) +
  theme(axis.text.x  = element_text(angle=90,hjust=1, vjust=0.5), # 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_line(linetype = 0)
        axis.ticks.x = element_line(linetype = x_ticks),
        strip.text.x = element_text(size = rel(1.1)),
        strip.text.y = element_text(size = rel(1.1))
  ) +
  labs(x = "Rehabilitation age (years)", y = "Similarity to Reference (%)") +
  facet_grid(rows = vars(study), cols = vars(facet_x), scales = "fixed", labeller = labeller(study = study_names) ) # study_labeller   as_labeller(study_names)

p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Restoration-trajectories--IlukaEneabba--ASV-Jaccard-Bray-with-Preds-Exclude-7yr.tiff"), width = 11, height = 10, units = "cm", dpi = 600, compression = "lzw",type="cairo")



str(traj_data)
unique(traj_data$study) # Iluka-Eneabba                         Iluka-Eneabba\n(exclude 7-year rehab)
unique(traj_data$group) # use "ref" only
unique(traj_data$facet_x)
#  Jaccard\nASV-level     Bray-Curtis\nASV-level


## "Iluka-Eneabba"
sel <- which(traj_data$study == "Iluka-Eneabba")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Jaccard\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 12.14   16.43   18.59   18.42   20.54   24.26 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Bray-Curtis\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 21.08   26.33   29.82   29.45   32.40   39.76 



str(pred_models_boot)

unique(pred_models_boot$study_and_metric)
# [1] "Iluka-Eneabba__Jaccard\nASV-level"                             "Iluka-Eneabba__Bray-Curtis\nASV-level"                        
# [3] "Iluka-Eneabba\n(exclude 7-year rehab)__Jaccard\nASV-level"     "Iluka-Eneabba\n(exclude 7-year rehab)__Bray-Curtis\nASV-level" 

pred_models_boot$study_and_metric <- factor(pred_models_boot$study_and_metric,
                                            levels = c(
                                              "Iluka-Eneabba__Jaccard\nASV-level"                         ,    "Iluka-Eneabba__Bray-Curtis\nASV-level"                  ,      
                                              "Iluka-Eneabba\n(exclude 7-year rehab)__Jaccard\nASV-level" ,    "Iluka-Eneabba\n(exclude 7-year rehab)__Bray-Curtis\nASV-level"
                                            ))

dim(pred_models_boot) # 3000    10

dat.sum <- pred_models_boot
head(dat.sum)
# study            facet_x                  study_and_metric group group_num similarity boot_no ref_median pred_time_to_ref_median
# 11 Iluka-Eneabba Jaccard\nASV-level Iluka-Eneabba__Jaccard\nASV-level     7         7   10.09930       1   18.59137                     115
# 2  Iluka-Eneabba Jaccard\nASV-level Iluka-Eneabba__Jaccard\nASV-level    10        10   11.18426       1   18.59137                     115
# 3  Iluka-Eneabba Jaccard\nASV-level Iluka-Eneabba__Jaccard\nASV-level    15        15   12.41765       1   18.59137                     115
# 4  Iluka-Eneabba Jaccard\nASV-level Iluka-Eneabba__Jaccard\nASV-level    19        19   13.13672       1   18.59137                     115
# 5  Iluka-Eneabba Jaccard\nASV-level Iluka-Eneabba__Jaccard\nASV-level    24        24   13.84735       1   18.59137                     115
# 6  Iluka-Eneabba Jaccard\nASV-level Iluka-Eneabba__Jaccard\nASV-level    30        30   14.52613       1   18.59137                     115
# pred_type
# 11 logarithmic
# 2  logarithmic
# 3  logarithmic
# 4  logarithmic
# 5  logarithmic
# 6  logarithmic

for (i in 1:length(levels(dat.sum$study_and_metric))) {
  #i<-1
  this_study_and_metric <- levels(dat.sum$study_and_metric)[i]
  sel <- which(dat.sum$study_and_metric == this_study_and_metric)
  #unique(dat.sum$ref_median[sel])
  temp <- dat.sum$pred_time_to_ref_median[sel]
  #quantile(temp, probs = c(0.05, 0.25, 0.5, 0.75, 0.95), na.rm = TRUE)
  #hist(temp)
  res <- quantile(temp, probs = c(0.05, 0.5, 0.95), na.rm = TRUE)
  names(res)
  #"5%"  "50%" "95%"
  #print(paste0(this_study_and_metric))
  print(paste0(this_study_and_metric,": ",res["50%"]," (",res["5%"],", ",res["95%"],")"))
}
# [1] "Iluka-Eneabba__Jaccard\nASV-level: 97 (73.9, 133.45)"
# [1] "Iluka-Eneabba__Bray-Curtis\nASV-level: 93.5 (69, 131.35)"
# [1] "Iluka-Eneabba\n(exclude 7-year rehab)__Jaccard\nASV-level: 61 (54, 71.05)"
# [1] "Iluka-Eneabba\n(exclude 7-year rehab)__Bray-Curtis\nASV-level: 60 (52.95, 71.05)"


levels( pred_models_boot$facet_x )
# "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"        "Unweighted UniFrac\nASV-level" "Weighted UniFrac\nASV-level"  




## plot time to target

names(pred_models_boot)
# [1] "study"                   "facet_x"                 "study_and_metric"        "group"                   "group_num"              
# [6] "similarity"              "boot_no"                 "ref_median"              "pred_time_to_ref_median" "pred_type"



df <- pred_models_boot
unique(df$study) #  Iluka-Eneabba                         Iluka-Eneabba\n(exclude 7-year rehab)
df$study <- factor(df$study,
                   levels = c("Iluka-Eneabba", "Iluka-Eneabba\n(exclude 7-year rehab)"),
                   labels = c("Eneabba", "Eneabba (exclude 7-year)"), ordered = TRUE)
df$facet_x
# Levels: Jaccard\nASV-level < Bray-Curtis\nASV-level < Unweighted UniFrac\nASV-level < Weighted UniFrac\nASV-level
df$facet_x <- factor(df$facet_x, 
                     levels = c("Jaccard\nASV-level",
                                "Bray-Curtis\nASV-level"),
                     labels = c("Jaccard",
                                "Bray-Curtis"), ordered = TRUE)

p <- ggplot(data=df, aes(x=facet_x, y=pred_time_to_ref_median)) +  # , color = facet_x
  geom_boxplot() +
  theme_bw() +
  facet_wrap(vars(study) , nrow=1) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = rel(1.1))
  ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Predicted-time-to-target--Iluka-Eneabba--ASV-Jaccard-Bray-exclude7yr.tiff"), width = 10, height = 8, units = "cm", dpi = 600, compression = "lzw",type="cairo")


#-------------------------


#### facet_grid trajectory plots + predict time to reach ref:
##   ASV -- Jaccard + Bray-Curtis -- South32 only - with & without youngest rehab_age group & associated southernmost Ref
#-------------------------

names(df_results)


names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"    
# [7] "group"          "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"  


keep_results <- c("South32-Jaccard-ASV-withTree" ,           "South32-Bray-ASV-withTree" ,
                  "South32-Jaccard-ASV-withTree-exclude2yr", "South32-Bray-ASV-withTree-exclude2yr"
)


keep_cols <- c("samp","compare_with","similarity","group","dist_measure","tax_level","study","corrected","facet_x")
traj_data <- df_results[[ keep_results[1] ]][ ,keep_cols]
for (i in 2:length(keep_results)) {
  traj_data <- rbind(traj_data,df_results[[ keep_results[i] ]][ ,keep_cols])
  print(paste0("completed ",i))
}
dim(traj_data) # 530    9
unique(traj_data$facet_x) # "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"

keep_cols_anno <- c("group","sig_letter","similarity","dist_measure","tax_level","study","corrected","facet_x")
anno_data <- df_anno[[ keep_results[1] ]][ ,keep_cols_anno]
for (i in 2:length(keep_results)) {
  anno_data <- rbind(anno_data,df_anno[[ keep_results[i] ]][ ,keep_cols_anno])
  print(paste0("completed ",i))
}
dim(anno_data) # 34  8
class(anno_data$facet_x) # "character"
unique(anno_data$facet_x) # "character"
# [1] "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"
anno_data$facet_x <- factor(anno_data$facet_x,
                            levels = c("Jaccard\nASV-level","Bray-Curtis\nASV-level"),
                            ordered = TRUE)

unique(traj_data$facet_x)
# [1] "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"
class(traj_data$facet_x) # "character"
traj_data$facet_x <- factor(traj_data$facet_x,
                            levels = c("Jaccard\nASV-level","Bray-Curtis\nASV-level"),
                            ordered = TRUE)

unique(traj_data$study)
# [1] "South32"                         "South32\n(exclude 2-year rehab)"
class(traj_data$study) # "character"
traj_data$study <- factor(traj_data$study,
                          levels = c("South32", "South32\n(exclude 2-year rehab)"),
                          ordered = TRUE)

unique(anno_data$study)
# [1] "South32"                         "South32\n(exclude 2-year rehab)"
class(anno_data$study) # "character"
anno_data$study <- factor(anno_data$study,
                          levels = c("South32"  ,  "South32\n(exclude 2-year rehab)"),
                          ordered = TRUE)

class(traj_data$group)
# "ordered" "factor" 

traj_data$group <- factor(traj_data$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)

levels(traj_data$group)
# [1] "0"   "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14" 
# [16] "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29" 
# [31] "30"  "31"  "32"  "33"  "34"  "35"  "36"  "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44" 
# [46] "45"  "Ref"

length(levels(traj_data$group)) #47

x_labels <- c("0","","","","","","","","","","10",
              "","","","","","","","","","20",
              "","","","","","","","","","30",
              "","","","","","","","","","40",
              "","","","","","Ref")
length(x_labels)
# 47
x_ticks <- c(1,0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,1)
length(x_ticks)
#47


unique(traj_data$study) # South32                         South32\n(exclude 2-year rehab)

unique(anno_data$study) # South32                         South32\n(exclude 2-year rehab)

 



## median reference horizontal lines ...

href_lines <- data.frame(
  study  = rep( c("South32" ,  "South32\n(exclude 2-year rehab)"), times = 2 ),
  facet_x  = rep( c("Jaccard\nASV-level","Bray-Curtis\nASV-level"), each = 2),
  similarity= NA
)

x <- traj_data

for (i in 1:dim(href_lines)[1]) {
  #i<-1
  sel <- which(x$study == href_lines$study[i] & x$facet_x == href_lines$facet_x[i] & x$group == "Ref")
  href_lines$similarity[i] <- median( x$similarity[sel] )
}
href_lines$facet_x <- factor(href_lines$facet_x,
                             levels = c("Jaccard\nASV-level","Bray-Curtis\nASV-level"),
                             ordered = TRUE)


## fit logarithmic models for each data type (facet_x)



## BOOTSTRAP MODELS ...

pred_models_boot <- data.frame(study=NA, # one of: "Alcoa", "Iluka-Eneabba", "South32"
                               facet_x=NA, # one of: c("Jaccard\nASV-level","Bray-Curtis\nASV-level", "Jaccard\nGenus-level","Bray-Curtis\nGenus-level")
                               study_and_metric=NA, # combine the two values above
                               group=NA,
                               group_num=NA,
                               similarity=NA,
                               boot_no=NA,                 # EXTRA
                               ref_median=NA,              # EXTRA
                               pred_time_to_ref_median=NA, # EXTRA
                               pred_type=NA)     # c("rectangular hyperbola", "logarithmic")
pred_models_boot.temp <- pred_models_boot

studies <- c("South32" ,  "South32\n(exclude 2-year rehab)")
metrics <- c("Jaccard\nASV-level","Bray-Curtis\nASV-level")
pred_types <- c("rectangular hyperbola", "negative exponential", "logarithmic")
k<-3

## run logarithmic model only !!!

x <- traj_data
dim(x) # 530   9

for (i in 1:length(studies)) {
  #i<-1
  this_study <- studies[i]
  for (j in 1:length(metrics)) {
    #j<-3
    this_metric <- metrics[j]
    
    # prepare data for modelling
    subsel <- which(x$facet_x == this_metric & x$study==this_study)
    newdat <- x[subsel, c("similarity","group") ]
    newdat$group_num <- as.numeric(as.character(newdat$group))
    # remove 'Ref' similarities which have now been converted to NA
    sel.na <- which(is.na(newdat$group_num))
    
    this_ref_median <- median(newdat$similarity[sel.na])
    
    newdat <- newdat[-sel.na, ]
    
    # prep output dataframe for prediction model to populate
    sel.dup <- which(duplicated(newdat$group)==TRUE)
    new_preds <- data.frame(matrix( nrow=dim(newdat[-sel.dup, ])[1], ncol = length(names(pred_models_boot.temp)) ))
    names(new_preds) <- names(pred_models_boot.temp)
    new_preds$group_num <- sort( newdat$group_num[-sel.dup] )
    new_preds$study <- this_study
    new_preds$facet_x <- this_metric
    new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$facet_x)
    new_preds$group <- as.character(new_preds$group_num)
    new_preds$ref_median <- this_ref_median
    
    #for (k in 1:length(pred_types)) {
    
    this_pred_model_type <- pred_types[k]
    
    ## NOTE:
    #  rectangular hyperbola ( fct = MM.2() from ‘drc' package; https://www.statforbiology.com/nonlinearregression/usefulequations#michaelis-menten_equation), and 
    #  negative exponential ( fct = DRC.negExp() from 'aomisc' package; https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function)
    #  models were trialled however these regularly failed to achieve model fits.
    
    
    # NOW using "logarithmic" model only !!!!!
    
    # negative exponential model, using fct = DRC.negExp() from 'aomisc' package
    # https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function
    
    
    for (b in 1:100) {
      #b<-1
      new_preds$boot_no <- b
      set.seed(b + 1234)
      sel.boot <- sample(x = c(1:dim(newdat)[1]), size = dim(newdat)[1], replace = TRUE)
      
      bootdat <- newdat[sel.boot, ]
      
      # exclude outlying data points from trajectory data-cloud ... 
      # as outliers are likely to throw out model-fitting
      bp <- boxplot(bootdat$similarity, print=FALSE)
      sel.outlier <- which(bootdat$similarity %in% bp$out)
      if (length(sel.outlier)>0) {
        bootdat <- bootdat[-sel.outlier, ]
      }
      
      #plot(bootdat$group_num, bootdat$similarity)
      
      model <- drm(similarity ~ group_num , fct = DRC.logCurve(), # DRC.negExp()
                   data = bootdat)
      
      new_preds$similarity <- predict(model, new_preds )
      new_preds$pred_type <- this_pred_model_type
      
      y <- data.frame(group_num=1:500, sim_pred=NA)
      y$sim_pred <- predict(model, y )
      y$reached_target <- y$sim_pred >= this_ref_median
      #plot(y$group_num, y$sim_pred)
      row.names(y)
      
      sel <- which(y$reached_target == TRUE)
      
      if (length(sel) > 0) {
        years_to_ref <- min(sel)
      } else {
        years_to_ref <- NA
      }
      new_preds$pred_time_to_ref_median <- years_to_ref
      
      # add predictions from end of last year to 40 years ...
      max_year <- max(new_preds$group_num)
      year <- max_year + 1
      while (year < 40) {
        new_preds <- rbind(new_preds, new_preds[ dim(new_preds)[1], ])
        year <- year + 1
        new_preds$group_num[dim(new_preds)[1]] <- year
        new_preds$group[dim(new_preds)[1]] <- as.character( new_preds$group_num[dim(new_preds)[1]] )
        new_preds$similarity[dim(new_preds)[1]] <- y[ year, "sim_pred"]
      }
      
      pred_models_boot <- rbind(pred_models_boot, new_preds)
      
    } # END bootstrap
    
    print(paste0("completed study: ",studies[i],"; metric: ",metrics[j],"; pred_type: ",pred_types[k]))  
    #} # END pred_types
  } # END metrics
} # END studies


# remove NA first row
pred_models_boot <- pred_models_boot[-1, ]

unique(pred_models_boot$facet_x)
pred_models_boot$facet_x <- factor(pred_models_boot$facet_x,
                                   levels = c("Jaccard\nASV-level", "Bray-Curtis\nASV-level"),
                                   ordered = TRUE)

unique(pred_models_boot$study)
pred_models_boot$study <- factor(pred_models_boot$study,
                                 levels = c("South32" ,  "South32\n(exclude 2-year rehab)"),
                                 ordered = TRUE)

dim(pred_models_boot) # 7400    10


# correct time prediction to ref; update NA to (>) 500 yrs
sel <- which(is.na(pred_models_boot$pred_time_to_ref_median))
pred_models_boot$pred_time_to_ref_median[sel] <- 500


pred_models_boot$group <- factor(pred_models_boot$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)
str(pred_models_boot)


str(traj_data)


temp.traj_data <- traj_data
temp.pred_models_boot <- pred_models_boot

unique(traj_data$study) #  South32                         South32\n(exclude 2-year rehab)

study_names <- c(
  South32="Worsley",
  `South32\n(exclude 2-year rehab)`="Worsley\n(exclude 2-year)"
)



p <- ggplot(data=traj_data, aes(x=group, y=similarity)) +
  geom_line(data=pred_models_boot, aes(  x=group, y=similarity, group = boot_no ), colour = "#e34a33", alpha = 0.02) + # "red"
  geom_hline(data = href_lines, aes(yintercept = similarity),  color= "#2c7fb8", linetype="longdash") + # "blue"
  #geom_line(data= pred_models_alldata, aes(  x=group, y=similarity, group = 1), colour = "#e34a33" ) + # "red"
  
  geom_boxplot(outlier.shape = NA, width = 2) +
  #geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  
  theme_bw() +
  geom_text(data=anno_data, aes(x=group, y=similarity, label=sig_letter),nudge_y = 2, size=2.75) +
  scale_x_discrete(drop=FALSE, expand = c(0.05,0.05),
                   labels = x_labels) +
  theme(axis.text.x  = element_text(angle=90,hjust=1, vjust=0.5), # 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_line(linetype = 0)
        axis.ticks.x = element_line(linetype = x_ticks),
        strip.text.x = element_text(size = rel(1.1)),
        strip.text.y = element_text(size = rel(1.1))
  ) +
  labs(x = "Rehabilitation age (years)", y = "Similarity to Reference (%)") +
  facet_grid(rows = vars(study), cols = vars(facet_x), scales = "fixed", labeller = labeller(study = study_names) ) # study_labeller   as_labeller(study_names)

p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Restoration-trajectories--Worsley--ASV-Jaccard-Bray-with-Preds-Exclude-2yr.tiff"), width = 11, height = 10, units = "cm", dpi = 600, compression = "lzw",type="cairo")



str(traj_data)
unique(traj_data$study) # South32                         South32\n(exclude 2-year rehab)
unique(traj_data$group) # use "ref" only
unique(traj_data$facet_x)
#  Jaccard\nASV-level     Bray-Curtis\nASV-level


## "South32"
sel <- which(traj_data$study == "South32")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Jaccard\nASV-level")
summary(traj_data$similarity[sel[subsel]])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 17.25   20.16   25.72   24.56   28.41   31.25 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$facet_x[sel] == "Bray-Curtis\nASV-level")
summary(traj_data$similarity[sel[subsel]])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 21.39   26.45   36.02   34.32   39.96   46.11 



str(pred_models_boot)

unique(pred_models_boot$study_and_metric)
# [1] "South32__Jaccard\nASV-level"                             "South32__Bray-Curtis\nASV-level"                        
# [3] "South32\n(exclude 2-year rehab)__Jaccard\nASV-level"     "South32\n(exclude 2-year rehab)__Bray-Curtis\nASV-level"

pred_models_boot$study_and_metric <- factor(pred_models_boot$study_and_metric,
                                            levels = c(
                                              "South32__Jaccard\nASV-level"                          ,   "South32__Bray-Curtis\nASV-level"         ,               
                                              "South32\n(exclude 2-year rehab)__Jaccard\nASV-level"   ,  "South32\n(exclude 2-year rehab)__Bray-Curtis\nASV-level"
                                            ))

dim(pred_models_boot) # 7400    10

dat.sum <- pred_models_boot
head(dat.sum)
# study            facet_x            study_and_metric group group_num similarity boot_no ref_median pred_time_to_ref_median   pred_type
# 11 South32 Jaccard\nASV-level South32__Jaccard\nASV-level     2         2   15.03478       1    25.7234                     192 logarithmic
# 2  South32 Jaccard\nASV-level South32__Jaccard\nASV-level     8         8   18.28406       1    25.7234                     192 logarithmic
# 3  South32 Jaccard\nASV-level South32__Jaccard\nASV-level    12        12   19.23441       1    25.7234                     192 logarithmic
# 4  South32 Jaccard\nASV-level South32__Jaccard\nASV-level    14        14   19.59572       1    25.7234                     192 logarithmic
# 5  South32 Jaccard\nASV-level South32__Jaccard\nASV-level    17        17   20.05079       1    25.7234                     192 logarithmic
# 6  South32 Jaccard\nASV-level South32__Jaccard\nASV-level    20        20   20.43171       1    25.7234                     192 logarithmic

for (i in 1:length(levels(dat.sum$study_and_metric))) {
  #i<-1
  this_study_and_metric <- levels(dat.sum$study_and_metric)[i]
  sel <- which(dat.sum$study_and_metric == this_study_and_metric)
  #unique(dat.sum$ref_median[sel])
  temp <- dat.sum$pred_time_to_ref_median[sel]
  #quantile(temp, probs = c(0.05, 0.25, 0.5, 0.75, 0.95), na.rm = TRUE)
  #hist(temp)
  res <- quantile(temp, probs = c(0.05, 0.5, 0.95), na.rm = TRUE)
  names(res)
  #"5%"  "50%" "95%"
  #print(paste0(this_study_and_metric))
  print(paste0(this_study_and_metric,": ",res["50%"]," (",res["5%"],", ",res["95%"],")"))
}
# [1] "South32__Jaccard\nASV-level: 131.5 (68.8, 313.25)"
# [1] "South32__Bray-Curtis\nASV-level: 103.5 (68.9, 173.549999999999)"
# [1] "South32\n(exclude 2-year rehab)__Jaccard\nASV-level: 36 (32, 42)"
# [1] "South32\n(exclude 2-year rehab)__Bray-Curtis\nASV-level: 40.5 (36, 47)"


levels( pred_models_boot$facet_x )
# "Jaccard\nASV-level"            "Bray-Curtis\nASV-level"        "Unweighted UniFrac\nASV-level" "Weighted UniFrac\nASV-level"  




## plot time to target

names(pred_models_boot)
# [1] "study"                   "facet_x"                 "study_and_metric"        "group"                   "group_num"              
# [6] "similarity"              "boot_no"                 "ref_median"              "pred_time_to_ref_median" "pred_type"



df <- pred_models_boot
unique(df$study) #  South32                         South32\n(exclude 2-year rehab)
df$study <- factor(df$study,
                   levels = c("South32", "South32\n(exclude 2-year rehab)"),
                   labels = c("Worsley", "Worsley (exclude 2-year)"), ordered = TRUE)
df$facet_x
# Levels: Jaccard\nASV-level < Bray-Curtis\nASV-level < Unweighted UniFrac\nASV-level < Weighted UniFrac\nASV-level
df$facet_x <- factor(df$facet_x, 
                     levels = c("Jaccard\nASV-level",
                                "Bray-Curtis\nASV-level"),
                     labels = c("Jaccard",
                                "Bray-Curtis"), ordered = TRUE)

p <- ggplot(data=df, aes(x=facet_x, y=pred_time_to_ref_median)) +  # , color = facet_x
  geom_boxplot() +
  theme_bw() +
  facet_wrap(vars(study) , nrow=1) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = rel(1.1))
  ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Predicted-time-to-target--Worsley--ASV-Jaccard-Bray-exclude2yr.tiff"), width = 10, height = 8, units = "cm", dpi = 600, compression = "lzw",type="cairo")


#-------------------------




#### facet_grid trajectory plots + predict time to reach ref:
##   ASV -- Jaccard + Bray-Curtis -- ASV, 99%, 97%, 95%, 90% OTUs -- Alcoa
#-------------------------

names(df_results)


names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"    
# [7] "group"          "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"  

names(df_results)[ grep("Alcoa-Jaccard-",x = names(df_results)) ]

keep_results <- c("Alcoa-Bray-ASV-withTree",
                  "Alcoa-Bray-otu99",
                  "Alcoa-Bray-otu97",
                  "Alcoa-Bray-otu95",                   
                  "Alcoa-Bray-otu90",
                  
                  "Alcoa-Jaccard-ASV-withTree",
                  "Alcoa-Jaccard-otu99",
                  "Alcoa-Jaccard-otu97",
                  "Alcoa-Jaccard-otu95",
                  "Alcoa-Jaccard-otu90"
)


keep_cols <- c("samp","compare_with","similarity","group","dist_measure","tax_level","study","corrected","facet_x")
traj_data <- df_results[[ keep_results[1] ]][ ,keep_cols]
for (i in 2:length(keep_results)) {
  traj_data <- rbind(traj_data,df_results[[ keep_results[i] ]][ ,keep_cols])
  print(paste0("completed ",i))
}
dim(traj_data) # 6300    9
unique(traj_data$facet_x) #
# [1] "Bray-Curtis\nASV-level" "Bray-Curtis\n99% OTUs"  "Bray-Curtis\n97% OTUs"  "Bray-Curtis\n95% OTUs"  "Bray-Curtis\n90% OTUs" 
# [6] "Jaccard\nASV-level"     "Jaccard\n99% OTUs"      "Jaccard\n97% OTUs"      "Jaccard\n95% OTUs"      "Jaccard\n90% OTUs" 

class(traj_data$facet_x) # "character"
traj_data$facet_x <- factor(traj_data$facet_x,
                            levels = c("Bray-Curtis\nASV-level",
                                       "Bray-Curtis\n99% OTUs",
                                       "Bray-Curtis\n97% OTUs",
                                       "Bray-Curtis\n95% OTUs",
                                       "Bray-Curtis\n90% OTUs",
                                       
                                       "Jaccard\nASV-level",
                                       "Jaccard\n99% OTUs",
                                       "Jaccard\n97% OTUs",
                                       "Jaccard\n95% OTUs",
                                       "Jaccard\n90% OTUs" 
                                       
                            ),
                            ordered = TRUE)

traj_data$NEW_newfacet_x <- NA
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
traj_data$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\n99% OTUs","Jaccard\n99% OTUs"))
traj_data$NEW_newfacet_x[sel] <- "99% OTUs"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\n97% OTUs","Jaccard\n97% OTUs"))
traj_data$NEW_newfacet_x[sel] <- "97% OTUs"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\n95% OTUs","Jaccard\n95% OTUs"))
traj_data$NEW_newfacet_x[sel] <- "95% OTUs"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\n90% OTUs","Jaccard\n90% OTUs"))
traj_data$NEW_newfacet_x[sel] <- "90% OTUs"

traj_data$NEW_newfacet_x <- factor(traj_data$NEW_newfacet_x,
                                   levels = c("ASV-level",
                                              "99% OTUs",
                                              "97% OTUs",
                                              "95% OTUs",
                                              "90% OTUs"
                                   ),
                                   ordered = TRUE)


keep_cols_anno <- c("group","sig_letter","similarity","dist_measure","tax_level","study","corrected","facet_x")
anno_data <- df_anno[[ keep_results[1] ]][ ,keep_cols_anno]
for (i in 2:length(keep_results)) {
  anno_data <- rbind(anno_data,df_anno[[ keep_results[i] ]][ ,keep_cols_anno])
  print(paste0("completed ",i))
}
dim(anno_data) # 70  8
class(anno_data$facet_x) # "character"
unique(anno_data$facet_x) # "character"
# [1] "Bray-Curtis\nASV-level" "Bray-Curtis\n99% OTUs"  "Bray-Curtis\n97% OTUs"  "Bray-Curtis\n95% OTUs"  "Bray-Curtis\n90% OTUs" 
# [6] "Jaccard\nASV-level"     "Jaccard\n99% OTUs"      "Jaccard\n97% OTUs"      "Jaccard\n95% OTUs"      "Jaccard\n90% OTUs" 


anno_data$NEW_newfacet_x <- NA
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
anno_data$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\n99% OTUs","Jaccard\n99% OTUs"))
anno_data$NEW_newfacet_x[sel] <- "99% OTUs"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\n97% OTUs","Jaccard\n97% OTUs"))
anno_data$NEW_newfacet_x[sel] <- "97% OTUs"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\n95% OTUs","Jaccard\n95% OTUs"))
anno_data$NEW_newfacet_x[sel] <- "95% OTUs"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\n90% OTUs","Jaccard\n90% OTUs"))
anno_data$NEW_newfacet_x[sel] <- "90% OTUs"

anno_data$NEW_newfacet_x <- factor(anno_data$NEW_newfacet_x,
                                   levels = c("ASV-level",
                                              "99% OTUs",
                                              "97% OTUs",
                                              "95% OTUs",
                                              "90% OTUs"
                                   ),
                                   ordered = TRUE)

names(anno_data)
#  "group"          "sig_letter"     "similarity"     "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"        "NEW_newfacet_x"

# use $dist_measure - i.e. Bray-Curtis or Jaccard as y-facet 


class(traj_data$group) # "ordered" "factor" 

traj_data$group <- factor(traj_data$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)

levels(traj_data$group)
# [1] "0"   "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14" 
# [16] "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29" 
# [31] "30"  "31"  "32"  "33"  "34"  "35"  "36"  "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44" 
# [46] "45"  "Ref"

length(levels(traj_data$group)) #47

x_labels <- c("0","","","","","","","","","","10",
              "","","","","","","","","","20",
              "","","","","","","","","","30",
              "","","","","","","","","","40",
              "","","","","","Ref")
length(x_labels)
# 47
x_ticks <- c(1,0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,1)
length(x_ticks)
#47


unique(traj_data$study) # "Alcoa"

unique(anno_data$study) # "Alcoa"

 



## median reference horizontal lines ...

href_lines <- data.frame(
  study = rep("Alcoa",times=10),
  facet_x = c("Bray-Curtis\nASV-level",
              "Bray-Curtis\n99% OTUs",
              "Bray-Curtis\n97% OTUs",
              "Bray-Curtis\n95% OTUs",
              "Bray-Curtis\n90% OTUs" ,
              
              "Jaccard\nASV-level",
              "Jaccard\n99% OTUs",
              "Jaccard\n97% OTUs",
              "Jaccard\n95% OTUs",
              "Jaccard\n90% OTUs"
  ),
  dist_measure = rep( c("Bray-Curtis","Jaccard"), each = 5 ),
  similarity= NA
)

x <- traj_data

for (i in 1:dim(href_lines)[1]) {
  #i<-1
  sel <- which(x$study == href_lines$study[i] & x$facet_x == href_lines$facet_x[i] & x$group == "Ref")
  href_lines$similarity[i] <- median( x$similarity[sel] )
}

href_lines$NEW_newfacet_x <- NA
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
href_lines$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\n99% OTUs","Jaccard\n99% OTUs"))
href_lines$NEW_newfacet_x[sel] <- "99% OTUs"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\n97% OTUs","Jaccard\n97% OTUs"))
href_lines$NEW_newfacet_x[sel] <- "97% OTUs"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\n95% OTUs","Jaccard\n95% OTUs"))
href_lines$NEW_newfacet_x[sel] <- "95% OTUs"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\n90% OTUs","Jaccard\n90% OTUs"))
href_lines$NEW_newfacet_x[sel] <- "90% OTUs"


href_lines$NEW_newfacet_x <- factor(href_lines$NEW_newfacet_x,
                                    levels = c("ASV-level",
                                               "99% OTUs",
                                               "97% OTUs",
                                               "95% OTUs",
                                               "90% OTUs"
                                    ),
                                    ordered = TRUE)


## fit logarithmic models for each data type (facet_x)



## BOOTSTRAP MODELS ...

pred_models_boot <- data.frame(study=NA, #
                               facet_x=NA, #
                               study_and_metric=NA, # combine the two values above
                               group=NA,
                               group_num=NA,
                               similarity=NA,
                               boot_no=NA,                 # EXTRA
                               ref_median=NA,              # EXTRA
                               pred_time_to_ref_median=NA, # EXTRA
                               pred_type=NA)     # c("rectangular hyperbola", "logarithmic")
pred_models_boot.temp <- pred_models_boot

#studies <- c("Alcoa", "Iluka-Eneabba", "South32")

studies <- c("Alcoa")

metrics <- c("Bray-Curtis\nASV-level",
             "Bray-Curtis\n99% OTUs",
             "Bray-Curtis\n97% OTUs",
             "Bray-Curtis\n95% OTUs",
             "Bray-Curtis\n90% OTUs" ,
             
             "Jaccard\nASV-level",
             "Jaccard\n99% OTUs",
             "Jaccard\n97% OTUs",
             "Jaccard\n95% OTUs",
             "Jaccard\n90% OTUs"
)
pred_types <- c("rectangular hyperbola", "negative exponential", "logarithmic")
## run logarithmic model only !!!
k<-3



x <- traj_data
dim(x) # 6300  10

for (i in 1:length(studies)) {
  #i<-1
  this_study <- studies[i]
  for (j in 1:length(metrics)) {
    #j<-1
    this_metric <- metrics[j]
    
    # prepare data for modelling
    subsel <- which(x$facet_x == this_metric & x$study==this_study)
    newdat <- x[subsel, c("similarity","group") ]
    newdat$group_num <- as.numeric(as.character(newdat$group))
    # remove 'Ref' similarities which have now been converted to NA
    sel.na <- which(is.na(newdat$group_num))
    
    this_ref_median <- median(newdat$similarity[sel.na])
    
    newdat <- newdat[-sel.na, ]
    
    # prep output dataframe for prediction model to populate
    sel.dup <- which(duplicated(newdat$group)==TRUE)
    new_preds <- data.frame(matrix( nrow=dim(newdat[-sel.dup, ])[1], ncol = length(names(pred_models_boot.temp)) ))
    names(new_preds) <- names(pred_models_boot.temp)
    new_preds$group_num <- sort( newdat$group_num[-sel.dup] )
    new_preds$study <- this_study
    new_preds$facet_x <- this_metric
    new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$facet_x)
    new_preds$group <- as.character(new_preds$group_num)
    new_preds$ref_median <- this_ref_median
    
    #for (k in 1:length(pred_types)) {
    
    
    this_pred_model_type <- pred_types[k]
    
    ## NOTE:
    #  rectangular hyperbola ( fct = MM.2() from ‘drc' package; https://www.statforbiology.com/nonlinearregression/usefulequations#michaelis-menten_equation), and 
    #  negative exponential ( fct = DRC.negExp() from 'aomisc' package; https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function)
    #  models were trialled however these regularly failed to achieve model fits.
    
    
    # NOW using "logarithmic" model only !!!!!
    
    # negative exponential model, using fct = DRC.negExp() from 'aomisc' package
    # https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function
    
    
    for (b in 1:100) {
      #b<-20
      new_preds$boot_no <- b
      set.seed(b + 1234)
      sel.boot <- sample(x = c(1:dim(newdat)[1]), size = dim(newdat)[1], replace = TRUE)
      
      bootdat <- newdat[sel.boot, ]
      
      # exclude outlying data points from trajectory data-cloud ... 
      # as outliers are likely to throw out model-fitting
      bp <- boxplot(bootdat$similarity, print=FALSE)
      sel.outlier <- which(bootdat$similarity %in% bp$out)
      if (length(sel.outlier)>0) {
        bootdat <- bootdat[-sel.outlier, ]
      }
      
      #plot(bootdat$group_num, bootdat$similarity)
      
      model <- drm(similarity ~ group_num , fct = DRC.logCurve(), # DRC.negExp()
                   data = bootdat)
      
      new_preds$similarity <- predict(model, new_preds )
      new_preds$pred_type <- this_pred_model_type
      
      y <- data.frame(group_num=1:500, sim_pred=NA)
      y$sim_pred <- predict(model, y )
      y$reached_target <- y$sim_pred >= this_ref_median
      #plot(y$group_num, y$sim_pred)
      row.names(y)
      
      sel <- which(y$reached_target == TRUE)
      
      if (length(sel) > 0) {
        years_to_ref <- min(sel)
      } else {
        years_to_ref <- NA
      }
      new_preds$pred_time_to_ref_median <- years_to_ref
      
      # add predictions from end of last year to 40 years ...
      max_year <- max(new_preds$group_num)
      year <- max_year + 1
      while (year < 40) {
        new_preds <- rbind(new_preds, new_preds[ dim(new_preds)[1], ])
        year <- year + 1
        new_preds$group_num[dim(new_preds)[1]] <- year
        new_preds$group[dim(new_preds)[1]] <- as.character( new_preds$group_num[dim(new_preds)[1]] )
        new_preds$similarity[dim(new_preds)[1]] <- y[ year, "sim_pred"]
      }
      
      pred_models_boot <- rbind(pred_models_boot, new_preds)
      
    } # END bootstrap
    

    print(paste0("completed study: ",studies[i],"; metric: ",metrics[j],"; pred_type: ",pred_types[k]))  
    #} # END pred_types
  } # END metrics
} # END studies


# remove NA first row
pred_models_boot <- pred_models_boot[-1, ]

unique(pred_models_boot$facet_x)
# [1] "Bray-Curtis\nASV-level" "Bray-Curtis\n99% OTUs"  "Bray-Curtis\n97% OTUs"  "Bray-Curtis\n95% OTUs"  "Bray-Curtis\n90% OTUs" 
# [6] "Jaccard\nASV-level"     "Jaccard\n99% OTUs"      "Jaccard\n97% OTUs"      "Jaccard\n95% OTUs"      "Jaccard\n90% OTUs" 

dim(pred_models_boot) # 16000  10

pred_models_boot$NEW_newfacet_x <- NA
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\n99% OTUs","Jaccard\n99% OTUs"))
pred_models_boot$NEW_newfacet_x[sel] <- "99% OTUs"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\n97% OTUs","Jaccard\n97% OTUs"))
pred_models_boot$NEW_newfacet_x[sel] <- "97% OTUs"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\n95% OTUs","Jaccard\n95% OTUs"))
pred_models_boot$NEW_newfacet_x[sel] <- "95% OTUs"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\n90% OTUs","Jaccard\n90% OTUs"))
pred_models_boot$NEW_newfacet_x[sel] <- "90% OTUs"

pred_models_boot$NEW_newfacet_x <- factor(pred_models_boot$NEW_newfacet_x,
                                          levels = c("ASV-level",
                                                     "99% OTUs",
                                                     "97% OTUs",
                                                     "95% OTUs",
                                                     "90% OTUs"
                                          ),
                                          ordered = TRUE)

# correct time prediction to ref; update NA to (>) 500 yrs
sel <- which(is.na(pred_models_boot$pred_time_to_ref_median))
if (length(sel)>1) { pred_models_boot$pred_time_to_ref_median[sel] <- 500 }

pred_models_boot$group <- factor(pred_models_boot$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)
table(pred_models_boot$group)

pred_models_boot$dist_measure <- NA
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nASV-level","Bray-Curtis\n99% OTUs","Bray-Curtis\n97% OTUs",
                                             "Bray-Curtis\n95% OTUs","Bray-Curtis\n90% OTUs"))
pred_models_boot$dist_measure[sel] <- "Bray-Curtis"
sel <- which(pred_models_boot$facet_x %in% c("Jaccard\nASV-level","Jaccard\n99% OTUs","Jaccard\n97% OTUs",
                                             "Jaccard\n95% OTUs","Jaccard\n90% OTUs"))
pred_models_boot$dist_measure[sel] <- "Jaccard"


str(pred_models_boot)


str(traj_data)



p <- ggplot(data=traj_data, aes(x=group, y=similarity)) +
  geom_line(data=pred_models_boot, aes(  x=group, y=similarity, group = boot_no ), colour = "#e34a33", alpha = 0.02) + # "red"
  geom_hline(data = href_lines, aes(yintercept = similarity),  color= "#2c7fb8", linetype="longdash") + # "blue"
  #geom_line(data= pred_models_alldata, aes(  x=group, y=similarity, group = 1), colour = "#e34a33" ) + # "red"
  
  geom_boxplot(outlier.shape = NA, width = 2) +
  #geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  
  theme_bw() +
  geom_text(data=anno_data, aes(x=group, y=similarity, label=sig_letter),nudge_y = 2, size=2.75) +
  scale_x_discrete(drop=FALSE, expand = c(0.05,0.05),
                   labels = x_labels) +
  theme(axis.text.x  = element_text(angle=90,hjust=1, vjust=0.5), # 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_line(linetype = 0)
        axis.ticks.x = element_line(linetype = x_ticks)
  ) +
  labs(x = "Rehabilitation age (years)", y = "Similarity to Reference (%)") +
  facet_grid(rows = vars(dist_measure), cols = vars(NEW_newfacet_x), scales = "fixed")

p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Restoration-trajectories--Alcoa--Bray-Jaccard-ASV-99-97-95-90-OTUs-with-Preds.tiff"), width = 18, height = 10, units = "cm", dpi = 600, compression = "lzw",type="cairo")



str(traj_data)
unique(traj_data$study) # "Alcoa"
unique(traj_data$group) # use "ref" only
unique(traj_data$NEW_newfacet_x)
# [1] ASV-level 99% OTUs  97% OTUs  95% OTUs  90% OTUs 
unique(traj_data$dist_measure) # "Bray-Curtis" "Jaccard"    

## Bray-Curtis
sel <- which(traj_data$dist_measure == "Bray-Curtis")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "ASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 22.38   31.28   35.55   35.79   39.93   60.91 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "99% OTUs")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 31.72   40.81   45.73   45.76   50.11   69.43 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "97% OTUs")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 42.14   52.78   57.03   56.98   60.77   77.42 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "95% OTUs")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 47.85   59.51   63.98   63.42   66.85   82.15 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "90% OTUs")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 58.35   71.08   73.99   73.78   76.93   88.58 


## Jaccard
sel <- which(traj_data$dist_measure == "Jaccard")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "ASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 9.503  17.824  20.345  20.092  23.213  35.809 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "99% OTUs")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 12.78   24.39   27.23   26.58   30.36   41.62 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "97% OTUs")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 18.09   35.01   37.33   36.15   40.15   49.97 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "95% OTUs")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 21.09   41.72   44.39   42.36   46.75   56.31 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "90% OTUs")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 26.56   54.52   57.09   54.42   59.73   65.40 




str(pred_models_boot)


unique(pred_models_boot$study_and_metric)
# [1] "Alcoa__Bray-Curtis\nASV-level" "Alcoa__Bray-Curtis\n99% OTUs"  "Alcoa__Bray-Curtis\n97% OTUs"  "Alcoa__Bray-Curtis\n95% OTUs" 
# [5] "Alcoa__Bray-Curtis\n90% OTUs"  "Alcoa__Jaccard\nASV-level"     "Alcoa__Jaccard\n99% OTUs"      "Alcoa__Jaccard\n97% OTUs"     
# [9] "Alcoa__Jaccard\n95% OTUs"      "Alcoa__Jaccard\n90% OTUs"  

pred_models_boot$study_and_metric <- factor(pred_models_boot$study_and_metric,
                                            levels = c(
                                              "Alcoa__Bray-Curtis\nASV-level",
                                              "Alcoa__Bray-Curtis\n99% OTUs",
                                              "Alcoa__Bray-Curtis\n97% OTUs",
                                              "Alcoa__Bray-Curtis\n95% OTUs",
                                              "Alcoa__Bray-Curtis\n90% OTUs",
                                              
                                              "Alcoa__Jaccard\nASV-level",
                                              "Alcoa__Jaccard\n99% OTUs",
                                              "Alcoa__Jaccard\n97% OTUs",
                                              "Alcoa__Jaccard\n95% OTUs",
                                              "Alcoa__Jaccard\n90% OTUs"
                                            ), ordered = TRUE)

dim(pred_models_boot) # 16000    12

dat.sum <- pred_models_boot
head(dat.sum)
# study                facet_x              study_and_metric group group_num similarity boot_no ref_median
# 11 Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level     2         2   9.389107       1   35.55047
# 2  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level     8         8  21.149581       1   35.55047
# 3  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    14        14  25.897020       1   35.55047
# 4  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    17        17  27.544121       1   35.55047
# 5  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    25        25  30.815846       1   35.55047
# 6  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    29        29  32.074951       1   35.55047
# pred_time_to_ref_median   pred_type NEW_newfacet_x dist_measure
# 11                      44 logarithmic      ASV-level  Bray-Curtis
# 2                       44 logarithmic      ASV-level  Bray-Curtis
# 3                       44 logarithmic      ASV-level  Bray-Curtis
# 4                       44 logarithmic      ASV-level  Bray-Curtis
# 5                       44 logarithmic      ASV-level  Bray-Curtis
# 6                       44 logarithmic      ASV-level  Bray-Curtis

for (i in 1:length(levels(dat.sum$study_and_metric))) {
  #i<-1
  this_study_and_metric <- levels(dat.sum$study_and_metric)[i]
  sel <- which(dat.sum$study_and_metric == this_study_and_metric)
  #unique(dat.sum$ref_median[sel])
  temp <- dat.sum$pred_time_to_ref_median[sel]
  #quantile(temp, probs = c(0.05, 0.25, 0.5, 0.75, 0.95), na.rm = TRUE)
  #hist(temp)
  res <- quantile(temp, probs = c(0.05, 0.5, 0.95), na.rm = TRUE)
  names(res)
  #"5%"  "50%" "95%"
  #print(paste0(this_study_and_metric))
  print(paste0(this_study_and_metric,": ",res["50%"]," (",res["5%"],", ",res["95%"],") yr."))
}
# [1] "Alcoa__Bray-Curtis\nASV-level: 43 (39, 49) yr."
# [1] "Alcoa__Bray-Curtis\n99% OTUs: 40 (37, 45) yr."
# [1] "Alcoa__Bray-Curtis\n97% OTUs: 37 (35, 42) yr."
# [1] "Alcoa__Bray-Curtis\n95% OTUs: 37 (34, 42) yr."
# [1] "Alcoa__Bray-Curtis\n90% OTUs: 37 (33, 42) yr."
# [1] "Alcoa__Jaccard\nASV-level: 60.5 (51, 79) yr."
# [1] "Alcoa__Jaccard\n99% OTUs: 63.5 (53, 85) yr."
# [1] "Alcoa__Jaccard\n97% OTUs: 81 (63.95, 117.35) yr."
# [1] "Alcoa__Jaccard\n95% OTUs: 93.5 (68.95, 156.25) yr."
# [1] "Alcoa__Jaccard\n90% OTUs: 133.5 (77.8, 441.749999999999) yr."


## plot time to target

names(pred_models_boot)
# [1] "study"                   "facet_x"                 "study_and_metric"        "group"                   "group_num"              
# [6] "similarity"              "boot_no"                 "ref_median"              "pred_time_to_ref_median" "pred_type"              
# [11] "NEW_newfacet_x"          "dist_measure" 


p <- ggplot(data=pred_models_boot, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median)) + #, color = dist_measure
  geom_boxplot() +
  #scale_color_manual(values = c("Bray-Curtis" = "#d73027", "Jaccard" = "#4575b4"), name = "Distance\nmeasure") +
  theme_bw() +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()#,
        #strip.background = element_blank(),
        #strip.text.x = element_blank()
  ) +
  facet_wrap(facets = vars(dist_measure), scales = "free_y" ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p


df <- pred_models_boot
df$dist_measure <- factor(df$dist_measure, levels = c("Jaccard","Bray-Curtis"),ordered = TRUE)

p <- ggplot(data=df, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median)) + #, color = dist_measure
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()#,
    ) +
  facet_wrap(facets = vars(dist_measure), scales = "free_y" ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p



grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Predicted-time-to-target--A--Alcoa--Bray-Jaccard-ASV-99-97-95-90-OTUs.tiff"), width = 12, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")



#-------------------------


#### facet_grid trajectory plots + predict time to reach ref:
##   ASV -- Jaccard + Bray-Curtis -- Genus + Family + ... Phylum -- Alcoa (PRUNED)
#-------------------------

names(df_results)


names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"    
# [7] "group"          "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"  

names(df_results)[ grep("Alcoa-Bray-ASV-withTree",x = names(df_results)) ]

keep_results <- c("Alcoa-Bray-ASV-withTree",
                  "Alcoa-Bray-Genus-withTree-Pruned",
                  "Alcoa-Bray-Family-withTree-Pruned" ,
                  "Alcoa-Bray-Order-withTree-Pruned" ,
                  "Alcoa-Bray-Class-withTree-Pruned" ,
                  "Alcoa-Bray-Phylum-withTree-Pruned",
                  
                  "Alcoa-Jaccard-ASV-withTree",
                  "Alcoa-Jaccard-Genus-withTree-Pruned",
                  "Alcoa-Jaccard-Family-withTree-Pruned" ,
                  "Alcoa-Jaccard-Order-withTree-Pruned",
                  "Alcoa-Jaccard-Class-withTree-Pruned",
                  "Alcoa-Jaccard-Phylum-withTree-Pruned"
                  )


keep_cols <- c("samp","compare_with","similarity","group","dist_measure","tax_level","study","corrected","facet_x")
traj_data <- df_results[[ keep_results[1] ]][ ,keep_cols]
for (i in 2:length(keep_results)) {
  traj_data <- rbind(traj_data,df_results[[ keep_results[i] ]][ ,keep_cols])
  print(paste0("completed ",i))
}
dim(traj_data) # 7560    9
unique(traj_data$facet_x) #
# [1] "Bray-Curtis\nASV-level"    "Bray-Curtis\nGenus-level"  "Bray-Curtis\nFamily-level" "Bray-Curtis\nOrder-level"  "Bray-Curtis\nClass-level" 
# [6] "Bray-Curtis\nPhylum-level" "Jaccard\nASV-level"        "Jaccard\nGenus-level"      "Jaccard\nFamily-level"     "Jaccard\nOrder-level"     
# [11] "Jaccard\nClass-level"      "Jaccard\nPhylum-level" 

class(traj_data$facet_x) # "character"
traj_data$facet_x <- factor(traj_data$facet_x,
                            levels = c("Bray-Curtis\nASV-level",
                                       "Bray-Curtis\nGenus-level",
                                       "Bray-Curtis\nFamily-level",
                                       "Bray-Curtis\nOrder-level",
                                       "Bray-Curtis\nClass-level" ,
                                       "Bray-Curtis\nPhylum-level",
                                       
                                       "Jaccard\nASV-level",
                                       "Jaccard\nGenus-level",
                                       "Jaccard\nFamily-level",
                                       "Jaccard\nOrder-level",
                                       "Jaccard\nClass-level" ,
                                       "Jaccard\nPhylum-level"
                                       ),
                            ordered = TRUE)

traj_data$NEW_newfacet_x <- NA
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
traj_data$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nGenus-level","Jaccard\nGenus-level"))
traj_data$NEW_newfacet_x[sel] <- "Genus-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nFamily-level","Jaccard\nFamily-level"))
traj_data$NEW_newfacet_x[sel] <- "Family-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nOrder-level","Jaccard\nOrder-level"))
traj_data$NEW_newfacet_x[sel] <- "Order-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nClass-level","Jaccard\nClass-level"))
traj_data$NEW_newfacet_x[sel] <- "Class-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nPhylum-level","Jaccard\nPhylum-level"))
traj_data$NEW_newfacet_x[sel] <- "Phylum-level"

traj_data$NEW_newfacet_x <- factor(traj_data$NEW_newfacet_x,
                                   levels = c("ASV-level",
                                              "Genus-level",
                                              "Family-level",
                                              "Order-level",
                                              "Class-level",
                                              "Phylum-level"
                                   ),
                                   ordered = TRUE)


keep_cols_anno <- c("group","sig_letter","similarity","dist_measure","tax_level","study","corrected","facet_x")
anno_data <- df_anno[[ keep_results[1] ]][ ,keep_cols_anno]
for (i in 2:length(keep_results)) {
  anno_data <- rbind(anno_data,df_anno[[ keep_results[i] ]][ ,keep_cols_anno])
  print(paste0("completed ",i))
}
dim(anno_data) # 84  8
class(anno_data$facet_x) # "character"
unique(anno_data$facet_x) # "character"
# [1] "Bray-Curtis\nASV-level"    "Bray-Curtis\nGenus-level"  "Bray-Curtis\nFamily-level" "Bray-Curtis\nOrder-level"  "Bray-Curtis\nClass-level" 
# [6] "Bray-Curtis\nPhylum-level" "Jaccard\nASV-level"        "Jaccard\nGenus-level"      "Jaccard\nFamily-level"     "Jaccard\nOrder-level"     
# [11] "Jaccard\nClass-level"      "Jaccard\nPhylum-level" 


anno_data$NEW_newfacet_x <- NA
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
anno_data$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nGenus-level","Jaccard\nGenus-level"))
anno_data$NEW_newfacet_x[sel] <- "Genus-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nFamily-level","Jaccard\nFamily-level"))
anno_data$NEW_newfacet_x[sel] <- "Family-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nOrder-level","Jaccard\nOrder-level"))
anno_data$NEW_newfacet_x[sel] <- "Order-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nClass-level","Jaccard\nClass-level"))
anno_data$NEW_newfacet_x[sel] <- "Class-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nPhylum-level","Jaccard\nPhylum-level"))
anno_data$NEW_newfacet_x[sel] <- "Phylum-level"

anno_data$NEW_newfacet_x <- factor(anno_data$NEW_newfacet_x,
                                   levels = c("ASV-level",
                                              "Genus-level",
                                              "Family-level",
                                              "Order-level",
                                              "Class-level",
                                              "Phylum-level"
                                   ),
                                   ordered = TRUE)

names(anno_data)
#  "group"          "sig_letter"     "similarity"     "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"        "NEW_newfacet_x"

# use $dist_measure - i.e. Bray-Curtis or Jaccard as y-facet 


class(traj_data$group) # "ordered" "factor" 

traj_data$group <- factor(traj_data$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)

levels(traj_data$group)
# [1] "0"   "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14" 
# [16] "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29" 
# [31] "30"  "31"  "32"  "33"  "34"  "35"  "36"  "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44" 
# [46] "45"  "Ref"

length(levels(traj_data$group)) #47

x_labels <- c("0","","","","","","","","","","10",
              "","","","","","","","","","20",
              "","","","","","","","","","30",
              "","","","","","","","","","40",
              "","","","","","Ref")
length(x_labels)
# 47
x_ticks <- c(1,0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,1)
length(x_ticks)
#47


unique(traj_data$study) # "Alcoa"

unique(anno_data$study) # "Alcoa"

 



## median reference horizontal lines ...

href_lines <- data.frame(
  #study  = rep( c("Alcoa", "Iluka-Eneabba", "South32"), times = 4 ),
  study = rep("Alcoa",times=12),
  # facet_x  = rep( c("Bray-Curtis\nASV-level",
  #                   ">0.001% - Bray-Curtis\nASV-level",
  #                   ">0.01% - Bray-Curtis\nASV-level",
  #                   ">0.1% - Bray-Curtis\nASV-level"), each = 3),
  facet_x = c("Bray-Curtis\nASV-level",
              "Bray-Curtis\nGenus-level",
              "Bray-Curtis\nFamily-level",
              "Bray-Curtis\nOrder-level",
              "Bray-Curtis\nClass-level" ,
              "Bray-Curtis\nPhylum-level",
              
              "Jaccard\nASV-level",
              "Jaccard\nGenus-level",
              "Jaccard\nFamily-level",
              "Jaccard\nOrder-level",
              "Jaccard\nClass-level" ,
              "Jaccard\nPhylum-level"
              ),
  dist_measure = rep( c("Bray-Curtis","Jaccard"), each = 6 ),
  similarity= NA
)

x <- traj_data

for (i in 1:dim(href_lines)[1]) {
  #i<-1
  sel <- which(x$study == href_lines$study[i] & x$facet_x == href_lines$facet_x[i] & x$group == "Ref")
  href_lines$similarity[i] <- median( x$similarity[sel] )
}

href_lines$NEW_newfacet_x <- NA
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
href_lines$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nGenus-level","Jaccard\nGenus-level"))
href_lines$NEW_newfacet_x[sel] <- "Genus-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nFamily-level","Jaccard\nFamily-level"))
href_lines$NEW_newfacet_x[sel] <- "Family-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nOrder-level","Jaccard\nOrder-level"))
href_lines$NEW_newfacet_x[sel] <- "Order-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nClass-level","Jaccard\nClass-level"))
href_lines$NEW_newfacet_x[sel] <- "Class-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nPhylum-level","Jaccard\nPhylum-level"))
href_lines$NEW_newfacet_x[sel] <- "Phylum-level"

href_lines$NEW_newfacet_x <- factor(href_lines$NEW_newfacet_x,
                                    levels = c("ASV-level",
                                               "Genus-level",
                                               "Family-level",
                                               "Order-level",
                                               "Class-level",
                                               "Phylum-level"
                                    ),
                                    ordered = TRUE)


## fit logarithmic models for each data type (facet_x)



## BOOTSTRAP MODELS ...

pred_models_boot <- data.frame(study=NA, #
                               facet_x=NA, #
                               study_and_metric=NA, # combine the two values above
                               group=NA,
                               group_num=NA,
                               similarity=NA,
                               boot_no=NA,                 # EXTRA
                               ref_median=NA,              # EXTRA
                               pred_time_to_ref_median=NA, # EXTRA
                               pred_type=NA)     # c("rectangular hyperbola", "logarithmic")
pred_models_boot.temp <- pred_models_boot


studies <- c("Alcoa")

metrics <- c("Bray-Curtis\nASV-level",
             "Bray-Curtis\nGenus-level",
             "Bray-Curtis\nFamily-level",
             "Bray-Curtis\nOrder-level",
             "Bray-Curtis\nClass-level" ,
             "Bray-Curtis\nPhylum-level",
             
             "Jaccard\nASV-level",
             "Jaccard\nGenus-level",
             "Jaccard\nFamily-level",
             "Jaccard\nOrder-level",
             "Jaccard\nClass-level" ,
             "Jaccard\nPhylum-level"
)
pred_types <- c("rectangular hyperbola", "negative exponential", "logarithmic")
## run logarithmic model only !!!
k<-3



x <- traj_data
dim(x) # 7560  10

for (i in 1:length(studies)) {
  #i<-1
  this_study <- studies[i]
  for (j in 1:length(metrics)) {
    #j<-1
    this_metric <- metrics[j]
    
    # prepare data for modelling
    subsel <- which(x$facet_x == this_metric & x$study==this_study)
    newdat <- x[subsel, c("similarity","group") ]
    newdat$group_num <- as.numeric(as.character(newdat$group))
    # remove 'Ref' similarities which have now been converted to NA
    sel.na <- which(is.na(newdat$group_num))
    
    this_ref_median <- median(newdat$similarity[sel.na])
    
    newdat <- newdat[-sel.na, ]
    

    # prep output dataframe for prediction model to populate
    sel.dup <- which(duplicated(newdat$group)==TRUE)
    new_preds <- data.frame(matrix( nrow=dim(newdat[-sel.dup, ])[1], ncol = length(names(pred_models_boot.temp)) ))
    names(new_preds) <- names(pred_models_boot.temp)
    new_preds$group_num <- sort( newdat$group_num[-sel.dup] )
    new_preds$study <- this_study
    new_preds$facet_x <- this_metric
    new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$facet_x)
    new_preds$group <- as.character(new_preds$group_num)
    new_preds$ref_median <- this_ref_median
    
    #for (k in 1:length(pred_types)) {

    
    this_pred_model_type <- pred_types[k]
    
    ## NOTE:
    #  rectangular hyperbola ( fct = MM.2() from ‘drc' package; https://www.statforbiology.com/nonlinearregression/usefulequations#michaelis-menten_equation), and 
    #  negative exponential ( fct = DRC.negExp() from 'aomisc' package; https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function)
    #  models were trialled however these regularly failed to achieve model fits.
    
    
    # NOW using "logarithmic" model only !!!!!
    
    # negative exponential model, using fct = DRC.negExp() from 'aomisc' package
    # https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function
    
    
    for (b in 1:100) {
      #b<-20
      new_preds$boot_no <- b
      set.seed(b + 1234)
      sel.boot <- sample(x = c(1:dim(newdat)[1]), size = dim(newdat)[1], replace = TRUE)
      
      bootdat <- newdat[sel.boot, ]
      
      # exclude outlying data points from trajectory data-cloud ... 
      # as outliers are likely to throw out model-fitting
      bp <- boxplot(bootdat$similarity, print=FALSE)
      sel.outlier <- which(bootdat$similarity %in% bp$out)
      if (length(sel.outlier)>0) {
        bootdat <- bootdat[-sel.outlier, ]
      }
      
      #plot(bootdat$group_num, bootdat$similarity)
      
      model <- drm(similarity ~ group_num , fct = DRC.logCurve(), # DRC.negExp()
                   data = bootdat)
      
      new_preds$similarity <- predict(model, new_preds )
      new_preds$pred_type <- this_pred_model_type
      
      y <- data.frame(group_num=1:500, sim_pred=NA)
      y$sim_pred <- predict(model, y )
      y$reached_target <- y$sim_pred >= this_ref_median
      #plot(y$group_num, y$sim_pred)
      row.names(y)
      
      sel <- which(y$reached_target == TRUE)
      
      if (length(sel) > 0) {
        years_to_ref <- min(sel)
      } else {
        years_to_ref <- NA
      }
      new_preds$pred_time_to_ref_median <- years_to_ref
      
      # add predictions from end of last year to 40 years ...
      max_year <- max(new_preds$group_num)
      year <- max_year + 1
      while (year < 40) {
        new_preds <- rbind(new_preds, new_preds[ dim(new_preds)[1], ])
        year <- year + 1
        new_preds$group_num[dim(new_preds)[1]] <- year
        new_preds$group[dim(new_preds)[1]] <- as.character( new_preds$group_num[dim(new_preds)[1]] )
        new_preds$similarity[dim(new_preds)[1]] <- y[ year, "sim_pred"]
      }
      
      pred_models_boot <- rbind(pred_models_boot, new_preds)
      
    } # END bootstrap
    

    print(paste0("completed study: ",studies[i],"; metric: ",metrics[j],"; pred_type: ",pred_types[k]))  
    #} # END pred_types
  } # END metrics
} # END studies


# remove NA first row
pred_models_boot <- pred_models_boot[-1, ]

unique(pred_models_boot$facet_x)
# [1] "Bray-Curtis\nASV-level"    "Bray-Curtis\nGenus-level"  "Bray-Curtis\nFamily-level" "Bray-Curtis\nOrder-level"  "Bray-Curtis\nClass-level" 
# [6] "Bray-Curtis\nPhylum-level" "Jaccard\nASV-level"        "Jaccard\nGenus-level"      "Jaccard\nFamily-level"     "Jaccard\nOrder-level"     
# [11] "Jaccard\nClass-level"      "Jaccard\nPhylum-level"   

dim(pred_models_boot) # 19200    10

pred_models_boot$NEW_newfacet_x <- NA
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nGenus-level","Jaccard\nGenus-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Genus-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nFamily-level","Jaccard\nFamily-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Family-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nOrder-level","Jaccard\nOrder-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Order-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nClass-level","Jaccard\nClass-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Class-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nPhylum-level","Jaccard\nPhylum-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Phylum-level"

pred_models_boot$NEW_newfacet_x <- factor(pred_models_boot$NEW_newfacet_x,
                                    levels = c("ASV-level",
                                               "Genus-level",
                                               "Family-level",
                                               "Order-level",
                                               "Class-level",
                                               "Phylum-level"
                                    ),
                                    ordered = TRUE)

# correct time prediction to ref; update NA to (>) 500 yrs
sel <- which(is.na(pred_models_boot$pred_time_to_ref_median))
if (length(sel)>1) { pred_models_boot$pred_time_to_ref_median[sel] <- 500 }

pred_models_boot$group <- factor(pred_models_boot$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)
table(pred_models_boot$group)

pred_models_boot$dist_measure <- NA
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nASV-level","Bray-Curtis\nGenus-level","Bray-Curtis\nFamily-level",
                                             "Bray-Curtis\nOrder-level","Bray-Curtis\nClass-level","Bray-Curtis\nPhylum-level"))
pred_models_boot$dist_measure[sel] <- "Bray-Curtis"
sel <- which(pred_models_boot$facet_x %in% c("Jaccard\nASV-level","Jaccard\nGenus-level","Jaccard\nFamily-level",
                                             "Jaccard\nOrder-level","Jaccard\nClass-level","Jaccard\nPhylum-level"))
pred_models_boot$dist_measure[sel] <- "Jaccard"


str(pred_models_boot)

str(traj_data)


p <- ggplot(data=traj_data, aes(x=group, y=similarity)) +
  geom_line(data=pred_models_boot, aes(  x=group, y=similarity, group = boot_no ), colour = "#e34a33", alpha = 0.02) + # "red"
  geom_hline(data = href_lines, aes(yintercept = similarity),  color= "#2c7fb8", linetype="longdash") + # "blue"
  #geom_line(data= pred_models_alldata, aes(  x=group, y=similarity, group = 1), colour = "#e34a33" ) + # "red"
  
  geom_boxplot(outlier.shape = NA, width = 2) +
  #geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  
  theme_bw() +
  geom_text(data=anno_data, aes(x=group, y=similarity, label=sig_letter),nudge_y = 2, size=2.75) +
  scale_x_discrete(drop=FALSE, expand = c(0.05,0.05),
                   labels = x_labels) +
  theme(axis.text.x  = element_text(angle=90,hjust=1, vjust=0.5), # 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_line(linetype = 0)
        axis.ticks.x = element_line(linetype = x_ticks)
  ) +
  labs(x = "Rehabilitation age (years)", y = "Similarity to Reference (%)") +
  facet_grid(rows = vars(dist_measure), cols = vars(NEW_newfacet_x), scales = "fixed")

p


grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Restoration-trajectories--Alcoa--Bray-Jaccard-ASV-Genus-Family-Order-Class-Phylum-PRUNED-with-Preds--A.tiff"), width = 18, height = 10, units = "cm", res = 600, compression = "lzw",type="cairo")



str(traj_data)
unique(traj_data$study) # "Alcoa"
unique(traj_data$group) # use "ref" only
unique(traj_data$dist_measure) # "Bray-Curtis" "Jaccard"
unique(traj_data$tax_level)
# [1] "ASV (with Tree)"             "Genus (with Tree & pruned)"  "Family (with Tree & pruned)" "Order (with Tree & pruned)" 
# [5] "Class (with Tree & pruned)"  "Phylum (with Tree & pruned)"

## Bray-Curtis
sel <- which(traj_data$dist_measure == "Bray-Curtis")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 22.38   31.28   35.55   35.79   39.93   60.91 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Genus (with Tree & pruned)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 53.17   68.23   72.66   72.17   76.16   89.95 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Family (with Tree & pruned)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 59.93   74.90   78.23   77.81   81.49   92.22 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Order (with Tree & pruned)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 67.74   79.65   82.06   81.89   85.15   92.87 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Class (with Tree & pruned)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 71.27   85.61   88.70   87.91   90.95   94.89
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Phylum (with Tree & pruned)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 75.39   88.67   90.84   90.25   92.95   96.79 

## Jaccard
sel <- which(traj_data$dist_measure == "Jaccard")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 9.503  17.824  20.345  20.092  23.213  35.809 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Genus (with Tree & pruned)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 41.26   65.28   68.01   65.82   70.07   76.09 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Family (with Tree & pruned)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 53.25   76.57   79.14   77.29   81.50   86.06 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Order (with Tree & pruned)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 51.88   75.20   77.94   75.76   79.71   85.94 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Class (with Tree & pruned)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 48.44   74.55   79.66   77.44   84.13   91.23 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Phylum (with Tree & pruned)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 58.62   75.86   79.31   79.26   83.33   96.15 



str(pred_models_boot)


unique(pred_models_boot$study_and_metric)
# [1] "Alcoa__Bray-Curtis\nASV-level"    "Alcoa__Bray-Curtis\nGenus-level"  "Alcoa__Bray-Curtis\nFamily-level" "Alcoa__Bray-Curtis\nOrder-level" 
# [5] "Alcoa__Bray-Curtis\nClass-level"  "Alcoa__Bray-Curtis\nPhylum-level" "Alcoa__Jaccard\nASV-level"        "Alcoa__Jaccard\nGenus-level"     
# [9] "Alcoa__Jaccard\nFamily-level"     "Alcoa__Jaccard\nOrder-level"      "Alcoa__Jaccard\nClass-level"      "Alcoa__Jaccard\nPhylum-level"

pred_models_boot$study_and_metric <- factor(pred_models_boot$study_and_metric,
                                            levels = c(
                                              "Alcoa__Bray-Curtis\nASV-level",
                                              "Alcoa__Bray-Curtis\nGenus-level",
                                              "Alcoa__Bray-Curtis\nFamily-level",
                                              "Alcoa__Bray-Curtis\nOrder-level" ,
                                              "Alcoa__Bray-Curtis\nClass-level" ,
                                              "Alcoa__Bray-Curtis\nPhylum-level",
                                              
                                              "Alcoa__Jaccard\nASV-level"  ,
                                              "Alcoa__Jaccard\nGenus-level" ,
                                              "Alcoa__Jaccard\nFamily-level" ,
                                              "Alcoa__Jaccard\nOrder-level" ,
                                              "Alcoa__Jaccard\nClass-level" ,
                                              "Alcoa__Jaccard\nPhylum-level"
                                            ), ordered = TRUE)

dim(pred_models_boot) # 19200    12

dat.sum <- pred_models_boot
head(dat.sum)
# study                facet_x              study_and_metric group group_num similarity boot_no ref_median
# 11 Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level     2         2   9.389107       1   35.55047
# 2  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level     8         8  21.149581       1   35.55047
# 3  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    14        14  25.897020       1   35.55047
# 4  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    17        17  27.544121       1   35.55047
# 5  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    25        25  30.815846       1   35.55047
# 6  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    29        29  32.074951       1   35.55047
# pred_time_to_ref_median   pred_type NEW_newfacet_x dist_measure
# 11                      44 logarithmic      ASV-level  Bray-Curtis
# 2                       44 logarithmic      ASV-level  Bray-Curtis
# 3                       44 logarithmic      ASV-level  Bray-Curtis
# 4                       44 logarithmic      ASV-level  Bray-Curtis
# 5                       44 logarithmic      ASV-level  Bray-Curtis
# 6                       44 logarithmic      ASV-level  Bray-Curtis

for (i in 1:length(levels(dat.sum$study_and_metric))) {
  #i<-1
  this_study_and_metric <- levels(dat.sum$study_and_metric)[i]
  sel <- which(dat.sum$study_and_metric == this_study_and_metric)
  #unique(dat.sum$ref_median[sel])
  temp <- dat.sum$pred_time_to_ref_median[sel]
  #quantile(temp, probs = c(0.05, 0.25, 0.5, 0.75, 0.95), na.rm = TRUE)
  #hist(temp)
  res <- quantile(temp, probs = c(0.05, 0.5, 0.95), na.rm = TRUE)
  names(res)
  #"5%"  "50%" "95%"
  #print(paste0(this_study_and_metric))
  print(paste0(this_study_and_metric,": ",res["50%"]," (",res["5%"],", ",res["95%"],") yr."))
}
# [1] "Alcoa__Bray-Curtis\nASV-level: 43 (39, 49) yr."
# [1] "Alcoa__Bray-Curtis\nGenus-level: 34 (31, 37) yr."
# [1] "Alcoa__Bray-Curtis\nFamily-level: 33 (31, 38.05) yr."
# [1] "Alcoa__Bray-Curtis\nOrder-level: 35 (32, 40) yr."
# [1] "Alcoa__Bray-Curtis\nClass-level: 50 (44.95, 61.05) yr."
# [1] "Alcoa__Bray-Curtis\nPhylum-level: 51 (44.95, 65) yr."
# [1] "Alcoa__Jaccard\nASV-level: 60.5 (51, 79) yr."
# [1] "Alcoa__Jaccard\nGenus-level: 487.5 (234.35, 500) yr."
# [1] "Alcoa__Jaccard\nFamily-level: 198 (139.25, 353.099999999998) yr."
# [1] "Alcoa__Jaccard\nOrder-level: 258 (155.85, 500) yr."
# [1] "Alcoa__Jaccard\nClass-level: 500 (500, 500) yr."
# [1] "Alcoa__Jaccard\nPhylum-level: 500 (95.8500000000001, 500) yr."


## plot time to target

names(pred_models_boot)
# [1] "study"                   "facet_x"                 "study_and_metric"        "group"                   "group_num"              
# [6] "similarity"              "boot_no"                 "ref_median"              "pred_time_to_ref_median" "pred_type"              
# [11] "NEW_newfacet_x"          "dist_measure" 

p <- ggplot(data=pred_models_boot, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median, color = dist_measure)) +
  geom_boxplot() +
  scale_color_manual(values = c("Bray-Curtis" = "#d73027", "Jaccard" = "#4575b4"), name = "Distance\nmeasure") +
  theme_bw() +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        
        strip.background = element_blank(),
        strip.text.x = element_blank()
  ) +
  facet_wrap(facets = vars(dist_measure), scales = "free_y" ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Predicted-time-to-target--Alcoa--Bray-Jaccard-ASV-Genus-Family-Order-Class-Phylum-PRUNED.tiff"), width = 14, height = 10, units = "cm", dpi = 600, compression = "lzw",type="cairo")


df <- pred_models_boot
df$dist_measure <- factor(df$dist_measure, levels=c("Jaccard","Bray-Curtis"), ordered = TRUE)
measures.new <- c(`Bray-Curtis`="Bray-Curtis\nPruned", Jaccard="Jaccard\nPruned" )


p <- ggplot(data=df, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median)) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()#,
       ) +
  facet_wrap(facets = vars(dist_measure), scales = "free_y" ,labeller = labeller(dist_measure = measures.new)) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p

grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Predicted-time-to-target--Alcoa--Bray-Jaccard-ASV-Genus-Family-Order-Class-Phylum-PRUNED.tiff"), width = 10, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")


#-------------------------

#### facet_grid trajectory plots + predict time to reach ref:
##   ASV -- Jaccard + Bray-Curtis -- Genus + Family + ... Phylum -- Alcoa (NOT PRUNED)
#-------------------------

names(df_results)

 

names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"    
# [7] "group"          "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"  


keep_results <- c("Alcoa-Bray-ASV-withTree",
                  "Alcoa-Bray-Genus-withTree",
                  "Alcoa-Bray-Family-withTree" ,
                  "Alcoa-Bray-Order-withTree" ,
                  "Alcoa-Bray-Class-withTree" ,
                  "Alcoa-Bray-Phylum-withTree",
                  
                  "Alcoa-Jaccard-ASV-withTree",
                  "Alcoa-Jaccard-Genus-withTree",
                  "Alcoa-Jaccard-Family-withTree" ,
                  "Alcoa-Jaccard-Order-withTree",
                  "Alcoa-Jaccard-Class-withTree",
                  "Alcoa-Jaccard-Phylum-withTree"
)


keep_cols <- c("samp","compare_with","similarity","group","dist_measure","tax_level","study","corrected","facet_x")
traj_data <- df_results[[ keep_results[1] ]][ ,keep_cols]
for (i in 2:length(keep_results)) {
  traj_data <- rbind(traj_data,df_results[[ keep_results[i] ]][ ,keep_cols])
  print(paste0("completed ",i))
}
dim(traj_data) # 7560    9
unique(traj_data$facet_x) #
# [1] "Bray-Curtis\nASV-level"    "Bray-Curtis\nGenus-level"  "Bray-Curtis\nFamily-level" "Bray-Curtis\nOrder-level"  "Bray-Curtis\nClass-level" 
# [6] "Bray-Curtis\nPhylum-level" "Jaccard\nASV-level"        "Jaccard\nGenus-level"      "Jaccard\nFamily-level"     "Jaccard\nOrder-level"     
# [11] "Jaccard\nClass-level"      "Jaccard\nPhylum-level"

class(traj_data$facet_x) # "character"
traj_data$facet_x <- factor(traj_data$facet_x,
                            levels = c("Bray-Curtis\nASV-level",
                                       "Bray-Curtis\nGenus-level",
                                       "Bray-Curtis\nFamily-level",
                                       "Bray-Curtis\nOrder-level",
                                       "Bray-Curtis\nClass-level" ,
                                       "Bray-Curtis\nPhylum-level",
                                       
                                       "Jaccard\nASV-level",
                                       "Jaccard\nGenus-level",
                                       "Jaccard\nFamily-level",
                                       "Jaccard\nOrder-level",
                                       "Jaccard\nClass-level" ,
                                       "Jaccard\nPhylum-level"
                            ),
                            ordered = TRUE)

traj_data$NEW_newfacet_x <- NA
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
traj_data$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nGenus-level","Jaccard\nGenus-level"))
traj_data$NEW_newfacet_x[sel] <- "Genus-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nFamily-level","Jaccard\nFamily-level"))
traj_data$NEW_newfacet_x[sel] <- "Family-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nOrder-level","Jaccard\nOrder-level"))
traj_data$NEW_newfacet_x[sel] <- "Order-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nClass-level","Jaccard\nClass-level"))
traj_data$NEW_newfacet_x[sel] <- "Class-level"
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nPhylum-level","Jaccard\nPhylum-level"))
traj_data$NEW_newfacet_x[sel] <- "Phylum-level"

traj_data$NEW_newfacet_x <- factor(traj_data$NEW_newfacet_x,
                                   levels = c("ASV-level",
                                              "Genus-level",
                                              "Family-level",
                                              "Order-level",
                                              "Class-level",
                                              "Phylum-level"
                                   ),
                                   ordered = TRUE)


keep_cols_anno <- c("group","sig_letter","similarity","dist_measure","tax_level","study","corrected","facet_x")
anno_data <- df_anno[[ keep_results[1] ]][ ,keep_cols_anno]
for (i in 2:length(keep_results)) {
  anno_data <- rbind(anno_data,df_anno[[ keep_results[i] ]][ ,keep_cols_anno])
  print(paste0("completed ",i))
}
dim(anno_data) # 84  8
class(anno_data$facet_x) # "character"
unique(anno_data$facet_x) # "character"
# [1] "Bray-Curtis\nASV-level"    "Bray-Curtis\nGenus-level"  "Bray-Curtis\nFamily-level" "Bray-Curtis\nOrder-level"  "Bray-Curtis\nClass-level" 
# [6] "Bray-Curtis\nPhylum-level" "Jaccard\nASV-level"        "Jaccard\nGenus-level"      "Jaccard\nFamily-level"     "Jaccard\nOrder-level"     
# [11] "Jaccard\nClass-level"      "Jaccard\nPhylum-level" 



anno_data$NEW_newfacet_x <- NA
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
anno_data$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nGenus-level","Jaccard\nGenus-level"))
anno_data$NEW_newfacet_x[sel] <- "Genus-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nFamily-level","Jaccard\nFamily-level"))
anno_data$NEW_newfacet_x[sel] <- "Family-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nOrder-level","Jaccard\nOrder-level"))
anno_data$NEW_newfacet_x[sel] <- "Order-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nClass-level","Jaccard\nClass-level"))
anno_data$NEW_newfacet_x[sel] <- "Class-level"
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nPhylum-level","Jaccard\nPhylum-level"))
anno_data$NEW_newfacet_x[sel] <- "Phylum-level"

anno_data$NEW_newfacet_x <- factor(anno_data$NEW_newfacet_x,
                                   levels = c("ASV-level",
                                              "Genus-level",
                                              "Family-level",
                                              "Order-level",
                                              "Class-level",
                                              "Phylum-level"
                                   ),
                                   ordered = TRUE)

names(anno_data)
#  "group"          "sig_letter"     "similarity"     "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"        "NEW_newfacet_x"

# use $dist_measure - i.e. Bray-Curtis or Jaccard as y-facet 


class(traj_data$group) # "ordered" "factor" 

traj_data$group <- factor(traj_data$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)

levels(traj_data$group)
# [1] "0"   "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14" 
# [16] "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29" 
# [31] "30"  "31"  "32"  "33"  "34"  "35"  "36"  "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44" 
# [46] "45"  "Ref"

length(levels(traj_data$group)) #47

x_labels <- c("0","","","","","","","","","","10",
              "","","","","","","","","","20",
              "","","","","","","","","","30",
              "","","","","","","","","","40",
              "","","","","","Ref")
length(x_labels)
# 47
x_ticks <- c(1,0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,1)
length(x_ticks)
#47


unique(traj_data$study) # "Alcoa"

unique(anno_data$study) # "Alcoa"

 



## median reference horizontal lines ...

href_lines <- data.frame(
  #study  = rep( c("Alcoa", "Iluka-Eneabba", "South32"), times = 4 ),
  study = rep("Alcoa",times=12),
  # facet_x  = rep( c("Bray-Curtis\nASV-level",
  #                   ">0.001% - Bray-Curtis\nASV-level",
  #                   ">0.01% - Bray-Curtis\nASV-level",
  #                   ">0.1% - Bray-Curtis\nASV-level"), each = 3),
  facet_x = c("Bray-Curtis\nASV-level",
              "Bray-Curtis\nGenus-level",
              "Bray-Curtis\nFamily-level",
              "Bray-Curtis\nOrder-level",
              "Bray-Curtis\nClass-level" ,
              "Bray-Curtis\nPhylum-level",
              
              "Jaccard\nASV-level",
              "Jaccard\nGenus-level",
              "Jaccard\nFamily-level",
              "Jaccard\nOrder-level",
              "Jaccard\nClass-level" ,
              "Jaccard\nPhylum-level"
  ),
  dist_measure = rep( c("Bray-Curtis","Jaccard"), each = 6 ),
  similarity= NA
)

x <- traj_data

for (i in 1:dim(href_lines)[1]) {
  #i<-1
  sel <- which(x$study == href_lines$study[i] & x$facet_x == href_lines$facet_x[i] & x$group == "Ref")
  href_lines$similarity[i] <- median( x$similarity[sel] )
}

href_lines$NEW_newfacet_x <- NA
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
href_lines$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nGenus-level","Jaccard\nGenus-level"))
href_lines$NEW_newfacet_x[sel] <- "Genus-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nFamily-level","Jaccard\nFamily-level"))
href_lines$NEW_newfacet_x[sel] <- "Family-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nOrder-level","Jaccard\nOrder-level"))
href_lines$NEW_newfacet_x[sel] <- "Order-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nClass-level","Jaccard\nClass-level"))
href_lines$NEW_newfacet_x[sel] <- "Class-level"
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nPhylum-level","Jaccard\nPhylum-level"))
href_lines$NEW_newfacet_x[sel] <- "Phylum-level"

href_lines$NEW_newfacet_x <- factor(href_lines$NEW_newfacet_x,
                                    levels = c("ASV-level",
                                               "Genus-level",
                                               "Family-level",
                                               "Order-level",
                                               "Class-level",
                                               "Phylum-level"
                                    ),
                                    ordered = TRUE)


## fit logarithmic models for each data type (facet_x)



## BOOTSTRAP MODELS ...

pred_models_boot <- data.frame(study=NA, #
                               facet_x=NA, #
                               study_and_metric=NA, # combine the two values above
                               group=NA,
                               group_num=NA,
                               similarity=NA,
                               boot_no=NA,                 # EXTRA
                               ref_median=NA,              # EXTRA
                               pred_time_to_ref_median=NA, # EXTRA
                               pred_type=NA)     # c("rectangular hyperbola", "logarithmic")
pred_models_boot.temp <- pred_models_boot


studies <- c("Alcoa")

metrics <- c("Bray-Curtis\nASV-level",
             "Bray-Curtis\nGenus-level",
             "Bray-Curtis\nFamily-level",
             "Bray-Curtis\nOrder-level",
             "Bray-Curtis\nClass-level" ,
             "Bray-Curtis\nPhylum-level",
             
             "Jaccard\nASV-level",
             "Jaccard\nGenus-level",
             "Jaccard\nFamily-level",
             "Jaccard\nOrder-level",
             "Jaccard\nClass-level" ,
             "Jaccard\nPhylum-level"
)
pred_types <- c("rectangular hyperbola", "negative exponential", "logarithmic")
## run logarithmic model only !!!
k<-3



x <- traj_data
dim(x) # 7560  10

for (i in 1:length(studies)) {
  #i<-1
  this_study <- studies[i]
  for (j in 1:length(metrics)) {
    #j<-1
    this_metric <- metrics[j]
    
    # prepare data for modelling
    subsel <- which(x$facet_x == this_metric & x$study==this_study)
    newdat <- x[subsel, c("similarity","group") ]
    newdat$group_num <- as.numeric(as.character(newdat$group))
    # remove 'Ref' similarities which have now been converted to NA
    sel.na <- which(is.na(newdat$group_num))
    
    this_ref_median <- median(newdat$similarity[sel.na])
    
    newdat <- newdat[-sel.na, ]
    

    # prep output dataframe for prediction model to populate
    sel.dup <- which(duplicated(newdat$group)==TRUE)
    new_preds <- data.frame(matrix( nrow=dim(newdat[-sel.dup, ])[1], ncol = length(names(pred_models_boot.temp)) ))
    names(new_preds) <- names(pred_models_boot.temp)
    new_preds$group_num <- sort( newdat$group_num[-sel.dup] )
    new_preds$study <- this_study
    new_preds$facet_x <- this_metric
    new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$facet_x)
    new_preds$group <- as.character(new_preds$group_num)
    new_preds$ref_median <- this_ref_median
    
    #for (k in 1:length(pred_types)) {

    this_pred_model_type <- pred_types[k]
    
    ## NOTE:
    #  rectangular hyperbola ( fct = MM.2() from ‘drc' package; https://www.statforbiology.com/nonlinearregression/usefulequations#michaelis-menten_equation), and 
    #  negative exponential ( fct = DRC.negExp() from 'aomisc' package; https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function)
    #  models were trialled however these regularly failed to achieve model fits.
    
    
    # NOW using "logarithmic" model only !!!!!
    
    # negative exponential model, using fct = DRC.negExp() from 'aomisc' package
    # https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function
    
    
    for (b in 1:100) {
      #b<-20
      new_preds$boot_no <- b
      set.seed(b + 1234)
      sel.boot <- sample(x = c(1:dim(newdat)[1]), size = dim(newdat)[1], replace = TRUE)
      
      bootdat <- newdat[sel.boot, ]
      
      # exclude outlying data points from trajectory data-cloud ... 
      # as outliers are likely to throw out model-fitting
      bp <- boxplot(bootdat$similarity, print=FALSE)
      sel.outlier <- which(bootdat$similarity %in% bp$out)
      if (length(sel.outlier)>0) {
        bootdat <- bootdat[-sel.outlier, ]
      }
      
      #plot(bootdat$group_num, bootdat$similarity)
      
      model <- drm(similarity ~ group_num , fct = DRC.logCurve(), # DRC.negExp()
                   data = bootdat)
      
      new_preds$similarity <- predict(model, new_preds )
      new_preds$pred_type <- this_pred_model_type
      
      y <- data.frame(group_num=1:500, sim_pred=NA)
      y$sim_pred <- predict(model, y )
      y$reached_target <- y$sim_pred >= this_ref_median
      #plot(y$group_num, y$sim_pred)
      row.names(y)
      
      sel <- which(y$reached_target == TRUE)
      
      if (length(sel) > 0) {
        years_to_ref <- min(sel)
      } else {
        years_to_ref <- NA
      }
      new_preds$pred_time_to_ref_median <- years_to_ref
      
      # add predictions from end of last year to 40 years ...
      max_year <- max(new_preds$group_num)
      year <- max_year + 1
      while (year < 40) {
        new_preds <- rbind(new_preds, new_preds[ dim(new_preds)[1], ])
        year <- year + 1
        new_preds$group_num[dim(new_preds)[1]] <- year
        new_preds$group[dim(new_preds)[1]] <- as.character( new_preds$group_num[dim(new_preds)[1]] )
        new_preds$similarity[dim(new_preds)[1]] <- y[ year, "sim_pred"]
      }
      
      pred_models_boot <- rbind(pred_models_boot, new_preds)
      
    } # END bootstrap
    

    print(paste0("completed study: ",studies[i],"; metric: ",metrics[j],"; pred_type: ",pred_types[k]))  
    #} # END pred_types
  } # END metrics
} # END studies


# remove NA first row
pred_models_boot <- pred_models_boot[-1, ]

unique(pred_models_boot$facet_x)
# [1] "Bray-Curtis\nASV-level"    "Bray-Curtis\nGenus-level"  "Bray-Curtis\nFamily-level" "Bray-Curtis\nOrder-level"  "Bray-Curtis\nClass-level" 
# [6] "Bray-Curtis\nPhylum-level" "Jaccard\nASV-level"        "Jaccard\nGenus-level"      "Jaccard\nFamily-level"     "Jaccard\nOrder-level"     
# [11] "Jaccard\nClass-level"      "Jaccard\nPhylum-level"   

dim(pred_models_boot) # 19200  10

pred_models_boot$NEW_newfacet_x <- NA
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "ASV-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nGenus-level","Jaccard\nGenus-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Genus-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nFamily-level","Jaccard\nFamily-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Family-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nOrder-level","Jaccard\nOrder-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Order-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nClass-level","Jaccard\nClass-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Class-level"
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nPhylum-level","Jaccard\nPhylum-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "Phylum-level"

pred_models_boot$NEW_newfacet_x <- factor(pred_models_boot$NEW_newfacet_x,
                                          levels = c("ASV-level",
                                                     "Genus-level",
                                                     "Family-level",
                                                     "Order-level",
                                                     "Class-level",
                                                     "Phylum-level"
                                          ),
                                          ordered = TRUE)

# correct time prediction to ref; update NA to (>) 500 yrs
sel <- which(is.na(pred_models_boot$pred_time_to_ref_median))
if (length(sel)>1) { pred_models_boot$pred_time_to_ref_median[sel] <- 500 }

pred_models_boot$group <- factor(pred_models_boot$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)
table(pred_models_boot$group)

pred_models_boot$dist_measure <- NA
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nASV-level","Bray-Curtis\nGenus-level","Bray-Curtis\nFamily-level",
                                             "Bray-Curtis\nOrder-level","Bray-Curtis\nClass-level","Bray-Curtis\nPhylum-level"))
pred_models_boot$dist_measure[sel] <- "Bray-Curtis"
sel <- which(pred_models_boot$facet_x %in% c("Jaccard\nASV-level","Jaccard\nGenus-level","Jaccard\nFamily-level",
                                             "Jaccard\nOrder-level","Jaccard\nClass-level","Jaccard\nPhylum-level"))
pred_models_boot$dist_measure[sel] <- "Jaccard"


str(pred_models_boot)


str(traj_data)



p <- ggplot(data=traj_data, aes(x=group, y=similarity)) +
  geom_line(data=pred_models_boot, aes(  x=group, y=similarity, group = boot_no ), colour = "#e34a33", alpha = 0.02) + # "red"
  geom_hline(data = href_lines, aes(yintercept = similarity),  color= "#2c7fb8", linetype="longdash") + # "blue"
  #geom_line(data= pred_models_alldata, aes(  x=group, y=similarity, group = 1), colour = "#e34a33" ) + # "red"
  
  geom_boxplot(outlier.shape = NA, width = 2) +
  #geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  
  theme_bw() +
  geom_text(data=anno_data, aes(x=group, y=similarity, label=sig_letter),nudge_y = 2, size=2.75) +
  scale_x_discrete(drop=FALSE, expand = c(0.05,0.05),
                   labels = x_labels) +
  theme(axis.text.x  = element_text(angle=90,hjust=1, vjust=0.5), # 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_line(linetype = 0)
        axis.ticks.x = element_line(linetype = x_ticks)
  ) +
  labs(x = "Rehabilitation age (years)", y = "Similarity to Reference (%)") +
  facet_grid(rows = vars(dist_measure), cols = vars(NEW_newfacet_x), scales = "fixed")

p


grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Restoration-trajectories--Alcoa--Bray-Jaccard-ASV-Genus-Family-Order-Class-Phylum-NON-PRUNED-B-with-Preds.tiff"), width = 18, height = 10, units = "cm", res = 600, compression = "lzw",type="cairo")


str(traj_data)
unique(traj_data$study) # "Alcoa"
unique(traj_data$group) # use "ref" only
unique(traj_data$dist_measure) # "Bray-Curtis" "Jaccard"
unique(traj_data$tax_level)
# "ASV (with Tree)"    "Genus (with Tree)"  "Family (with Tree)" "Order (with Tree)"  "Class (with Tree)"  "Phylum (with Tree)"

## Bray-Curtis
sel <- which(traj_data$dist_measure == "Bray-Curtis")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 22.38   31.28   35.55   35.79   39.93   60.91 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Genus (with Tree)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 55.91   69.41   73.62   72.88   76.54   89.52 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Family (with Tree)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 59.66   73.73   77.44   76.82   80.40   91.30 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Order (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 66.71   79.02   81.61   81.42   84.30   92.22 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Class (with Tree)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 71.59   85.44   88.34   87.64   90.69   94.76 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Phylum (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 75.39   88.67   90.84   90.25   92.95   96.79 

## Jaccard
sel <- which(traj_data$dist_measure == "Jaccard")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 9.503  17.824  20.345  20.092  23.213  35.809 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Genus (with Tree)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 41.65   66.60   68.49   66.49   70.59   76.39 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Family (with Tree)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 49.19   71.71   73.96   72.05   76.38   82.22 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Order (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 48.26   70.99   74.32   71.84   76.24   83.72 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Class (with Tree)" )
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 46.58   71.64   76.06   73.85   78.95   90.32 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "Phylum (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 58.62   75.86   79.31   79.26   83.33   96.15 



str(pred_models_boot)

unique(pred_models_boot$study_and_metric)
# [1] "Alcoa__Bray-Curtis\nASV-level"    "Alcoa__Bray-Curtis\nGenus-level"  "Alcoa__Bray-Curtis\nFamily-level" "Alcoa__Bray-Curtis\nOrder-level" 
# [5] "Alcoa__Bray-Curtis\nClass-level"  "Alcoa__Bray-Curtis\nPhylum-level" "Alcoa__Jaccard\nASV-level"        "Alcoa__Jaccard\nGenus-level"     
# [9] "Alcoa__Jaccard\nFamily-level"     "Alcoa__Jaccard\nOrder-level"      "Alcoa__Jaccard\nClass-level"      "Alcoa__Jaccard\nPhylum-level"

pred_models_boot$study_and_metric <- factor(pred_models_boot$study_and_metric,
                                            levels = c(
                                              "Alcoa__Bray-Curtis\nASV-level",
                                              "Alcoa__Bray-Curtis\nGenus-level",
                                              "Alcoa__Bray-Curtis\nFamily-level",
                                              "Alcoa__Bray-Curtis\nOrder-level" ,
                                              "Alcoa__Bray-Curtis\nClass-level" ,
                                              "Alcoa__Bray-Curtis\nPhylum-level",
                                              
                                              "Alcoa__Jaccard\nASV-level"  ,
                                              "Alcoa__Jaccard\nGenus-level" ,
                                              "Alcoa__Jaccard\nFamily-level" ,
                                              "Alcoa__Jaccard\nOrder-level" ,
                                              "Alcoa__Jaccard\nClass-level" ,
                                              "Alcoa__Jaccard\nPhylum-level"
                                            ), ordered = TRUE)

dim(pred_models_boot) # 19200    12

dat.sum <- pred_models_boot
head(dat.sum)
# study                facet_x              study_and_metric group group_num similarity boot_no ref_median
# 11 Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level     2         2   9.389107       1   35.55047
# 2  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level     8         8  21.149581       1   35.55047
# 3  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    14        14  25.897020       1   35.55047
# 4  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    17        17  27.544121       1   35.55047
# 5  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    25        25  30.815846       1   35.55047
# 6  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    29        29  32.074951       1   35.55047
# pred_time_to_ref_median   pred_type NEW_newfacet_x dist_measure
# 11                      44 logarithmic      ASV-level  Bray-Curtis
# 2                       44 logarithmic      ASV-level  Bray-Curtis
# 3                       44 logarithmic      ASV-level  Bray-Curtis
# 4                       44 logarithmic      ASV-level  Bray-Curtis
# 5                       44 logarithmic      ASV-level  Bray-Curtis
# 6                       44 logarithmic      ASV-level  Bray-Curtis

for (i in 1:length(levels(dat.sum$study_and_metric))) {
  #i<-1
  this_study_and_metric <- levels(dat.sum$study_and_metric)[i]
  sel <- which(dat.sum$study_and_metric == this_study_and_metric)
  #unique(dat.sum$ref_median[sel])
  temp <- dat.sum$pred_time_to_ref_median[sel]
  #quantile(temp, probs = c(0.05, 0.25, 0.5, 0.75, 0.95), na.rm = TRUE)
  #hist(temp)
  res <- quantile(temp, probs = c(0.05, 0.5, 0.95), na.rm = TRUE)
  names(res)
  #"5%"  "50%" "95%"
  #print(paste0(this_study_and_metric))
  print(paste0(this_study_and_metric,": ",res["50%"]," (",res["5%"],", ",res["95%"],") yr."))
}
# [1] "Alcoa__Bray-Curtis\nASV-level: 43 (39, 49) yr."
# [1] "Alcoa__Bray-Curtis\nGenus-level: 37 (34, 42) yr."
# [1] "Alcoa__Bray-Curtis\nFamily-level: 36 (32.95, 41) yr."
# [1] "Alcoa__Bray-Curtis\nOrder-level: 35 (32, 39.05) yr."
# [1] "Alcoa__Bray-Curtis\nClass-level: 47 (42, 57) yr."
# [1] "Alcoa__Bray-Curtis\nPhylum-level: 51 (44.95, 65) yr."
# [1] "Alcoa__Jaccard\nASV-level: 60.5 (51, 79) yr."
# [1] "Alcoa__Jaccard\nGenus-level: 224 (120.8, 500) yr."
# [1] "Alcoa__Jaccard\nFamily-level: 263.5 (154.95, 500) yr."
# [1] "Alcoa__Jaccard\nOrder-level: 292 (150.95, 500) yr."
# [1] "Alcoa__Jaccard\nClass-level: 500 (409.9, 500) yr."
# [1] "Alcoa__Jaccard\nPhylum-level: 500 (95.8500000000001, 500) yr."


## plot time to target

names(pred_models_boot)
# [1] "study"                   "facet_x"                 "study_and_metric"        "group"                   "group_num"              
# [6] "similarity"              "boot_no"                 "ref_median"              "pred_time_to_ref_median" "pred_type"              
# [11] "NEW_newfacet_x"          "dist_measure" 

p <- ggplot(data=pred_models_boot, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median, color = dist_measure)) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        
        strip.background = element_blank(),
        strip.text.x = element_blank()
  ) +
  facet_wrap(facets = vars(dist_measure), scales = "free_y" ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Predicted-time-to-target--Alcoa--Bray-Jaccard-ASV-Genus-Family-Order-Class-PhylumNON-PRUNED.tiff"), width = 14, height = 10, units = "cm", dpi = 600, compression = "lzw",type="cairo")

df <- pred_models_boot
df$dist_measure <- factor(df$dist_measure, levels=c("Jaccard","Bray-Curtis"), ordered = TRUE)
measures.new <- c(`Bray-Curtis`="Bray-Curtis\nNon-pruned", Jaccard="Jaccard\nNon-pruned" )


p <- ggplot(data=df, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median)) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()#,
  ) +
  facet_wrap(facets = vars(dist_measure), scales = "free_y" ,labeller = labeller(dist_measure = measures.new)) +
  #labs(x = NULL, y = "Predicted time to reach target (yr)")
  labs(x = NULL, y = NULL)
p

#grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Predicted-time-to-target--Alcoa--Bray-Jaccard-ASV-Genus-Family-Order-Class-Phylum-NON-PRUNED.tiff"), width = 10, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")


#-------------------------

#### facet_grid trajectory plots + predict time to reach ref:
##   ASV -- Jaccard + Bray-Curtis -- All + >0.001 + >0.01 + >0.1 -- Alcoa
#-------------------------

names(df_results)



names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"    
# [7] "group"          "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"  


keep_results <- c("Alcoa-Bray-ASV-withTree" , 
                  "Alcoa-Bray-ASV-withTree->0.001%" ,
                  "Alcoa-Bray-ASV-withTree->0.01%" ,
                  "Alcoa-Bray-ASV-withTree->0.1%" ,
                  
                  "Alcoa-Jaccard-ASV-withTree",
                  "Alcoa-Jaccard-ASV-withTree->0.001%",
                  "Alcoa-Jaccard-ASV-withTree->0.01%",     
                  "Alcoa-Jaccard-ASV-withTree->0.1%"
)


keep_cols <- c("samp","compare_with","similarity","group","dist_measure","tax_level","study","corrected","facet_x")
traj_data <- df_results[[ keep_results[1] ]][ ,keep_cols]
for (i in 2:length(keep_results)) {
  traj_data <- rbind(traj_data,df_results[[ keep_results[i] ]][ ,keep_cols])
  print(paste0("completed ",i))
}
dim(traj_data) # 5040    9
unique(traj_data$facet_x) #
# [1] "Bray-Curtis\nASV-level"           ">0.001% - Bray-Curtis\nASV-level" ">0.01% - Bray-Curtis\nASV-level"  ">0.1% - Bray-Curtis\nASV-level"  
# [5] "Jaccard\nASV-level"               ">0.001% - Jaccard\nASV-level"     ">0.01% - Jaccard\nASV-level"      ">0.1% - Jaccard\nASV-level"

class(traj_data$facet_x) # "character"
traj_data$facet_x <- factor(traj_data$facet_x,
                            levels = c("Bray-Curtis\nASV-level",
                                       ">0.001% - Bray-Curtis\nASV-level",
                                       ">0.01% - Bray-Curtis\nASV-level",
                                       ">0.1% - Bray-Curtis\nASV-level",
                                       
                                       "Jaccard\nASV-level",
                                       ">0.001% - Jaccard\nASV-level",
                                       ">0.01% - Jaccard\nASV-level",
                                       ">0.1% - Jaccard\nASV-level"
                            ),
                            ordered = TRUE)

traj_data$NEW_newfacet_x <- NA
sel <- which(traj_data$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
traj_data$NEW_newfacet_x[sel] <- "All ASVs"
sel <- which(traj_data$facet_x %in% c(">0.001% - Bray-Curtis\nASV-level",">0.001% - Jaccard\nASV-level"))
traj_data$NEW_newfacet_x[sel] <- "ASVs >0.001%"
sel <- which(traj_data$facet_x %in% c(">0.01% - Bray-Curtis\nASV-level",">0.01% - Jaccard\nASV-level"))
traj_data$NEW_newfacet_x[sel] <- "ASVs >0.01%"
sel <- which(traj_data$facet_x %in% c(">0.1% - Bray-Curtis\nASV-level",">0.1% - Jaccard\nASV-level"))
traj_data$NEW_newfacet_x[sel] <- "ASVs >0.1%"

traj_data$NEW_newfacet_x <- factor(traj_data$NEW_newfacet_x,
                                   levels = c("All ASVs",
                                              "ASVs >0.001%",
                                              "ASVs >0.01%",
                                              "ASVs >0.1%"
                                   ),
                                   ordered = TRUE)


keep_cols_anno <- c("group","sig_letter","similarity","dist_measure","tax_level","study","corrected","facet_x")
anno_data <- df_anno[[ keep_results[1] ]][ ,keep_cols_anno]
for (i in 2:length(keep_results)) {
  anno_data <- rbind(anno_data,df_anno[[ keep_results[i] ]][ ,keep_cols_anno])
  print(paste0("completed ",i))
}
dim(anno_data) # 56  8
class(anno_data$facet_x) # "character"
unique(anno_data$facet_x) # "character"
# [1] "Bray-Curtis\nASV-level"           ">0.001% - Bray-Curtis\nASV-level" ">0.01% - Bray-Curtis\nASV-level"  ">0.1% - Bray-Curtis\nASV-level"  
# [5] "Jaccard\nASV-level"               ">0.001% - Jaccard\nASV-level"     ">0.01% - Jaccard\nASV-level"      ">0.1% - Jaccard\nASV-level"  


anno_data$NEW_newfacet_x <- NA
sel <- which(anno_data$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
anno_data$NEW_newfacet_x[sel] <- "All ASVs"
sel <- which(anno_data$facet_x %in% c(">0.001% - Bray-Curtis\nASV-level",">0.001% - Jaccard\nASV-level"))
anno_data$NEW_newfacet_x[sel] <- "ASVs >0.001%"
sel <- which(anno_data$facet_x %in% c(">0.01% - Bray-Curtis\nASV-level",">0.01% - Jaccard\nASV-level"))
anno_data$NEW_newfacet_x[sel] <- "ASVs >0.01%"
sel <- which(anno_data$facet_x %in% c(">0.1% - Bray-Curtis\nASV-level",">0.1% - Jaccard\nASV-level"))
anno_data$NEW_newfacet_x[sel] <- "ASVs >0.1%"

anno_data$NEW_newfacet_x <- factor(anno_data$NEW_newfacet_x,
                                   levels = c("All ASVs",
                                              "ASVs >0.001%",
                                              "ASVs >0.01%",
                                              "ASVs >0.1%"
                                   ),
                                   ordered = TRUE)

names(anno_data)
#  "group"          "sig_letter"     "similarity"     "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"        "NEW_newfacet_x"

# use $dist_measure - i.e. Bray-Curtis or Jaccard as y-facet 


class(traj_data$group) # "ordered" "factor" 

traj_data$group <- factor(traj_data$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)

levels(traj_data$group)
# [1] "0"   "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14" 
# [16] "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29" 
# [31] "30"  "31"  "32"  "33"  "34"  "35"  "36"  "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44" 
# [46] "45"  "Ref"

length(levels(traj_data$group)) #47

x_labels <- c("0","","","","","","","","","","10",
              "","","","","","","","","","20",
              "","","","","","","","","","30",
              "","","","","","","","","","40",
              "","","","","","Ref")
length(x_labels)
# 47
x_ticks <- c(1,0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,1)
length(x_ticks)
#47


unique(traj_data$study) # "Alcoa"

unique(anno_data$study) # "Alcoa"

 



## median reference horizontal lines ...

href_lines <- data.frame(
  #study  = rep( c("Alcoa", "Iluka-Eneabba", "South32"), times = 4 ),
  study = rep("Alcoa",times=8),
  # facet_x  = rep( c("Bray-Curtis\nASV-level",
  #                   ">0.001% - Bray-Curtis\nASV-level",
  #                   ">0.01% - Bray-Curtis\nASV-level",
  #                   ">0.1% - Bray-Curtis\nASV-level"), each = 3),
  facet_x = c("Bray-Curtis\nASV-level",
              ">0.001% - Bray-Curtis\nASV-level",
              ">0.01% - Bray-Curtis\nASV-level",
              ">0.1% - Bray-Curtis\nASV-level",
              "Jaccard\nASV-level",
              ">0.001% - Jaccard\nASV-level",
              ">0.01% - Jaccard\nASV-level",
              ">0.1% - Jaccard\nASV-level"
  ),
  dist_measure = rep( c("Bray-Curtis","Jaccard"), each = 4 ),
  similarity= NA
)

x <- traj_data

for (i in 1:dim(href_lines)[1]) {
  #i<-1
  sel <- which(x$study == href_lines$study[i] & x$facet_x == href_lines$facet_x[i] & x$group == "Ref")
  href_lines$similarity[i] <- median( x$similarity[sel] )
}

href_lines$NEW_newfacet_x <- NA
sel <- which(href_lines$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
href_lines$NEW_newfacet_x[sel] <- "All ASVs"
sel <- which(href_lines$facet_x %in% c(">0.001% - Bray-Curtis\nASV-level",">0.001% - Jaccard\nASV-level"))
href_lines$NEW_newfacet_x[sel] <- "ASVs >0.001%"
sel <- which(href_lines$facet_x %in% c(">0.01% - Bray-Curtis\nASV-level",">0.01% - Jaccard\nASV-level"))
href_lines$NEW_newfacet_x[sel] <- "ASVs >0.01%"
sel <- which(href_lines$facet_x %in% c(">0.1% - Bray-Curtis\nASV-level",">0.1% - Jaccard\nASV-level"))
href_lines$NEW_newfacet_x[sel] <- "ASVs >0.1%"

href_lines$NEW_newfacet_x <- factor(href_lines$NEW_newfacet_x,
                                    levels = c("All ASVs",
                                               "ASVs >0.001%",
                                               "ASVs >0.01%",
                                               "ASVs >0.1%"
                                    ),
                                    ordered = TRUE)


## fit logarithmic models for each data type (facet_x)



## BOOTSTRAP MODELS ...

pred_models_boot <- data.frame(study=NA, # one of: "Alcoa", "Iluka-Eneabba", "South32"
                               facet_x=NA, # one of: c("Jaccard\nASV-level","Bray-Curtis\nASV-level", "Jaccard\nGenus-level","Bray-Curtis\nGenus-level")
                               study_and_metric=NA, # combine the two values above
                               group=NA,
                               group_num=NA,
                               similarity=NA,
                               boot_no=NA,                 # EXTRA
                               ref_median=NA,              # EXTRA
                               pred_time_to_ref_median=NA, # EXTRA
                               pred_type=NA)     # c("rectangular hyperbola", "logarithmic")
pred_models_boot.temp <- pred_models_boot

#studies <- c("Alcoa", "Iluka-Eneabba", "South32")

studies <- c("Alcoa")

metrics <- c("Bray-Curtis\nASV-level",
             ">0.001% - Bray-Curtis\nASV-level",
             ">0.01% - Bray-Curtis\nASV-level",
             ">0.1% - Bray-Curtis\nASV-level",
             
             "Jaccard\nASV-level",
             ">0.001% - Jaccard\nASV-level",
             ">0.01% - Jaccard\nASV-level",
             ">0.1% - Jaccard\nASV-level"
)
pred_types <- c("rectangular hyperbola", "negative exponential", "logarithmic")
k<-3

## run logarithmic model only !!!

x <- traj_data
dim(x) # 5040  10

for (i in 1:length(studies)) {
  #i<-1
  this_study <- studies[i]
  for (j in 1:length(metrics)) {
    #j<-1
    this_metric <- metrics[j]
    
    # prepare data for modelling
    subsel <- which(x$facet_x == this_metric & x$study==this_study)
    newdat <- x[subsel, c("similarity","group") ]
    newdat$group_num <- as.numeric(as.character(newdat$group))
    # remove 'Ref' similarities which have now been converted to NA
    sel.na <- which(is.na(newdat$group_num))
    
    this_ref_median <- median(newdat$similarity[sel.na])
    
    newdat <- newdat[-sel.na, ]
    

    # prep output dataframe for prediction model to populate
    sel.dup <- which(duplicated(newdat$group)==TRUE)
    new_preds <- data.frame(matrix( nrow=dim(newdat[-sel.dup, ])[1], ncol = length(names(pred_models_boot.temp)) ))
    names(new_preds) <- names(pred_models_boot.temp)
    new_preds$group_num <- sort( newdat$group_num[-sel.dup] )
    new_preds$study <- this_study
    new_preds$facet_x <- this_metric
    new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$facet_x)
    new_preds$group <- as.character(new_preds$group_num)
    new_preds$ref_median <- this_ref_median
    
    #for (k in 1:length(pred_types)) {

    
    this_pred_model_type <- pred_types[k]
    
    ## NOTE:
    #  rectangular hyperbola ( fct = MM.2() from ‘drc' package; https://www.statforbiology.com/nonlinearregression/usefulequations#michaelis-menten_equation), and 
    #  negative exponential ( fct = DRC.negExp() from 'aomisc' package; https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function)
    #  models were trialled however these regularly failed to achieve model fits.
    
    
    # NOW using "logarithmic" model only !!!!!
    
    # negative exponential model, using fct = DRC.negExp() from 'aomisc' package
    # https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function
    
    
    for (b in 1:100) {
      #b<-20
      new_preds$boot_no <- b
      set.seed(b + 1234)
      sel.boot <- sample(x = c(1:dim(newdat)[1]), size = dim(newdat)[1], replace = TRUE)
      
      bootdat <- newdat[sel.boot, ]
      
      # exclude outlying data points from trajectory data-cloud ... 
      # as outliers are likely to throw out model-fitting
      bp <- boxplot(bootdat$similarity, print=FALSE)
      sel.outlier <- which(bootdat$similarity %in% bp$out)
      if (length(sel.outlier)>0) {
        bootdat <- bootdat[-sel.outlier, ]
      }
      
      #plot(bootdat$group_num, bootdat$similarity)
      
      model <- drm(similarity ~ group_num , fct = DRC.logCurve(), # DRC.negExp()
                   data = bootdat)
      
      new_preds$similarity <- predict(model, new_preds )
      new_preds$pred_type <- this_pred_model_type
      
      y <- data.frame(group_num=1:500, sim_pred=NA)
      y$sim_pred <- predict(model, y )
      y$reached_target <- y$sim_pred >= this_ref_median
      #plot(y$group_num, y$sim_pred)
      row.names(y)
      
      sel <- which(y$reached_target == TRUE)
      
      if (length(sel) > 0) {
        years_to_ref <- min(sel)
      } else {
        years_to_ref <- NA
      }
      new_preds$pred_time_to_ref_median <- years_to_ref
      
      # add predictions from end of last year to 40 years ...
      max_year <- max(new_preds$group_num)
      year <- max_year + 1
      while (year < 40) {
        new_preds <- rbind(new_preds, new_preds[ dim(new_preds)[1], ])
        year <- year + 1
        new_preds$group_num[dim(new_preds)[1]] <- year
        new_preds$group[dim(new_preds)[1]] <- as.character( new_preds$group_num[dim(new_preds)[1]] )
        new_preds$similarity[dim(new_preds)[1]] <- y[ year, "sim_pred"]
      }
      
      pred_models_boot <- rbind(pred_models_boot, new_preds)
      
    } # END bootstrap
    

    print(paste0("completed study: ",studies[i],"; metric: ",metrics[j],"; pred_type: ",pred_types[k]))  
    #} # END pred_types
  } # END metrics
} # END studies


# remove NA first row
pred_models_boot <- pred_models_boot[-1, ]

unique(pred_models_boot$facet_x)
# [1] "Bray-Curtis\nASV-level"           ">0.001% - Bray-Curtis\nASV-level" ">0.01% - Bray-Curtis\nASV-level"  ">0.1% - Bray-Curtis\nASV-level"  
# [5] "Jaccard\nASV-level"               ">0.001% - Jaccard\nASV-level"     ">0.01% - Jaccard\nASV-level"      ">0.1% - Jaccard\nASV-level"    

dim(pred_models_boot) # 12800  10

pred_models_boot$NEW_newfacet_x <- NA
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nASV-level","Jaccard\nASV-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "All ASVs"
sel <- which(pred_models_boot$facet_x %in% c(">0.001% - Bray-Curtis\nASV-level",">0.001% - Jaccard\nASV-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "ASVs >0.001%"
sel <- which(pred_models_boot$facet_x %in% c(">0.01% - Bray-Curtis\nASV-level",">0.01% - Jaccard\nASV-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "ASVs >0.01%"
sel <- which(pred_models_boot$facet_x %in% c(">0.1% - Bray-Curtis\nASV-level",">0.1% - Jaccard\nASV-level"))
pred_models_boot$NEW_newfacet_x[sel] <- "ASVs >0.1%"

pred_models_boot$NEW_newfacet_x <- factor(pred_models_boot$NEW_newfacet_x,
                                          levels = c("All ASVs",
                                                     "ASVs >0.001%",
                                                     "ASVs >0.01%",
                                                     "ASVs >0.1%"
                                          ),
                                          ordered = TRUE)

# correct time prediction to ref; update NA to (>) 500 yrs
sel <- which(is.na(pred_models_boot$pred_time_to_ref_median))
if (length(sel)>1) { pred_models_boot$pred_time_to_ref_median[sel] <- 500 }

pred_models_boot$group <- factor(pred_models_boot$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)
table(pred_models_boot$group)

pred_models_boot$dist_measure <- NA
sel <- which(pred_models_boot$facet_x %in% c("Bray-Curtis\nASV-level",">0.001% - Bray-Curtis\nASV-level", ">0.01% - Bray-Curtis\nASV-level",">0.1% - Bray-Curtis\nASV-level"))
pred_models_boot$dist_measure[sel] <- "Bray-Curtis"
sel <- which(pred_models_boot$facet_x %in% c("Jaccard\nASV-level",">0.001% - Jaccard\nASV-level",">0.01% - Jaccard\nASV-level",">0.1% - Jaccard\nASV-level"))
pred_models_boot$dist_measure[sel] <- "Jaccard"


str(pred_models_boot)


str(traj_data)




p <- ggplot(data=traj_data, aes(x=group, y=similarity)) +
  geom_line(data=pred_models_boot, aes(  x=group, y=similarity, group = boot_no ), colour = "#e34a33", alpha = 0.02) + # "red"
  geom_hline(data = href_lines, aes(yintercept = similarity),  color= "#2c7fb8", linetype="longdash") + # "blue"
  #geom_line(data= pred_models_alldata, aes(  x=group, y=similarity, group = 1), colour = "#e34a33" ) + # "red"
  
  geom_boxplot(outlier.shape = NA, width = 2) +
  #geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  
  theme_bw() +
  geom_text(data=anno_data, aes(x=group, y=similarity, label=sig_letter),nudge_y = 2, size=2.75) +
  scale_x_discrete(drop=FALSE, expand = c(0.05,0.05),
                   labels = x_labels) +
  theme(axis.text.x  = element_text(angle=90,hjust=1, vjust=0.5), # 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_line(linetype = 0)
        axis.ticks.x = element_line(linetype = x_ticks)
  ) +
  labs(x = "Rehabilitation age (years)", y = "Similarity to Reference (%)") +
  facet_grid(rows = vars(dist_measure), cols = vars(NEW_newfacet_x), scales = "fixed")

p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Restoration-trajectories--Alcoa--ASV-Bray-Jaccard-Dominant-Taxa-with-Preds.tiff"), width = 18, height = 10, units = "cm", dpi = 600, compression = "lzw",type="cairo")



str(traj_data)
unique(traj_data$study) # "Alcoa"
unique(traj_data$dist_measure) # "Bray-Curtis" "Jaccard"    
unique(traj_data$group) # use "ref" only
unique(traj_data$tax_level)
# "ASV (with Tree)"           ">0.001% - ASV (with Tree)" ">0.01% - ASV (with Tree)"  ">0.1% - ASV (with Tree)" 

## Huntly-Alcoa
sel <- which(traj_data$dist_measure == "Bray-Curtis")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 22.38   31.28   35.55   35.79   39.93   60.91 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == ">0.001% - ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 23.95   33.69   38.26   38.41   43.28   63.96 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == ">0.01% - ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 37.64   48.66   54.30   53.76   59.31   79.84 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == ">0.1% - ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 48.99   62.30   67.66   67.18   72.14   84.80 

sel <- which(traj_data$dist_measure == "Jaccard")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == "ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 9.503  17.824  20.345  20.092  23.213  35.809 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == ">0.001% - ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 11.33   21.68   24.90   24.49   28.32   41.15 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == ">0.01% - ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 44.70   59.23   63.79   62.71   67.19   78.27 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$tax_level[sel] == ">0.1% - ASV (with Tree)")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 79.17   88.57   91.67   91.01   94.29   98.55 


str(pred_models_boot)


unique(pred_models_boot$study_and_metric)
# [1] "Alcoa__Bray-Curtis\nASV-level"           "Alcoa__>0.001% - Bray-Curtis\nASV-level" "Alcoa__>0.01% - Bray-Curtis\nASV-level" 
# [4] "Alcoa__>0.1% - Bray-Curtis\nASV-level"   "Alcoa__Jaccard\nASV-level"               "Alcoa__>0.001% - Jaccard\nASV-level"    
# [7] "Alcoa__>0.01% - Jaccard\nASV-level"      "Alcoa__>0.1% - Jaccard\nASV-level"   

pred_models_boot$study_and_metric <- factor(pred_models_boot$study_and_metric,
                                            levels = c(
                                              "Alcoa__Bray-Curtis\nASV-level",
                                              "Alcoa__>0.001% - Bray-Curtis\nASV-level",
                                              "Alcoa__>0.01% - Bray-Curtis\nASV-level", 
                                              "Alcoa__>0.1% - Bray-Curtis\nASV-level" ,
                                              
                                              "Alcoa__Jaccard\nASV-level" ,
                                              "Alcoa__>0.001% - Jaccard\nASV-level" ,
                                              "Alcoa__>0.01% - Jaccard\nASV-level" ,
                                              "Alcoa__>0.1% - Jaccard\nASV-level"
                                            ))

dim(pred_models_boot) # 12800    12

dat.sum <- pred_models_boot
head(dat.sum)
#   study                facet_x              study_and_metric group group_num similarity boot_no ref_median
# 11 Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level     2         2   9.389107       1   35.55047
# 2  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level     8         8  21.149581       1   35.55047
# 3  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    14        14  25.897020       1   35.55047
# 4  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    17        17  27.544121       1   35.55047
# 5  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    25        25  30.815846       1   35.55047
# 6  Alcoa Bray-Curtis\nASV-level Alcoa__Bray-Curtis\nASV-level    29        29  32.074951       1   35.55047
# pred_time_to_ref_median   pred_type NEW_newfacet_x dist_measure
# 11                      44 logarithmic       All ASVs  Bray-Curtis
# 2                       44 logarithmic       All ASVs  Bray-Curtis
# 3                       44 logarithmic       All ASVs  Bray-Curtis
# 4                       44 logarithmic       All ASVs  Bray-Curtis
# 5                       44 logarithmic       All ASVs  Bray-Curtis
# 6                       44 logarithmic       All ASVs  Bray-Curtis

for (i in 1:length(levels(dat.sum$study_and_metric))) {
  #i<-1
  this_study_and_metric <- levels(dat.sum$study_and_metric)[i]
  sel <- which(dat.sum$study_and_metric == this_study_and_metric)
  #unique(dat.sum$ref_median[sel])
  temp <- dat.sum$pred_time_to_ref_median[sel]
  #quantile(temp, probs = c(0.05, 0.25, 0.5, 0.75, 0.95), na.rm = TRUE)
  #hist(temp)
  res <- quantile(temp, probs = c(0.05, 0.5, 0.95), na.rm = TRUE)
  names(res)
  #"5%"  "50%" "95%"
  #print(paste0(this_study_and_metric))
  print(paste0(this_study_and_metric,": ",res["50%"]," (",res["5%"],", ",res["95%"],")"))
}
# [1] "Alcoa__Bray-Curtis\nASV-level: 43 (39, 49)"
# [1] "Alcoa__>0.001% - Bray-Curtis\nASV-level: 45 (40, 51)"
# [1] "Alcoa__>0.01% - Bray-Curtis\nASV-level: 48 (43, 55)"
# [1] "Alcoa__>0.1% - Bray-Curtis\nASV-level: 38 (34.95, 43)"
# [1] "Alcoa__Jaccard\nASV-level: 60.5 (51, 79)"
# [1] "Alcoa__>0.001% - Jaccard\nASV-level: 72.5 (60, 99.1499999999999)"
# [1] "Alcoa__>0.01% - Jaccard\nASV-level: 56 (48, 68.0999999999999)"
# [1] "Alcoa__>0.1% - Jaccard\nASV-level: 29 (27, 31.05)"


## plot time to target

names(pred_models_boot)
# [1] "study"                   "facet_x"                 "study_and_metric"        "group"                   "group_num"              
# [6] "similarity"              "boot_no"                 "ref_median"              "pred_time_to_ref_median" "pred_type"              
# [11] "NEW_newfacet_x"          "dist_measure" 

p <- ggplot(data=pred_models_boot, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median, color = dist_measure)) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()
  ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Predicted-time-to-target--Alcoa--ASV-Bray-Jaccard-Dominant-Taxa.tiff"), width = 14, height = 10, units = "cm", dpi = 600, compression = "lzw",type="cairo")


df <- pred_models_boot
df$dist_measure <- factor(df$dist_measure, levels=c("Jaccard","Bray-Curtis"), ordered = TRUE)
unique(df$NEW_newfacet_x)
# [1] All ASVs     ASVs >0.001% ASVs >0.01%  ASVs >0.1%  
# Levels: All ASVs < ASVs >0.001% < ASVs >0.01% < ASVs >0.1%

df$NEW_newfacet_x <- factor(df$NEW_newfacet_x,
                            levels=c("All ASVs", "ASVs >0.001%", "ASVs >0.01%", "ASVs >0.1%"),
                            labels = c("All ASVs", ">0.001%", ">0.01%", ">0.1%"), ordered = TRUE)

p <- ggplot(data=df, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median)) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()#,
  ) +
  facet_wrap(facets = vars(dist_measure), scales = "free_y" ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)")
p

grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Predicted-time-to-target--Alcoa--ASV-Bray-Jaccard-Dominant-Taxa-C.tiff"), width = 12, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")


#-------------------------


#### Investigating spatial auto-correlation -- Alcoa -- Bray-Curtis 
#-------------------------

names(df_results)

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485 
ntaxa(r1.ps) # 30751
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30751 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 30751 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 30751 tips and 30749 internal nodes ]


r1.ps@sam_data$`Date since change in Land Use`
r1.ps@sam_data$Notes

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)




df_out <- df_results[[ "Alcoa-Bray-ASV-withTree" ]]


names(r1.ps@sam_data) # includes ...
# [1] "FULL_ID"                                               "Sample_id"                                            
# [3] "Date sampled"                                          "lat raw"                                              
# [5] "lon raw"                                               "lat (-) in decimal degrees"                           
# [7] "long in decimal degrees"           

#r1.ps@sam_data$`lat raw`
r1.ps@sam_data$`lat (-) in decimal degrees`
#r1.ps@sam_data$`lon raw`
r1.ps@sam_data$`long in decimal degrees`

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")

identical(colnames(dist.bray),rownames(dist.bray)) # TRUE


## explore spatial auto-correlation

# data.frame containing longitude and latitude coordinates

long_lats <- data.frame( xlong=r1.ps@sam_data$`long in decimal degrees`,
                         ylat =r1.ps@sam_data$`lat (-) in decimal degrees`
                         )
row.names(long_lats) <- row.names(r1.ps@sam_data)

# calculate matrix of pairwise distances between all points
geo_dist.m <- geodist(long_lats, measure = "haversine")

geo_dist <- as.data.frame( geo_dist.m )
row.names(geo_dist) <- row.names(r1.ps@sam_data)
colnames(geo_dist) <- row.names(r1.ps@sam_data)

# check that colnames and row.names align
identical(row.names(geo_dist),row.names(dist.bray)) # TRUE
identical(colnames(geo_dist),colnames(dist.bray)) # TRUE

identical(colnames(dist.bray), row.names(dist.bray)) # TRUE


df_eco <- dist.bray

df_ecogeocor <- data.frame(samp= rep( x = colnames(df_eco), each=length( colnames(df_eco) ) ),
                           compare_with = rep( x = colnames(df_eco), times = length(colnames(df_eco))),
                           
                           samp_type=NA,
                           compare_with_type=NA,
                           
                           dist_eco=NA,
                           dist_geo=NA
)

for (i in 1:dim(df_ecogeocor)[1]) {
  #i<-1
  this_samp <- df_ecogeocor$samp[i]
  this_compare <- df_ecogeocor$compare_with[i]
  sel1 <- which( row.names(r1.ps@sam_data) == this_samp)
  df_ecogeocor$samp_type[i] <- as.character( r1.ps@sam_data$group[sel1] )
  
  sel2 <- which( row.names(r1.ps@sam_data) == this_compare)
  df_ecogeocor$compare_with_type[i] <- as.character( r1.ps@sam_data$group[sel2] )
  
  row <- which(rownames(df_eco)==this_samp)
  col <- which(colnames(df_eco)==this_compare)
  df_ecogeocor$dist_eco[i] <- df_eco[row,col]
  
  row <- which(rownames(geo_dist)==this_samp)
  col <- which(colnames(geo_dist)==this_compare)
  df_ecogeocor$dist_geo[i] <- geo_dist[row,col]
  
  print(paste0("completed ",i))
}

dim(df_ecogeocor) #  1296    6

# remove remnant self-comparisons
sel <- which(df_ecogeocor$samp==df_ecogeocor$compare_with) # qty 36
identical( which(df_ecogeocor$samp==df_ecogeocor$compare_with), which(df_ecogeocor$dist_eco==0) ) # TRUE
df_ecogeocor <- df_ecogeocor[-sel, ]
dim(df_ecogeocor) # 1260    6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)
summary(df_ecogeocor$dist_eco)

head(df_ecogeocor)
#     samp compare_with samp_type compare_with_type  dist_eco  dist_geo
# 2 X39041       X39043         2               Ref 0.8689162   73.7641
# 3 X39041       X39045         2                 2 0.6026308 3879.6268
# 4 X39041       X39047         2               Ref 0.8663426 3927.1052
# 5 X39041       X39049         2                 2 0.6113812 2828.6415
# 6 X39041       X39051         2               Ref 0.8569059 2797.7893
# 7 X39041       X39053         2                 8 0.7341150 2408.0242


# consider only compare with type 'Ref' as that is focus 
# and potential for problematic study design and analysis

sel <- which(df_ecogeocor$compare_with_type=="Ref") # 630
df_ecogeocor <- df_ecogeocor[sel, ]
dim(df_ecogeocor) # 630   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)

names(df_ecogeocor)
# [1] "samp"              "compare_with"      "samp_type"         "compare_with_type" "dist_eco"         
# [6] "dist_geo" 


unique(df_ecogeocor$samp_type)
# "2"   "Ref" "8"   "25"  "29"  "14"  "17" 

df_ecogeocor$group <- df_ecogeocor$samp_type
df_ecogeocor$group <- factor(df_ecogeocor$group, 
                             levels=c("2","8","14","17","25","29","Ref"),
                             labels=c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref"),
                             ordered = TRUE)
length(levels(df_ecogeocor$group)) # 7

# library(viridis); packageVersion("viridis") # ‘0.5.1’
# cols.group <- viridis(n=length(levels(df_ecogeocor$group)))
# names(cols.group) <- levels(df_ecogeocor$group)

cols.group.alcoa
# 2 yr      8 yr     14 yr     17 yr     25 yr     29 yr       Ref 
# "#9e0142" "#d53e4f" "#fdae61" "#e6f598" "#abdda4" "#66c2a5" "#5e4fa2" 

cols.group <- cols.group.alcoa

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "25 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "29 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.4, label = "Original data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +

  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.8)),
    legend.text = element_text(size=rel(0.7))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Bray-Curtis)")
p
grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Bray-Curtis-ALCOA-A.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")


## for each rehab age determine if there is excess spatial auto-correlation??
## expect increasing eco-distance with increasing geo-distance
## if this occurs, does it occur at a rate greater 

( reveg_groups <- levels(df_ecogeocor$group) )
# "2 yr"  "8 yr"  "14 yr" "17 yr" "25 yr" "29 yr" "Ref" 



dim(df_ecogeocor) # 630  7
dim(df_out)       # 630  12

head(df_ecogeocor)
# samp compare_with samp_type compare_with_type  dist_eco  dist_geo group
# 2  X39041       X39043         2               Ref 0.8689162   73.7641  2 yr
# 4  X39041       X39047         2               Ref 0.8663426 3927.1052  2 yr
# 6  X39041       X39051         2               Ref 0.8569059 2797.7893  2 yr
# 8  X39041       X39055         2               Ref 0.8862454 2264.7422  2 yr
# 10 X39041       X39059         2               Ref 0.8625679 9779.8129  2 yr
# 12 X39041       X39063         2               Ref 0.8996854 8736.8677  2 yr

head(df_out)


identical(df_ecogeocor$samp,df_out$samp) # TRUE
identical(df_ecogeocor$compare_with,df_out$compare_with) # TRUE


df_ecogeocor$similarity <- df_out$similarity
df_ecogeocor$tax_level <- df_out$tax_level
df_ecogeocor$study <- df_out$study
df_ecogeocor$dist_measure <- df_out$dist_measure
df_ecogeocor$facet_x <- df_out$facet_x
df_ecogeocor$similarity_corr <- NA
df_ecogeocor$dist_eco_corr <- NA


preds.temp <- data.frame(dist_geo = sort( unique(df_ecogeocor$dist_geo)))
diff_models <- list()


for (i in 1:(length(reveg_groups)-1)) {
  #i<-1
  this_group <- reveg_groups[i]
  
  x <- df_ecogeocor %>% filter(group == reveg_groups[i])
  x_ref <- df_ecogeocor %>% filter(group == "Ref")
  #dim(x) # 54 14
  #dim(x_ref) # 306  14
  # mean centre eco-distances
  x$dist_eco_mc <- x$dist_eco - mean(x$dist_eco)
  #plot(x$dist_geo,x$dist_eco_mc)
  x_ref$dist_eco_mc <- x_ref$dist_eco - mean(x_ref$dist_eco)
  #plot(x_ref$dist_geo,x_ref$dist_eco_mc)
  
  #model each spatial trend
  m <- lm(data=x, formula = dist_eco_mc ~ poly(dist_geo, 2))
  m_ref <- lm(data=x_ref, formula = dist_eco_mc ~ poly(dist_geo, 2))
  preds <- preds.temp
  ## limit predictions to max dist_geo
  #sel <- which(preds$dist_geo <= max(x$dist_geo))
  # limit predictions to extent of dist_geo
  sel <- which(preds$dist_geo >= min(x$dist_geo) & preds$dist_geo <= max(x$dist_geo))
  
  preds <- as.data.frame( preds[sel, ] )
  names(preds)[1] <- "dist_geo" # need to rename field; it is lost in previous step
  preds$m <- predict(m, newdata = preds)
  preds$m_ref <- predict(m_ref, newdata = preds)
  #plot(preds$dist_geo,preds$m,col="red")
  #points(preds$dist_geo,preds$m_ref,col="green")
  preds$diff <- preds$m - preds$m_ref
  preds$group <- this_group
  #points(preds$dist_geo,preds$diff,col="orange")
  plot(preds$dist_geo,preds$diff,col="orange", main = paste0("Diff Reveg age: ",this_group))
  # apply correction to similarity
  sel.df <- which(df_ecogeocor$group == this_group)
  for (j in 1:length(sel.df)) {
    #j<-1
    sel.pred <- which(preds$dist_geo == df_ecogeocor$dist_geo[ sel.df[j] ])
    this_diff_dist_eco <- preds$diff[sel.pred]
    # similarity = 100*(1 - distance)
    # corrected similarity = 100*[1 -  (distance - distance-correction)]
    df_ecogeocor$similarity[ sel.df[j] ] # 13.10838
    df_ecogeocor$similarity_corr[ sel.df[j] ] <- 100*(1 - (df_ecogeocor$dist_eco[ sel.df[j] ] - this_diff_dist_eco))
    df_ecogeocor$dist_eco_corr[ sel.df[j] ] <- 1 - (1/100)*df_ecogeocor$similarity_corr[ sel.df[j] ]
  }
  plot(df_ecogeocor$similarity[ sel.df ], df_ecogeocor$similarity_corr[ sel.df ], main = paste0("Group: ",this_group))
  diff_models[[this_group]] <- preds
  print(paste0("completed ",i))
}

## but note there will be no 'corrected' values for the "Ref" samples - so just use previous values 
sel <- which(df_ecogeocor$group == "Ref")
df_ecogeocor$similarity_corr[sel] # all NAs
df_ecogeocor$dist_eco_corr[sel] # all NAs
df_ecogeocor$similarity_corr[sel] <- df_ecogeocor$similarity[sel]
df_ecogeocor$dist_eco_corr[sel]   <- df_ecogeocor$dist_eco[sel]


## plot modelled data for excess spatial auto-correlation 

names(diff_models) # "2 yr"  "8 yr"  "14 yr" "17 yr" "25 yr" "29 yr"

pdata <- diff_models[[1]]
for (i in 2:length(diff_models)) {
  pdata <- rbind(pdata,diff_models[[i]])
}
dim(pdata) # 2362    5
names(pdata) # "dist_geo" "m"        "m_ref"    "diff"     "group"
unique(pdata$group) # "2 yr"  "8 yr"  "14 yr" "17 yr" "25 yr" "29 yr"
pdata$group <- factor(pdata$group, levels=c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr"),
                      ordered=TRUE)

p <- ggplot(data = pdata, aes(x=dist_geo,y=diff,color=factor(group)) ) +  #,color=factor(group)
  #geom_point( alpha = 0.6) +
  #geom_point() +# pdata %>% filter(group == "3 yr"),aes(x=dist_geo,y=diff)) +
  geom_line(size=1) +
  geom_hline(yintercept = 0, size=0.5, linetype="dashed", colour="grey") +
  theme_bw() +
  scale_color_manual(values=cols.group[-length(cols.group)]) +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.8)),
    legend.text = element_text(size=rel(0.7))
  ) +
  #ylim(0,0.2) +
  guides(colour = guide_legend(title = "Rehab\nage",override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "\U0394 ecol. dist. (Bray-Curtis)\nRehab vs. Ref samples")
p
grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","DIFF-in-Ecodist-vs-Geodist-Bray-Curtis-ALCOA-B.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")



## plot corrected distributions ...

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco_corr, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "25 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "29 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.4, label = "Corrected data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.8)),
    legend.text = element_text(size=rel(0.7))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Bray-Curtis)")
p
grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Bray-Curtis-ALCOA-SACORRECTED-C.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")



names(df_results)
names(df_results[[40]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"         
# [8] "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"

names(df_ecogeocor)

## make names of df_ecogeocor compatible with df_out
head(df_out)

head(df_ecogeocor)


# confirm these refer to the same sites and comparison sites
identical( paste0(df_out$samp,"_",df_out$compare_with) , paste0(df_ecogeocor$samp,"_",df_ecogeocor$compare_with) ) # TRUE
dim(df_out) # 630  12

df_new <- df_out
df_new$similarity <- df_ecogeocor$similarity_corr
df_new$corrected <- "yes"

head(df_new)
# samp compare_with     site             site_full_desc year_rehab similarity group dist_measure       tax_level study
# 1 X39041       X39043 karri 11 karri 11 (post mine rehab)       2014   15.72438     2  Bray-Curtis ASV (with Tree) Alcoa
# 2 X39041       X39047 karri 11 karri 11 (post mine rehab)       2014   15.66220     2  Bray-Curtis ASV (with Tree) Alcoa
# 3 X39041       X39051 karri 11 karri 11 (post mine rehab)       2014   16.79160     2  Bray-Curtis ASV (with Tree) Alcoa
# 4 X39041       X39055 karri 11 karri 11 (post mine rehab)       2014   13.91881     2  Bray-Curtis ASV (with Tree) Alcoa
# 5 X39041       X39059 karri 11 karri 11 (post mine rehab)       2014   13.85382     2  Bray-Curtis ASV (with Tree) Alcoa
# 6 X39041       X39063 karri 11 karri 11 (post mine rehab)       2014   10.68174     2  Bray-Curtis ASV (with Tree) Alcoa
# corrected                facet_x
# 1       yes Bray-Curtis\nASV-level
# 2       yes Bray-Curtis\nASV-level
# 3       yes Bray-Curtis\nASV-level
# 4       yes Bray-Curtis\nASV-level
# 5       yes Bray-Curtis\nASV-level
# 6       yes Bray-Curtis\nASV-level

unique(df_new$group)
# [1] 2   Ref 8   25  29  14  17 
# Levels: 2 < 8 < 14 < 17 < 25 < 29 < Ref

# Alcoa (sampled in 2016; rehab 2-29 yr old)


table(df_new$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 


# only make this once
#df_results <- list()
length(df_results) #
names(df_results)

df_results[["Alcoa-Bray-ASV-withTree-SAcorrected"]] <- df_new
#df_out <- df_results[[1]]


p <- ggplot(data=df_new, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity\nSA-corrected") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p
ggsave(plot=p, filename = paste0(workdir,"/plots/","Bray-Curtis-Similarity--Alcoa-Site-with-SA-correction.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw", type = "cairo")



## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_new)
# 'data.frame':	630 obs. of  12 variables:


df_new$group


# Kruskal-Wallis test
 
kt <- kruskal.test( similarity ~ group, df_new) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 352.65, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_new$group ) # 2   Ref 8   25  29  14  17 

pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) # "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      c         c 
# 6     8     ac       a c 
# 1    14      a       a   
# 2    17      b        b  
# 4    25      b        b  
# 5    29     bd        b d
# 7   Ref      d          d

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_new)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(22,30,39,48,48,51,55),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "yes" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-Bray-ASV-withTree-SAcorrected"]] <- anno


#-------------------------

#### Investigating spatial auto-correlation -- Alcoa -- Jaccard
#-------------------------

names(df_results)

phy_in <- phy.alcoa.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 1772249

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485 
ntaxa(r1.ps) # 30751
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30751 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 30751 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 30751 tips and 30749 internal nodes ]


r1.ps@sam_data$`Date since change in Land Use`
r1.ps@sam_data$Notes

r1.ps@sam_data$group <- r1.ps@sam_data$`Date since change in Land Use`
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)




df_out <- df_results[[ "Alcoa-Jaccard-ASV-withTree" ]]


names(r1.ps@sam_data) # includes ...
# [1] "FULL_ID"                                               "Sample_id"                                            
# [3] "Date sampled"                                          "lat raw"                                              
# [5] "lon raw"                                               "lat (-) in decimal degrees"                           
# [7] "long in decimal degrees"           

#r1.ps@sam_data$`lat raw`
r1.ps@sam_data$`lat (-) in decimal degrees`
#r1.ps@sam_data$`lon raw`
r1.ps@sam_data$`long in decimal degrees`

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")

identical(colnames(dist.jacc),rownames(dist.jacc)) # TRUE


## determine spatial auto-correlation

# data.frame containing longitude and latitude coordinates

long_lats <- data.frame( xlong=r1.ps@sam_data$`long in decimal degrees`,
                         ylat =r1.ps@sam_data$`lat (-) in decimal degrees`
)
row.names(long_lats) <- row.names(r1.ps@sam_data)

# calculate matrix of pairwise distances between all points
geo_dist.m <- geodist(long_lats, measure = "haversine")

geo_dist <- as.data.frame( geo_dist.m )
row.names(geo_dist) <- row.names(r1.ps@sam_data)
colnames(geo_dist) <- row.names(r1.ps@sam_data)

# check that colnames and row.names align
identical(row.names(geo_dist),row.names(dist.bray)) # TRUE
identical(colnames(geo_dist),colnames(dist.bray)) # TRUE

identical(colnames(dist.bray), row.names(dist.bray)) # TRUE


df_eco <- dist.jacc

df_ecogeocor <- data.frame(samp= rep( x = colnames(df_eco), each=length( colnames(df_eco) ) ),
                           compare_with = rep( x = colnames(df_eco), times = length(colnames(df_eco))),
                           
                           samp_type=NA,
                           compare_with_type=NA,
                           
                           dist_eco=NA,
                           dist_geo=NA
)

for (i in 1:dim(df_ecogeocor)[1]) {
  #i<-1
  this_samp <- df_ecogeocor$samp[i]
  this_compare <- df_ecogeocor$compare_with[i]
  sel1 <- which( row.names(r1.ps@sam_data) == this_samp)
  df_ecogeocor$samp_type[i] <- as.character( r1.ps@sam_data$group[sel1] )
  
  sel2 <- which( row.names(r1.ps@sam_data) == this_compare)
  df_ecogeocor$compare_with_type[i] <- as.character( r1.ps@sam_data$group[sel2] )
  
  row <- which(rownames(df_eco)==this_samp)
  col <- which(colnames(df_eco)==this_compare)
  df_ecogeocor$dist_eco[i] <- df_eco[row,col]
  
  row <- which(rownames(geo_dist)==this_samp)
  col <- which(colnames(geo_dist)==this_compare)
  df_ecogeocor$dist_geo[i] <- geo_dist[row,col]
  
  print(paste0("completed ",i))
}

dim(df_ecogeocor) #  1296    6

# remove remnant self-comparisons
sel <- which(df_ecogeocor$samp==df_ecogeocor$compare_with) # qty 36
identical( which(df_ecogeocor$samp==df_ecogeocor$compare_with), which(df_ecogeocor$dist_eco==0) ) # TRUE
df_ecogeocor <- df_ecogeocor[-sel, ]
dim(df_ecogeocor) # 1260    6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)
summary(df_ecogeocor$dist_eco)

head(df_ecogeocor)
# samp compare_with samp_type compare_with_type  dist_eco  dist_geo
# 2 X39041       X39043         2               Ref 0.8883514   73.7641
# 3 X39041       X39045         2                 2 0.6712861 3879.6268
# 4 X39041       X39047         2               Ref 0.8808063 3927.1052
# 5 X39041       X39049         2                 2 0.6708695 2828.6415
# 6 X39041       X39051         2               Ref 0.8836938 2797.7893
# 7 X39041       X39053         2                 8 0.7701849 2408.0242


# consider only compare with type 'Ref' as that is focus 
# and potential for problematic study design and analysis

sel <- which(df_ecogeocor$compare_with_type=="Ref") # 630
df_ecogeocor <- df_ecogeocor[sel, ]
dim(df_ecogeocor) # 630   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)

names(df_ecogeocor)
# [1] "samp"              "compare_with"      "samp_type"         "compare_with_type" "dist_eco"         
# [6] "dist_geo" 


unique(df_ecogeocor$samp_type)
# "2"   "Ref" "8"   "25"  "29"  "14"  "17" 

df_ecogeocor$group <- df_ecogeocor$samp_type
df_ecogeocor$group <- factor(df_ecogeocor$group, 
                             levels=c("2","8","14","17","25","29","Ref"),
                             labels=c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref"),
                             ordered = TRUE)
length(levels(df_ecogeocor$group)) # 7

# #library(viridis); packageVersion("viridis") # ‘0.5.1’
# cols.group <- viridis(n=length(levels(df_ecogeocor$group)))
# names(cols.group) <- levels(df_ecogeocor$group)

cols.group <- cols.group.alcoa

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "25 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "29 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.4, label = "Original data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.8)),
    legend.text = element_text(size=rel(0.7))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Jaccard)")
p
grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Jaccard-ALCOA-A.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")


## for each rehab age determine if there is excess spatial auto-correlation??
## expect increasing eco-distance with increasing geo-distance
## if this occurs, does it occur at a rate greater 

( reveg_groups <- levels(df_ecogeocor$group) )
# "2 yr"  "8 yr"  "14 yr" "17 yr" "25 yr" "29 yr" "Ref"



dim(df_ecogeocor) # 630  7
dim(df_out)       # 630  12

head(df_ecogeocor)
# samp compare_with samp_type compare_with_type  dist_eco  dist_geo group
# 2  X39041       X39043         2               Ref 0.8883514   73.7641  2 yr
# 4  X39041       X39047         2               Ref 0.8808063 3927.1052  2 yr
# 6  X39041       X39051         2               Ref 0.8836938 2797.7893  2 yr
# 8  X39041       X39055         2               Ref 0.8975366 2264.7422  2 yr
# 10 X39041       X39059         2               Ref 0.8794687 9779.8129  2 yr
# 12 X39041       X39063         2               Ref 0.9128551 8736.8677  2 yr

head(df_out)
# samp compare_with     site             site_full_desc year_rehab similarity group dist_measure       tax_level study corrected            facet_x
# 1 X39041       X39043 karri 11 karri 11 (post mine rehab)       2014  11.164864     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 2 X39041       X39047 karri 11 karri 11 (post mine rehab)       2014  11.919369     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 3 X39041       X39051 karri 11 karri 11 (post mine rehab)       2014  11.630619     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 4 X39041       X39055 karri 11 karri 11 (post mine rehab)       2014  10.246344     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 5 X39041       X39059 karri 11 karri 11 (post mine rehab)       2014  12.053126     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 6 X39041       X39063 karri 11 karri 11 (post mine rehab)       2014   8.714487     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level

identical(df_ecogeocor$samp,df_out$samp) # TRUE
identical(df_ecogeocor$compare_with,df_out$compare_with) # TRUE


df_ecogeocor$similarity <- df_out$similarity
df_ecogeocor$tax_level <- df_out$tax_level
df_ecogeocor$study <- df_out$study
df_ecogeocor$dist_measure <- df_out$dist_measure
df_ecogeocor$facet_x <- df_out$facet_x
df_ecogeocor$similarity_corr <- NA
df_ecogeocor$dist_eco_corr <- NA


preds.temp <- data.frame(dist_geo = sort( unique(df_ecogeocor$dist_geo)))
diff_models <- list()


for (i in 1:(length(reveg_groups)-1)) {
  #i<-1
  this_group <- reveg_groups[i]
  
  x <- df_ecogeocor %>% filter(group == reveg_groups[i])
  x_ref <- df_ecogeocor %>% filter(group == "Ref")
  #dim(x) # 54 14
  #dim(x_ref) # 306  14
  # mean centre eco-distances
  x$dist_eco_mc <- x$dist_eco - mean(x$dist_eco)
  #plot(x$dist_geo,x$dist_eco_mc)
  x_ref$dist_eco_mc <- x_ref$dist_eco - mean(x_ref$dist_eco)
  #plot(x_ref$dist_geo,x_ref$dist_eco_mc)
  
  #model each spatial trend
  m <- lm(data=x, formula = dist_eco_mc ~ poly(dist_geo, 2))
  m_ref <- lm(data=x_ref, formula = dist_eco_mc ~ poly(dist_geo, 2))
  preds <- preds.temp
  # # limit predictions to max dist_geo
  # sel <- which(preds$dist_geo <= max(x$dist_geo))
  # limit predictions to extent of dist_geo
  sel <- which(preds$dist_geo >= min(x$dist_geo) & preds$dist_geo <= max(x$dist_geo))
  
  preds <- as.data.frame( preds[sel, ] )
  names(preds)[1] <- "dist_geo" # need to rename field; it is lost in previous step
  preds$m <- predict(m, newdata = preds)
  preds$m_ref <- predict(m_ref, newdata = preds)
  #plot(preds$dist_geo,preds$m,col="red")
  #points(preds$dist_geo,preds$m_ref,col="green")
  preds$diff <- preds$m - preds$m_ref
  preds$group <- this_group
  #points(preds$dist_geo,preds$diff,col="orange")
  plot(preds$dist_geo,preds$diff,col="orange", main = paste0("Diff Reveg age: ",this_group))
  # apply correction to similarity
  sel.df <- which(df_ecogeocor$group == this_group)
  for (j in 1:length(sel.df)) {
    #j<-1
    sel.pred <- which(preds$dist_geo == df_ecogeocor$dist_geo[ sel.df[j] ])
    this_diff_dist_eco <- preds$diff[sel.pred]
    # similarity = 100*(1 - distance)
    # corrected similarity = 100*[1 -  (distance - distance-correction)]
    df_ecogeocor$similarity[ sel.df[j] ] # 13.10838
    df_ecogeocor$similarity_corr[ sel.df[j] ] <- 100*(1 - (df_ecogeocor$dist_eco[ sel.df[j] ] - this_diff_dist_eco))
    df_ecogeocor$dist_eco_corr[ sel.df[j] ] <- 1 - (1/100)*df_ecogeocor$similarity_corr[ sel.df[j] ]
  }
  plot(df_ecogeocor$similarity[ sel.df ], df_ecogeocor$similarity_corr[ sel.df ], main = paste0("Group: ",this_group))
  diff_models[[this_group]] <- preds
  print(paste0("completed ",i))
}

## but note there will be no 'corrected' values for the "Ref" samples - so just use previous values 
sel <- which(df_ecogeocor$group == "Ref")
df_ecogeocor$similarity_corr[sel] # all NAs
df_ecogeocor$dist_eco_corr[sel] # all NAs
df_ecogeocor$similarity_corr[sel] <- df_ecogeocor$similarity[sel]
df_ecogeocor$dist_eco_corr[sel]   <- df_ecogeocor$dist_eco[sel]


## plot modelled data for excess spatial auto-correlation 

names(diff_models) # "3 yr"  "9 yr"  "15 yr" "18 yr" "26 yr" "30 yr"

pdata <- diff_models[[1]]
for (i in 2:length(diff_models)) {
  pdata <- rbind(pdata,diff_models[[i]])
}
dim(pdata) # 2362    5
names(pdata) # "dist_geo" "m"        "m_ref"    "diff"     "group"
unique(pdata$group) # "2 yr"  "8 yr"  "14 yr" "17 yr" "25 yr" "29 yr"
pdata$group <- factor(pdata$group, levels=c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr"),
                      ordered=TRUE)

p <- ggplot(data = pdata, aes(x=dist_geo,y=diff,color=factor(group)) ) +  #,color=factor(group)
  #geom_point( alpha = 0.6) +
  #geom_point() +# pdata %>% filter(group == "3 yr"),aes(x=dist_geo,y=diff)) +
  geom_line(size=1) +
  geom_hline(yintercept = 0, size=0.5, linetype="dashed", colour="grey") +
  theme_bw() +
  scale_color_manual(values=cols.group[-length(cols.group)]) +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.8)),
    legend.text = element_text(size=rel(0.7))
  ) +
  #ylim(0,0.2) +
  guides(colour = guide_legend(title = "Rehab\nage",override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "\U0394 ecol. dist. (Jaccard)\nRehab vs. Ref samples")
p
grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","DIFF-in-Ecodist-vs-Geodist-Jaccard-ALCOA-B.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")



## plot corrected distributions ...

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco_corr, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "25 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "29 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.43, label = "Corrected data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.8)),
    legend.text = element_text(size=rel(0.7))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Jaccard)")
p
grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Jaccard-ALCOA-SACORRECTED-C.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")



names(df_results)
names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"         
# [8] "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"

names(df_ecogeocor)

## make names of df_ecogeocor compatible with df_out
head(df_out)
# samp compare_with     site             site_full_desc year_rehab similarity group dist_measure       tax_level study corrected            facet_x
# 1 X39041       X39043 karri 11 karri 11 (post mine rehab)       2014  11.164864     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 2 X39041       X39047 karri 11 karri 11 (post mine rehab)       2014  11.919369     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 3 X39041       X39051 karri 11 karri 11 (post mine rehab)       2014  11.630619     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 4 X39041       X39055 karri 11 karri 11 (post mine rehab)       2014  10.246344     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 5 X39041       X39059 karri 11 karri 11 (post mine rehab)       2014  12.053126     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level
# 6 X39041       X39063 karri 11 karri 11 (post mine rehab)       2014   8.714487     2      Jaccard ASV (with Tree) Alcoa        no Jaccard\nASV-level

head(df_ecogeocor)
# samp compare_with samp_type compare_with_type  dist_eco  dist_geo group similarity       tax_level study dist_measure            facet_x similarity_corr dist_eco_corr
# 2  X39041       X39043         2               Ref 0.8883514   73.7641  2 yr  11.164864 ASV (with Tree) Alcoa      Jaccard Jaccard\nASV-level       12.286528     0.8771347
# 4  X39041       X39047         2               Ref 0.8808063 3927.1052  2 yr  11.919369 ASV (with Tree) Alcoa      Jaccard Jaccard\nASV-level       12.971661     0.8702834
# 6  X39041       X39051         2               Ref 0.8836938 2797.7893  2 yr  11.630619 ASV (with Tree) Alcoa      Jaccard Jaccard\nASV-level       12.742994     0.8725701
# 8  X39041       X39055         2               Ref 0.8975366 2264.7422  2 yr  10.246344 ASV (with Tree) Alcoa      Jaccard Jaccard\nASV-level       11.375629     0.8862437
# 10 X39041       X39059         2               Ref 0.8794687 9779.8129  2 yr  12.053126 ASV (with Tree) Alcoa      Jaccard Jaccard\nASV-level       12.265989     0.8773401
# 12 X39041       X39063         2               Ref 0.9128551 8736.8677  2 yr   8.714487 ASV (with Tree) Alcoa      Jaccard Jaccard\nASV-level        9.141757     0.9085824

# confirm these refer to the same sites and comparison sites
identical( paste0(df_out$samp,"_",df_out$compare_with) , paste0(df_ecogeocor$samp,"_",df_ecogeocor$compare_with) ) # TRUE
dim(df_out) # 630  12

df_new <- df_out
df_new$similarity <- df_ecogeocor$similarity_corr
df_new$corrected <- "yes"

head(df_new)
# samp compare_with     site             site_full_desc year_rehab similarity group dist_measure       tax_level study corrected            facet_x
# 1 X39041       X39043 karri 11 karri 11 (post mine rehab)       2014  12.286528     2      Jaccard ASV (with Tree) Alcoa       yes Jaccard\nASV-level
# 2 X39041       X39047 karri 11 karri 11 (post mine rehab)       2014  12.971661     2      Jaccard ASV (with Tree) Alcoa       yes Jaccard\nASV-level
# 3 X39041       X39051 karri 11 karri 11 (post mine rehab)       2014  12.742994     2      Jaccard ASV (with Tree) Alcoa       yes Jaccard\nASV-level
# 4 X39041       X39055 karri 11 karri 11 (post mine rehab)       2014  11.375629     2      Jaccard ASV (with Tree) Alcoa       yes Jaccard\nASV-level
# 5 X39041       X39059 karri 11 karri 11 (post mine rehab)       2014  12.265989     2      Jaccard ASV (with Tree) Alcoa       yes Jaccard\nASV-level
# 6 X39041       X39063 karri 11 karri 11 (post mine rehab)       2014   9.141757     2      Jaccard ASV (with Tree) Alcoa       yes Jaccard\nASV-level

unique(df_new$group)
# [1] 2   Ref 8   25  29  14  17



table(df_new$group)
# 2   8  14  17  25  29 Ref 
# 54  54  54  54  54  54 306 


# only make this once
#df_results <- list()
length(df_results) #
names(df_results)

df_results[["Alcoa-Jaccard-ASV-withTree-SAcorrected"]] <- df_new
#df_out <- df_results[[1]]


p <- ggplot(data=df_new, aes(x=group, similarity)) + ggtitle("Jaccard Similarity\nSA-corrected") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p
ggsave(plot=p, filename = paste0(workdir,"/plots/","Jaccard-Similarity--Alcoa-Site-with-SA-correction.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw", type = "cairo")




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_new)
# 'data.frame':	630 obs. of  12 variables:


df_new$group


# Kruskal-Wallis test
kt <- kruskal.test( similarity ~ group, df_new) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 239.82, df = 6, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_new$group ) # 2   Ref 8   25  29  14  17 


pt <- dunnTest( similarity ~ group, data = df_out,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)

unique(sig$Group) #  "14"  "17"  "2"   "25"  "29"  "8"   "Ref"

sig$Group <- factor( sig$Group, levels=c("2","8","14","17","25","29","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 3     2      d         d 
# 6     8     bd       b d 
# 1    14     ab      ab   
# 2    17     ac      a c  
# 4    25     ac      a c  
# 5    29     ce        c e
# 7   Ref      e          e

sig <- sig[ order(sig$Group), ]

levels(sig$Group) #  "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

## store annotations for later facet_grid ggplot
names(df_new)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(20,22,25,30,31,35,40),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Alcoa", times = dim(sig)[1]),
                   corrected = rep( "yes" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Alcoa-Jaccard-ASV-withTree-SAcorrected"]] <- anno


#-------------------------

#### Investigating spatial auto-correlation -- Iluka-Eneabba -- Bray-Curtis 
#-------------------------

names(df_results)

phy_in <- phy.iluka_eneabba.withTree
min(taxa_sums(phy_in)) # 1
sum(sample_sums(phy_in)) # 2155211

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142
ntaxa(r1.ps) # 27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 27115 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 27115 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 27115 tips and 27114 internal nodes ]


str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$year.of.rehab.or.ref
unique(r1.ps@sam_data$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                               labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)



names(df_results)

df_out <- df_results[[ "Iluka-Eneabba-Bray-ASV-withTree" ]]


names(r1.ps@sam_data) # includes ...
# [1] "Sample_ID"                                             "NCBI_Submission"                                      
# [3] "NCBI_Sample_Accession"                                 "Organism"                                             
# [5] "Tax_ID"                                                "SampleName_depth"                                     
# [7] "NCBI_BioProject"                                       "Date_sampled..YYYY.MM.DD."                            
# [9] "Time_sampled..hh.mm."                                  "Latitude..decimal.degrees."                           
# [11] "Longitude..decimal.degrees."                

r1.ps@sam_data$Latitude..decimal.degrees.
r1.ps@sam_data$Longitude..decimal.degrees.

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")

identical(colnames(dist.bray),rownames(dist.bray)) # TRUE


## determine spatial auto-correlation

# data.frame containing longitude and latitude coordinates

long_lats <- data.frame( xlong=r1.ps@sam_data$Longitude..decimal.degrees.,
                         ylat =r1.ps@sam_data$Latitude..decimal.degrees.
)
row.names(long_lats) <- row.names(r1.ps@sam_data)

# calculate matrix of pairwise distances between all points
geo_dist.m <- geodist(long_lats, measure = "haversine")

geo_dist <- as.data.frame( geo_dist.m )
row.names(geo_dist) <- row.names(r1.ps@sam_data)
colnames(geo_dist) <- row.names(r1.ps@sam_data)

# check that colnames and row.names align
identical(row.names(geo_dist),row.names(dist.bray)) # TRUE
identical(colnames(geo_dist),colnames(dist.bray)) # TRUE

identical(colnames(dist.bray), row.names(dist.bray)) # TRUE


df_eco <- dist.bray

df_ecogeocor <- data.frame(samp= rep( x = colnames(df_eco), each=length( colnames(df_eco) ) ),
                           compare_with = rep( x = colnames(df_eco), times = length(colnames(df_eco))),
                           
                           samp_type=NA,
                           compare_with_type=NA,
                           
                           dist_eco=NA,
                           dist_geo=NA
)

for (i in 1:dim(df_ecogeocor)[1]) {
  #i<-1
  this_samp <- df_ecogeocor$samp[i]
  this_compare <- df_ecogeocor$compare_with[i]
  sel1 <- which( row.names(r1.ps@sam_data) == this_samp)
  df_ecogeocor$samp_type[i] <- as.character( r1.ps@sam_data$group[sel1] )
  
  sel2 <- which( row.names(r1.ps@sam_data) == this_compare)
  df_ecogeocor$compare_with_type[i] <- as.character( r1.ps@sam_data$group[sel2] )
  
  row <- which(rownames(df_eco)==this_samp)
  col <- which(colnames(df_eco)==this_compare)
  df_ecogeocor$dist_eco[i] <- df_eco[row,col]
  
  row <- which(rownames(geo_dist)==this_samp)
  col <- which(colnames(geo_dist)==this_compare)
  df_ecogeocor$dist_geo[i] <- geo_dist[row,col]
  
  print(paste0("completed ",i))
}

dim(df_ecogeocor) #  676   6

# remove remnant self-comparisons
sel <- which(df_ecogeocor$samp==df_ecogeocor$compare_with) # qty 26
identical( which(df_ecogeocor$samp==df_ecogeocor$compare_with), which(df_ecogeocor$dist_eco==0) ) # TRUE
df_ecogeocor <- df_ecogeocor[-sel, ]
dim(df_ecogeocor) # 650   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)
summary(df_ecogeocor$dist_eco)

head(df_ecogeocor)
#      samp compare_with samp_type compare_with_type  dist_eco  dist_geo
# 2 X138408      X138410       Ref               Ref 0.7032144  146.3397
# 3 X138408      X138412       Ref               Ref 0.7545849 4708.6686
# 4 X138408      X138418       Ref                30 0.7757839 6069.0393
# 5 X138408      X138424       Ref                24 0.8074344 7423.1019
# 6 X138408      X138426       Ref                24 0.8142378 7596.8801
# 7 X138408      X138428       Ref                38 0.7947150 9336.2228


# consider only compare with type 'Ref' as that is focus 
# and potential for problematic study design and analysis

sel <- which(df_ecogeocor$compare_with_type=="Ref") # 225
df_ecogeocor <- df_ecogeocor[sel, ]
dim(df_ecogeocor) # 225   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)

names(df_ecogeocor)
# [1] "samp"              "compare_with"      "samp_type"         "compare_with_type" "dist_eco"         
# [6] "dist_geo" 


unique(df_ecogeocor$samp_type)
# "Ref" "30"  "24"  "38"  "7"   "10"  "15"  "19" 

df_ecogeocor$group <- df_ecogeocor$samp_type
df_ecogeocor$group <- factor(df_ecogeocor$group, 
                             levels=c("7", "10", "15", "19", "24", "30", "38", "Ref"),
                             labels=c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr","Ref"),
                             ordered = TRUE)
length(levels(df_ecogeocor$group)) # 8

# #library(viridis); packageVersion("viridis") # ‘0.5.1’
# cols.group <- viridis(n=length(levels(df_ecogeocor$group)))
# names(cols.group) <- levels(df_ecogeocor$group)
# cols.group
# #      7 yr       10 yr       15 yr       19 yr       24 yr       30 yr       38 yr         Ref 
# # "#440154FF" "#46337EFF" "#365C8DFF" "#277F8EFF" "#1FA187FF" "#4AC16DFF" "#9FDA3AFF" "#FDE725FF" 

cols.group <- cols.group.iluka



p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "7 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "10 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "15 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "19 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "24 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "30 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "38 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.56, label = "Original data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.6))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Bray-Curtis)")
p
grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Bray-Curtis-ILUKA-A.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")


## for each rehab age determine if there is excess spatial auto-correlation??
## expect increasing eco-distance with increasing geo-distance
## if this occurs, does it occur at a rate greater 

( reveg_groups <- levels(df_ecogeocor$group) )
# "7 yr"  "10 yr" "15 yr" "19 yr" "24 yr" "30 yr" "38 yr" "Ref"  


dim(df_ecogeocor) # 225   7
dim(df_out)       # 225   10

head(df_ecogeocor)
#       samp compare_with samp_type compare_with_type  dist_eco  dist_geo group
# 2  X138408      X138410       Ref               Ref 0.7032144  146.3397   Ref
# 3  X138408      X138412       Ref               Ref 0.7545849 4708.6686   Ref
# 8  X138408      X138430       Ref               Ref 0.7632617 9736.9107   Ref
# 9  X138408      X138432       Ref               Ref 0.7492605 9129.7606   Ref
# 12 X138408      X138438       Ref               Ref 0.7736147 9220.6055   Ref
# 13 X138408      X138444       Ref               Ref 0.7373299 3732.5476   Ref

head(df_out)
#      samp compare_with year_rehab similarity group dist_measure       tax_level         study corrected                facet_x
# 2 X138408      X138410        REF   29.67856   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 3 X138408      X138412        REF   24.54151   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 4 X138408      X138430        REF   23.67383   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 5 X138408      X138432        REF   25.07395   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 6 X138408      X138438        REF   22.63853   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 7 X138408      X138444        REF   26.26701   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level

identical(df_ecogeocor$samp,df_out$samp) # TRUE
identical(df_ecogeocor$compare_with,df_out$compare_with) # TRUE


df_ecogeocor$similarity <- df_out$similarity
df_ecogeocor$tax_level <- df_out$tax_level
df_ecogeocor$study <- df_out$study
df_ecogeocor$dist_measure <- df_out$dist_measure
df_ecogeocor$facet_x <- df_out$facet_x
df_ecogeocor$similarity_corr <- NA
df_ecogeocor$dist_eco_corr <- NA


preds.temp <- data.frame(dist_geo = sort( unique(df_ecogeocor$dist_geo)))
diff_models <- list()


for (i in 1:(length(reveg_groups)-1)) {
  #i<-1
  this_group <- reveg_groups[i]
  
  x <- df_ecogeocor %>% filter(group == reveg_groups[i])
  x_ref <- df_ecogeocor %>% filter(group == "Ref")
  #dim(x) # 54 14
  #dim(x_ref) # 306  14
  # mean centre eco-distances
  x$dist_eco_mc <- x$dist_eco - mean(x$dist_eco)
  #plot(x$dist_geo,x$dist_eco_mc)
  x_ref$dist_eco_mc <- x_ref$dist_eco - mean(x_ref$dist_eco)
  #plot(x_ref$dist_geo,x_ref$dist_eco_mc)
  
  #model each spatial trend
  m <- lm(data=x, formula = dist_eco_mc ~ poly(dist_geo, 2))
  m_ref <- lm(data=x_ref, formula = dist_eco_mc ~ poly(dist_geo, 2))
  preds <- preds.temp
  # # limit predictions to max dist_geo
  # sel <- which(preds$dist_geo <= max(x$dist_geo))
  # limit predictions to extent of dist_geo
  sel <- which(preds$dist_geo >= min(x$dist_geo) & preds$dist_geo <= max(x$dist_geo))
  
  preds <- as.data.frame( preds[sel, ] )
  names(preds)[1] <- "dist_geo" # need to rename field; it is lost in previous step
  preds$m <- predict(m, newdata = preds)
  preds$m_ref <- predict(m_ref, newdata = preds)
  #plot(preds$dist_geo,preds$m,col="red")
  #points(preds$dist_geo,preds$m_ref,col="green")
  preds$diff <- preds$m - preds$m_ref
  preds$group <- this_group
  #points(preds$dist_geo,preds$diff,col="orange")
  plot(preds$dist_geo,preds$diff,col="orange", main = paste0("Diff Reveg age: ",this_group))
  # apply correction to similarity
  sel.df <- which(df_ecogeocor$group == this_group)
  for (j in 1:length(sel.df)) {
    #j<-1
    sel.pred <- which(preds$dist_geo == df_ecogeocor$dist_geo[ sel.df[j] ])
    this_diff_dist_eco <- preds$diff[sel.pred]
    # similarity = 100*(1 - distance)
    # corrected similarity = 100*[1 -  (distance - distance-correction)]
    df_ecogeocor$similarity[ sel.df[j] ] # 13.10838
    df_ecogeocor$similarity_corr[ sel.df[j] ] <- 100*(1 - (df_ecogeocor$dist_eco[ sel.df[j] ] - this_diff_dist_eco))
    df_ecogeocor$dist_eco_corr[ sel.df[j] ] <- 1 - (1/100)*df_ecogeocor$similarity_corr[ sel.df[j] ]
  }
  plot(df_ecogeocor$similarity[ sel.df ], df_ecogeocor$similarity_corr[ sel.df ], main = paste0("Group: ",this_group))
  diff_models[[this_group]] <- preds
  print(paste0("completed ",i))
}

## but note there will be no 'corrected' values for the "Ref" samples - so just use previous values 
sel <- which(df_ecogeocor$group == "Ref")
df_ecogeocor$similarity_corr[sel] # all NAs
df_ecogeocor$dist_eco_corr[sel] # all NAs
df_ecogeocor$similarity_corr[sel] <- df_ecogeocor$similarity[sel]
df_ecogeocor$dist_eco_corr[sel]   <- df_ecogeocor$dist_eco[sel]


## plot modelled data for excess spatial auto-correlation 

names(diff_models) # "7 yr"  "10 yr" "15 yr" "19 yr" "24 yr" "30 yr" "38 yr"

pdata <- diff_models[[1]]
for (i in 2:length(diff_models)) {
  pdata <- rbind(pdata,diff_models[[i]])
}
dim(pdata) # 1014    5
names(pdata) # "dist_geo" "m"        "m_ref"    "diff"     "group"
unique(pdata$group) # "7 yr"  "10 yr" "15 yr" "19 yr" "24 yr" "30 yr" "38 yr"
pdata$group <- factor(pdata$group, levels=c("7 yr",  "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr"),
                      ordered=TRUE)

p <- ggplot(data = pdata, aes(x=dist_geo,y=diff,color=factor(group)) ) +  #,color=factor(group)
  #geom_point( alpha = 0.6) +
  #geom_point() +# pdata %>% filter(group == "3 yr"),aes(x=dist_geo,y=diff)) +
  geom_line(size=1) +
  geom_hline(yintercept = 0, size=0.5, linetype="dashed", colour="grey") +
  theme_bw() +
  scale_color_manual(values=cols.group[-length(cols.group)]) +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.6)),
    legend.text = element_text(size=rel(0.56))
  ) +
  #ylim(0,0.2) +
  guides(colour = guide_legend(title = "Rehab\nage",override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "\U0394 ecol. dist. (Bray-Curtis)\nRehab vs. Ref samples")
p
grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","DIFF-in-Ecodist-vs-Geodist-Bray-Curtis-ILUKA-B.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")



## plot corrected distributions ...

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco_corr, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "7 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "10 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "15 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "19 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "24 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "30 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "38 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.56, label = "Corrected data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.6))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Bray-Curtis)")
p
grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Bray-Curtis-ILUKA-SACORRECTED-C.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")



names(df_results)
names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"         
# [8] "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"

names(df_ecogeocor)

## make names of df_ecogeocor compatible with df_out
head(df_out)
#      samp compare_with year_rehab similarity group dist_measure       tax_level         study corrected                facet_x
# 2 X138408      X138410        REF   29.67856   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 3 X138408      X138412        REF   24.54151   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 4 X138408      X138430        REF   23.67383   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 5 X138408      X138432        REF   25.07395   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 6 X138408      X138438        REF   22.63853   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level
# 7 X138408      X138444        REF   26.26701   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba        no Bray-Curtis\nASV-level

head(df_ecogeocor)
#       samp compare_with samp_type compare_with_type  dist_eco  dist_geo group similarity       tax_level         study dist_measure                facet_x similarity_corr dist_eco_corr
# 2  X138408      X138410       Ref               Ref 0.7032144  146.3397   Ref   29.67856 ASV (with Tree) Iluka-Eneabba  Bray-Curtis Bray-Curtis\nASV-level        29.67856     0.7032144
# 3  X138408      X138412       Ref               Ref 0.7545849 4708.6686   Ref   24.54151 ASV (with Tree) Iluka-Eneabba  Bray-Curtis Bray-Curtis\nASV-level        24.54151     0.7545849
# 8  X138408      X138430       Ref               Ref 0.7632617 9736.9107   Ref   23.67383 ASV (with Tree) Iluka-Eneabba  Bray-Curtis Bray-Curtis\nASV-level        23.67383     0.7632617
# 9  X138408      X138432       Ref               Ref 0.7492605 9129.7606   Ref   25.07395 ASV (with Tree) Iluka-Eneabba  Bray-Curtis Bray-Curtis\nASV-level        25.07395     0.7492605
# 12 X138408      X138438       Ref               Ref 0.7736147 9220.6055   Ref   22.63853 ASV (with Tree) Iluka-Eneabba  Bray-Curtis Bray-Curtis\nASV-level        22.63853     0.7736147
# 13 X138408      X138444       Ref               Ref 0.7373299 3732.5476   Ref   26.26701 ASV (with Tree) Iluka-Eneabba  Bray-Curtis Bray-Curtis\nASV-level        26.26701     0.7373299

# confirm these refer to the same sites and comparison sites
identical( paste0(df_out$samp,"_",df_out$compare_with) , paste0(df_ecogeocor$samp,"_",df_ecogeocor$compare_with) ) # TRUE
dim(df_out) # 225  10

df_new <- df_out
df_new$similarity <- df_ecogeocor$similarity_corr
df_new$corrected <- "yes"

head(df_new)
#      samp compare_with year_rehab similarity group dist_measure       tax_level         study corrected                facet_x
# 2 X138408      X138410        REF   29.67856   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba       yes Bray-Curtis\nASV-level
# 3 X138408      X138412        REF   24.54151   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba       yes Bray-Curtis\nASV-level
# 4 X138408      X138430        REF   23.67383   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba       yes Bray-Curtis\nASV-level
# 5 X138408      X138432        REF   25.07395   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba       yes Bray-Curtis\nASV-level
# 6 X138408      X138438        REF   22.63853   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba       yes Bray-Curtis\nASV-level
# 7 X138408      X138444        REF   26.26701   Ref  Bray-Curtis ASV (with Tree) Iluka-Eneabba       yes Bray-Curtis\nASV-level

unique(df_new$group)
# Ref 30  24  38  7   10  15  19 

df_new$group <- factor(df_new$group, levels=c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


table(df_new$group)
# 7  10  15  19  24  30  38 Ref 
# 18  18  27  18  27  18  27  72 


# only make this once
#df_results <- list()
length(df_results) #
names(df_results)

df_results[["Iluka-Eneabba-Bray-ASV-withTree-SAcorrected"]] <- df_new
#df_out <- df_results[[1]]


p <- ggplot(data=df_new, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity\nSA-corrected") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p
ggsave(plot=p, filename = paste0(workdir,"/plots/","Bray-Curtis-Similarity--Iluka-Site-with-SA-correction.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw", type = "cairo")




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_new)
# 'data.frame':	225 obs. of  10 variables:


df_new$group


# Kruskal-Wallis test
 
kt <- kruskal.test( similarity ~ group, df_new) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 141.88, df = 7, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_new$group ) 
# Ref 30  24  38  7   10  15  19 
# Levels: 7 < 10 < 15 < 19 < 24 < 30 < 38 < Ref


# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_new
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("7","10","15","19","24","30","38","Ref"),
                     labels = c("7","ten","15","19","24","thirty","38","Ref"),ordered = TRUE)


 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	8 obs. of  3 variables:

unique(sig$Group) # "15"     "19"     "24"     "38"     "7"      "Ref"    "ten"    "thirty"

sig$Group <- factor( sig$Group, levels=c("7","ten","15","19","24","thirty","38","Ref"),
                     labels = c("7","10","15","19","24","30","38","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 5     7     ab        ab 
# 7    10      b         b 
# 1    15     ab        ab 
# 2    19      a        a  
# 3    24     ab        ab 
# 8    30     ac        a c
# 4    38      c          c
# 6   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "7"   "10"  "15"  "19"  "24"  "30"  "38"  "Ref"

## store annotations for later facet_grid ggplot
names(df_new)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(28,25,30,31,32,35,39,45),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Iluka-Eneabba", times = dim(sig)[1]),
                   corrected = rep( "yes" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Iluka-Eneabba-Bray-ASV-withTree-SAcorrected"]] <- anno


#-------------------------

#### Investigating spatial auto-correlation -- Iluka-Eneabba -- Jaccard
#-------------------------

phy_in <- phy.iluka_eneabba.withTree
min(taxa_sums(phy_in)) # 1
sum(sample_sums(phy_in)) # 2155211

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142
ntaxa(r1.ps) # 27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 27115 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 27115 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 27115 tips and 27114 internal nodes ]


str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$year.of.rehab.or.ref
unique(r1.ps@sam_data$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                               labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)



names(df_results)

df_out <- df_results[[ "Iluka-Eneabba-Jaccard-ASV-withTree" ]]


names(r1.ps@sam_data) # includes ...
# [1] "Sample_ID"                                             "NCBI_Submission"                                      
# [3] "NCBI_Sample_Accession"                                 "Organism"                                             
# [5] "Tax_ID"                                                "SampleName_depth"                                     
# [7] "NCBI_BioProject"                                       "Date_sampled..YYYY.MM.DD."                            
# [9] "Time_sampled..hh.mm."                                  "Latitude..decimal.degrees."                           
# [11] "Longitude..decimal.degrees."                

r1.ps@sam_data$Latitude..decimal.degrees.
r1.ps@sam_data$Longitude..decimal.degrees.

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")

identical(colnames(dist.jacc),rownames(dist.jacc)) # TRUE


## determine spatial auto-correlation

# data.frame containing longitude and latitude coordinates

long_lats <- data.frame( xlong=r1.ps@sam_data$Longitude..decimal.degrees.,
                         ylat =r1.ps@sam_data$Latitude..decimal.degrees.
)
row.names(long_lats) <- row.names(r1.ps@sam_data)

# calculate matrix of pairwise distances between all points
geo_dist.m <- geodist(long_lats, measure = "haversine")

geo_dist <- as.data.frame( geo_dist.m )
row.names(geo_dist) <- row.names(r1.ps@sam_data)
colnames(geo_dist) <- row.names(r1.ps@sam_data)

# check that colnames and row.names align
identical(row.names(geo_dist),row.names(dist.jacc)) # TRUE
identical(colnames(geo_dist),colnames(dist.jacc)) # TRUE

identical(colnames(dist.jacc), row.names(dist.jacc)) # TRUE


df_eco <- dist.jacc

df_ecogeocor <- data.frame(samp= rep( x = colnames(df_eco), each=length( colnames(df_eco) ) ),
                           compare_with = rep( x = colnames(df_eco), times = length(colnames(df_eco))),
                           
                           samp_type=NA,
                           compare_with_type=NA,
                           
                           dist_eco=NA,
                           dist_geo=NA
)

for (i in 1:dim(df_ecogeocor)[1]) {
  #i<-1
  this_samp <- df_ecogeocor$samp[i]
  this_compare <- df_ecogeocor$compare_with[i]
  sel1 <- which( row.names(r1.ps@sam_data) == this_samp)
  df_ecogeocor$samp_type[i] <- as.character( r1.ps@sam_data$group[sel1] )
  
  sel2 <- which( row.names(r1.ps@sam_data) == this_compare)
  df_ecogeocor$compare_with_type[i] <- as.character( r1.ps@sam_data$group[sel2] )
  
  row <- which(rownames(df_eco)==this_samp)
  col <- which(colnames(df_eco)==this_compare)
  df_ecogeocor$dist_eco[i] <- df_eco[row,col]
  
  row <- which(rownames(geo_dist)==this_samp)
  col <- which(colnames(geo_dist)==this_compare)
  df_ecogeocor$dist_geo[i] <- geo_dist[row,col]
  
  print(paste0("completed ",i))
}

dim(df_ecogeocor) #  676   6

# remove remnant self-comparisons
sel <- which(df_ecogeocor$samp==df_ecogeocor$compare_with) # qty 26
identical( which(df_ecogeocor$samp==df_ecogeocor$compare_with), which(df_ecogeocor$dist_eco==0) ) # TRUE
df_ecogeocor <- df_ecogeocor[-sel, ]
dim(df_ecogeocor) # 650   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)
summary(df_ecogeocor$dist_eco)

head(df_ecogeocor)
#      samp compare_with samp_type compare_with_type  dist_eco  dist_geo
# 2 X138408      X138410       Ref               Ref 0.8133140  146.3397
# 3 X138408      X138412       Ref               Ref 0.8339100 4708.6686
# 4 X138408      X138418       Ref                30 0.8577062 6069.0393
# 5 X138408      X138424       Ref                24 0.8766965 7423.1019
# 6 X138408      X138426       Ref                24 0.8791935 7596.8801
# 7 X138408      X138428       Ref                38 0.8728149 9336.2228


# consider only compare with type 'Ref' as that is focus 
# and potential for problematic study design and analysis

sel <- which(df_ecogeocor$compare_with_type=="Ref") # 225
df_ecogeocor <- df_ecogeocor[sel, ]
dim(df_ecogeocor) # 225   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)

names(df_ecogeocor)
# [1] "samp"              "compare_with"      "samp_type"         "compare_with_type" "dist_eco"         
# [6] "dist_geo" 


unique(df_ecogeocor$samp_type)
# "Ref" "30"  "24"  "38"  "7"   "10"  "15"  "19" 

df_ecogeocor$group <- df_ecogeocor$samp_type
df_ecogeocor$group <- factor(df_ecogeocor$group, 
                             levels=c("7", "10", "15", "19", "24", "30", "38", "Ref"),
                             labels=c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr","Ref"),
                             ordered = TRUE)
length(levels(df_ecogeocor$group)) # 8

# #library(viridis); packageVersion("viridis") # ‘0.5.1’
# cols.group <- viridis(n=length(levels(df_ecogeocor$group)))
# names(cols.group) <- levels(df_ecogeocor$group)
# cols.group
# #      7 yr       10 yr       15 yr       19 yr       24 yr       30 yr       38 yr         Ref 
# # "#440154FF" "#46337EFF" "#365C8DFF" "#277F8EFF" "#1FA187FF" "#4AC16DFF" "#9FDA3AFF" "#FDE725FF" 

cols.group <- cols.group.iluka

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "7 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "10 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "15 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "19 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "24 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "30 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "38 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.73, label = "Original data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.6))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Jaccard)")
p
grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Jaccard-ILUKA-A.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")


## for each rehab age determine if there is excess spatial auto-correlation??
## expect increasing eco-distance with increasing geo-distance
## if this occurs, does it occur at a rate greater 

( reveg_groups <- levels(df_ecogeocor$group) )
# "7 yr"  "10 yr" "15 yr" "19 yr" "24 yr" "30 yr" "38 yr" "Ref"  


dim(df_ecogeocor) # 225   7
dim(df_out)       # 225   10

head(df_ecogeocor)
#       samp compare_with samp_type compare_with_type  dist_eco  dist_geo group
# 2  X138408      X138410       Ref               Ref 0.8133140  146.3397   Ref
# 3  X138408      X138412       Ref               Ref 0.8339100 4708.6686   Ref
# 8  X138408      X138430       Ref               Ref 0.8401989 9736.9107   Ref
# 9  X138408      X138432       Ref               Ref 0.8531169 9129.7606   Ref
# 12 X138408      X138438       Ref               Ref 0.8660997 9220.6055   Ref
# 13 X138408      X138444       Ref               Ref 0.8294949 3732.5476   Ref

head(df_out)
#      samp compare_with year_rehab similarity group dist_measure       tax_level         study corrected            facet_x
# 2 X138408      X138410        REF   18.66860   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 3 X138408      X138412        REF   16.60900   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 4 X138408      X138430        REF   15.98011   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 5 X138408      X138432        REF   14.68831   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 6 X138408      X138438        REF   13.39003   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 7 X138408      X138444        REF   17.05051   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level

identical(df_ecogeocor$samp,df_out$samp) # TRUE
identical(df_ecogeocor$compare_with,df_out$compare_with) # TRUE


df_ecogeocor$similarity <- df_out$similarity
df_ecogeocor$tax_level <- df_out$tax_level
df_ecogeocor$study <- df_out$study
df_ecogeocor$dist_measure <- df_out$dist_measure
df_ecogeocor$facet_x <- df_out$facet_x
df_ecogeocor$similarity_corr <- NA
df_ecogeocor$dist_eco_corr <- NA


preds.temp <- data.frame(dist_geo = sort( unique(df_ecogeocor$dist_geo)))
diff_models <- list()


for (i in 1:(length(reveg_groups)-1)) {
  #i<-1
  this_group <- reveg_groups[i]
  
  x <- df_ecogeocor %>% filter(group == reveg_groups[i])
  x_ref <- df_ecogeocor %>% filter(group == "Ref")
  #dim(x) # 54 14
  #dim(x_ref) # 306  14
  # mean centre eco-distances
  x$dist_eco_mc <- x$dist_eco - mean(x$dist_eco)
  #plot(x$dist_geo,x$dist_eco_mc)
  x_ref$dist_eco_mc <- x_ref$dist_eco - mean(x_ref$dist_eco)
  #plot(x_ref$dist_geo,x_ref$dist_eco_mc)
  
  #model each spatial trend
  m <- lm(data=x, formula = dist_eco_mc ~ poly(dist_geo, 2))
  m_ref <- lm(data=x_ref, formula = dist_eco_mc ~ poly(dist_geo, 2))
  preds <- preds.temp
  # # limit predictions to max dist_geo
  # sel <- which(preds$dist_geo <= max(x$dist_geo))
  # limit predictions to extent of dist_geo
  sel <- which(preds$dist_geo >= min(x$dist_geo) & preds$dist_geo <= max(x$dist_geo))
  
  preds <- as.data.frame( preds[sel, ] )
  names(preds)[1] <- "dist_geo" # need to rename field; it is lost in previous step
  preds$m <- predict(m, newdata = preds)
  preds$m_ref <- predict(m_ref, newdata = preds)
  #plot(preds$dist_geo,preds$m,col="red")
  #points(preds$dist_geo,preds$m_ref,col="green")
  preds$diff <- preds$m - preds$m_ref
  preds$group <- this_group
  #points(preds$dist_geo,preds$diff,col="orange")
  plot(preds$dist_geo,preds$diff,col="orange", main = paste0("Diff Reveg age: ",this_group))
  # apply correction to similarity
  sel.df <- which(df_ecogeocor$group == this_group)
  for (j in 1:length(sel.df)) {
    #j<-1
    sel.pred <- which(preds$dist_geo == df_ecogeocor$dist_geo[ sel.df[j] ])
    this_diff_dist_eco <- preds$diff[sel.pred]
    # similarity = 100*(1 - distance)
    # corrected similarity = 100*[1 -  (distance - distance-correction)]
    df_ecogeocor$similarity[ sel.df[j] ] # 13.10838
    df_ecogeocor$similarity_corr[ sel.df[j] ] <- 100*(1 - (df_ecogeocor$dist_eco[ sel.df[j] ] - this_diff_dist_eco))
    df_ecogeocor$dist_eco_corr[ sel.df[j] ] <- 1 - (1/100)*df_ecogeocor$similarity_corr[ sel.df[j] ]
  }
  plot(df_ecogeocor$similarity[ sel.df ], df_ecogeocor$similarity_corr[ sel.df ], main = paste0("Group: ",this_group))
  diff_models[[this_group]] <- preds
  print(paste0("completed ",i))
}

## but note there will be no 'corrected' values for the "Ref" samples - so just use previous values 
sel <- which(df_ecogeocor$group == "Ref")
df_ecogeocor$similarity_corr[sel] # all NAs
df_ecogeocor$dist_eco_corr[sel] # all NAs
df_ecogeocor$similarity_corr[sel] <- df_ecogeocor$similarity[sel]
df_ecogeocor$dist_eco_corr[sel]   <- df_ecogeocor$dist_eco[sel]


## plot modelled data for excess spatial auto-correlation 

names(diff_models) # "7 yr"  "10 yr" "15 yr" "19 yr" "24 yr" "30 yr" "38 yr"

pdata <- diff_models[[1]]
for (i in 2:length(diff_models)) {
  pdata <- rbind(pdata,diff_models[[i]])
}
dim(pdata) # 1118    5
names(pdata) # "dist_geo" "m"        "m_ref"    "diff"     "group"
unique(pdata$group) # "7 yr"  "10 yr" "15 yr" "19 yr" "24 yr" "30 yr" "38 yr"
pdata$group <- factor(pdata$group, levels=c("7 yr",  "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr"),
                      ordered=TRUE)

p <- ggplot(data = pdata, aes(x=dist_geo,y=diff,color=factor(group)) ) +  #,color=factor(group)
  #geom_point( alpha = 0.6) +
  #geom_point() +# pdata %>% filter(group == "3 yr"),aes(x=dist_geo,y=diff)) +
  geom_line(size=1) +
  geom_hline(yintercept = 0, size=0.5, linetype="dashed", colour="grey") +
  theme_bw() +
  scale_color_manual(values=cols.group[-length(cols.group)]) +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.6)),
    legend.text = element_text(size=rel(0.56))
  ) +
  #ylim(0,0.2) +
  guides(colour = guide_legend(title = "Rehab\nage",override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "\U0394 ecol. dist. (Jaccard)\nRehab vs. Ref samples")
p
grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","DIFF-in-Ecodist-vs-Geodist-Jaccard-ILUKA-B.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")



## plot corrected distributions ...

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco_corr, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "7 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "10 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "15 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "19 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "24 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "30 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "38 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.73, label = "Corrected data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.6))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA))) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Jaccard)")
p
grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Jaccard-ILUKA-SACORRECTED-C.tiff"), width = 8, height = 10, units = "cm", res=600, compression="lzw",type="cairo")



names(df_results)
names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"         
# [8] "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"

names(df_ecogeocor)

## make names of df_ecogeocor compatible with df_out
head(df_out)
#      samp compare_with year_rehab similarity group dist_measure       tax_level         study corrected            facet_x
# 2 X138408      X138410        REF   18.66860   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 3 X138408      X138412        REF   16.60900   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 4 X138408      X138430        REF   15.98011   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 5 X138408      X138432        REF   14.68831   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 6 X138408      X138438        REF   13.39003   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level
# 7 X138408      X138444        REF   17.05051   Ref      Jaccard ASV (with tree) Iluka-Eneabba        no Jaccard\nASV-level

head(df_ecogeocor)
# samp compare_with samp_type compare_with_type  dist_eco  dist_geo group similarity       tax_level         study dist_measure            facet_x
# 2  X138408      X138410       Ref               Ref 0.8133140  146.3397   Ref   18.66860 ASV (with tree) Iluka-Eneabba      Jaccard Jaccard\nASV-level
# 3  X138408      X138412       Ref               Ref 0.8339100 4708.6686   Ref   16.60900 ASV (with tree) Iluka-Eneabba      Jaccard Jaccard\nASV-level
# 8  X138408      X138430       Ref               Ref 0.8401989 9736.9107   Ref   15.98011 ASV (with tree) Iluka-Eneabba      Jaccard Jaccard\nASV-level
# 9  X138408      X138432       Ref               Ref 0.8531169 9129.7606   Ref   14.68831 ASV (with tree) Iluka-Eneabba      Jaccard Jaccard\nASV-level
# 12 X138408      X138438       Ref               Ref 0.8660997 9220.6055   Ref   13.39003 ASV (with tree) Iluka-Eneabba      Jaccard Jaccard\nASV-level
# 13 X138408      X138444       Ref               Ref 0.8294949 3732.5476   Ref   17.05051 ASV (with tree) Iluka-Eneabba      Jaccard Jaccard\nASV-level
# similarity_corr dist_eco_corr
# 2         18.66860     0.8133140
# 3         16.60900     0.8339100
# 8         15.98011     0.8401989
# 9         14.68831     0.8531169
# 12        13.39003     0.8660997
# 13        17.05051     0.8294949

# confirm these refer to the same sites and comparison sites
identical( paste0(df_out$samp,"_",df_out$compare_with) , paste0(df_ecogeocor$samp,"_",df_ecogeocor$compare_with) ) # TRUE
dim(df_out) # 225  10

df_new <- df_out
df_new$similarity <- df_ecogeocor$similarity_corr
df_new$corrected <- "yes"

head(df_new)
#      samp compare_with year_rehab similarity group dist_measure       tax_level         study corrected            facet_x
# 2 X138408      X138410        REF   18.66860   Ref      Jaccard ASV (with tree) Iluka-Eneabba       yes Jaccard\nASV-level
# 3 X138408      X138412        REF   16.60900   Ref      Jaccard ASV (with tree) Iluka-Eneabba       yes Jaccard\nASV-level
# 4 X138408      X138430        REF   15.98011   Ref      Jaccard ASV (with tree) Iluka-Eneabba       yes Jaccard\nASV-level
# 5 X138408      X138432        REF   14.68831   Ref      Jaccard ASV (with tree) Iluka-Eneabba       yes Jaccard\nASV-level
# 6 X138408      X138438        REF   13.39003   Ref      Jaccard ASV (with tree) Iluka-Eneabba       yes Jaccard\nASV-level
# 7 X138408      X138444        REF   17.05051   Ref      Jaccard ASV (with tree) Iluka-Eneabba       yes Jaccard\nASV-level

unique(df_new$group)
# [1] Ref 30  24  38  7   10  15  19 
# Levels: 7 < 10 < 15 < 19 < 24 < 30 < 38 < Ref

# df_new$group <- factor(df_new$group, levels=c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# # Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


table(df_new$group)
# 7  10  15  19  24  30  38 Ref 
# 18  18  27  18  27  18  27  72 


# only make this once
#df_results <- list()
length(df_results) #
names(df_results)

df_results[["Iluka-Eneabba-Jaccard-ASV-withTree-SAcorrected"]] <- df_new
#df_out <- df_results[[1]]


p <- ggplot(data=df_new, aes(x=group, similarity)) + ggtitle("Jaccard Similarity\nSA-corrected") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p
ggsave(plot=p, filename = paste0(workdir,"/plots/","Jaccard-Similarity--Iluka-Site-with-SA-correction.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw", type = "cairo")




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_new)
# 'data.frame':	225 obs. of  10 variables:


df_new$group


# Kruskal-Wallis test
 
kt <- kruskal.test( similarity ~ group, df_new) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 155.71, df = 7, p-value < 2.2e-16


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_new$group ) 
# Ref 30  24  38  7   10  15  19 
# Levels: 7 < 10 < 15 < 19 < 24 < 30 < 38 < Ref


# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_new
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("7","10","15","19","24","30","38","Ref"),
                     labels = c("7","ten","15","19","24","thirty","38","Ref"),ordered = TRUE)


 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	8 obs. of  3 variables:

unique(sig$Group) # "15"     "19"     "24"     "38"     "7"      "Ref"    "ten"    "thirty"

sig$Group <- factor( sig$Group, levels=c("7","ten","15","19","24","thirty","38","Ref"),
                     labels = c("7","10","15","19","24","30","38","Ref"),
                     ordered=TRUE )

sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 5     7     ab       ab  
# 7    10      b        b  
# 1    15     ab       ab  
# 2    19      a       a   
# 3    24     ab       ab  
# 8    30     ad       a  d
# 4    38     cd         cd
# 6   Ref      c         c 

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "7"   "10"  "15"  "19"  "24"  "30"  "38"  "Ref"

## store annotations for later facet_grid ggplot
names(df_new)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(18,15,18,20,21,22,24,25),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "Iluka-Eneabba", times = dim(sig)[1]),
                   corrected = rep( "yes" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["Iluka-Eneabba-Jaccard-ASV-withTree-SAcorrected"]] <- anno


#-------------------------

#### Investigating spatial auto-correlation -- South32 -- Bray-Curtis 
#-------------------------

names(df_results)

phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122
ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]



str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`year of rehab or ref`
unique(r1.ps@sam_data$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                               labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)



names(df_results)

df_out <- df_results[[ "South32-Bray-ASV-withTree" ]]


names(r1.ps@sam_data) # includes ...
# [1] "Sample_ID"                                             "NCBI_Submission"                                      
# [3] "NCBI_Sample_Accession"                                 "Organism"                                             
# [5] "Tax_ID"                                                "SampleName_depth"                                     
# [7] "NCBI_BioProject"                                       "Date_sampled [YYYY-MM-DD]"                            
# [9] "Time_sampled [hh:mm]"                                  "Latitude [decimal degrees]"                           
# [11] "Longitude [decimal degrees]"     

r1.ps@sam_data$`Latitude [decimal degrees]`
r1.ps@sam_data$`Longitude [decimal degrees]`

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")

identical(colnames(dist.bray),rownames(dist.bray)) # TRUE


## determine spatial auto-correlation

# data.frame containing longitude and latitude coordinates

long_lats <- data.frame( xlong=r1.ps@sam_data$`Longitude [decimal degrees]`,
                         ylat =r1.ps@sam_data$`Latitude [decimal degrees]`
)
row.names(long_lats) <- row.names(r1.ps@sam_data)

# calculate matrix of pairwise distances between all points
geo_dist.m <- geodist(long_lats, measure = "haversine")

geo_dist <- as.data.frame( geo_dist.m )
row.names(geo_dist) <- row.names(r1.ps@sam_data)
colnames(geo_dist) <- row.names(r1.ps@sam_data)

# check that colnames and row.names align
identical(row.names(geo_dist),row.names(dist.bray)) # TRUE
identical(colnames(geo_dist),colnames(dist.bray)) # TRUE

identical(colnames(dist.bray), row.names(dist.bray)) # TRUE


df_eco <- dist.bray

df_ecogeocor <- data.frame(samp= rep( x = colnames(df_eco), each=length( colnames(df_eco) ) ),
                           compare_with = rep( x = colnames(df_eco), times = length(colnames(df_eco))),
                           
                           samp_type=NA,
                           compare_with_type=NA,
                           
                           dist_eco=NA,
                           dist_geo=NA
)

for (i in 1:dim(df_ecogeocor)[1]) {
  #i<-1
  this_samp <- df_ecogeocor$samp[i]
  this_compare <- df_ecogeocor$compare_with[i]
  sel1 <- which( row.names(r1.ps@sam_data) == this_samp)
  df_ecogeocor$samp_type[i] <- as.character( r1.ps@sam_data$group[sel1] )
  
  sel2 <- which( row.names(r1.ps@sam_data) == this_compare)
  df_ecogeocor$compare_with_type[i] <- as.character( r1.ps@sam_data$group[sel2] )
  
  row <- which(rownames(df_eco)==this_samp)
  col <- which(colnames(df_eco)==this_compare)
  df_ecogeocor$dist_eco[i] <- df_eco[row,col]
  
  row <- which(rownames(geo_dist)==this_samp)
  col <- which(colnames(geo_dist)==this_compare)
  df_ecogeocor$dist_geo[i] <- geo_dist[row,col]
  
  print(paste0("completed ",i))
}

dim(df_ecogeocor) #  625   6

# remove remnant self-comparisons
sel <- which(df_ecogeocor$samp==df_ecogeocor$compare_with) #
identical( which(df_ecogeocor$samp==df_ecogeocor$compare_with), which(df_ecogeocor$dist_eco==0) ) # TRUE
df_ecogeocor <- df_ecogeocor[-sel, ]
dim(df_ecogeocor) # 600   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)
summary(df_ecogeocor$dist_eco)

head(df_ecogeocor)
#      samp compare_with samp_type compare_with_type  dist_eco  dist_geo
# 2 X138358      X138360        23                23 0.5328332  698.9024
# 3 X138358      X138362        23               Ref 0.6198404  991.2395
# 4 X138358      X138364        23                 2 0.7639962 1212.3960
# 5 X138358      X138366        23               Ref 0.7285577 1291.3074
# 6 X138358      X138368        23                17 0.6351576 1679.4491
# 7 X138358      X138370        23                12 0.6750120 1201.3421


# consider only compare with type 'Ref' as that is focus 
# and potential for problematic study design and analysis

sel <- which(df_ecogeocor$compare_with_type=="Ref") # qty 144
df_ecogeocor <- df_ecogeocor[sel, ]
dim(df_ecogeocor) # 144  6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)

names(df_ecogeocor)
# [1] "samp"              "compare_with"      "samp_type"         "compare_with_type" "dist_eco"         
# [6] "dist_geo" 


unique(df_ecogeocor$samp_type)
# "23"  "Ref" "2"   "17"  "12"  "20"  "8"   "14"  "28" 

df_ecogeocor$group <- df_ecogeocor$samp_type
df_ecogeocor$group <- factor(df_ecogeocor$group, 
                             levels=c("2","8","12","14","17","20","23","28","Ref"),
                             labels=c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"),
                             ordered = TRUE)
length(levels(df_ecogeocor$group)) # 9

# #library(viridis); packageVersion("viridis") # ‘0.5.1’
# cols.group <- viridis(n=length(levels(df_ecogeocor$group)))
# names(cols.group) <- levels(df_ecogeocor$group)
# cols.group
# # 2 yr        8 yr       12 yr       14 yr       17 yr       20 yr       23 yr       28 yr         Ref 
# # "#440154FF" "#472D7BFF" "#3B528BFF" "#2C728EFF" "#21908CFF" "#27AD81FF" "#5DC863FF" "#AADC32FF" "#FDE725FF" 

cols.group <- cols.group.south32

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "12 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "20 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "23 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "28 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.51, label = "Original data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.65))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Bray-Curtis)")
p
grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Bray-Curtis-SOUTH32-A.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")


## for each rehab age determine if there is excess spatial auto-correlation??
## expect increasing eco-distance with increasing geo-distance
## if this occurs, does it occur at a rate greater 

( reveg_groups <- levels(df_ecogeocor$group) )
# "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr" "Ref"  


dim(df_ecogeocor) # 144   7
dim(df_out)       # 144   10

head(df_ecogeocor)
# samp compare_with samp_type compare_with_type  dist_eco  dist_geo group
# 3  X138358      X138362        23               Ref 0.6198404  991.2395 23 yr
# 5  X138358      X138366        23               Ref 0.7285577 1291.3074 23 yr
# 9  X138358      X138374        23               Ref 0.6956136 2447.0999 23 yr
# 10 X138358      X138376        23               Ref 0.7995085 2692.4727 23 yr
# 11 X138358      X138378        23               Ref 0.7671557 4442.1090 23 yr
# 24 X138358      X138404        23               Ref 0.7496951 5697.4216 23 yr

head(df_out)
# samp compare_with year_rehab similarity group dist_measure       tax_level   study corrected                facet_x
# 1 X138358      X138362       1996   38.01596    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 2 X138358      X138366       1996   27.14423    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 3 X138358      X138374       1996   30.43864    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 4 X138358      X138376       1996   20.04915    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 5 X138358      X138378       1996   23.28443    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 6 X138358      X138404       1996   25.03049    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level

identical(df_ecogeocor$samp,df_out$samp) # TRUE
identical(df_ecogeocor$compare_with,df_out$compare_with) # TRUE


df_ecogeocor$similarity <- df_out$similarity
df_ecogeocor$tax_level <- df_out$tax_level
df_ecogeocor$study <- df_out$study
df_ecogeocor$dist_measure <- df_out$dist_measure
df_ecogeocor$facet_x <- df_out$facet_x
df_ecogeocor$similarity_corr <- NA
df_ecogeocor$dist_eco_corr <- NA


preds.temp <- data.frame(dist_geo = sort( unique(df_ecogeocor$dist_geo)))
diff_models <- list()


for (i in 1:(length(reveg_groups)-1)) {
  #i<-1
  this_group <- reveg_groups[i]
  
  x <- df_ecogeocor %>% filter(group == reveg_groups[i])
  x_ref <- df_ecogeocor %>% filter(group == "Ref")
  #dim(x) # 54 14
  #dim(x_ref) # 306  14
  # mean centre eco-distances
  x$dist_eco_mc <- x$dist_eco - mean(x$dist_eco)
  #plot(x$dist_geo,x$dist_eco_mc)
  x_ref$dist_eco_mc <- x_ref$dist_eco - mean(x_ref$dist_eco)
  #plot(x_ref$dist_geo,x_ref$dist_eco_mc)
  
  #model each spatial trend
  m <- lm(data=x, formula = dist_eco_mc ~ poly(dist_geo, 2))
  m_ref <- lm(data=x_ref, formula = dist_eco_mc ~ poly(dist_geo, 2))
  preds <- preds.temp
  # # limit predictions to max dist_geo
  # sel <- which(preds$dist_geo <= max(x$dist_geo))
  # limit predictions to extent of dist_geo
  sel <- which(preds$dist_geo >= min(x$dist_geo) & preds$dist_geo <= max(x$dist_geo))
  
  preds <- as.data.frame( preds[sel, ] )
  names(preds)[1] <- "dist_geo" # need to rename field; it is lost in previous step
  preds$m <- predict(m, newdata = preds)
  preds$m_ref <- predict(m_ref, newdata = preds)
  #plot(preds$dist_geo,preds$m,col="red")
  #points(preds$dist_geo,preds$m_ref,col="green")
  preds$diff <- preds$m - preds$m_ref
  preds$group <- this_group
  #points(preds$dist_geo,preds$diff,col="orange")
  plot(preds$dist_geo,preds$diff,col="orange", main = paste0("Diff Reveg age: ",this_group))
  # apply correction to similarity
  sel.df <- which(df_ecogeocor$group == this_group)
  for (j in 1:length(sel.df)) {
    #j<-1
    sel.pred <- which(preds$dist_geo == df_ecogeocor$dist_geo[ sel.df[j] ])
    this_diff_dist_eco <- preds$diff[sel.pred]
    # similarity = 100*(1 - distance)
    # corrected similarity = 100*[1 -  (distance - distance-correction)]
    df_ecogeocor$similarity[ sel.df[j] ] # 13.10838
    df_ecogeocor$similarity_corr[ sel.df[j] ] <- 100*(1 - (df_ecogeocor$dist_eco[ sel.df[j] ] - this_diff_dist_eco))
    df_ecogeocor$dist_eco_corr[ sel.df[j] ] <- 1 - (1/100)*df_ecogeocor$similarity_corr[ sel.df[j] ]
  }
  plot(df_ecogeocor$similarity[ sel.df ], df_ecogeocor$similarity_corr[ sel.df ], main = paste0("Group: ",this_group))
  diff_models[[this_group]] <- preds
  print(paste0("completed ",i))
}

## but note there will be no 'corrected' values for the "Ref" samples - so just use previous values 
sel <- which(df_ecogeocor$group == "Ref")
df_ecogeocor$similarity_corr[sel] # all NAs
df_ecogeocor$dist_eco_corr[sel] # all NAs
df_ecogeocor$similarity_corr[sel] <- df_ecogeocor$similarity[sel]
df_ecogeocor$dist_eco_corr[sel]   <- df_ecogeocor$dist_eco[sel]


## plot modelled data for excess spatial auto-correlation 

names(diff_models) # "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr"

pdata <- diff_models[[1]]
for (i in 2:length(diff_models)) {
  pdata <- rbind(pdata,diff_models[[i]])
}
dim(pdata) # 937    5
names(pdata) # "dist_geo" "m"        "m_ref"    "diff"     "group"
unique(pdata$group) # "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr"
pdata$group <- factor(pdata$group, levels=c("2 yr",  "8 yr",  "12 yr", "14 yr", "17 yr", "20 yr", "23 yr", "28 yr"),
                      ordered=TRUE)

p <- ggplot(data = pdata, aes(x=dist_geo,y=diff,color=factor(group)) ) +  #,color=factor(group)
  #geom_point( alpha = 0.6) +
  #geom_point() +# pdata %>% filter(group == "3 yr"),aes(x=dist_geo,y=diff)) +
  geom_line(size=1) +
  geom_hline(yintercept = 0, size=0.5, linetype="dashed", colour="grey") +
  theme_bw() +
  scale_color_manual(values=cols.group[-length(cols.group)]) +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.7)),
    legend.text = element_text(size=rel(0.65))
  ) +
  #ylim(0,0.2) +
  guides(colour = guide_legend(title = "Rehab\nage",override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "\U0394 ecol. dist. (Bray-Curtis)\nRehab vs. Ref samples")
p
grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","DIFF-in-Ecodist-vs-Geodist-Bray-Curtis-SOUTH32-B.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")



## plot corrected distributions ...

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco_corr, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "12 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "20 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "23 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "28 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.51, label = "Corrected data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.65))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Bray-Curtis)")
p
grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Bray-Curtis-SOUTH32-SACORRECTED-C.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")



names(df_results)
names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"         
# [8] "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"

names(df_ecogeocor)

## make names of df_ecogeocor compatible with df_out
head(df_out)
# samp compare_with year_rehab similarity group dist_measure       tax_level   study corrected                facet_x
# 1 X138358      X138362       1996   38.01596    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 2 X138358      X138366       1996   27.14423    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 3 X138358      X138374       1996   30.43864    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 4 X138358      X138376       1996   20.04915    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 5 X138358      X138378       1996   23.28443    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 6 X138358      X138404       1996   25.03049    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level

head(df_ecogeocor)
# samp compare_with samp_type compare_with_type  dist_eco  dist_geo group similarity       tax_level   study dist_measure                facet_x similarity_corr
# 3  X138358      X138362        23               Ref 0.6198404  991.2395 23 yr   38.01596 ASV (with Tree) South32  Bray-Curtis Bray-Curtis\nASV-level        40.53385
# 5  X138358      X138366        23               Ref 0.7285577 1291.3074 23 yr   27.14423 ASV (with Tree) South32  Bray-Curtis Bray-Curtis\nASV-level        29.95435
# 9  X138358      X138374        23               Ref 0.6956136 2447.0999 23 yr   30.43864 ASV (with Tree) South32  Bray-Curtis Bray-Curtis\nASV-level        33.84204
# 10 X138358      X138376        23               Ref 0.7995085 2692.4727 23 yr   20.04915 ASV (with Tree) South32  Bray-Curtis Bray-Curtis\nASV-level        23.46973
# 11 X138358      X138378        23               Ref 0.7671557 4442.1090 23 yr   23.28443 ASV (with Tree) South32  Bray-Curtis Bray-Curtis\nASV-level        25.72323
# 24 X138358      X138404        23               Ref 0.7496951 5697.4216 23 yr   25.03049 ASV (with Tree) South32  Bray-Curtis Bray-Curtis\nASV-level        25.57148
# dist_eco_corr
# 3      0.5946615
# 5      0.7004565
# 9      0.6615796
# 10     0.7653027
# 11     0.7427677
# 24     0.7442852

# confirm these refer to the same sites and comparison sites
identical( paste0(df_out$samp,"_",df_out$compare_with) , paste0(df_ecogeocor$samp,"_",df_ecogeocor$compare_with) ) # TRUE
dim(df_out) # 144  10

df_new <- df_out
df_new$similarity <- df_ecogeocor$similarity_corr
df_new$corrected <- "yes"

head(df_new)
# samp compare_with year_rehab similarity group dist_measure       tax_level   study corrected                facet_x
# 1 X138358      X138362       1996   40.53385    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 2 X138358      X138366       1996   29.95435    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 3 X138358      X138374       1996   33.84204    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 4 X138358      X138376       1996   23.46973    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 5 X138358      X138378       1996   25.72323    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 6 X138358      X138404       1996   25.57148    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level

unique(df_new$group)
# [1] 23  Ref 2   17  12  20  8   14  28 
# Levels: 2 < 8 < 12 < 14 < 17 < 20 < 23 < 28 < Ref



table(df_new$group)
# 2   8  12  14  17  20  23  28 Ref 
# 18  18   6  12  12  12  24  12  30 


# only make this once
#df_results <- list()
length(df_results) #
names(df_results)

df_results[["South32-Bray-ASV-withTree-SAcorrected"]] <- df_new
#df_out <- df_results[[1]]


p <- ggplot(data=df_new, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity\nSA-corrected") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p
ggsave(plot=p, filename = paste0(workdir,"/plots/","Bray-Curtis-Similarity--South32-Site-with-SA-correction.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw", type = "cairo")




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_new)
# 'data.frame':	225 obs. of  10 variables:


df_new$group


# Kruskal-Wallis test
 
kt <- kruskal.test( similarity ~ group, df_new) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 71.72, df = 8, p-value = 2.232e-12


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_new$group ) 
# [1] 23  Ref 2   17  12  20  8   14  28 
# Levels: 2 < 8 < 12 < 14 < 17 < 20 < 23 < 28 < Ref


# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_new
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("2","8","12","14","17","20","23","28","Ref"),
                     labels = c("2","8","12","14","17","twenty","23","28","Ref"),ordered = TRUE)


 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	9 obs. of  3 variables:

unique(sig$Group) # "12"     "14"     "17"     "2"      "23"     "28"     "8"      "Ref"    "twenty"

sig$Group <- factor( sig$Group, levels=c("2","8","12","14","17","twenty","23","28","Ref"),
                     labels = c("2","8","12","14","17","20","23","28","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 4     2      a         a 
# 7     8      a         a 
# 1    12     ab         ab
# 2    14     ab         ab
# 3    17     ab         ab
# 9    20      b          b
# 5    23      b          b
# 6    28      b          b
# 8   Ref      b          b

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "12"  "14"  "17"  "20"  "23"  "28"  "Ref"

## store annotations for later facet_grid ggplot
names(df_new)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(32,32,36,37,38,50,48,49,50),
                   dist_measure = rep( "Bray-Curtis", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "South32", times = dim(sig)[1]),
                   corrected = rep( "yes" , times = dim(sig)[1]),
                   facet_x  = rep( "Bray-Curtis\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["South32-Bray-ASV-withTree-SAcorrected"]] <- anno


#-------------------------

#### Investigating spatial auto-correlation -- South32 -- Jaccard
#-------------------------

phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122
ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]



str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`year of rehab or ref`
unique(r1.ps@sam_data$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                               labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)



names(df_results)

df_out <- df_results[[ "South32-Jaccard-ASV-withTree" ]]


names(r1.ps@sam_data) # includes ...
# [1] "Sample_ID"                                             "NCBI_Submission"                                      
# [3] "NCBI_Sample_Accession"                                 "Organism"                                             
# [5] "Tax_ID"                                                "SampleName_depth"                                     
# [7] "NCBI_BioProject"                                       "Date_sampled [YYYY-MM-DD]"                            
# [9] "Time_sampled [hh:mm]"                                  "Latitude [decimal degrees]"                           
# [11] "Longitude [decimal degrees]"     

r1.ps@sam_data$`Latitude [decimal degrees]`
r1.ps@sam_data$`Longitude [decimal degrees]`

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")

identical(colnames(dist.jacc),rownames(dist.jacc)) # TRUE


## determine spatial auto-correlation

# data.frame containing longitude and latitude coordinates

long_lats <- data.frame( xlong=r1.ps@sam_data$`Longitude [decimal degrees]`,
                         ylat =r1.ps@sam_data$`Latitude [decimal degrees]`
)
row.names(long_lats) <- row.names(r1.ps@sam_data)

# calculate matrix of pairwise distances between all points
geo_dist.m <- geodist(long_lats, measure = "haversine")

geo_dist <- as.data.frame( geo_dist.m )
row.names(geo_dist) <- row.names(r1.ps@sam_data)
colnames(geo_dist) <- row.names(r1.ps@sam_data)

# check that colnames and row.names align
identical(row.names(geo_dist),row.names(dist.jacc)) # TRUE
identical(colnames(geo_dist),colnames(dist.jacc)) # TRUE

identical(colnames(dist.jacc), row.names(dist.jacc)) # TRUE


df_eco <- dist.jacc

df_ecogeocor <- data.frame(samp= rep( x = colnames(df_eco), each=length( colnames(df_eco) ) ),
                           compare_with = rep( x = colnames(df_eco), times = length(colnames(df_eco))),
                           
                           samp_type=NA,
                           compare_with_type=NA,
                           
                           dist_eco=NA,
                           dist_geo=NA
)

for (i in 1:dim(df_ecogeocor)[1]) {
  #i<-1
  this_samp <- df_ecogeocor$samp[i]
  this_compare <- df_ecogeocor$compare_with[i]
  sel1 <- which( row.names(r1.ps@sam_data) == this_samp)
  df_ecogeocor$samp_type[i] <- as.character( r1.ps@sam_data$group[sel1] )
  
  sel2 <- which( row.names(r1.ps@sam_data) == this_compare)
  df_ecogeocor$compare_with_type[i] <- as.character( r1.ps@sam_data$group[sel2] )
  
  row <- which(rownames(df_eco)==this_samp)
  col <- which(colnames(df_eco)==this_compare)
  df_ecogeocor$dist_eco[i] <- df_eco[row,col]
  
  row <- which(rownames(geo_dist)==this_samp)
  col <- which(colnames(geo_dist)==this_compare)
  df_ecogeocor$dist_geo[i] <- geo_dist[row,col]
  
  print(paste0("completed ",i))
}

dim(df_ecogeocor) #  625   6

# remove remnant self-comparisons
sel <- which(df_ecogeocor$samp==df_ecogeocor$compare_with) #
identical( which(df_ecogeocor$samp==df_ecogeocor$compare_with), which(df_ecogeocor$dist_eco==0) ) # TRUE
df_ecogeocor <- df_ecogeocor[-sel, ]
dim(df_ecogeocor) # 600   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)
summary(df_ecogeocor$dist_eco)

head(df_ecogeocor)
#      samp compare_with samp_type compare_with_type  dist_eco  dist_geo
# 2 X138358      X138360        23                23 0.6668722  698.9024
# 3 X138358      X138362        23               Ref 0.7011121  991.2395
# 4 X138358      X138364        23                 2 0.7854806 1212.3960
# 5 X138358      X138366        23               Ref 0.7874227 1291.3074
# 6 X138358      X138368        23                17 0.7072853 1679.4491
# 7 X138358      X138370        23                12 0.7463683 1201.3421


# consider only compare with type 'Ref' as that is focus 
# and potential for problematic study design and analysis

sel <- which(df_ecogeocor$compare_with_type=="Ref") # qty 144
df_ecogeocor <- df_ecogeocor[sel, ]
dim(df_ecogeocor) # 144  6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)

names(df_ecogeocor)
# [1] "samp"              "compare_with"      "samp_type"         "compare_with_type" "dist_eco"         
# [6] "dist_geo" 


unique(df_ecogeocor$samp_type)
# "23"  "Ref" "2"   "17"  "12"  "20"  "8"   "14"  "28" 

df_ecogeocor$group <- df_ecogeocor$samp_type
df_ecogeocor$group <- factor(df_ecogeocor$group, 
                             levels=c("2","8","12","14","17","20","23","28","Ref"),
                             labels=c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"),
                             ordered = TRUE)
length(levels(df_ecogeocor$group)) # 9

# #library(viridis); packageVersion("viridis") # ‘0.5.1’
# cols.group <- viridis(n=length(levels(df_ecogeocor$group)))
# names(cols.group) <- levels(df_ecogeocor$group)
# cols.group
# # 2 yr        8 yr       12 yr       14 yr       17 yr       20 yr       23 yr       28 yr         Ref 
# # "#440154FF" "#472D7BFF" "#3B528BFF" "#2C728EFF" "#21908CFF" "#27AD81FF" "#5DC863FF" "#AADC32FF" "#FDE725FF" 

cols.group <- cols.group.south32

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "12 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "20 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "23 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "28 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.66, label = "Original data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.65))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Jaccard)")
p
grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Jaccard-SOUTH32-A.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")


## for each rehab age determine if there is excess spatial auto-correlation??
## expect increasing eco-distance with increasing geo-distance
## if this occurs, does it occur at a rate greater 

( reveg_groups <- levels(df_ecogeocor$group) )
# "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr" "Ref"  


dim(df_ecogeocor) # 144   7
dim(df_out)       # 144   10

head(df_ecogeocor)
#       samp compare_with samp_type compare_with_type  dist_eco  dist_geo group
# 3  X138358      X138362        23               Ref 0.7011121  991.2395 23 yr
# 5  X138358      X138366        23               Ref 0.7874227 1291.3074 23 yr
# 9  X138358      X138374        23               Ref 0.7614333 2447.0999 23 yr
# 10 X138358      X138376        23               Ref 0.8451676 2692.4727 23 yr
# 11 X138358      X138378        23               Ref 0.8135325 4442.1090 23 yr
# 24 X138358      X138404        23               Ref 0.7958911 5697.4216 23 yr

head(df_out)
#      samp compare_with year_rehab similarity group dist_measure       tax_level   study corrected            facet_x
# 1 X138358      X138362       1996   29.88879    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 2 X138358      X138366       1996   21.25773    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 3 X138358      X138374       1996   23.85667    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 4 X138358      X138376       1996   15.48324    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 5 X138358      X138378       1996   18.64675    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 6 X138358      X138404       1996   20.41089    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level

identical(df_ecogeocor$samp,df_out$samp) # TRUE
identical(df_ecogeocor$compare_with,df_out$compare_with) # TRUE


df_ecogeocor$similarity <- df_out$similarity
df_ecogeocor$tax_level <- df_out$tax_level
df_ecogeocor$study <- df_out$study
df_ecogeocor$dist_measure <- df_out$dist_measure
df_ecogeocor$facet_x <- df_out$facet_x
df_ecogeocor$similarity_corr <- NA
df_ecogeocor$dist_eco_corr <- NA


preds.temp <- data.frame(dist_geo = sort( unique(df_ecogeocor$dist_geo)))
diff_models <- list()


for (i in 1:(length(reveg_groups)-1)) {
  #i<-1
  this_group <- reveg_groups[i]
  
  x <- df_ecogeocor %>% filter(group == reveg_groups[i])
  x_ref <- df_ecogeocor %>% filter(group == "Ref")
  #dim(x) # 54 14
  #dim(x_ref) # 306  14
  # mean centre eco-distances
  x$dist_eco_mc <- x$dist_eco - mean(x$dist_eco)
  #plot(x$dist_geo,x$dist_eco_mc)
  x_ref$dist_eco_mc <- x_ref$dist_eco - mean(x_ref$dist_eco)
  #plot(x_ref$dist_geo,x_ref$dist_eco_mc)
  
  #model each spatial trend
  m <- lm(data=x, formula = dist_eco_mc ~ poly(dist_geo, 2))
  m_ref <- lm(data=x_ref, formula = dist_eco_mc ~ poly(dist_geo, 2))
  preds <- preds.temp
  # # limit predictions to max dist_geo
  # sel <- which(preds$dist_geo <= max(x$dist_geo))
  # limit predictions to extent of dist_geo
  sel <- which(preds$dist_geo >= min(x$dist_geo) & preds$dist_geo <= max(x$dist_geo))
  
  preds <- as.data.frame( preds[sel, ] )
  names(preds)[1] <- "dist_geo" # need to rename field; it is lost in previous step
  preds$m <- predict(m, newdata = preds)
  preds$m_ref <- predict(m_ref, newdata = preds)
  #plot(preds$dist_geo,preds$m,col="red")
  #points(preds$dist_geo,preds$m_ref,col="green")
  preds$diff <- preds$m - preds$m_ref
  preds$group <- this_group
  #points(preds$dist_geo,preds$diff,col="orange")
  plot(preds$dist_geo,preds$diff,col="orange", main = paste0("Diff Reveg age: ",this_group))
  # apply correction to similarity
  sel.df <- which(df_ecogeocor$group == this_group)
  for (j in 1:length(sel.df)) {
    #j<-1
    sel.pred <- which(preds$dist_geo == df_ecogeocor$dist_geo[ sel.df[j] ])
    this_diff_dist_eco <- preds$diff[sel.pred]
    # similarity = 100*(1 - distance)
    # corrected similarity = 100*[1 -  (distance - distance-correction)]
    df_ecogeocor$similarity[ sel.df[j] ] # 13.10838
    df_ecogeocor$similarity_corr[ sel.df[j] ] <- 100*(1 - (df_ecogeocor$dist_eco[ sel.df[j] ] - this_diff_dist_eco))
    df_ecogeocor$dist_eco_corr[ sel.df[j] ] <- 1 - (1/100)*df_ecogeocor$similarity_corr[ sel.df[j] ]
  }
  plot(df_ecogeocor$similarity[ sel.df ], df_ecogeocor$similarity_corr[ sel.df ], main = paste0("Group: ",this_group))
  diff_models[[this_group]] <- preds
  print(paste0("completed ",i))
}

## but note there will be no 'corrected' values for the "Ref" samples - so just use previous values 
sel <- which(df_ecogeocor$group == "Ref")
df_ecogeocor$similarity_corr[sel] # all NAs
df_ecogeocor$dist_eco_corr[sel] # all NAs
df_ecogeocor$similarity_corr[sel] <- df_ecogeocor$similarity[sel]
df_ecogeocor$dist_eco_corr[sel]   <- df_ecogeocor$dist_eco[sel]


## plot modelled data for excess spatial auto-correlation 

names(diff_models) # "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr"

pdata <- diff_models[[1]]
for (i in 2:length(diff_models)) {
  pdata <- rbind(pdata,diff_models[[i]])
}
dim(pdata) # 937    5
names(pdata) # "dist_geo" "m"        "m_ref"    "diff"     "group"
unique(pdata$group) # "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr"
pdata$group <- factor(pdata$group, levels=c("2 yr",  "8 yr",  "12 yr", "14 yr", "17 yr", "20 yr", "23 yr", "28 yr"),
                      ordered=TRUE)

p <- ggplot(data = pdata, aes(x=dist_geo,y=diff,color=factor(group)) ) +  #,color=factor(group)
  #geom_point( alpha = 0.6) +
  #geom_point() +# pdata %>% filter(group == "3 yr"),aes(x=dist_geo,y=diff)) +
  geom_line(size=1) +
  geom_hline(yintercept = 0, size=0.5, linetype="dashed", colour="grey") +
  theme_bw() +
  scale_color_manual(values=cols.group[-length(cols.group)]) +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.7)),
    legend.text = element_text(size=rel(0.65))
  ) +
  #ylim(0,0.2) +
  guides(colour = guide_legend(title = "Rehab\nage",override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "\U0394 ecol. dist. (Jaccard)\nRehab vs. Ref samples")
p
grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","DIFF-in-Ecodist-vs-Geodist-Jaccard-SOUTH32-B.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")



## plot corrected distributions ...

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco_corr, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "12 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "20 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "23 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "28 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.66, label = "Corrected data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.65))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Jaccard)")
p
grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Jaccard-SOUTH32-SACORRECTED-C.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")



names(df_results)
names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"         
# [8] "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"

names(df_ecogeocor)

## make names of df_ecogeocor compatible with df_out
head(df_out)
# samp compare_with year_rehab similarity group dist_measure       tax_level   study corrected            facet_x
# 1 X138358      X138362       1996   29.88879    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 2 X138358      X138366       1996   21.25773    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 3 X138358      X138374       1996   23.85667    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 4 X138358      X138376       1996   15.48324    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 5 X138358      X138378       1996   18.64675    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level
# 6 X138358      X138404       1996   20.41089    23      Jaccard ASV (with Tree) South32        no Jaccard\nASV-level

head(df_ecogeocor)
# samp compare_with samp_type compare_with_type  dist_eco  dist_geo group similarity       tax_level   study dist_measure            facet_x similarity_corr
# 3  X138358      X138362        23               Ref 0.7011121  991.2395 23 yr   29.88879 ASV (with Tree) South32      Jaccard Jaccard\nASV-level        30.38463
# 5  X138358      X138366        23               Ref 0.7874227 1291.3074 23 yr   21.25773 ASV (with Tree) South32      Jaccard Jaccard\nASV-level        22.15341
# 9  X138358      X138374        23               Ref 0.7614333 2447.0999 23 yr   23.85667 ASV (with Tree) South32      Jaccard Jaccard\nASV-level        25.81715
# 10 X138358      X138376        23               Ref 0.8451676 2692.4727 23 yr   15.48324 ASV (with Tree) South32      Jaccard Jaccard\nASV-level        17.57267
# 11 X138358      X138378        23               Ref 0.8135325 4442.1090 23 yr   18.64675 ASV (with Tree) South32      Jaccard Jaccard\nASV-level        20.66966
# 24 X138358      X138404        23               Ref 0.7958911 5697.4216 23 yr   20.41089 ASV (with Tree) South32      Jaccard Jaccard\nASV-level        21.32058
# dist_eco_corr
# 3      0.6961537
# 5      0.7784659
# 9      0.7418285
# 10     0.8242733
# 11     0.7933034
# 24     0.7867942

# confirm these refer to the same sites and comparison sites
identical( paste0(df_out$samp,"_",df_out$compare_with) , paste0(df_ecogeocor$samp,"_",df_ecogeocor$compare_with) ) # TRUE
dim(df_out) # 144  10

df_new <- df_out
df_new$similarity <- df_ecogeocor$similarity_corr
df_new$corrected <- "yes"

head(df_new)
# samp compare_with year_rehab similarity group dist_measure       tax_level   study corrected            facet_x
# 1 X138358      X138362       1996   30.38463    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 2 X138358      X138366       1996   22.15341    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 3 X138358      X138374       1996   25.81715    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 4 X138358      X138376       1996   17.57267    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 5 X138358      X138378       1996   20.66966    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 6 X138358      X138404       1996   21.32058    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level

unique(df_new$group)
# [1] 23  Ref 2   17  12  20  8   14  28 
# Levels: 2 < 8 < 12 < 14 < 17 < 20 < 23 < 28 < Ref



table(df_new$group)
# 2   8  12  14  17  20  23  28 Ref 
# 18  18   6  12  12  12  24  12  30 


# only make this once
#df_results <- list()
length(df_results) #
names(df_results)

df_results[["South32-Jaccard-ASV-withTree-SAcorrected"]] <- df_new
#df_out <- df_results[[1]]


p <- ggplot(data=df_new, aes(x=group, similarity)) + ggtitle("Jaccard Similarity\nSA-corrected") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p
ggsave(plot=p, filename = paste0(workdir,"/plots/","Jaccard-Similarity--South32-Site-with-SA-correction.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw", type = "cairo")




## Kruskal-Wallis multiple comparison test 
## with post-hoc Dunn test

str(df_new)
# 'data.frame':	144 obs. of  10 variables:


df_new$group


# Kruskal-Wallis test
 
kt <- kruskal.test( similarity ~ group, df_new) # Kruskal Wallis test
kt
# Kruskal-Wallis rank sum test
# data:  distance by group
# Kruskal-Wallis chi-squared = 68.528, df = 8, p-value = 9.645e-12


# library(FSA); packageVersion("FSA") # '0.8.30'
# library(rcompanion); packageVersion("rcompanion") # '2.3.25'

## Dunn Test uses factor vector or non-numeric vector that can be coerced to a factor vector

unique( df_new$group ) 
# [1] 23  Ref 2   17  12  20  8   14  28 
# Levels: 2 < 8 < 12 < 14 < 17 < 20 < 23 < 28 < Ref


# post-hoc Dunn test below has a glitch that throws out zeros from grouping labels ... to perform work-around
temp <- df_new
temp$group <- as.character(temp$group)
temp$group <- factor(temp$group, levels = c("2","8","12","14","17","20","23","28","Ref"),
                     labels = c("2","8","12","14","17","twenty","23","28","Ref"),ordered = TRUE)


 
pt <- dunnTest( similarity ~ group, data = temp,
                method = "bonferroni" )
pt

pt$dtres
pt <- pt$res
pt

cldList(comparison = pt$Comparison,
        p.value    = pt$P.adj,
        threshold  = 0.05)

sig <- cldList(comparison = pt$Comparison,
               p.value    = pt$P.adj,
               threshold  = 0.05)

str(sig)
# 'data.frame':	9 obs. of  3 variables:

unique(sig$Group) # "12"     "14"     "17"     "2"      "23"     "28"     "8"      "Ref"    "twenty"

sig$Group <- factor( sig$Group, levels=c("2","8","12","14","17","twenty","23","28","Ref"),
                     labels = c("2","8","12","14","17","20","23","28","Ref"),
                     ordered=TRUE )


sig[ order(sig$Group), ]
# Group Letter MonoLetter
# 4     2      a        a  
# 7     8      a        a  
# 1    12    abc        abc
# 2    14     ab        ab 
# 3    17    abc        abc
# 9    20     bc         bc
# 5    23     bc         bc
# 6    28     bc         bc
# 8   Ref      c          c

sig <- sig[ order(sig$Group), ]

levels(sig$Group) # "2"   "8"   "12"  "14"  "17"  "20"  "23"  "28"  "Ref"

## store annotations for later facet_grid ggplot
names(df_new)
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"          "dist_measure"   "tax_level"     
# [10] "study"          "corrected"      "facet_x"  

anno <- data.frame(group=sig$Group,
                   sig_letter = sig$Letter,
                   similarity= c(25,24,28,26,34,36,32,33,36),
                   dist_measure = rep( "Jaccard", times = dim(sig)[1]),
                   tax_level  = rep( "ASV (with Tree)", times = dim(sig)[1]),
                   study  = rep( "South32", times = dim(sig)[1]),
                   corrected = rep( "yes" , times = dim(sig)[1]),
                   facet_x  = rep( "Jaccard\nASV-level", times = dim(sig)[1])
)

# only make this once
#df_anno <- list()
length(df_anno) # 0
names(df_anno)

df_anno[["South32-Jaccard-ASV-withTree-SAcorrected"]] <- anno


#-------------------------


### NOW EXCLUDE DISTANT SITES IN SOUTH32 CASE STUDY

#### Investigating spatial auto-correlation -- South32 -- Bray-Curtis -- ExcludeFarSouth
#-------------------------

names(df_results)

phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122
ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]



str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`year of rehab or ref`
unique(r1.ps@sam_data$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                               labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)



names(df_results)

df_out <- df_results[[ "South32-Bray-ASV-withTree" ]]


names(r1.ps@sam_data) # includes ...
# [1] "Sample_ID"                                             "NCBI_Submission"                                      
# [3] "NCBI_Sample_Accession"                                 "Organism"                                             
# [5] "Tax_ID"                                                "SampleName_depth"                                     
# [7] "NCBI_BioProject"                                       "Date_sampled [YYYY-MM-DD]"                            
# [9] "Time_sampled [hh:mm]"                                  "Latitude [decimal degrees]"                           
# [11] "Longitude [decimal degrees]"     

r1.ps@sam_data$`Latitude [decimal degrees]`
r1.ps@sam_data$`Longitude [decimal degrees]`

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.bray <- vegan::vegdist(x = otu_tab, method = "bray")
str(dist.bray)
dist.bray <- as(dist.bray, "matrix")

identical(colnames(dist.bray),rownames(dist.bray)) # TRUE


## determine spatial auto-correlation

# data.frame containing longitude and latitude coordinates

long_lats <- data.frame( xlong=r1.ps@sam_data$`Longitude [decimal degrees]`,
                         ylat =r1.ps@sam_data$`Latitude [decimal degrees]`
)
row.names(long_lats) <- row.names(r1.ps@sam_data)

# calculate matrix of pairwise distances between all points
geo_dist.m <- geodist(long_lats, measure = "haversine")

geo_dist <- as.data.frame( geo_dist.m )
row.names(geo_dist) <- row.names(r1.ps@sam_data)
colnames(geo_dist) <- row.names(r1.ps@sam_data)

# check that colnames and row.names align
identical(row.names(geo_dist),row.names(dist.bray)) # TRUE
identical(colnames(geo_dist),colnames(dist.bray)) # TRUE

identical(colnames(dist.bray), row.names(dist.bray)) # TRUE


df_eco <- dist.bray

df_ecogeocor <- data.frame(samp= rep( x = colnames(df_eco), each=length( colnames(df_eco) ) ),
                           compare_with = rep( x = colnames(df_eco), times = length(colnames(df_eco))),
                           
                           samp_type=NA,
                           compare_with_type=NA,
                           
                           dist_eco=NA,
                           dist_geo=NA
)

for (i in 1:dim(df_ecogeocor)[1]) {
  #i<-1
  this_samp <- df_ecogeocor$samp[i]
  this_compare <- df_ecogeocor$compare_with[i]
  sel1 <- which( row.names(r1.ps@sam_data) == this_samp)
  df_ecogeocor$samp_type[i] <- as.character( r1.ps@sam_data$group[sel1] )
  
  sel2 <- which( row.names(r1.ps@sam_data) == this_compare)
  df_ecogeocor$compare_with_type[i] <- as.character( r1.ps@sam_data$group[sel2] )
  
  row <- which(rownames(df_eco)==this_samp)
  col <- which(colnames(df_eco)==this_compare)
  df_ecogeocor$dist_eco[i] <- df_eco[row,col]
  
  row <- which(rownames(geo_dist)==this_samp)
  col <- which(colnames(geo_dist)==this_compare)
  df_ecogeocor$dist_geo[i] <- geo_dist[row,col]
  
  print(paste0("completed ",i))
}

dim(df_ecogeocor) #  625   6

# remove remnant self-comparisons
sel <- which(df_ecogeocor$samp==df_ecogeocor$compare_with) #
identical( which(df_ecogeocor$samp==df_ecogeocor$compare_with), which(df_ecogeocor$dist_eco==0) ) # TRUE
df_ecogeocor <- df_ecogeocor[-sel, ]
dim(df_ecogeocor) # 600   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)
summary(df_ecogeocor$dist_eco)

head(df_ecogeocor)
#      samp compare_with samp_type compare_with_type  dist_eco  dist_geo
# 2 X138358      X138360        23                23 0.5328332  698.9024
# 3 X138358      X138362        23               Ref 0.6198404  991.2395
# 4 X138358      X138364        23                 2 0.7639962 1212.3960
# 5 X138358      X138366        23               Ref 0.7285577 1291.3074
# 6 X138358      X138368        23                17 0.6351576 1679.4491
# 7 X138358      X138370        23                12 0.6750120 1201.3421


# consider only compare with type 'Ref' as that is focus 
# and potential for problematic study design and analysis

sel <- which(df_ecogeocor$compare_with_type=="Ref") # qty 144
df_ecogeocor <- df_ecogeocor[sel, ]
dim(df_ecogeocor) # 144  6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)


## what is distribution of geological distance?
hist(df_ecogeocor$dist_geo)
hist(r1.ps@sam_data$`Latitude [decimal degrees]`, breaks = 30)
# southern cluster is > 32.94
sel <- which(r1.ps@sam_data$`Latitude [decimal degrees]` > 32.94) # qty 3
row.names(r1.ps@sam_data)[sel] # "X138402" "X138404" "X138406"
r1.ps@sam_data$Sample_ID[sel] # "102.100.100/138402" "102.100.100/138404" "102.100.100/138406"
r1.ps@sam_data$group[sel] # 2   Ref 2  

# remove these from df_ecogeocor
names(df_ecogeocor)
#"samp"              "compare_with"      "samp_type"         "compare_with_type" "dist_eco"          "dist_geo"   

sel.rm <- which(df_ecogeocor$samp %in% c("X138402", "X138404", "X138406") | df_ecogeocor$compare_with %in% c("X138402", "X138404", "X138406")) # qty 39 rows
df_ecogeocor[sel.rm, ]
#        samp compare_with samp_type compare_with_type  dist_eco   dist_geo
# 24  X138358      X138404        23               Ref 0.7496951  5697.4216
# 49  X138360      X138404        23               Ref 0.7753224  5150.2787
# 74  X138362      X138404       Ref               Ref 0.7110602  4892.8740
# 99  X138364      X138404         2               Ref 0.8574517  5299.2474
# 124 X138366      X138404       Ref               Ref 0.7860574  5278.2006
# 149 X138368      X138404        17               Ref 0.7694838  7261.6294
# 174 X138370      X138404        12               Ref 0.8147149  6672.6095
# 199 X138372      X138404        23               Ref 0.7382950  6889.9139
# 224 X138374      X138404       Ref               Ref 0.6749751  7839.1925
# 249 X138376      X138404       Ref               Ref 0.7584346  8270.3423
# 274 X138378      X138404       Ref               Ref 0.7436902 10025.0451
# 299 X138380      X138404        20               Ref 0.7004545  9725.9159
# 324 X138382      X138404         8               Ref 0.7799786  9930.3255
# 349 X138384      X138404         8               Ref 0.7573260  9261.6138
# 374 X138386      X138404        14               Ref 0.7897343  8844.5744
# 399 X138388      X138404        20               Ref 0.6386867  7333.4042
# 424 X138390      X138404        23               Ref 0.7172869  7804.4292
# 449 X138392      X138404        17               Ref 0.6993644  6750.0638
# 474 X138394      X138404         8               Ref 0.8073981  6691.6399
# 499 X138396      X138404        14               Ref 0.7471638  6283.9716
# 524 X138398      X138404        28               Ref 0.6988470  5318.3271
# 549 X138400      X138404        28               Ref 0.6334947  5312.5879
# 553 X138402      X138362         2               Ref 0.7123166  4969.6985
# 555 X138402      X138366         2               Ref 0.7709804  5326.2428
# 559 X138402      X138374         2               Ref 0.7194856  7891.3114
# 560 X138402      X138376         2               Ref 0.7766343  8328.1321
# 561 X138402      X138378         2               Ref 0.7525221 10081.1935
# 574 X138402      X138404         2               Ref 0.7531688   107.7767
# 578 X138404      X138362       Ref               Ref 0.7110602  4892.8740
# 580 X138404      X138366       Ref               Ref 0.7860574  5278.2006
# 584 X138404      X138374       Ref               Ref 0.6749751  7839.1925
# 585 X138404      X138376       Ref               Ref 0.7584346  8270.3423
# 586 X138404      X138378       Ref               Ref 0.7436902 10025.0451
# 603 X138406      X138362         2               Ref 0.8239348  7166.2478
# 605 X138406      X138366         2               Ref 0.8808618  7440.3262
# 609 X138406      X138374         2               Ref 0.8430028 10017.5409
# 610 X138406      X138376         2               Ref 0.8899893 10480.4286
# 611 X138406      X138378         2               Ref 0.8743764 12223.8767
# 624 X138406      X138404         2               Ref 0.7982151  2274.9200


dim(df_ecogeocor) # 144   6
df_ecogeocor <- df_ecogeocor[-sel.rm, ]
dim(df_ecogeocor) # 105   6

names(df_ecogeocor)
# [1] "samp"              "compare_with"      "samp_type"         "compare_with_type" "dist_eco"         
# [6] "dist_geo" 


unique(df_ecogeocor$samp_type)
# "23"  "Ref" "2"   "17"  "12"  "20"  "8"   "14"  "28" 

df_ecogeocor$group <- df_ecogeocor$samp_type
df_ecogeocor$group <- factor(df_ecogeocor$group, 
                             levels=c("2","8","12","14","17","20","23","28","Ref"),
                             labels=c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"),
                             ordered = TRUE)
length(levels(df_ecogeocor$group)) # 9

table(df_ecogeocor$group)
# 2 yr  8 yr 12 yr 14 yr 17 yr 20 yr 23 yr 28 yr   Ref 
#    5    15     5    10    10    10    20    10    20 

# #library(viridis); packageVersion("viridis") # ‘0.5.1’
# cols.group <- viridis(n=length(levels(df_ecogeocor$group)))
# names(cols.group) <- levels(df_ecogeocor$group)
# cols.group
# # 2 yr        8 yr       12 yr       14 yr       17 yr       20 yr       23 yr       28 yr         Ref 
# # "#440154FF" "#472D7BFF" "#3B528BFF" "#2C728EFF" "#21908CFF" "#27AD81FF" "#5DC863FF" "#AADC32FF" "#FDE725FF" 

cols.group <- cols.group.south32

dim(df_ecogeocor) # 105   7

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,y=dist_eco, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "12 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "20 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "23 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "28 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.45, label = "Filtered data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.65))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Bray-Curtis)")
p
grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Bray-Curtis-SOUTH32-ExclFarSouthSites-A.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")


## for each rehab age determine if there is excess spatial auto-correlation??
## expect increasing eco-distance with increasing geo-distance
## if this occurs, does it occur at a rate greater 

( reveg_groups <- levels(df_ecogeocor$group) )
# "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr" "Ref"  


dim(df_ecogeocor) # 105   7
dim(df_out)       # 144   10

# also need to filter out far southern sites in df_out
sel.rm <- which(df_out$samp %in% c("X138402", "X138404", "X138406") | df_out$compare_with %in% c("X138402", "X138404", "X138406")) # qty 39 rows
df_out[sel.rm, ]
df_out <- df_out[-sel.rm, ]

head(df_ecogeocor)
#       samp compare_with samp_type compare_with_type  dist_eco  dist_geo group
# 3  X138358      X138362        23               Ref 0.6198404  991.2395 23 yr
# 5  X138358      X138366        23               Ref 0.7285577 1291.3074 23 yr
# 9  X138358      X138374        23               Ref 0.6956136 2447.0999 23 yr
# 10 X138358      X138376        23               Ref 0.7995085 2692.4727 23 yr
# 11 X138358      X138378        23               Ref 0.7671557 4442.1090 23 yr
# 28 X138360      X138362        23               Ref 0.6778759  296.7652 23 yr

head(df_out)
#      samp compare_with year_rehab similarity group dist_measure       tax_level   study corrected                facet_x
# 1 X138358      X138362       1996   38.01596    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 2 X138358      X138366       1996   27.14423    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 3 X138358      X138374       1996   30.43864    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 4 X138358      X138376       1996   20.04915    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 5 X138358      X138378       1996   23.28443    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level
# 7 X138360      X138362       1996   32.21241    23  Bray-Curtis ASV (with Tree) South32        no Bray-Curtis\nASV-level

identical(df_ecogeocor$samp,df_out$samp) # TRUE
identical(df_ecogeocor$compare_with,df_out$compare_with) # TRUE


df_ecogeocor$similarity <- df_out$similarity
df_ecogeocor$tax_level <- df_out$tax_level
df_ecogeocor$study <- df_out$study
df_ecogeocor$dist_measure <- df_out$dist_measure
df_ecogeocor$facet_x <- df_out$facet_x
df_ecogeocor$similarity_corr <- NA
df_ecogeocor$dist_eco_corr <- NA


preds.temp <- data.frame(dist_geo = sort( unique(df_ecogeocor$dist_geo)))
diff_models <- list()


for (i in 1:(length(reveg_groups)-1)) {
  #i<-3
  this_group <- reveg_groups[i]
  
  x <- df_ecogeocor %>% filter(group == reveg_groups[i])
  x_ref <- df_ecogeocor %>% filter(group == "Ref")
  #dim(x) # 54 14
  #dim(x_ref) # 306  14
  # mean centre eco-distances
  x$dist_eco_mc <- x$dist_eco - mean(x$dist_eco)
  #plot(x$dist_geo,x$dist_eco_mc)
  x_ref$dist_eco_mc <- x_ref$dist_eco - mean(x_ref$dist_eco)
  #plot(x_ref$dist_geo,x_ref$dist_eco_mc)
  
  #model each spatial trend
  m <- lm(data=x, formula = dist_eco_mc ~ poly(dist_geo, 2))
  m_ref <- lm(data=x_ref, formula = dist_eco_mc ~ poly(dist_geo, 2))
  preds <- preds.temp
  # limit predictions to extent of dist_geo
  sel <- which(preds$dist_geo >= min(x$dist_geo) & preds$dist_geo <= max(x$dist_geo))
  preds <- as.data.frame( preds[sel, ] )
  names(preds)[1] <- "dist_geo" # need to rename field; it is lost in previous step
  preds$m <- predict(m, newdata = preds)
  preds$m_ref <- predict(m_ref, newdata = preds)
  #plot(preds$dist_geo,preds$m,col="red")
  #points(preds$dist_geo,preds$m_ref,col="green")
  preds$diff <- preds$m - preds$m_ref
  preds$group <- this_group
  #points(preds$dist_geo,preds$diff,col="orange")
  plot(preds$dist_geo,preds$diff,col="orange", main = paste0("Diff Reveg age: ",this_group))
  # apply correction to similarity
  sel.df <- which(df_ecogeocor$group == this_group)
  for (j in 1:length(sel.df)) {
    #j<-1
    sel.pred <- which(preds$dist_geo == df_ecogeocor$dist_geo[ sel.df[j] ])
    this_diff_dist_eco <- preds$diff[sel.pred]
    # similarity = 100*(1 - distance)
    # corrected similarity = 100*[1 -  (distance - distance-correction)]
    df_ecogeocor$similarity[ sel.df[j] ] # 13.10838
    df_ecogeocor$similarity_corr[ sel.df[j] ] <- 100*(1 - (df_ecogeocor$dist_eco[ sel.df[j] ] - this_diff_dist_eco))
    df_ecogeocor$dist_eco_corr[ sel.df[j] ] <- 1 - (1/100)*df_ecogeocor$similarity_corr[ sel.df[j] ]
  }
  plot(df_ecogeocor$similarity[ sel.df ], df_ecogeocor$similarity_corr[ sel.df ], main = paste0("Group: ",this_group))
  diff_models[[this_group]] <- preds
  print(paste0("completed ",i))
}

## but note there will be no 'corrected' values for the "Ref" samples - so just use previous values 
sel <- which(df_ecogeocor$group == "Ref")
df_ecogeocor$similarity_corr[sel] # all NAs
df_ecogeocor$dist_eco_corr[sel] # all NAs
df_ecogeocor$similarity_corr[sel] <- df_ecogeocor$similarity[sel]
df_ecogeocor$dist_eco_corr[sel]   <- df_ecogeocor$dist_eco[sel]


## plot modelled data for excess spatial auto-correlation 

names(diff_models) # "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr"

pdata <- diff_models[[1]]
for (i in 2:length(diff_models)) {
  pdata <- rbind(pdata,diff_models[[i]])
}
dim(pdata) # 636    5
names(pdata) # "dist_geo" "m"        "m_ref"    "diff"     "group"
unique(pdata$group) # "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr"
pdata$group <- factor(pdata$group, levels=c("2 yr",  "8 yr",  "12 yr", "14 yr", "17 yr", "20 yr", "23 yr", "28 yr"),
                      ordered=TRUE)

p <- ggplot(data = pdata, aes(x=dist_geo,y=diff,color=factor(group)) ) +  #,color=factor(group)
  #geom_point( alpha = 0.6) +
  #geom_point() +# pdata %>% filter(group == "3 yr"),aes(x=dist_geo,y=diff)) +
  geom_line(size=1) +
  geom_hline(yintercept = 0, size=0.5, linetype="dashed", colour="grey") +
  theme_bw() +
  scale_color_manual(values=cols.group[-length(cols.group)]) +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.7)),
    legend.text = element_text(size=rel(0.65))
  ) +
  #ylim(0,0.2) +
  guides(colour = guide_legend(title = "Rehab\nage",override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "\U0394 ecol. dist. (Bray-Curtis)\nRehab vs. Ref samples")
p
grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","DIFF-in-Ecodist-vs-Geodist-Bray-Curtis-SOUTH32-ExclFarSouthSites-B.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")



## plot corrected distributions ...

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco_corr, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "12 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "20 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "23 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "28 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.45, label = "Filtered &\ncorrected data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.65))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Bray-Curtis)")
p
grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Bray-Curtis-SOUTH32-SACORRECTED-ExclFarSouthSites-C.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")



names(df_results)
names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"         
# [8] "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"

names(df_ecogeocor)

## make names of df_ecogeocor compatible with df_out
head(df_out)

head(df_ecogeocor)

# confirm these refer to the same sites and comparison sites
identical( paste0(df_out$samp,"_",df_out$compare_with) , paste0(df_ecogeocor$samp,"_",df_ecogeocor$compare_with) ) # TRUE
dim(df_out) # 105  10

df_new <- df_out
df_new$similarity <- df_ecogeocor$similarity_corr
df_new$corrected <- "yes"

head(df_new)
# samp compare_with year_rehab similarity group dist_measure       tax_level   study corrected                facet_x
# 1 X138358      X138362       1996   36.20741    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 2 X138358      X138366       1996   25.92499    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 3 X138358      X138374       1996   31.03079    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 4 X138358      X138376       1996   20.93217    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 5 X138358      X138378       1996   25.29034    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level
# 7 X138360      X138362       1996   28.85174    23  Bray-Curtis ASV (with Tree) South32       yes Bray-Curtis\nASV-level

unique(df_new$group)
# [1] 23  Ref 2   17  12  20  8   14  28 
# Levels: 2 < 8 < 12 < 14 < 17 < 20 < 23 < 28 < Ref



table(df_new$group)
# 2   8  12  14  17  20  23  28 Ref 
# 5  15   5  10  10  10  20  10  20 


# only make this once
#df_results <- list()
length(df_results) #
names(df_results)

df_results[["South32-Bray-ASV-withTree-SAcorrected-ExclFarSouthSites"]] <- df_new
#df_out <- df_results[[1]]


p <- ggplot(data=df_new, aes(x=group, similarity)) + ggtitle("Bray Curtis Similarity\nSA-corrected - Excluding far south sites") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p
ggsave(plot=p, filename = paste0(workdir,"/plots/","Bray-Curtis-Similarity--South32-Site-with-SA-correction-ExclFarSouthSites.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw", type = "cairo")




#-------------------------

#### Investigating spatial auto-correlation -- South32 -- Jaccard -- ExcludeFarSouth
#-------------------------

phy_in <- phy.south32.withTree
min(taxa_sums(phy_in)) # 2
sum(sample_sums(phy_in)) # 2049625

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122
ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]



str(r1.ps@sam_data)
names(sample_data(r1.ps))

r1.ps@sam_data$group <- r1.ps@sam_data$`year of rehab or ref`
unique(r1.ps@sam_data$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                               labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)



names(df_results)

df_out <- df_results[[ "South32-Jaccard-ASV-withTree" ]]


names(r1.ps@sam_data) # includes ...
# [1] "Sample_ID"                                             "NCBI_Submission"                                      
# [3] "NCBI_Sample_Accession"                                 "Organism"                                             
# [5] "Tax_ID"                                                "SampleName_depth"                                     
# [7] "NCBI_BioProject"                                       "Date_sampled [YYYY-MM-DD]"                            
# [9] "Time_sampled [hh:mm]"                                  "Latitude [decimal degrees]"                           
# [11] "Longitude [decimal degrees]"     

r1.ps@sam_data$`Latitude [decimal degrees]`
r1.ps@sam_data$`Longitude [decimal degrees]`

head(r1.ps@otu_table)
otu_tab <- t( as(r1.ps@otu_table, "matrix"))

dist.jacc <- vegan::vegdist(x = otu_tab, method = "jaccard", binary = TRUE)
str(dist.jacc)
dist.jacc <- as(dist.jacc, "matrix")

identical(colnames(dist.jacc),rownames(dist.jacc)) # TRUE


## determine spatial auto-correlation

# data.frame containing longitude and latitude coordinates

long_lats <- data.frame( xlong=r1.ps@sam_data$`Longitude [decimal degrees]`,
                         ylat =r1.ps@sam_data$`Latitude [decimal degrees]`
)
row.names(long_lats) <- row.names(r1.ps@sam_data)

# calculate matrix of pairwise distances between all points
geo_dist.m <- geodist(long_lats, measure = "haversine")

geo_dist <- as.data.frame( geo_dist.m )
row.names(geo_dist) <- row.names(r1.ps@sam_data)
colnames(geo_dist) <- row.names(r1.ps@sam_data)

# check that colnames and row.names align
identical(row.names(geo_dist),row.names(dist.jacc)) # TRUE
identical(colnames(geo_dist),colnames(dist.jacc)) # TRUE

identical(colnames(dist.jacc), row.names(dist.jacc)) # TRUE


df_eco <- dist.jacc

df_ecogeocor <- data.frame(samp= rep( x = colnames(df_eco), each=length( colnames(df_eco) ) ),
                           compare_with = rep( x = colnames(df_eco), times = length(colnames(df_eco))),
                           
                           samp_type=NA,
                           compare_with_type=NA,
                           
                           dist_eco=NA,
                           dist_geo=NA
)

for (i in 1:dim(df_ecogeocor)[1]) {
  #i<-1
  this_samp <- df_ecogeocor$samp[i]
  this_compare <- df_ecogeocor$compare_with[i]
  sel1 <- which( row.names(r1.ps@sam_data) == this_samp)
  df_ecogeocor$samp_type[i] <- as.character( r1.ps@sam_data$group[sel1] )
  
  sel2 <- which( row.names(r1.ps@sam_data) == this_compare)
  df_ecogeocor$compare_with_type[i] <- as.character( r1.ps@sam_data$group[sel2] )
  
  row <- which(rownames(df_eco)==this_samp)
  col <- which(colnames(df_eco)==this_compare)
  df_ecogeocor$dist_eco[i] <- df_eco[row,col]
  
  row <- which(rownames(geo_dist)==this_samp)
  col <- which(colnames(geo_dist)==this_compare)
  df_ecogeocor$dist_geo[i] <- geo_dist[row,col]
  
  print(paste0("completed ",i))
}

dim(df_ecogeocor) #  625   6

# remove remnant self-comparisons
sel <- which(df_ecogeocor$samp==df_ecogeocor$compare_with) #
identical( which(df_ecogeocor$samp==df_ecogeocor$compare_with), which(df_ecogeocor$dist_eco==0) ) # TRUE
df_ecogeocor <- df_ecogeocor[-sel, ]
dim(df_ecogeocor) # 600   6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)
summary(df_ecogeocor$dist_eco)

head(df_ecogeocor)
#      samp compare_with samp_type compare_with_type  dist_eco  dist_geo
# 2 X138358      X138360        23                23 0.6668722  698.9024
# 3 X138358      X138362        23               Ref 0.7011121  991.2395
# 4 X138358      X138364        23                 2 0.7854806 1212.3960
# 5 X138358      X138366        23               Ref 0.7874227 1291.3074
# 6 X138358      X138368        23                17 0.7072853 1679.4491
# 7 X138358      X138370        23                12 0.7463683 1201.3421


# consider only compare with type 'Ref' as that is focus 
# and potential for problematic study design and analysis

sel <- which(df_ecogeocor$compare_with_type=="Ref") # qty 144
df_ecogeocor <- df_ecogeocor[sel, ]
dim(df_ecogeocor) # 144  6

plot(x = df_ecogeocor$dist_geo, y = df_ecogeocor$dist_eco)


## what is distribution of geological distance?
hist(df_ecogeocor$dist_geo)
hist(r1.ps@sam_data$`Latitude [decimal degrees]`, breaks = 30)
# southern cluster is > 32.94
sel <- which(r1.ps@sam_data$`Latitude [decimal degrees]` > 32.94) # qty 3
row.names(r1.ps@sam_data)[sel] # "X138402" "X138404" "X138406"
r1.ps@sam_data$Sample_ID[sel] # "102.100.100/138402" "102.100.100/138404" "102.100.100/138406"
r1.ps@sam_data$group[sel] # 2   Ref 2  

# remove these from df_ecogeocor
names(df_ecogeocor)
# [1] "samp"              "compare_with"      "samp_type"         "compare_with_type" "dist_eco"         
# [6] "dist_geo" 

sel.rm <- which(df_ecogeocor$samp %in% c("X138402", "X138404", "X138406") | df_ecogeocor$compare_with %in% c("X138402", "X138404", "X138406")) # qty 39 rows
df_ecogeocor[sel.rm, ]

dim(df_ecogeocor) # 144   6
df_ecogeocor <- df_ecogeocor[-sel.rm, ]
dim(df_ecogeocor) # 105   6


unique(df_ecogeocor$samp_type)
# "23"  "Ref" "2"   "17"  "12"  "20"  "8"   "14"  "28" 

df_ecogeocor$group <- df_ecogeocor$samp_type
df_ecogeocor$group <- factor(df_ecogeocor$group, 
                             levels=c("2","8","12","14","17","20","23","28","Ref"),
                             labels=c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"),
                             ordered = TRUE)
length(levels(df_ecogeocor$group)) # 9

table(df_ecogeocor$group)
# 2 yr  8 yr 12 yr 14 yr 17 yr 20 yr 23 yr 28 yr   Ref 
#    5    15     5    10    10    10    20    10    20


# #library(viridis); packageVersion("viridis") # ‘0.5.1’
# cols.group <- viridis(n=length(levels(df_ecogeocor$group)))
# names(cols.group) <- levels(df_ecogeocor$group)
# cols.group
# # 2 yr        8 yr       12 yr       14 yr       17 yr       20 yr       23 yr       28 yr         Ref 
# # "#440154FF" "#472D7BFF" "#3B528BFF" "#2C728EFF" "#21908CFF" "#27AD81FF" "#5DC863FF" "#AADC32FF" "#FDE725FF" 

cols.group <- cols.group.south32

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "12 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "20 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "23 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "28 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.6, label = "Filtered data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.65))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Jaccard)")
p
grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Jaccard-SOUTH32-ExclFarSouthSites-A.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")


## for each rehab age determine if there is excess spatial auto-correlation??
## expect increasing eco-distance with increasing geo-distance
## if this occurs, does it occur at a rate greater 

( reveg_groups <- levels(df_ecogeocor$group) )
# "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr" "Ref"  


dim(df_ecogeocor) # 105   7
dim(df_out)       # 144   10

# also need to filter out far southern sites in df_out
sel.rm <- which(df_out$samp %in% c("X138402", "X138404", "X138406") | df_out$compare_with %in% c("X138402", "X138404", "X138406")) # qty 39 rows
df_out[sel.rm, ]
df_out <- df_out[-sel.rm, ]
dim(df_out)       # 105   10

head(df_ecogeocor)

head(df_out)


identical(df_ecogeocor$samp,df_out$samp) # TRUE
identical(df_ecogeocor$compare_with,df_out$compare_with) # TRUE


df_ecogeocor$similarity <- df_out$similarity
df_ecogeocor$tax_level <- df_out$tax_level
df_ecogeocor$study <- df_out$study
df_ecogeocor$dist_measure <- df_out$dist_measure
df_ecogeocor$facet_x <- df_out$facet_x
df_ecogeocor$similarity_corr <- NA
df_ecogeocor$dist_eco_corr <- NA


preds.temp <- data.frame(dist_geo = sort( unique(df_ecogeocor$dist_geo)))
diff_models <- list()


for (i in 1:(length(reveg_groups)-1)) {
  #i<-1
  this_group <- reveg_groups[i]
  
  x <- df_ecogeocor %>% filter(group == reveg_groups[i])
  x_ref <- df_ecogeocor %>% filter(group == "Ref")
  #dim(x) # 54 14
  #dim(x_ref) # 306  14
  # mean centre eco-distances
  x$dist_eco_mc <- x$dist_eco - mean(x$dist_eco)
  #plot(x$dist_geo,x$dist_eco_mc)
  x_ref$dist_eco_mc <- x_ref$dist_eco - mean(x_ref$dist_eco)
  #plot(x_ref$dist_geo,x_ref$dist_eco_mc)
  
  #model each spatial trend
  m <- lm(data=x, formula = dist_eco_mc ~ poly(dist_geo, 2))
  m_ref <- lm(data=x_ref, formula = dist_eco_mc ~ poly(dist_geo, 2))
  preds <- preds.temp
  ## limit predictions to max dist_geo
  #sel <- which(preds$dist_geo <= max(x$dist_geo))
  # limit predictions to extent of dist_geo
  sel <- which(preds$dist_geo >= min(x$dist_geo) & preds$dist_geo <= max(x$dist_geo))
  preds <- as.data.frame( preds[sel, ] )
  names(preds)[1] <- "dist_geo" # need to rename field; it is lost in previous step
  preds$m <- predict(m, newdata = preds)
  preds$m_ref <- predict(m_ref, newdata = preds)
  #plot(preds$dist_geo,preds$m,col="red")
  #points(preds$dist_geo,preds$m_ref,col="green")
  preds$diff <- preds$m - preds$m_ref
  preds$group <- this_group
  #points(preds$dist_geo,preds$diff,col="orange")
  plot(preds$dist_geo,preds$diff,col="orange", main = paste0("Diff Reveg age: ",this_group))
  # apply correction to similarity
  sel.df <- which(df_ecogeocor$group == this_group)
  for (j in 1:length(sel.df)) {
    #j<-1
    sel.pred <- which(preds$dist_geo == df_ecogeocor$dist_geo[ sel.df[j] ])
    this_diff_dist_eco <- preds$diff[sel.pred]
    # similarity = 100*(1 - distance)
    # corrected similarity = 100*[1 -  (distance - distance-correction)]
    df_ecogeocor$similarity[ sel.df[j] ] # 13.10838
    df_ecogeocor$similarity_corr[ sel.df[j] ] <- 100*(1 - (df_ecogeocor$dist_eco[ sel.df[j] ] - this_diff_dist_eco))
    df_ecogeocor$dist_eco_corr[ sel.df[j] ] <- 1 - (1/100)*df_ecogeocor$similarity_corr[ sel.df[j] ]
  }
  plot(df_ecogeocor$similarity[ sel.df ], df_ecogeocor$similarity_corr[ sel.df ], main = paste0("Group: ",this_group))
  diff_models[[this_group]] <- preds
  print(paste0("completed ",i))
}

## but note there will be no 'corrected' values for the "Ref" samples - so just use previous values 
sel <- which(df_ecogeocor$group == "Ref")
df_ecogeocor$similarity_corr[sel] # all NAs
df_ecogeocor$dist_eco_corr[sel] # all NAs
df_ecogeocor$similarity_corr[sel] <- df_ecogeocor$similarity[sel]
df_ecogeocor$dist_eco_corr[sel]   <- df_ecogeocor$dist_eco[sel]


## plot modelled data for excess spatial auto-correlation 

names(diff_models) # "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr"

pdata <- diff_models[[1]]
for (i in 2:length(diff_models)) {
  pdata <- rbind(pdata,diff_models[[i]])
}
dim(pdata) # 636    5
names(pdata) # "dist_geo" "m"        "m_ref"    "diff"     "group"
unique(pdata$group) # "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr"
pdata$group <- factor(pdata$group, levels=c("2 yr",  "8 yr",  "12 yr", "14 yr", "17 yr", "20 yr", "23 yr", "28 yr"),
                      ordered=TRUE)

p <- ggplot(data = pdata, aes(x=dist_geo,y=diff,color=factor(group)) ) +  #,color=factor(group)
  #geom_point( alpha = 0.6) +
  #geom_point() +# pdata %>% filter(group == "3 yr"),aes(x=dist_geo,y=diff)) +
  geom_line(size=1) +
  geom_hline(yintercept = 0, size=0.5, linetype="dashed", colour="grey") +
  theme_bw() +
  scale_color_manual(values=cols.group[-length(cols.group)]) +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.7)),
    legend.text = element_text(size=rel(0.65))
  ) +
  #ylim(0,0.2) +
  guides(colour = guide_legend(title = "Rehab\nage",override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "\U0394 ecol. dist. (Jaccard)\nRehab vs. Ref samples")
p
grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","DIFF-in-Ecodist-vs-Geodist-Jaccard-SOUTH32-ExclFarSouthSites-B.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")



## plot corrected distributions ...

p <- ggplot(data=df_ecogeocor, aes(x=dist_geo,dist_eco_corr, colour = factor(group) )) + 
  geom_point( alpha = 0.6) +
  theme_bw() +
  scale_color_manual(values=cols.group) + 
  
  stat_smooth(data = df_ecogeocor %>% filter(group == "2 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "8 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "12 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "14 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "17 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "20 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "23 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "28 yr"),method = "lm", formula = y ~ poly(x, 2)) +
  stat_smooth(data = df_ecogeocor %>% filter(group == "Ref"),method = "lm", formula = y ~ poly(x, 2)) +
  
  annotate(geom="text", x= 0, y= 0.6, label = "Filtered &\ncorrected data", hjust=0, vjust=0, size = 3.25 , col = "grey50") +
  
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size=rel(0.75)),
    legend.text = element_text(size=rel(0.65))
  ) +
  guides(colour = guide_legend(title = "Rehab\nage", override.aes = list(alpha = 1, fill=NA), nrow = 3)) +
  labs(x = "Geographic distance (m)", y = "Ecological distance (Jaccard)")
p
grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Ecodist-vs-Geodist-Jaccard-SOUTH32-SACORRECTED-ExclFarSouthSites-C.tiff"), width = 8, height = 11, units = "cm", res=600, compression="lzw",type="cairo")



names(df_results)
names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"     "group"         
# [8] "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"

names(df_ecogeocor)

## make names of df_ecogeocor compatible with df_out
head(df_out)


head(df_ecogeocor)

# confirm these refer to the same sites and comparison sites
identical( paste0(df_out$samp,"_",df_out$compare_with) , paste0(df_ecogeocor$samp,"_",df_ecogeocor$compare_with) ) # TRUE
dim(df_out) # 105  10

df_new <- df_out
df_new$similarity <- df_ecogeocor$similarity_corr
df_new$corrected <- "yes"

head(df_new)
#      samp compare_with year_rehab similarity group dist_measure       tax_level   study corrected            facet_x
# 1 X138358      X138362       1996   27.61936    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 2 X138358      X138366       1996   19.96079    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 3 X138358      X138374       1996   25.07977    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 4 X138358      X138376       1996   16.99089    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 5 X138358      X138378       1996   19.64054    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level
# 7 X138360      X138362       1996   21.00862    23      Jaccard ASV (with Tree) South32       yes Jaccard\nASV-level

unique(df_new$group)
# [1] 23  Ref 2   17  12  20  8   14  28 
# Levels: 2 < 8 < 12 < 14 < 17 < 20 < 23 < 28 < Ref



table(df_new$group)
# 2   8  12  14  17  20  23  28 Ref 
# 5  15   5  10  10  10  20  10  20 


# only make this once
#df_results <- list()
length(df_results) #
names(df_results)

df_results[["South32-Jaccard-ASV-withTree-SAcorrected-ExclFarSouthSites"]] <- df_new
#df_out <- df_results[[1]]


p <- ggplot(data=df_new, aes(x=group, similarity)) + ggtitle("Jaccard Similarity\nSA-corrected - Excluded far south sites") + #x=group
  geom_boxplot(outlier.shape = NA)+
  #geom_point() +
  geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  #geom_hline(yintercept = 70, color="red") +
  theme_bw() +
  theme(#axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Similarity to Remnants (%)")
p
ggsave(plot=p, filename = paste0(workdir,"/plots/","Jaccard-Similarity--South32-Site-with-SA-correction-ExclFarSouthSites.tiff"), width = 10, height = 10, units = "cm", dpi = 600, compression = "lzw", type = "cairo")




#-------------------------



#### facet_grid trajectory plots + predict time to reach ref:
##   ASV -- Jaccard + Bray-Curtis -- Original + SA-corrected -- Alcoa + Iluka-Eneabba + South32
#-------------------------

names(df_results)



names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"    
# [7] "group"          "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"  


keep_results <- c("Alcoa-Jaccard-ASV-withTree"     ,  "Alcoa-Jaccard-ASV-withTree-SAcorrected" ,
                  "Alcoa-Bray-ASV-withTree" ,   "Alcoa-Bray-ASV-withTree-SAcorrected" ,
                  
                  
                  "Iluka-Eneabba-Jaccard-ASV-withTree" , "Iluka-Eneabba-Jaccard-ASV-withTree-SAcorrected" ,
                  "Iluka-Eneabba-Bray-ASV-withTree" , "Iluka-Eneabba-Bray-ASV-withTree-SAcorrected" ,
                  
                  
                  "South32-Jaccard-ASV-withTree"  ,  "South32-Jaccard-ASV-withTree-SAcorrected" ,
                  "South32-Bray-ASV-withTree" ,      "South32-Bray-ASV-withTree-SAcorrected"  
                   
)


keep_cols <- c("samp","compare_with","similarity","group","dist_measure","tax_level","study","corrected","facet_x")
traj_data <- df_results[[ keep_results[1] ]][ ,keep_cols]
for (i in 2:length(keep_results)) {
  traj_data <- rbind(traj_data,df_results[[ keep_results[i] ]][ ,keep_cols])
  print(paste0("completed ",i))
}
dim(traj_data) # 3996    9
unique(traj_data$facet_x) # "Jaccard\nASV-level"     "Bray-Curtis\nASV-level"

traj_data$NEW_newfacet_x <- traj_data$facet_x
sel <- which(traj_data$corrected == "no")
traj_data$NEW_newfacet_x[sel] <- paste0("Original data\n",traj_data$NEW_newfacet_x[sel])
sel <- which(traj_data$corrected == "yes")
traj_data$NEW_newfacet_x[sel] <- paste0("Corrected data\n",traj_data$NEW_newfacet_x[sel])
unique(traj_data$NEW_newfacet_x)
# [1] "Original data\nJaccard\nASV-level"      "Corrected data\nJaccard\nASV-level"     "Original data\nBray-Curtis\nASV-level" 
# [4] "Corrected data\nBray-Curtis\nASV-level"
traj_data$NEW_newfacet_x <- factor(traj_data$NEW_newfacet_x, levels = c("Original data\nJaccard\nASV-level",
                                                                        "Corrected data\nJaccard\nASV-level",
                                                                        "Original data\nBray-Curtis\nASV-level",
                                                                        "Corrected data\nBray-Curtis\nASV-level"
                                                                        ), ordered = TRUE)

keep_cols_anno <- c("group","sig_letter","similarity","dist_measure","tax_level","study","corrected","facet_x")
anno_data <- df_anno[[ keep_results[1] ]][ ,keep_cols_anno]
print(paste0(" - add rows:",dim(df_anno[[ keep_results[1] ]])[1]," - new dim: ",dim(anno_data)[1]," x ",dim(anno_data)[2]))
for (i in 2:length(keep_results)) {
  anno_data <- rbind(anno_data,df_anno[[ keep_results[i] ]][ ,keep_cols_anno])
  print(paste0("completed ",i," - add rows:",dim(df_anno[[ keep_results[i] ]])[1]," - new dim: ",dim(anno_data)[1]," x ",dim(anno_data)[2]))
}
dim(anno_data) # 96  8
class(anno_data$facet_x) # "character"
unique(anno_data$facet_x) # "character"
# "Jaccard\nASV-level"     "Bray-Curtis\nASV-level"

anno_data$NEW_newfacet_x <- anno_data$facet_x
sel <- which(anno_data$corrected == "no")
anno_data$NEW_newfacet_x[sel] <- paste0("Original data\n",anno_data$NEW_newfacet_x[sel])
sel <- which(anno_data$corrected == "yes")
anno_data$NEW_newfacet_x[sel] <- paste0("Corrected data\n",anno_data$NEW_newfacet_x[sel])
unique(anno_data$NEW_newfacet_x)
# [1] "Original data\nJaccard\nASV-level"      "Corrected data\nJaccard\nASV-level"     "Original data\nBray-Curtis\nASV-level" 
# [4] "Corrected data\nBray-Curtis\nASV-level"
anno_data$NEW_newfacet_x <- factor(anno_data$NEW_newfacet_x, levels = c("Original data\nJaccard\nASV-level",
                                                                        "Corrected data\nJaccard\nASV-level",
                                                                        "Original data\nBray-Curtis\nASV-level",
                                                                        "Corrected data\nBray-Curtis\nASV-level"),
                                   ordered = TRUE)

class(traj_data$group) # "ordered" "factor" 

traj_data$group <- factor(traj_data$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)

levels(traj_data$group)
# [1] "0"   "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14" 
# [16] "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29" 
# [31] "30"  "31"  "32"  "33"  "34"  "35"  "36"  "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44" 
# [46] "45"  "Ref"

length(levels(traj_data$group)) #47

x_labels <- c("0","","","","","","","","","","10",
              "","","","","","","","","","20",
              "","","","","","","","","","30",
              "","","","","","","","","","40",
              "","","","","","Ref")
length(x_labels)
# 47
x_ticks <- c(1,0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,0,0,0,0,1,
             0,0,0,0,0,1)
length(x_ticks)
#47


unique(traj_data$study) # "Alcoa"         "Iluka-Eneabba" "South32" 

unique(anno_data$study) # "Alcoa"         "Iluka-Eneabba" "South32"

 



## median reference horizontal lines ...

href_lines <- data.frame(
  study  = rep( c("Alcoa", "Iluka-Eneabba", "South32"), times = 4 ),
  # facet_x  = rep( c("Jaccard\nASV-level","Bray-Curtis\nASV-level",
  #                   "Unweighted UniFrac\nASV-level","Weighted UniFrac\nASV-level"), each = 3),
  
  NEW_newfacet_x <- rep( c("Original data\nJaccard\nASV-level",
                           "Corrected data\nJaccard\nASV-level",
                           "Original data\nBray-Curtis\nASV-level",
                           "Corrected data\nBray-Curtis\nASV-level"), each = 3),
  similarity= NA
)

x <- traj_data

for (i in 1:dim(href_lines)[1]) {
  #i<-1
  #sel <- which(x$study == href_lines$study[i] & x$facet_x == href_lines$facet_x[i] & x$group == "Ref")
  sel <- which(x$study == href_lines$study[i] & x$NEW_newfacet_x == href_lines$NEW_newfacet_x[i] & x$group == "Ref")
  href_lines$similarity[i] <- median( x$similarity[sel] )
}
href_lines$NEW_newfacet_x <- factor(href_lines$NEW_newfacet_x,
                             levels = c("Original data\nJaccard\nASV-level",
                                        "Corrected data\nJaccard\nASV-level",
                                        "Original data\nBray-Curtis\nASV-level",
                                        "Corrected data\nBray-Curtis\nASV-level"),
                             ordered = TRUE)


## fit logarithmic models for each data type (facet_x)



## BOOTSTRAP MODELS ...

pred_models_boot <- data.frame(study=NA, # one of: "Alcoa", "Iluka-Eneabba", "South32"
                               #facet_x=NA, # one of: c("Jaccard\nASV-level","Bray-Curtis\nASV-level", "Jaccard\nGenus-level","Bray-Curtis\nGenus-level")
                               NEW_newfacet_x=NA,
                               study_and_metric=NA, # combine the two values above
                               group=NA,
                               group_num=NA,
                               similarity=NA,
                               boot_no=NA,                 # EXTRA
                               ref_median=NA,              # EXTRA
                               pred_time_to_ref_median=NA, # EXTRA
                               pred_type=NA)     # c("rectangular hyperbola", "logarithmic")
pred_models_boot.temp <- pred_models_boot

studies <- c("Alcoa", "Iluka-Eneabba", "South32")
metrics <- c("Original data\nJaccard\nASV-level",
             "Corrected data\nJaccard\nASV-level",
             "Original data\nBray-Curtis\nASV-level",
             "Corrected data\nBray-Curtis\nASV-level")
pred_types <- c("rectangular hyperbola", "negative exponential", "logarithmic")
k<-3

## run logarithmic model only !!!

x <- traj_data
dim(x) # 3996    10

for (i in 1:length(studies)) {
  #i<-1
  this_study <- studies[i]
  for (j in 1:length(metrics)) {
    #j<-3
    this_metric <- metrics[j]
    
    # prepare data for modelling
    #subsel <- which(x$facet_x == this_metric & x$study==this_study)
    subsel <- which(x$NEW_newfacet_x == this_metric & x$study==this_study)
    
    newdat <- x[subsel, c("similarity","group") ]
    newdat$group_num <- as.numeric(as.character(newdat$group))
    # remove 'Ref' similarities which have now been converted to NA
    sel.na <- which(is.na(newdat$group_num))
    
    this_ref_median <- median(newdat$similarity[sel.na])
    
    newdat <- newdat[-sel.na, ]
    

    # prep output dataframe for prediction model to populate
    sel.dup <- which(duplicated(newdat$group)==TRUE)
    new_preds <- data.frame(matrix( nrow=dim(newdat[-sel.dup, ])[1], ncol = length(names(pred_models_boot.temp)) ))
    names(new_preds) <- names(pred_models_boot.temp)
    new_preds$group_num <- sort( newdat$group_num[-sel.dup] )
    new_preds$study <- this_study
    #new_preds$facet_x <- this_metric
    new_preds$NEW_newfacet_x <- this_metric
    #new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$facet_x)
    new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$NEW_newfacet_x)
    new_preds$group <- as.character(new_preds$group_num)
    new_preds$ref_median <- this_ref_median
    
    #for (k in 1:length(pred_types)) {

    this_pred_model_type <- pred_types[k]
    
    
    ## NOTE:
    #  rectangular hyperbola ( fct = MM.2() from ‘drc' package; https://www.statforbiology.com/nonlinearregression/usefulequations#michaelis-menten_equation), and 
    #  negative exponential ( fct = DRC.negExp() from 'aomisc' package; https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function)
    #  models were trialled however these regularly failed to achieve model fits.
    
    
    # NOW using "logarithmic" model only !!!!!
    
    # negative exponential model, using fct = DRC.negExp() from 'aomisc' package
    # https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function
    
    
    for (b in 1:100) {
      #b<-1
      new_preds$boot_no <- b
      set.seed(b + 1234)
      sel.boot <- sample(x = c(1:dim(newdat)[1]), size = dim(newdat)[1], replace = TRUE)
      
      bootdat <- newdat[sel.boot, ]
      
      # exclude outlying data points from trajectory data-cloud ... 
      # as outliers are likely to throw out model-fitting
      bp <- boxplot(bootdat$similarity, print=FALSE)
      sel.outlier <- which(bootdat$similarity %in% bp$out)
      if (length(sel.outlier)>0) {
        bootdat <- bootdat[-sel.outlier, ]
      }
      
      #plot(bootdat$group_num, bootdat$similarity)
      
      model <- drm(similarity ~ group_num , fct = DRC.logCurve(), # DRC.negExp()
                   data = bootdat)
      
      new_preds$similarity <- predict(model, new_preds )
      new_preds$pred_type <- this_pred_model_type
      
      y <- data.frame(group_num=1:500, sim_pred=NA)
      y$sim_pred <- predict(model, y )
      y$reached_target <- y$sim_pred >= this_ref_median
      #plot(y$group_num, y$sim_pred)
      row.names(y)
      
      sel <- which(y$reached_target == TRUE)
      
      if (length(sel) > 0) {
        years_to_ref <- min(sel)
      } else {
        years_to_ref <- NA
      }
      new_preds$pred_time_to_ref_median <- years_to_ref
      
      # add predictions from end of last year to 40 years ...
      max_year <- max(new_preds$group_num)
      year <- max_year + 1
      while (year < 40) {
        new_preds <- rbind(new_preds, new_preds[ dim(new_preds)[1], ])
        year <- year + 1
        new_preds$group_num[dim(new_preds)[1]] <- year
        new_preds$group[dim(new_preds)[1]] <- as.character( new_preds$group_num[dim(new_preds)[1]] )
        new_preds$similarity[dim(new_preds)[1]] <- y[ year, "sim_pred"]
      }
      
      pred_models_boot <- rbind(pred_models_boot, new_preds)
      
    } # END bootstrap
    

    print(paste0("completed study: ",studies[i],"; metric: ",metrics[j],"; pred_type: ",pred_types[k]))  
    #} # END pred_types
  } # END metrics
} # END studies


# remove NA first row
pred_models_boot <- pred_models_boot[-1, ]

#unique(pred_models_boot$facet_x)
unique(pred_models_boot$NEW_newfacet_x)
# [1] "Original data\nJaccard\nASV-level"      "Corrected data\nJaccard\nASV-level"     "Original data\nBray-Curtis\nASV-level" 
# [4] "Corrected data\nBray-Curtis\nASV-level"

# pred_models_boot$facet_x <- factor(pred_models_boot$facet_x,
#                                    levels = c("Jaccard\nASV-level", "Bray-Curtis\nASV-level", 
#                                               "Unweighted UniFrac\nASV-level", "Weighted UniFrac\nASV-level"),
#                                    ordered = TRUE)
pred_models_boot$NEW_newfacet_x <- factor(pred_models_boot$NEW_newfacet_x,
                                   levels = c("Original data\nJaccard\nASV-level",
                                              "Corrected data\nJaccard\nASV-level",
                                              "Original data\nBray-Curtis\nASV-level",
                                              "Corrected data\nBray-Curtis\nASV-level"),
                                   ordered = TRUE)


dim(pred_models_boot) # 17200  10


# correct time prediction to ref; update NA to (>) 500 yrs
sel <- which(is.na(pred_models_boot$pred_time_to_ref_median))
if (length(sel)>0) { pred_models_boot$pred_time_to_ref_median[sel] <- 500 }


pred_models_boot$group <- factor(pred_models_boot$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)
str(pred_models_boot)


str(traj_data)


study_names <- c(
  Alcoa="Huntly",
  `Iluka-Eneabba`="Eneabba",
  South32="Worsley"
)


p <- ggplot(data=traj_data, aes(x=group, y=similarity)) +
  geom_line(data=pred_models_boot, aes(  x=group, y=similarity, group = boot_no ), colour = "#e34a33", alpha = 0.02) + # "red"
  geom_hline(data = href_lines, aes(yintercept = similarity),  color= "#2c7fb8", linetype="longdash") + # "blue"
  #geom_line(data= pred_models_alldata, aes(  x=group, y=similarity, group = 1), colour = "#e34a33" ) + # "red"
  
  geom_boxplot(outlier.shape = NA, width = 2) +
  #geom_jitter(size=1.5,width = 0.15, alpha=0.2) +
  
  theme_bw() +
  geom_text(data=anno_data, aes(x=group, y=similarity, label=sig_letter),nudge_y = 2, size=2.75) +
  scale_x_discrete(drop=FALSE, expand = c(0.05,0.05),
                   labels = x_labels) +
  theme(axis.text.x  = element_text(angle=90,hjust=1, vjust=0.5), # 
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_line(linetype = 0)
        axis.ticks.x = element_line(linetype = x_ticks),
        strip.text.x = element_text(size = rel(1.1)),
        strip.text.y = element_text(size = rel(1.1))
        ) +
  labs(x = "Rehabilitation age (years)", y = "Similarity to Reference (%)") +
  facet_grid(rows = vars(study), cols = vars(NEW_newfacet_x), scales = "fixed", labeller = labeller(study = study_names))

p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Restoration-trajectories--Alcoa-Iluka-South32--ASV-Jaccard-Bray-Original-SA-Corrected-with-Preds.tiff"), width = 18, height = 16, units = "cm", dpi = 600, compression = "lzw",type="cairo")



str(traj_data)
unique(traj_data$study) # "Alcoa"         "Iluka-Eneabba" "South32"
unique(traj_data$group) # use "ref" only
unique(traj_data$NEW_newfacet_x)
# [1] Original data\nJaccard\nASV-level      Corrected data\nJaccard\nASV-level     Original data\nBray-Curtis\nASV-level 
# [4] Corrected data\nBray-Curtis\nASV-level

## Huntly-Alcoa
sel <- which(traj_data$study == "Alcoa")
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "Original data\nJaccard\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 9.503  17.824  20.345  20.092  23.213  35.809 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "Corrected data\nJaccard\nASV-level")
summary(traj_data$similarity[sel[subsel]])
# same
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "Original data\nBray-Curtis\nASV-level")
summary(traj_data$similarity[sel[subsel]])
#  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 22.38   31.28   35.55   35.79   39.93   60.91 
subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "Corrected data\nBray-Curtis\nASV-level")
summary(traj_data$similarity[sel[subsel]])
# same


str(pred_models_boot)

unique(pred_models_boot$study_and_metric)
# [1] "Alcoa__Original data\nJaccard\nASV-level"              "Alcoa__Corrected data\nJaccard\nASV-level"            
# [3] "Alcoa__Original data\nBray-Curtis\nASV-level"          "Alcoa__Corrected data\nBray-Curtis\nASV-level"        
# [5] "Iluka-Eneabba__Original data\nJaccard\nASV-level"      "Iluka-Eneabba__Corrected data\nJaccard\nASV-level"    
# [7] "Iluka-Eneabba__Original data\nBray-Curtis\nASV-level"  "Iluka-Eneabba__Corrected data\nBray-Curtis\nASV-level"
# [9] "South32__Original data\nJaccard\nASV-level"            "South32__Corrected data\nJaccard\nASV-level"          
# [11] "South32__Original data\nBray-Curtis\nASV-level"        "South32__Corrected data\nBray-Curtis\nASV-level" 

pred_models_boot$study_and_metric <- factor(pred_models_boot$study_and_metric,
                                            levels = c(
                                              "Alcoa__Original data\nJaccard\nASV-level"          ,    "Alcoa__Corrected data\nJaccard\nASV-level"            ,
                                              "Alcoa__Original data\nBray-Curtis\nASV-level"       ,   "Alcoa__Corrected data\nBray-Curtis\nASV-level"        ,
                                              "Iluka-Eneabba__Original data\nJaccard\nASV-level"    ,  "Iluka-Eneabba__Corrected data\nJaccard\nASV-level"    ,
                                              "Iluka-Eneabba__Original data\nBray-Curtis\nASV-level" , "Iluka-Eneabba__Corrected data\nBray-Curtis\nASV-level",
                                              "South32__Original data\nJaccard\nASV-level"          ,  "South32__Corrected data\nJaccard\nASV-level"          ,
                                              "South32__Original data\nBray-Curtis\nASV-level"      ,  "South32__Corrected data\nBray-Curtis\nASV-level" 
                                            ))

dim(pred_models_boot) # 17200    10

dat.sum <- pred_models_boot
head(dat.sum)
#    study                    NEW_newfacet_x                         study_and_metric group group_num similarity boot_no ref_median pred_time_to_ref_median   pred_type
# 11 Alcoa Original data\nJaccard\nASV-level Alcoa__Original data\nJaccard\nASV-level     2         2   8.819282       1   20.34483                      66 logarithmic
# 2  Alcoa Original data\nJaccard\nASV-level Alcoa__Original data\nJaccard\nASV-level     8         8  13.392202       1   20.34483                      66 logarithmic
# 3  Alcoa Original data\nJaccard\nASV-level Alcoa__Original data\nJaccard\nASV-level    14        14  15.238187       1   20.34483                      66 logarithmic
# 4  Alcoa Original data\nJaccard\nASV-level Alcoa__Original data\nJaccard\nASV-level    17        17  15.878643       1   20.34483                      66 logarithmic
# 5  Alcoa Original data\nJaccard\nASV-level Alcoa__Original data\nJaccard\nASV-level    25        25  17.150814       1   20.34483                      66 logarithmic
# 6  Alcoa Original data\nJaccard\nASV-level Alcoa__Original data\nJaccard\nASV-level    29        29  17.640402       1   20.34483                      66 logarithmic

for (i in 1:length(levels(dat.sum$study_and_metric))) {
  #i<-1
  this_study_and_metric <- levels(dat.sum$study_and_metric)[i]
  sel <- which(dat.sum$study_and_metric == this_study_and_metric)
  #unique(dat.sum$ref_median[sel])
  temp <- dat.sum$pred_time_to_ref_median[sel]
  #quantile(temp, probs = c(0.05, 0.25, 0.5, 0.75, 0.95), na.rm = TRUE)
  #hist(temp)
  res <- quantile(temp, probs = c(0.05, 0.5, 0.95), na.rm = TRUE)
  names(res)
  #"5%"  "50%" "95%"
  #print(paste0(this_study_and_metric))
  print(paste0(this_study_and_metric,": ",res["50%"]," (",res["5%"],", ",res["95%"]," yr)"))
}
# [1] "Alcoa__Original data\nJaccard\nASV-level: 60.5 (51, 79 yr)"
# [1] "Alcoa__Corrected data\nJaccard\nASV-level: 57 (48, 73 yr)"
# [1] "Alcoa__Original data\nBray-Curtis\nASV-level: 43 (39, 49 yr)"
# [1] "Alcoa__Corrected data\nBray-Curtis\nASV-level: 40.5 (37, 46 yr)"
# [1] "Iluka-Eneabba__Original data\nJaccard\nASV-level: 97 (73.9, 133.45 yr)"
# [1] "Iluka-Eneabba__Corrected data\nJaccard\nASV-level: 90 (69.95, 124 yr)"
# [1] "Iluka-Eneabba__Original data\nBray-Curtis\nASV-level: 93.5 (69, 131.35 yr)"
# [1] "Iluka-Eneabba__Corrected data\nBray-Curtis\nASV-level: 86.5 (64.95, 118.05 yr)"
# [1] "South32__Original data\nJaccard\nASV-level: 131.5 (68.8, 313.25 yr)"
# [1] "South32__Corrected data\nJaccard\nASV-level: 57 (41.95, 105.1 yr)"
# [1] "South32__Original data\nBray-Curtis\nASV-level: 103.5 (68.9, 173.549999999999 yr)"
# [1] "South32__Corrected data\nBray-Curtis\nASV-level: 50 (39, 76.05 yr)"


levels( pred_models_boot$NEW_newfacet_x )
# [1] "Original data\nJaccard\nASV-level"      "Corrected data\nJaccard\nASV-level"     "Original data\nBray-Curtis\nASV-level" 
# [4] "Corrected data\nBray-Curtis\nASV-level"



## plot time to target

names(pred_models_boot)
# [1] "study"                   "facet_x"                 "study_and_metric"        "group"                   "group_num"              
# [6] "similarity"              "boot_no"                 "ref_median"              "pred_time_to_ref_median" "pred_type"

p <- ggplot(data=pred_models_boot, aes(x=study, y=pred_time_to_ref_median, color = NEW_newfacet_x)) +
  geom_boxplot() +
  theme_bw() +
  #scale_color_manual(
  scale_colour_viridis_d(direction = -1, name  ="Distance measure", guide="legend",
                         limits = c("Original data\nJaccard\nASV-level",
                                    "Corrected data\nJaccard\nASV-level",
                                    "Original data\nBray-Curtis\nASV-level",
                                    "Corrected data\nBray-Curtis\nASV-level"),
                         labels = c("Original data\nJaccard",
                                    "Corrected data\nJaccard",
                                    "Original data\nBray-Curtis",
                                    "Corrected data\nBray-Curtis")
  ) +
  #guides(color = guide_legend(title="Distance measure")) +
  theme(#axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    legend.key.height = unit(30, "pt")
    #legend.margin = margin(t = 20,r = 0,b = 0,l = 20,"pt")
  ) +
  labs(x = "Study", y = "Predicted time to reach target (yr)")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Predicted-time-to-target--Alcoa-Iluka-South32--ASV-Jaccard-Bray-Original-SACorrected.tiff"), width = 16, height = 10, units = "cm", dpi = 600, compression = "lzw",type="cairo")


df <- pred_models_boot
unique( df$NEW_newfacet_x )
# [1] Original data\nJaccard\nASV-level      Corrected data\nJaccard\nASV-level     Original data\nBray-Curtis\nASV-level 
# [4] Corrected data\nBray-Curtis\nASV-level
# 4 Levels: Original data\nJaccard\nASV-level < Corrected data\nJaccard\nASV-level < ... < Corrected data\nBray-Curtis\nASV-level

unique( df$study ) # "Alcoa"         "Iluka-Eneabba" "South32"  
df$study <- factor(df$study, levels = c("Alcoa", "Iluka-Eneabba", "South32"  ),
                   labels = c("Huntly", "Eneabba", "Worsley"), ordered = TRUE)
df$NEW_newfacet_x <- factor(df$NEW_newfacet_x, 
                            levels = c("Original data\nJaccard\nASV-level",
                                       "Corrected data\nJaccard\nASV-level",
                                       "Original data\nBray-Curtis\nASV-level",
                                       "Corrected data\nBray-Curtis\nASV-level"),
                            labels = c("Original data\nJaccard",
                                       "Corrected data\nJaccard",
                                       "Original data\nBray-Curtis",
                                       "Corrected data\nBray-Curtis"),
                            ordered = TRUE)

p <- ggplot(data=df, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median)) + #, color = NEW_newfacet_x
  geom_boxplot() +
  theme_bw() +
  #guides(color = guide_legend(title="Distance measure")) +
  theme(
    axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    legend.key.height = unit(30, "pt")
    #legend.margin = margin(t = 20,r = 0,b = 0,l = 20,"pt")
  ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)") + 
  facet_wrap(facets = vars(study), scales = "free_y")
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Predicted-time-to-target--Alcoa-Iluka-South32--ASV-Jaccard-Bray-Original-SACorrected-facet-study.tiff"), width = 16, height = 8, units = "cm", dpi = 600, compression = "lzw",type="cairo")


#-------------------------


#### Predicted time to target for South32 - Excluding far south sites
#-------------------------

names(df_results)


names(df_results[[1]])
# [1] "samp"           "compare_with"   "site"           "site_full_desc" "year_rehab"     "similarity"    
# [7] "group"          "dist_measure"   "tax_level"      "study"          "corrected"      "facet_x"  


keep_results <- c( "South32-Jaccard-ASV-withTree",
                   "South32-Jaccard-ASV-withTree-SAcorrected-ExclFarSouthSites",
                   "South32-Bray-ASV-withTree",
                   "South32-Bray-ASV-withTree-SAcorrected-ExclFarSouthSites"
)


keep_cols <- c("samp","compare_with","similarity","group","dist_measure","tax_level","study","corrected","facet_x")
traj_data <- df_results[[ keep_results[1] ]][ ,keep_cols]
for (i in 2:length(keep_results)) {
  traj_data <- rbind(traj_data,df_results[[ keep_results[i] ]][ ,keep_cols])
  print(paste0("completed ",i))
}
dim(traj_data) # 498   9
unique(traj_data$facet_x) # "Jaccard\nASV-level"     "Bray-Curtis\nASV-level"

# now check there are NONE of the far south sites in this data:

sel.rm <- which(traj_data$samp %in% c("X138402", "X138404", "X138406") | traj_data$compare_with %in% c("X138402", "X138404", "X138406")) # qty 78 rows
traj_data$corrected[sel.rm]
# [1] "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no"
# [27] "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no"
# [53] "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no" "no"

# i.e. need to remove far south sites from the Original data component in this combined dataset

traj_data <- traj_data[-sel.rm, ]


traj_data$NEW_newfacet_x <- traj_data$facet_x
sel <- which(traj_data$corrected == "no")
traj_data$NEW_newfacet_x[sel] <- paste0("Original data\n",traj_data$NEW_newfacet_x[sel])
sel <- which(traj_data$corrected == "yes")
traj_data$NEW_newfacet_x[sel] <- paste0("Corrected data\n",traj_data$NEW_newfacet_x[sel])
unique(traj_data$NEW_newfacet_x)
# [1] "Original data\nJaccard\nASV-level"      "Corrected data\nJaccard\nASV-level"     "Original data\nBray-Curtis\nASV-level" 
# [4] "Corrected data\nBray-Curtis\nASV-level"
traj_data$NEW_newfacet_x <- factor(traj_data$NEW_newfacet_x, levels = c("Original data\nJaccard\nASV-level",
                                                                        "Corrected data\nJaccard\nASV-level",
                                                                        "Original data\nBray-Curtis\nASV-level",
                                                                        "Corrected data\nBray-Curtis\nASV-level"
), ordered = TRUE)



class(traj_data$group) # "ordered" "factor" 

traj_data$group <- factor(traj_data$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)

levels(traj_data$group)
# [1] "0"   "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14" 
# [16] "15"  "16"  "17"  "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29" 
# [31] "30"  "31"  "32"  "33"  "34"  "35"  "36"  "37"  "38"  "39"  "40"  "41"  "42"  "43"  "44" 
# [46] "45"  "Ref"

length(levels(traj_data$group)) #47

# x_labels <- c("0","","","","","","","","","","10",
#               "","","","","","","","","","20",
#               "","","","","","","","","","30",
#               "","","","","","","","","","40",
#               "","","","","","Ref")
# length(x_labels)
# # 47
# x_ticks <- c(1,0,0,0,0,0,0,0,0,0,1,
#              0,0,0,0,0,0,0,0,0,1,
#              0,0,0,0,0,0,0,0,0,1,
#              0,0,0,0,0,0,0,0,0,1,
#              0,0,0,0,0,1)
# length(x_ticks)
# #47


unique(traj_data$study) # "South32" 

#unique(anno_data$study) # "Alcoa"         "Iluka-Eneabba" "South32"

 




x <- traj_data




## fit logarithmic models for each data type (facet_x)



## BOOTSTRAP MODELS ...

pred_models_boot <- data.frame(study=NA, # one of: "Alcoa", "Iluka-Eneabba", "South32"
                               #facet_x=NA, # one of: c("Jaccard\nASV-level","Bray-Curtis\nASV-level", "Jaccard\nGenus-level","Bray-Curtis\nGenus-level")
                               NEW_newfacet_x=NA,
                               study_and_metric=NA, # combine the two values above
                               group=NA,
                               group_num=NA,
                               similarity=NA,
                               boot_no=NA,                 # EXTRA
                               ref_median=NA,              # EXTRA
                               pred_time_to_ref_median=NA, # EXTRA
                               pred_type=NA)     # c("rectangular hyperbola", "logarithmic")
pred_models_boot.temp <- pred_models_boot

studies <- c("South32")
metrics <- c("Original data\nJaccard\nASV-level",
             "Corrected data\nJaccard\nASV-level",
             "Original data\nBray-Curtis\nASV-level",
             "Corrected data\nBray-Curtis\nASV-level")
pred_types <- c("rectangular hyperbola", "negative exponential", "logarithmic")
k<-3

## run logarithmic model only !!!

x <- traj_data
dim(x) # 420    10

for (i in 1:length(studies)) {
  #i<-1
  this_study <- studies[i]
  for (j in 1:length(metrics)) {
    #j<-3
    this_metric <- metrics[j]
    
    # prepare data for modelling
    #subsel <- which(x$facet_x == this_metric & x$study==this_study)
    subsel <- which(x$NEW_newfacet_x == this_metric & x$study==this_study)
    
    newdat <- x[subsel, c("similarity","group") ]
    newdat$group_num <- as.numeric(as.character(newdat$group))
    # remove 'Ref' similarities which have now been converted to NA
    sel.na <- which(is.na(newdat$group_num))
    
    this_ref_median <- median(newdat$similarity[sel.na])
    
    newdat <- newdat[-sel.na, ]
    

    # prep output dataframe for prediction model to populate
    sel.dup <- which(duplicated(newdat$group)==TRUE)
    new_preds <- data.frame(matrix( nrow=dim(newdat[-sel.dup, ])[1], ncol = length(names(pred_models_boot.temp)) ))
    names(new_preds) <- names(pred_models_boot.temp)
    new_preds$group_num <- sort( newdat$group_num[-sel.dup] )
    new_preds$study <- this_study
    #new_preds$facet_x <- this_metric
    new_preds$NEW_newfacet_x <- this_metric
    #new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$facet_x)
    new_preds$study_and_metric <- paste0(new_preds$study,"__",new_preds$NEW_newfacet_x)
    new_preds$group <- as.character(new_preds$group_num)
    new_preds$ref_median <- this_ref_median
    
    #for (k in 1:length(pred_types)) {

    
    this_pred_model_type <- pred_types[k]
    
    ## NOTE:
    #  rectangular hyperbola ( fct = MM.2() from ‘drc' package; https://www.statforbiology.com/nonlinearregression/usefulequations#michaelis-menten_equation), and 
    #  negative exponential ( fct = DRC.negExp() from 'aomisc' package; https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function)
    #  models were trialled however these regularly failed to achieve model fits.
    
    
    # NOW using "logarithmic" model only !!!!!
    
    # negative exponential model, using fct = DRC.negExp() from 'aomisc' package
    # https://www.statforbiology.com/2020/stat_nls_usefulfunctions/#asymptotic-function
    
    
    for (b in 1:100) {
      #b<-1
      new_preds$boot_no <- b
      set.seed(b + 1234)
      sel.boot <- sample(x = c(1:dim(newdat)[1]), size = dim(newdat)[1], replace = TRUE)
      
      bootdat <- newdat[sel.boot, ]
      
      # exclude outlying data points from trajectory data-cloud ... 
      # as outliers are likely to throw out model-fitting
      bp <- boxplot(bootdat$similarity, print=FALSE)
      sel.outlier <- which(bootdat$similarity %in% bp$out)
      if (length(sel.outlier)>0) {
        bootdat <- bootdat[-sel.outlier, ]
      }
      
      #plot(bootdat$group_num, bootdat$similarity)
      
      model <- drm(similarity ~ group_num , fct = DRC.logCurve(), # DRC.negExp()
                   data = bootdat)
      
      new_preds$similarity <- predict(model, new_preds )
      new_preds$pred_type <- this_pred_model_type
      
      y <- data.frame(group_num=1:500, sim_pred=NA)
      y$sim_pred <- predict(model, y )
      y$reached_target <- y$sim_pred >= this_ref_median
      #plot(y$group_num, y$sim_pred)
      row.names(y)
      
      sel <- which(y$reached_target == TRUE)
      
      if (length(sel) > 0) {
        years_to_ref <- min(sel)
      } else {
        years_to_ref <- NA
      }
      new_preds$pred_time_to_ref_median <- years_to_ref
      
      # add predictions from end of last year to 40 years ...
      max_year <- max(new_preds$group_num)
      year <- max_year + 1
      while (year < 40) {
        new_preds <- rbind(new_preds, new_preds[ dim(new_preds)[1], ])
        year <- year + 1
        new_preds$group_num[dim(new_preds)[1]] <- year
        new_preds$group[dim(new_preds)[1]] <- as.character( new_preds$group_num[dim(new_preds)[1]] )
        new_preds$similarity[dim(new_preds)[1]] <- y[ year, "sim_pred"]
      }
      
      pred_models_boot <- rbind(pred_models_boot, new_preds)
      
    } # END bootstrap
    

    print(paste0("completed study: ",studies[i],"; metric: ",metrics[j],"; pred_type: ",pred_types[k]))  
    #} # END pred_types
  } # END metrics
} # END studies


# remove NA first row
pred_models_boot <- pred_models_boot[-1, ]

#unique(pred_models_boot$facet_x)
unique(pred_models_boot$NEW_newfacet_x)
# [1] "Original data\nJaccard\nASV-level"      "Corrected data\nJaccard\nASV-level"     "Original data\nBray-Curtis\nASV-level" 
# [4] "Corrected data\nBray-Curtis\nASV-level"

# pred_models_boot$facet_x <- factor(pred_models_boot$facet_x,
#                                    levels = c("Jaccard\nASV-level", "Bray-Curtis\nASV-level", 
#                                               "Unweighted UniFrac\nASV-level", "Weighted UniFrac\nASV-level"),
#                                    ordered = TRUE)
pred_models_boot$NEW_newfacet_x <- factor(pred_models_boot$NEW_newfacet_x,
                                          levels = c("Original data\nJaccard\nASV-level",
                                                     "Corrected data\nJaccard\nASV-level",
                                                     "Original data\nBray-Curtis\nASV-level",
                                                     "Corrected data\nBray-Curtis\nASV-level"),
                                          ordered = TRUE)


dim(pred_models_boot) # 7600  10


# correct time prediction to ref; update NA to (>) 500 yrs
sel <- which(is.na(pred_models_boot$pred_time_to_ref_median))
if (length(sel)>0) { pred_models_boot$pred_time_to_ref_median[sel] <- 500 }


pred_models_boot$group <- factor(pred_models_boot$group, levels=as.character(c(0:45,"Ref")),ordered=TRUE)
str(pred_models_boot)
# 'data.frame':	7600 obs. of  10 variables:
#   $ study                  : chr  "South32" "South32" "South32" "South32" ...
# $ NEW_newfacet_x         : Ord.factor w/ 4 levels "Original data\nJaccard\nASV-level"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ study_and_metric       : chr  "South32__Original data\nJaccard\nASV-level" "South32__Original data\nJaccard\nASV-level" "South32__Original data\nJaccard\nASV-level" "South32__Original data\nJaccard\nASV-level" ...
# $ group                  : Ord.factor w/ 47 levels "0"<"1"<"2"<"3"<..: 3 9 13 15 18 21 24 29 31 32 ...
# $ group_num              : num  2 8 12 14 17 20 23 28 30 31 ...
# $ similarity             : num  15 18.3 19.2 19.6 20.1 ...
# $ boot_no                : int  1 1 1 1 1 1 1 1 1 1 ...
# $ ref_median             : num  25.7 25.7 25.7 25.7 25.7 ...
# $ pred_time_to_ref_median: num  192 192 192 192 192 192 192 192 192 192 ...
# $ pred_type              : chr  "logarithmic" "logarithmic" "logarithmic" "logarithmic" ...

str(traj_data)
# 'data.frame':	420 obs. of  10 variables:
#   $ samp          : chr  "X138358" "X138358" "X138358" "X138358" ...
# $ compare_with  : chr  "X138362" "X138366" "X138374" "X138376" ...
# $ similarity    : num  29.9 21.3 23.9 15.5 18.6 ...
# $ group         : Ord.factor w/ 47 levels "0"<"1"<"2"<"3"<..: 24 24 24 24 24 24 24 24 24 24 ...
# $ dist_measure  : chr  "Jaccard" "Jaccard" "Jaccard" "Jaccard" ...
# $ tax_level     : chr  "ASV (with Tree)" "ASV (with Tree)" "ASV (with Tree)" "ASV (with Tree)" ...
# $ study         : chr  "South32" "South32" "South32" "South32" ...
# $ corrected     : chr  "no" "no" "no" "no" ...
# $ facet_x       : chr  "Jaccard\nASV-level" "Jaccard\nASV-level" "Jaccard\nASV-level" "Jaccard\nASV-level" ...
# $ NEW_newfacet_x: Ord.factor w/ 4 levels "Original data\nJaccard\nASV-level"<..: 1 1 1 1 1 1 1 1 1 1 ...




str(traj_data)
unique(traj_data$study) #  "South32"
unique(traj_data$group) # use "ref" only
unique(traj_data$NEW_newfacet_x)
# [1] Original data\nJaccard\nASV-level      Corrected data\nJaccard\nASV-level     Original data\nBray-Curtis\nASV-level 
# [4] Corrected data\nBray-Curtis\nASV-level
# 4 Levels: Original data\nJaccard\nASV-level < Corrected data\nJaccard\nASV-level < ... < Corrected data\nBray-Curtis\nASV-level

## study
sel <- which(traj_data$study == "South32")

subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "Original data\nJaccard\nASV-level")
summary(traj_data$similarity[sel[subsel]])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 18.03   25.34   27.60   26.60   28.77   31.25

subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "Corrected data\nJaccard\nASV-level")
summary(traj_data$similarity[sel[subsel]])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 18.03   25.34   27.60   26.60   28.77   31.25

subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "Original data\nBray-Curtis\nASV-level")
summary(traj_data$similarity[sel[subsel]])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 25.62   36.02   38.21   38.23   42.93   46.11 

subsel <- which(traj_data$group[sel] == "Ref" & traj_data$NEW_newfacet_x[sel] == "Corrected data\nBray-Curtis\nASV-level")
summary(traj_data$similarity[sel[subsel]])
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 25.62   36.02   38.21   38.23   42.93   46.11 


str(pred_models_boot)
# 'data.frame':	7600 obs. of  10 variables:
#   $ study                  : chr  "South32" "South32" "South32" "South32" ...
# $ NEW_newfacet_x         : Ord.factor w/ 4 levels "Original data\nJaccard\nASV-level"<..: 1 1 1 1 1 1 1 1 1 1 ...
# $ study_and_metric       : chr  "South32__Original data\nJaccard\nASV-level" "South32__Original data\nJaccard\nASV-level" "South32__Original data\nJaccard\nASV-level" "South32__Original data\nJaccard\nASV-level" ...
# $ group                  : Ord.factor w/ 47 levels "0"<"1"<"2"<"3"<..: 3 9 13 15 18 21 24 29 31 32 ...
# $ group_num              : num  2 8 12 14 17 20 23 28 30 31 ...
# $ similarity             : num  15 18.3 19.2 19.6 20.1 ...
# $ boot_no                : int  1 1 1 1 1 1 1 1 1 1 ...
# $ ref_median             : num  25.7 25.7 25.7 25.7 25.7 ...
# $ pred_time_to_ref_median: num  192 192 192 192 192 192 192 192 192 192 ...
# $ pred_type              : chr  "logarithmic" "logarithmic" "logarithmic" "logarithmic" ...

unique(pred_models_boot$study_and_metric)
# [1] "South32__Original data\nJaccard\nASV-level"      "South32__Corrected data\nJaccard\nASV-level"    
# [3] "South32__Original data\nBray-Curtis\nASV-level"  "South32__Corrected data\nBray-Curtis\nASV-level"

pred_models_boot$study_and_metric <- factor(pred_models_boot$study_and_metric,
                                            levels = c(
                                              "South32__Original data\nJaccard\nASV-level"    ,  "South32__Corrected data\nJaccard\nASV-level"  ,  
                                              "South32__Original data\nBray-Curtis\nASV-level" , "South32__Corrected data\nBray-Curtis\nASV-level"
                                              ), ordered = TRUE)

dim(pred_models_boot) # 7600    10

dat.sum <- pred_models_boot
head(dat.sum)

for (i in 1:length(levels(dat.sum$study_and_metric))) {
  #i<-1
  this_study_and_metric <- levels(dat.sum$study_and_metric)[i]
  sel <- which(dat.sum$study_and_metric == this_study_and_metric)
  #unique(dat.sum$ref_median[sel])
  temp <- dat.sum$pred_time_to_ref_median[sel]
  #quantile(temp, probs = c(0.05, 0.25, 0.5, 0.75, 0.95), na.rm = TRUE)
  #hist(temp)
  res <- quantile(temp, probs = c(0.05, 0.5, 0.95), na.rm = TRUE)
  names(res)
  #"5%"  "50%" "95%"
  #print(paste0(this_study_and_metric))
  print(paste0(this_study_and_metric,": ",res["50%"]," (",res["5%"],", ",res["95%"]," yr)"))
}
# [1] "South32__Original data\nJaccard\nASV-level: 119.5 (52.9, 500 yr)"
# [1] "South32__Corrected data\nJaccard\nASV-level: 137.5 (54.9, 500 yr)"
# [1] "South32__Original data\nBray-Curtis\nASV-level: 80 (49.9, 166.1 yr)"
# [1] "South32__Corrected data\nBray-Curtis\nASV-level: 81 (51.95, 163.15 yr)"


levels( pred_models_boot$NEW_newfacet_x )
# [1] "Original data\nJaccard\nASV-level"      "Corrected data\nJaccard\nASV-level"     "Original data\nBray-Curtis\nASV-level" 
# [4] "Corrected data\nBray-Curtis\nASV-level"



## plot time to target

names(pred_models_boot)
# [1] "study"                   "facet_x"                 "study_and_metric"        "group"                   "group_num"              
# [6] "similarity"              "boot_no"                 "ref_median"              "pred_time_to_ref_median" "pred_type"



df <- pred_models_boot
unique( df$NEW_newfacet_x )
# [1] Corrected data\nJaccard\nASV-level     Corrected data\nBray-Curtis\nASV-level
# Levels: Corrected data\nJaccard\nASV-level < Corrected data\nBray-Curtis\nASV-level


# unique( df$study ) # "Alcoa"         "Iluka-Eneabba" "South32"  
# df$study <- factor(df$study, levels = c("Alcoa", "Iluka-Eneabba", "South32"  ),
#                    labels = c("Huntly", "Eneabba", "Worsley"), ordered = TRUE)


df$NEW_newfacet_x <- factor(df$NEW_newfacet_x, 
                            levels = c("Original data\nJaccard\nASV-level",
                                       "Corrected data\nJaccard\nASV-level",
                                       "Original data\nBray-Curtis\nASV-level",
                                       "Corrected data\nBray-Curtis\nASV-level"),
                            labels = c("Filtered data\nJaccard",
                                       "Filtered & corrected\ndata; Jaccard",
                                       "Filtered data\nBray-Curtis",
                                       "Filtered & corrected\ndata; Bray-Curtis"),
                            ordered = TRUE)

study_name <- c(South32="Worsley (filtered)")

p <- ggplot(data=df, aes(x=NEW_newfacet_x, y=pred_time_to_ref_median)) + #, color = NEW_newfacet_x
  geom_boxplot() +
  theme_bw() +
  #guides(color = guide_legend(title="Distance measure")) +
  theme(
    axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    
    legend.key.height = unit(30, "pt")
    #legend.margin = margin(t = 20,r = 0,b = 0,l = 20,"pt")
  ) +
  labs(x = NULL, y = "Predicted time to reach target (yr)") + 
  facet_wrap(facets = vars(study), scales = "free_y", labeller = labeller(study=study_name))
p

ggsave(plot=p, filename = paste0(workdir,"/plots/","Predicted-time-to-target--South32--ASV-Jaccard-Bray-SACorrected-ExclFarSouthSites.tiff"), width = 6.5, height = 9, units = "cm", dpi = 600, compression = "lzw",type="cairo")





#-------------------------



#### Constrained correspondence analysis (CCA) - Alcoa
#-------------------------

# In Ramette 2007: 
# (CCA:) The approach is very similar to that of RDA, except that
# CCA is based on unimodal species-environment relationships
# whereas RDA is based on linear models (ter Braak, 1986).

# Also see Constrained Ordination here:
#http://deneflab.github.io/MicrobeMiseq/demos/mothur_2_phyloseq.html


phy_in <- phy.alcoa.withTree

sample_names(phy_in)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063" "X39065" "X39067"
# [15] "X39069" "X39071" "X39073" "X39075" "X39077" "X39079" "X39081" "X39083" "X39085" "X39087" "X39089" "X39091" "X39093" "X39095"
# [29] "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

## identify columns with all same data?
dat.temp <- as( sample_data(phy_in) , "data.frame" )
uniquevals <- list()
for (i in 1:dim(dat.temp)[2]) {
  #i<-30
  vals <- dat.temp[ ,i]
  uniquevals[[i]] <- length(unique(vals, na.rm = TRUE))
  print(paste0("field ",i," has - ",uniquevals[[i]]," - unique values"))
}
sel <- which(unlist(uniquevals) == 1 | is.na(unlist(uniquevals)) ) # qty 24

names(dat.temp)[sel]
# [1] "Depth"                                                 "Horizon.controlled.vocab..1."                         
# [3] "General.Ecological.Zone.controlled.vocab..3."          "Vegetation.Type..descriptive."                        
# [5] "Vegetation.Type.Controlled.vocab..4."                  "Vegetation.Dom..Trees...."                            
# [7] "Vegetation.Dom..Shrubs...."                            "Vegetation.Dom..Grasses...."                          
# [9] "Elevation..m."                                         "Slope...."                                            
# [11] "Slope.Aspect..Direction.or.degrees..e.g...NW.or.315.." "Profile.Position.controlled.vocab..5."                
# [13] "Method.of.Australian.soil.classification"              "Method.of.FAO.soil.classification"                    
# [15] "Immediate.Previous.Land.Use.controlled.vocab..2."      "X1yr.since.present"                                   
# [17] "X2yrs.since.present"                                   "X3yrs.since.present"                                  
# [19] "X4yrs.since.present"                                   "X5yrs.since.present"                                  
# [21] "Fire"                                                  "Fire.intensity.if.known"                              
# [23] "Flooding"                                              "Soil.moisture...."  

dat.temp[ ,sel]

## remove these fields - they don't offer any explanatory value
dat.temp2 <- dat.temp[ , -sel ]

### join back to phyloseq object!
identical( row.names(phy_in@sam_data), row.names(dat.temp2) ) # TRUE
sample_data( phy_in ) <- dat.temp2

sample_variables( phy_in )
# [1] "FULL_ID"                                             "Sample_id"                                          
# [3] "Date.sampled"                                        "lat.raw"                                            
# [5] "lon.raw"                                             "lat.....in.decimal.degrees"                         
# [7] "long.in.decimal.degrees"                             "Notes"                                              
# [9] "Location.description"                                "Current.land.use.controlled.vocab..2."              
# [11] "Vegetation.Total.cover...."                          "Other.comments"                                     
# [13] "Australian.Soil.Classification.controlled.vocab..6." "FAO.soil.classification.controlled.vocab..7."       
# [15] "Date.since.change.in.Land.Use"                       "Agrochemical.Additions"                             
# [17] "Tillage.controlled.vocab..9."                        "Extreme.Events"                                     
# [19] "Sample.ID"                                           "Color.controlled.vocab..10."                        
# [21] "Gravel..........2.0.mm."                             "Texture..."                                         
# [23] "Course.Sand......200.2000.µm."                       "Fine.Sand........20.200.µm."                        
# [25] "Sand...."                                            "Silt.......2.20.µm."                                
# [27] "Clay.......2.µm."                                    "Ammonium.Nitrogen..mg.Kg."                          
# [29] "Nitrate.Nitrogen..mg.Kg."                            "Phosphorus.Colwell..mg.Kg."                         
# [31] "Potassium.Colwell..mg.Kg."                           "Sulphur..mg.Kg."                                    
# [33] "Organic.Carbon...."                                  "Conductivity..dS.m."                                
# [35] "pH.Level..CaCl2...pH."                               "pH.Level..H2O...pH."                                
# [37] "DTPA.Copper..mg.Kg."                                 "DTPA.Iron..mg.Kg."                                  
# [39] "DTPA.Manganese..mg.Kg."                              "DTPA.Zinc..mg.Kg."                                  
# [41] "Exc..Aluminium..meq.100g."                           "Exc..Calcium..meq.100g."                            
# [43] "Exc..Magnesium..meq.100g."                           "Exc..Potassium..meq.100g."                          
# [45] "Exc..Sodium..meq.100g."                              "Boron.Hot.CaCl2..mg.Kg."   


soil_vars <- c(
  "lat.....in.decimal.degrees"                ,         
  "long.in.decimal.degrees"                    ,
  "Gravel..........2.0.mm."                    ,         "Texture..."                      ,                   
  "Course.Sand......200.2000.µm."             ,          "Fine.Sand........20.200.µm."    ,                    
  "Sand...."                                  ,          "Silt.......2.20.µm."            ,                    
  "Clay.......2.µm."                          ,          "Ammonium.Nitrogen..mg.Kg."      ,                    
  "Nitrate.Nitrogen..mg.Kg."                  ,          "Phosphorus.Colwell..mg.Kg."     ,                    
  "Potassium.Colwell..mg.Kg."                 ,          "Sulphur..mg.Kg."                ,                    
  "Organic.Carbon...."                        ,          "Conductivity..dS.m."            ,                    
  "pH.Level..CaCl2...pH."                     ,          "pH.Level..H2O...pH."            ,                    
  "DTPA.Copper..mg.Kg."                       ,          "DTPA.Iron..mg.Kg."              ,                    
  "DTPA.Manganese..mg.Kg."                    ,          "DTPA.Zinc..mg.Kg."              ,                    
  "Exc..Aluminium..meq.100g."                 ,          "Exc..Calcium..meq.100g."        ,                    
  "Exc..Magnesium..meq.100g."                 ,          "Exc..Potassium..meq.100g."      ,                    
  "Exc..Sodium..meq.100g."                    ,          "Boron.Hot.CaCl2..mg.Kg."   
)



## start with evenly sampled data

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  30751
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30751 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 46 sample variables ]
# tax_table()   Taxonomy Table:    [ 30751 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 30751 tips and 30749 internal nodes ]

r1.ps@sam_data$group <- r1.ps@sam_data$Date.since.change.in.Land.Use
sel <- which(r1.ps@sam_data$Notes == "adjacent natural") # qty 18
r1.ps@sam_data$group[sel] <- "Ref"
unique(r1.ps@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

r1.ps@sam_data$group <- factor( r1.ps@sam_data$group, 
                                levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                labels = c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)



## rescore gravel content classes

unique( r1.ps@sam_data$Gravel..........2.0.mm. )
# "5"     "0"     "60-65" "55-60" "75-80" "70-75" "65-70"
sel <- which(r1.ps@sam_data$Gravel..........2.0.mm. == "60-65" ) # qty 2
r1.ps@sam_data$Gravel..........2.0.mm.[sel] <- "62.5"
sel <- which(r1.ps@sam_data$Gravel..........2.0.mm. == "55-60" ) # qty 4
r1.ps@sam_data$Gravel..........2.0.mm.[sel] <- "57.5"
sel <- which(r1.ps@sam_data$Gravel..........2.0.mm. == "75-80" ) # qty 21
r1.ps@sam_data$Gravel..........2.0.mm.[sel] <- "77.5"
sel <- which(r1.ps@sam_data$Gravel..........2.0.mm. == "70-75" ) # qty 4
r1.ps@sam_data$Gravel..........2.0.mm.[sel] <- "72.5"
sel <- which(r1.ps@sam_data$Gravel..........2.0.mm. == "65-70" ) # qty 1
r1.ps@sam_data$Gravel..........2.0.mm.[sel] <- "67.5"
# change to numeric
class(r1.ps@sam_data$Gravel..........2.0.mm.) # "character"
r1.ps@sam_data$Gravel..........2.0.mm. <- as.numeric(r1.ps@sam_data$Gravel..........2.0.mm.)
summary(r1.ps@sam_data$Gravel..........2.0.mm.)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.00   62.50   77.50   65.42   77.50   77.50


## rescore below detection thresholds to zero
## set values below detection limit to zero

dat.temp <- as( sample_data(r1.ps), "data.frame")
str(dat.temp)

sel <- lapply(dat.temp, function(x) grepl(pattern = "[$<]", x=x))

sel[[1]]

for (i in 1:length(sel)) {
  #i <- 1
  temp1 <- dat.temp[ ,i]
  sel.fix <- which( sel[[i]] == TRUE)
  if (length(sel.fix) >=1 ) { temp1[ sel.fix ] <- "0" }
  dat.temp[ ,i] <- temp1
  print(paste0("completed ",i))
}

dat.temp
str(dat.temp)

## change to numeric data
dat.temp[ , soil_vars] <- lapply(X = dat.temp[ , soil_vars], FUN = as.numeric)
str(dat.temp)
soil_vars
# [1] "lat.....in.decimal.degrees"    "long.in.decimal.degrees"       "Gravel..........2.0.mm."       "Texture..."                   
# [5] "Course.Sand......200.2000.µm." "Fine.Sand........20.200.µm."   "Sand...."                      "Silt.......2.20.µm."          
# [9] "Clay.......2.µm."              "Ammonium.Nitrogen..mg.Kg."     "Nitrate.Nitrogen..mg.Kg."      "Phosphorus.Colwell..mg.Kg."   
# [13] "Potassium.Colwell..mg.Kg."     "Sulphur..mg.Kg."               "Organic.Carbon...."            "Conductivity..dS.m."          
# [17] "pH.Level..CaCl2...pH."         "pH.Level..H2O...pH."           "DTPA.Copper..mg.Kg."           "DTPA.Iron..mg.Kg."            
# [21] "DTPA.Manganese..mg.Kg."        "DTPA.Zinc..mg.Kg."             "Exc..Aluminium..meq.100g."     "Exc..Calcium..meq.100g."      
# [25] "Exc..Magnesium..meq.100g."     "Exc..Potassium..meq.100g."     "Exc..Sodium..meq.100g."        "Boron.Hot.CaCl2..mg.Kg." 

# rename soil vars

soil_vars_new_names <- c(
  "Latitude"    ,"Longitude"     ,  "Gravel"    ,   "Texture"             ,      
  "Course_sand" ,"Fine_sand",   "Sand"                  ,    "Silt"     ,     
  "Clay"              ,"Ammonium_nitrogen"  ,   "Nitrate_nitrogen"  ,    "Phosphorus_Colwell",   
  "Potassium_Colwell"     ,"Sulphur"            ,   "Organic_carbon"        ,    "Conductivity"      ,    
  "pH_CaCl2"         ,"pH_H2O"        ,   "Copper"       ,    "Iron"         ,   
  "Manganese"        ,"Zinc"          ,   "Exc_Aluminium" ,    "Exc_Calcium"  ,    
  "Exc_Magnesium"     ,"Exc_Potassium"  ,   "Exc_Sodium"    ,    "Boron" 
)

length(soil_vars);length(soil_vars_new_names) # 28 28

names(dat.temp[ ,soil_vars])
for (i in 1:length(soil_vars)) {
  #i<-1
  sel.df <- which(names(dat.temp) == soil_vars[i])
  names(dat.temp)[sel.df] <- soil_vars_new_names[i]
  print(paste0("completed ",i))
}
  
names(dat.temp)
# [1] "FULL_ID"                                             "Sample_id"                                           "Date.sampled"                                       
# [4] "lat.raw"                                             "lon.raw"                                             "Latitude"                                           
# [7] "Longitude"                                           "Notes"                                               "Location.description"                               
# [10] "Current.land.use.controlled.vocab..2."               "Vegetation.Total.cover...."                          "Other.comments"                                     
# [13] "Australian.Soil.Classification.controlled.vocab..6." "FAO.soil.classification.controlled.vocab..7."        "Date.since.change.in.Land.Use"                      
# [16] "Agrochemical.Additions"                              "Tillage.controlled.vocab..9."                        "Extreme.Events"                                     
# [19] "Sample.ID"                                           "Color.controlled.vocab..10."                         "Gravel"                                             
# [22] "Texture"                                             "Course_sand"                                         "Fine_sand"                                          
# [25] "Sand"                                                "Silt"                                                "Clay"                                               
# [28] "Ammonium_nitrogen"                                   "Nitrate_nitrogen"                                    "Phosphorus_Colwell"                                 
# [31] "Potassium_Colwell"                                   "Sulphur"                                             "Organic_carbon"                                     
# [34] "Conductivity"                                        "pH_CaCl2"                                            "pH_H2O"                                             
# [37] "Copper"                                              "Iron"                                                "Manganese"                                          
# [40] "Zinc"                                                "Exc_Aluminium"                                       "Exc_Calcium"                                        
# [43] "Exc_Magnesium"                                       "Exc_Potassium"                                       "Exc_Sodium"                                         
# [46] "Boron"                                              

sample_data( r1.ps ) <- dat.temp
r1.ps@sam_data


vars_all <- r1.ps@sam_data[ , soil_vars_new_names ]
str(vars_all)
vars_all <- as(vars_all,"data.frame")
str(vars_all)
class(vars_all)
dim(vars_all) # 36  28
names_vars_all <- names(vars_all)


# scale data
vars_all <- scale(vars_all)
str(vars_all)
class(vars_all) # "matrix" "array" 
vars_all <- as.data.frame(vars_all)
str(vars_all)

otu_all <- r1.ps@otu_table
otu_all <- as.data.frame(otu_all)
str(head(otu_all))
class(otu_all)
dim(otu_all) # 30751    36



## Are there highly correlated predictors?
## Methodology from caret package (see Applied Predictive Modeling text)
## to identify correlated predictors

# Determine correlations between numeric candidate predictors (independent variables)

pairs( vars_all, gap=0 )

correlations <- cor( vars_all  )
correlations
write.csv(x = correlations, file = "alcoa-soil-vars-all-correlations.csv", row.names=TRUE)

# findCorrelation {caret}: This function searches through a correlation matrix and returns a vector of integers corresponding to columns to remove to reduce pair-wise correlations.
highCorr <- findCorrelation(correlations, cutoff = .75)
length(highCorr) # 10

highCorr
#  15 27 28 25 24 26 18  7  5 17

names( vars_all )[highCorr] #
# [1] "Organic_carbon" "Exc_Sodium"     "Boron"          "Exc_Magnesium"  "Exc_Calcium"   
# [6] "Exc_Potassium"  "pH_H2O"         "Sand"           "Course_sand"    "pH_CaCl2" 

# inspect correlations table exported to csv > Excel
# Organic carbon - correlates with: Manganese, Exc_Calcium, Exc_Magnesium, Exc_Sodium, Boron
# Exc_Potassium - correlates with: Potassium_Colwell, Exc_Magnesium, Exc_Sodium, Boron
# pH_H20 - correlates with: pH_CaCl2, and negatively with Exc_Aluminium
# Sand - correlates with: Coarse_sand, and negatively with Silt and Clay

## prepare correlation plot
#M <- cor( cases2[, c(outcome, neworder.names)], use="na.or.complete")
colnames(correlations)
# [1] "Latitude"           "Longitude"          "Gravel"             "Texture"            "Course_sand"       
# [6] "Fine_sand"          "Sand"               "Silt"               "Clay"               "Ammonium_nitrogen" 
# [11] "Nitrate_nitrogen"   "Phosphorus_Colwell" "Potassium_Colwell"  "Sulphur"            "Organic_carbon"    
# [16] "Conductivity"       "pH_CaCl2"           "pH_H2O"             "Copper"             "Iron"              
# [21] "Manganese"          "Zinc"               "Exc_Aluminium"      "Exc_Calcium"        "Exc_Magnesium"     
# [26] "Exc_Potassium"      "Exc_Sodium"         "Boron" 
rownames(correlations)
# same as above

corrplot(corr = correlations, method="ellipse", type="upper") #  method="ellipse"  method="circle"
# saved w550, h700

dev.print(tiff, filename = paste0(workdir,"/plots/", "correlation-plot-soil-vars-Alcoa.tiff"),
          width = 16, height = 16, units = "cm", res=600,
          compression = "lzw", type="cairo")


vars_trim <- vars_all[ ,-highCorr]

names(vars_trim)
# [1] "Latitude"           "Longitude"          "Gravel"             "Texture"           
# [5] "Fine_sand"          "Silt"               "Clay"               "Ammonium_nitrogen" 
# [9] "Nitrate_nitrogen"   "Phosphorus_Colwell" "Potassium_Colwell"  "Sulphur"           
# [13] "Conductivity"       "Copper"             "Iron"               "Manganese"         
# [17] "Zinc"               "Exc_Aluminium" 

# ## choose alternative variables 
# choose_vars <- c(
#   "Gravel_perc_0_10"   ,   "BD_fine_earth_2_7"  ,  "OC_percent"     , "P_Colwell_mgperkg",  "K_Colwell_mgperkg",
#   "Ammonium_N_mgperkg" ,   "Nitrate_N_mgperkg"  ,  "Sulfur_mgperkg" , "EC_dSperm"        ,  "pH_H2O"  ,  "PBI"  
# )
# vars_trim2 <- vars_all[ , choose_vars]

# http://www.sthda.com/english/wiki/scatter-plot-matrices-r-base-graphs

# Correlation panel
panel.cor <- function(x, y){
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- round(cor(x, y), digits=2)
  txt <- paste0("r = ", r)
  cex.cor <- 0.8/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex.cor * r)
}

pairs( vars_trim, gap=0, lower.panel = panel.cor )


 
m0 <- cca( t(otu_all) ~ 1, vars_trim) # X = community matrix, Y = constraining matrix
# ~1 means fit an intercept only

#plot(m0)

fig <- ordiplot(m0, labels=TRUE)         # display = sp (species), = sites (=wa), =cn (centroids)
#identify(fig, "sites", cex=0.8) # in this case "species" means locations


 
m1 <- cca( formula = t(otu_all) ~ ., data = vars_trim )


# formula: Model formula, where the left hand side gives the community data matrix, 
# right hand side gives the constraining variables, and conditioning variables can be given within a special function Condition.
# data: data frame containing the variables on the right hand side of the model formula.


# http://cc.oulu.fi/~jarioksa/opetus/metodi/sessio2.pdf

# m <- step(m0, scope=formula(m1), test="perm", trace = 1)
# m$anova
# #                   Step Df Deviance Resid. Df Resid. Dev      AIC
# # 1                      NA       NA        35    3285810 413.1778
# # 2       + Conductivity -1 222810.5        34    3063000 412.6499
# # 3 + Phosphorus_Colwell -1 167912.7        33    2895087 412.6202
# # 4      + Exc_Aluminium -1 160065.8        32    2735021 412.5727
# # 5       - Conductivity  1 153039.7        33    2888061 412.5328
# 
# m$call
# # cca(formula = t(otu_all) ~ Phosphorus_Colwell + Exc_Aluminium, 
# #     data = vars_trim)

m2 <- ordiR2step(m0, scope=formula(m1))


# ...
# Step: R2.adj= 0.1077354 
# Call: t(otu_all) ~ Conductivity + Phosphorus_Colwell + Exc_Aluminium +      Copper 
# 
# R2.adjusted
# <All variables>       0.1530551
# + Potassium_Colwell   0.1127107
# + Texture             0.1124949
# + Longitude           0.1124459
# + Zinc                0.1123645
# + Gravel              0.1113063
# + Silt                0.1106636
# + Iron                0.1105032
# + Sulphur             0.1100548
# + Manganese           0.1099927
# + Ammonium_nitrogen   0.1095123
# + Clay                0.1090383
# + Latitude            0.1081903
# + Fine_sand           0.1077819
# <none>                0.1077354
# + Nitrate_nitrogen    0.1048447
# 
# Df    AIC      F Pr(>F)  
# + Potassium_Colwell  1 413.31 1.2001  0.072 .
# ---
#   Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

m2$call

# cca(formula = t(otu_all) ~ Conductivity + Phosphorus_Colwell + 
#       Exc_Aluminium + Copper, data = vars_trim)


m2.alcoa <- m2


m2 <- cca( formula = t(otu_all) ~ Conductivity + Phosphorus_Colwell + 
             Exc_Aluminium + Copper, data = vars_trim )



row.names(vars_trim)
# [1] "X39041" "X39043" "X39045" "X39047" "X39049" "X39051" "X39053" "X39055" "X39057" "X39059" "X39061" "X39063" "X39065" "X39067" "X39069" "X39071" "X39073" "X39075"
# [19] "X39077" "X39079" "X39081" "X39083" "X39085" "X39087" "X39089" "X39091" "X39093" "X39095" "X39097" "X39099" "X39161" "X39163" "X39165" "X39191" "X39193" "X39195"

r1.ps@sam_data$group
# [1] 3 yr  Ref   3 yr  Ref   3 yr  Ref   9 yr  Ref   9 yr  Ref   9 yr  Ref   26 yr Ref   30 yr Ref   30 yr Ref   30 yr Ref   26 yr Ref   15 yr Ref   15 yr Ref   18 yr Ref  
# [29] 26 yr Ref   18 yr Ref   15 yr Ref   18 yr Ref  

x <- cbind( as.data.frame(row.names(r1.ps@sam_data)), as.data.frame(r1.ps@sam_data$group))
x
# copy, paste, edit the output below to apply colours to the sample labels



cols.group.alcoa
# 2 yr      8 yr     14 yr     17 yr     25 yr     29 yr       Ref 
# "#9e0142" "#d53e4f" "#fdae61" "#e6f598" "#abdda4" "#66c2a5" "#5e4fa2" 

cols.group <- cols.group.alcoa

cols.ordi <- c(
  
  #row.names(r1.ps@sam_data) r1.ps@sam_data$group
  "X39041"    = "#9e0142",          #   2 yr
  "X39043"    = "#5e4fa2",          #    Ref
  "X39045"    = "#9e0142",          #   2 yr
  "X39047"    = "#5e4fa2",          #    Ref
  "X39049"    = "#9e0142",          #   2 yr
  "X39051"    = "#5e4fa2",          #    Ref
  "X39053"    = "#d53e4f",          #   8 yr
  "X39055"     = "#5e4fa2",         #    Ref
  "X39057"     = "#d53e4f",         #   8 yr
  "X39059"     = "#5e4fa2",         #    Ref
  "X39061"     = "#d53e4f",         #   8 yr
  "X39063"    = "#5e4fa2",          #    Ref
  "X39065"    = "#abdda4",          #  25 yr
  "X39067"     = "#5e4fa2",         #    Ref
  "X39069"    = "#66c2a5",          #  29 yr
  "X39071"    = "#5e4fa2",          #    Ref
  "X39073"    = "#66c2a5",          #  29 yr
  "X39075"    = "#5e4fa2",          #    Ref
  "X39077"    = "#66c2a5",          #  29 yr
  "X39079"    = "#5e4fa2",          #    Ref
  "X39081"    = "#abdda4",          #  25 yr
  "X39083"    = "#5e4fa2",          #    Ref
  "X39085"    = "#fdae61",          #  14 yr
  "X39087"    = "#5e4fa2",          #    Ref
  "X39089"    = "#fdae61",          #  14 yr
  "X39091"    = "#5e4fa2",          #    Ref
  "X39093"    = "#e6f598",          #  17 yr
  "X39095"    = "#5e4fa2",          #    Ref
  "X39097"    = "#abdda4",          #  25 yr
  "X39099"    = "#5e4fa2",          #    Ref
  "X39161"    = "#e6f598",          #  17 yr
  "X39163"    = "#5e4fa2",          #    Ref
  "X39165"    = "#fdae61",          #  14 yr
  "X39191"    = "#5e4fa2",          #    Ref
  "X39193"    = "#e6f598",          #  17 yr
  "X39195"    = "#5e4fa2"          #    Ref
)

#plot(m2)
plot(m2, type = "n")
points(m2, display = "spec", pch = 3, cex = 0.7, col = "grey")
points(x = m2, display = "sites", pch=21, bg = cols.ordi)
points(m2, display = "bp", col = "blue")
text(m2, display = "bp", col = "blue", cex = 0.85)
text(m2, "sites", cex=0.7, pos=1)

legend(x = "bottomleft", legend=c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref")
       ,pt.bg=cols.group,pch=21,cex=0.9,pt.cex=0.9, bty = "n") # bty="o",


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"

dev.print(tiff, filename = paste0(workdir,"/plots/", "CCA-16S-OTUs-and-Soil-vars-Alcoa-grey.tiff"),
          width = 18, height = 18, units = "cm", res=600,
          compression = "lzw", type="cairo")


anova(m2, by = "term", permutations = 999)
# Permutation test for cca under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# Model: cca(formula = t(otu_all) ~ Conductivity + Phosphorus_Colwell + Exc_Aluminium + Copper, data = vars_trim)
# Df ChiSquare      F Pr(>F)    
# Conductivity        1    0.3540 2.6589  0.001 ***
#   Phosphorus_Colwell  1    0.2668 2.0038  0.001 ***
#   Exc_Aluminium       1    0.2543 1.9102  0.002 ** 
#   Copper              1    0.2181 1.6387  0.014 *  
#   Residual           31    4.1269                  
# ---
# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1





# Plot variation in these soil vars by Rehab age:
# Alcoa: 
# Conductivity
# Phosphorus_Colwell
# Exc_Aluminium
# Copper

df <- dat.temp
names(df)

df$group <- df$`Date.since.change.in.Land.Use`
sel <- which(df$Notes == "adjacent natural") # qty 18
df$group[sel] <- "Ref"
unique(df$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"
df$group <- factor( df$group,
                    levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                    labels = c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)

table( df$group )
# 2 yr  8 yr 14 yr 17 yr 25 yr 29 yr   Ref 
#    3     3     3     3     3     3    18 

names(df)
df <- df[ ,c("Sample.ID","group", "Conductivity", "Phosphorus_Colwell", "Exc_Aluminium", "Copper" )]

df.melt <- melt(data = df, id.vars = c("Sample.ID","group"))
levels(df.melt$group)
names(df.melt) # "Sample.ID" "group"     "variable"  "value"

unique(df.melt$variable) # Potassium_Colwell Exc_Aluminium     Conductivity      Copper 
var_names <- c(
  Conductivity = "Conductivity (dS/m)",
  Phosphorus_Colwell = "Phosphorus Colwell (mg/kg)",
  Exc_Aluminium = "Exc. Aluminium (meq/100g)",
  Copper = "DTPA Copper (mg/kg)"
  )

p <- ggplot(df.melt, aes(x=group, y=value )) +
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(size=1.5,width = 0.15, alpha=0.4) +
  theme_bw() +
  facet_wrap(facets = vars(variable), nrow = 1, scales = "free_y", 
             strip.position = "left",
             labeller = labeller(variable = var_names)) +
  theme(
    strip.background = element_blank(),
    strip.placement = "outside",
    axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = NULL)
p

grid.text(label = "(A)", x = unit(0.025, "npc") , y = unit(0.94,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Key-soil-variables-from-CCA-Huntly--A.tiff"), width = 18, height = 7, units = "cm", res=600, compression="lzw",type="cairo")
dev.print(tiff, file = paste0(workdir,"/plots/","Key-soil-variables-from-CCA-Huntly.tiff"), width = 18, height = 7, units = "cm", res=600, compression="lzw",type="cairo")

#-------------------------

#### Constrained correspondence analysis (CCA) - Iluka-Eneabba
#-------------------------

# In Ramette 2007: 
# (CCA:) The approach is very similar to that of RDA, except that
# CCA is based on unimodal species-environment relationships
# whereas RDA is based on linear models (ter Braak, 1986).

# Also see Constrained Ordination here:
#http://deneflab.github.io/MicrobeMiseq/demos/mothur_2_phyloseq.html


phy_in <- phy.iluka_eneabba.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 33636 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 33636 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 33636 tips and 33634 internal nodes ]

sample_names(phy_in)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438" "X138444"
# [14] "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468" "X138470" "X138472"

## identify columns with all same data?
dat.temp <- as( sample_data(phy_in) , "data.frame" )
uniquevals <- list()
for (i in 1:dim(dat.temp)[2]) {
  #i<-30
  vals <- dat.temp[ ,i]
  uniquevals[[i]] <- length(unique(vals, na.rm = TRUE))
  print(paste0("field ",i," has - ",uniquevals[[i]]," - unique values"))
}
sel <- which(unlist(uniquevals) == 1 | is.na(unlist(uniquevals)) ) # qty 102

names(dat.temp)[sel]
# [1] "NCBI_Submission"                                       "NCBI_Sample_Accession"                                 "Organism"                                             
# [4] "Tax_ID"                                                "SampleName_depth"                                      "NCBI_BioProject"                                      
# [7] "Depth..m."                                             "geo_loc..country.subregion."                           "age.of.rehab"                                         
# [10] "NOTES"                                                 "Sample_attribution"                                    "Sample_descriptor"                                    
# [13] "Citation"                                              "Funding_source"                                        "Sample_collection_device.method"                      
# [16] "Sample_material_processing"                            "Sample_storage_method"                                 "DNA_extraction_method"                                
# [19] "Notes..incl..phys.chem.measurement.method.variation."  "Environment..free.living."                             "Env_material..control.vocab.0."                       
# [22] "Horizon..control.vocab.1."                             "General_Env_feature..control.vocab.3."                 "Env_biome..control.vocab.4."                          
# [25] "Vegetation.Total.cover...."                            "Vegetation.Dom..Trees...."                             "Vegetation.Dom..Shrubs...."                           
# [28] "Vegetation.Dom..Grasses...."                           "Slope...."                                             "Slope.Aspect..Direction.or.degrees..e.g...NW.or.315.."
# [31] "Profile.Position..control.vocab.5."                    "Australian.Soil.Classification..control.vocab.6."      "FAO.soil.classification..control.vocab.7."            
# [34] "Immediate.Previous.Land.Use..control.vocab.2."         "Date.since.change.in.Land.Use"                         "Crop.rotation.1yr.since.present"                      
# [37] "Crop.rotation.2yrs.since.present"                      "Crop.rotation.3yrs.since.present"                      "Crop.rotation.4yrs.since.present"                     
# [40] "Crop.rotation.5yrs.since.present"                      "Agrochemical.Additions"                                "Tillage..control.vocab.9."                            
# [43] "Fire"                                                  "fire.intensity.if.known"                               "Flooding"                                             
# [46] "Extreme.Events"                                        "Soil.moisture...."                                     "Water.Holding.capacity"                               
# [49] "bulk.density..g.cm3.or.kg.m3."                         "Microbial.Biomass"                                     "Phosphorus_Colwell..mg.Kg."                           
# [52] "Cation.Exchange.Capacity"                              "Arsenic...90"                                          "Scandium"                                             
# [55] "Vanadium"                                              "Chromium"                                              "Cobalt"                                               
# [58] "Nickel"                                                "Gallium"                                               "Germanium"                                            
# [61] "Arsenic...98"                                          "Selenium"                                              "Rubidium"                                             
# [64] "Strontium"                                             "Yttrium"                                               "Zirconium"                                            
# [67] "Niobium..Columbium."                                   "Molybdenum"                                            "Ruthenium"                                            
# [70] "Rhodium"                                               "Palladium"                                             "Silver"                                               
# [73] "Cadmium"                                               "Tin"                                                   "Antimony"                                             
# [76] "Cesium"                                                "Barium"                                                "Lanthanum"                                            
# [79] "Cerium"                                                "Praseodymium"                                          "Neodymium"                                            
# [82] "Samarium"                                              "Europium"                                              "Gadolinium"                                           
# [85] "Terbium"                                               "Dysprosium"                                            "Holmium"                                              
# [88] "Erbium"                                                "Thulium"                                               "Ytterbium"                                            
# [91] "Lutetium"                                              "Hafnium"                                               "Tantalum"                                             
# [94] "Tungsten.or.Wolfram"                                   "Osmium"                                                "Iridium"                                              
# [97] "Platinum"                                              "Gold"                                                  "Lead"                                                 
# [100] "Thorium"                                               "Uranium"                                               "site"

dat.temp[ ,sel]

## remove these fields - they don't offer any explanatory value
dat.temp2 <- dat.temp[ , -sel ]

### join back to phyloseq object!
identical( row.names(phy_in@sam_data), row.names(dat.temp2) ) # TRUE
sample_data( phy_in ) <- dat.temp2

sample_variables( phy_in )
# [1] "Sample_ID"                           "Date_sampled..YYYY.MM.DD."           "Time_sampled..hh.mm."                "Latitude..decimal.degrees."         
# [5] "Longitude..decimal.degrees."         "Sample_Site..location_description."  "year.of.rehab.or.ref"                "Broad_land_use..control.vocab.2."   
# [9] "Detailed_land_use..control.vocab.2." "Temperature..deg.C."                 "Elevation..m."                       "Color..control.vocab.10."           
# [13] "Gravel.........2.0.mm."              "Texture..."                          "Course.Sand......200.2000.µm."       "Fine.Sand......20.200.µm."          
# [17] "Sand...."                            "Silt......2.20.µm."                  "Clay.......2.µm."                    "Ammonium.Nitrogen..mg.Kg."          
# [21] "Nitrate_Nitrogen..mg.Kg."            "Potassium_Colwell..mg.Kg."           "Sulphur..mg.Kg."                     "Organic_Carbon...."                 
# [25] "Conductivity..dS.m."                 "Ph_Level_CaCl2..pH."                 "pH_Level_H2O..pH."                   "DTPA_Copper..mg.Kg."                
# [29] "DTPA_Iron..mg.Kg."                   "DTPA_Manganese..mg.Kg."              "DTPA_Zinc..mg.Kg."                   "Exc_Aluminium..meq.100g."           
# [33] "Exc_Calcium..meq.100g."              "Exc_Magnesium..meq.100g."            "Exc_Potassium..meq.100g."            "Exc_Sodium..meq.100g."              
# [37] "Boron_Hot_CaCl2..mg.Kg."  


soil_vars <- c(
  
  "Latitude..decimal.degrees."    ,     
  "Longitude..decimal.degrees."   ,           
  "Gravel.........2.0.mm."        ,      "Texture..."                    ,      "Course.Sand......200.2000.µm."   ,    "Fine.Sand......20.200.µm."     ,     
  "Sand...."                      ,      "Silt......2.20.µm."           ,       "Clay.......2.µm."                ,    "Ammonium.Nitrogen..mg.Kg."    ,      
  "Nitrate_Nitrogen..mg.Kg."      ,      "Potassium_Colwell..mg.Kg."    ,       "Sulphur..mg.Kg."                 ,    "Organic_Carbon...."           ,      
  "Conductivity..dS.m."           ,      "Ph_Level_CaCl2..pH."          ,       "pH_Level_H2O..pH."               ,    "DTPA_Copper..mg.Kg."          ,      
  "DTPA_Iron..mg.Kg."             ,      "DTPA_Manganese..mg.Kg."       ,       "DTPA_Zinc..mg.Kg."               ,    "Exc_Aluminium..meq.100g."     ,      
  "Exc_Calcium..meq.100g."        ,      "Exc_Magnesium..meq.100g."     ,       "Exc_Potassium..meq.100g."        ,    "Exc_Sodium..meq.100g."        ,      
  "Boron_Hot_CaCl2..mg.Kg."  
  
)



## start with evenly sampled data

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142

ntaxa(r1.ps) #  27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 27115 taxa and 26 samples ]
# sample_data() Sample Data:       [ 26 samples by 37 sample variables ]
# tax_table()   Taxonomy Table:    [ 27115 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 27115 tips and 27114 internal nodes ]

r1.ps@sam_data$group <- r1.ps@sam_data$year.of.rehab.or.ref
unique(r1.ps@sam_data$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                               labels = c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


r1.ps@sam_data


## rescore gravel content classes

unique( r1.ps@sam_data$Gravel.........2.0.mm.)
#  "0"    "5-10" "5" 
sel <- which(r1.ps@sam_data$Gravel.........2.0.mm. == "5-10" ) # qty 2
r1.ps@sam_data$Gravel.........2.0.mm.[sel] <- "7.5"
# sel <- which(r1.ps@sam_data$Gravel..........2.0.mm. == "55-60" ) # qty 4
# r1.ps@sam_data$Gravel..........2.0.mm.[sel] <- "57.5"
# sel <- which(r1.ps@sam_data$Gravel..........2.0.mm. == "75-80" ) # qty 21
# r1.ps@sam_data$Gravel..........2.0.mm.[sel] <- "77.5"
# sel <- which(r1.ps@sam_data$Gravel..........2.0.mm. == "70-75" ) # qty 4
# r1.ps@sam_data$Gravel..........2.0.mm.[sel] <- "72.5"
# sel <- which(r1.ps@sam_data$Gravel..........2.0.mm. == "65-70" ) # qty 1
# r1.ps@sam_data$Gravel..........2.0.mm.[sel] <- "67.5"

# change to numeric
class(r1.ps@sam_data$Gravel.........2.0.mm.) # "character"
r1.ps@sam_data$Gravel.........2.0.mm. <- as.numeric(r1.ps@sam_data$Gravel.........2.0.mm.)
summary(r1.ps@sam_data$Gravel.........2.0.mm.)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.000   0.000   0.000   2.308   5.000   7.500 


## rescore below detection thresholds to zero
## set values below detection limit to zero

dat.temp <- as( sample_data(r1.ps), "data.frame")
str(dat.temp)

sel <- lapply(dat.temp, function(x) grepl(pattern = "[$<]", x=x))

sel[[1]]

for (i in 1:length(sel)) {
  #i <- 1
  temp1 <- dat.temp[ ,i]
  sel.fix <- which( sel[[i]] == TRUE)
  if (length(sel.fix) >=1 ) { temp1[ sel.fix ] <- "0" }
  dat.temp[ ,i] <- temp1
  print(paste0("completed ",i))
}

dat.temp
str(dat.temp)

## change to numeric data
dat.temp[ , soil_vars] <- lapply(X = dat.temp[ , soil_vars], FUN = as.numeric)
str(dat.temp)
soil_vars
# [1] "Latitude..decimal.degrees."    "Longitude..decimal.degrees."   "Gravel.........2.0.mm."        "Texture..."                    "Course.Sand......200.2000.µm."
# [6] "Fine.Sand......20.200.µm."     "Sand...."                      "Silt......2.20.µm."            "Clay.......2.µm."              "Ammonium.Nitrogen..mg.Kg."    
# [11] "Nitrate_Nitrogen..mg.Kg."      "Potassium_Colwell..mg.Kg."     "Sulphur..mg.Kg."               "Organic_Carbon...."            "Conductivity..dS.m."          
# [16] "Ph_Level_CaCl2..pH."           "pH_Level_H2O..pH."             "DTPA_Copper..mg.Kg."           "DTPA_Iron..mg.Kg."             "DTPA_Manganese..mg.Kg."       
# [21] "DTPA_Zinc..mg.Kg."             "Exc_Aluminium..meq.100g."      "Exc_Calcium..meq.100g."        "Exc_Magnesium..meq.100g."      "Exc_Potassium..meq.100g."     
# [26] "Exc_Sodium..meq.100g."         "Boron_Hot_CaCl2..mg.Kg."      

# rename soil vars

soil_vars_new_names <- c(
  "Latitude"  ,  "Longitude" ,  "Gravel"   ,     "Texture"   ,                 "Course_sand",
  "Fine_sand"  ,   "Sand"     ,                 "Silt"       ,     "Clay"    ,          "Ammonium_nitrogen"    ,
  "Nitrate_nitrogen"  ,    "Potassium_Colwell" ,    "Sulphur"  ,             "Organic_carbon"        ,    "Conductivity"   ,       
  "pH_CaCl2"    ,       "pH_H2O"       ,      "Copper"     ,      "Iron"   ,          "Manganese"     ,  
  "Zinc"         ,    "Exc_Aluminium"  ,    "Exc_Calcium"    ,    "Exc_Magnesium"   ,   "Exc_Potassium"     ,
  "Exc_Sodium"    ,     "Boron"      
  )

length(soil_vars);length(soil_vars_new_names) # 27 27

names(dat.temp[ ,soil_vars])
for (i in 1:length(soil_vars)) {
  #i<-1
  sel.df <- which(names(dat.temp) == soil_vars[i])
  names(dat.temp)[sel.df] <- soil_vars_new_names[i]
  print(paste0("completed ",i))
}

names(dat.temp)
# [1] "Sample_ID"                           "Date_sampled..YYYY.MM.DD."           "Time_sampled..hh.mm."                "Latitude"                           
# [5] "Longitude"                           "Sample_Site..location_description."  "year.of.rehab.or.ref"                "Broad_land_use..control.vocab.2."   
# [9] "Detailed_land_use..control.vocab.2." "Temperature..deg.C."                 "Elevation..m."                       "Color..control.vocab.10."           
# [13] "Gravel"                              "Texture"                             "Course_sand"                         "Fine_sand"                          
# [17] "Sand"                                "Silt"                                "Clay"                                "Ammonium_nitrogen"                  
# [21] "Nitrate_nitrogen"                    "Potassium_Colwell"                   "Sulphur"                             "Organic_carbon"                     
# [25] "Conductivity"                        "pH_CaCl2"                            "pH_H2O"                              "Copper"                             
# [29] "Iron"                                "Manganese"                           "Zinc"                                "Exc_Aluminium"                      
# [33] "Exc_Calcium"                         "Exc_Magnesium"                       "Exc_Potassium"                       "Exc_Sodium"                         
# [37] "Boron"                               "group"                                                     

sample_data( r1.ps ) <- dat.temp
r1.ps@sam_data


vars_all <- r1.ps@sam_data[ , soil_vars_new_names ]
str(vars_all)
vars_all <- as(vars_all,"data.frame")
str(vars_all)
class(vars_all)
dim(vars_all) # 26 27
names_vars_all <- names(vars_all)


# scale data
vars_all <- scale(vars_all)
str(vars_all)
class(vars_all) # "matrix" "array" 
vars_all <- as.data.frame(vars_all)
str(vars_all)

otu_all <- r1.ps@otu_table
otu_all <- as.data.frame(otu_all)
str(head(otu_all))
class(otu_all)
dim(otu_all) # 27115    26



## Are there highly correlated predictors?
## Methodology from caret package (see Applied Predictive Modeling text)
## to identify correlated predictors

# Determine correlations between numeric candidate predictors (independent variables)

pairs( vars_all, gap=0 )

correlations <- cor( vars_all  )
correlations
write.csv(x = correlations, file = "iluka-soil-vars-all-correlations.csv", row.names=TRUE)

# findCorrelation {caret}: This function searches through a correlation matrix and returns a vector of integers corresponding to columns to remove to reduce pair-wise correlations.
highCorr <- findCorrelation(correlations, cutoff = .75)
length(highCorr) # 6

highCorr
#  13  5 24  7 15 16

names( vars_all )[highCorr] #
# "Sulphur"       "Course_sand"   "Exc_Magnesium" "Sand"          "Conductivity"  "pH_CaCl2"     

# inspect correlations table exported to csv > Excel
# Sulphur - correlates with: Conductivity
# Coarse_sand - correlates with: Fine_sand 
# Exc_Magnesium" - correlates with: Sand, 
# "Sand" - correlates with: 
# "Conductivity" - correlates with: Sulphur, Exc_Sodium
# "pH_CaCl2"  - correlates with: pH_H2O

## prepare correlation plot
colnames(correlations)
# [1] "Latitude"          "Longitude"         "Gravel"            "Texture"           "Course_sand"       "Fine_sand"        
# [7] "Sand"              "Silt"              "Clay"              "Ammonium_nitrogen" "Nitrate_nitrogen"  "Potassium_Colwell"
# [13] "Sulphur"           "Organic_carbon"    "Conductivity"      "pH_CaCl2"          "pH_H2O"            "Copper"           
# [19] "Iron"              "Manganese"         "Zinc"              "Exc_Aluminium"     "Exc_Calcium"       "Exc_Magnesium"    
# [25] "Exc_Potassium"     "Exc_Sodium"        "Boron"      
rownames(correlations)
# same as above

corrplot(corr = correlations, method="ellipse", type="upper") #  method="ellipse"  method="circle"

dev.print(tiff, filename = paste0(workdir,"/plots/", "correlation-plot-soil-vars-Iluka.tiff"),
          width = 16, height = 16, units = "cm", res=600,
          compression = "lzw", type="cairo")


vars_trim <- vars_all[ ,-highCorr]

names(vars_trim)
# [1] "Latitude"          "Longitude"         "Gravel"            "Texture"           "Fine_sand"         "Silt"             
# [7] "Clay"              "Ammonium_nitrogen" "Nitrate_nitrogen"  "Potassium_Colwell" "Organic_carbon"    "pH_H2O"           
# [13] "Copper"            "Iron"              "Manganese"         "Zinc"              "Exc_Aluminium"     "Exc_Calcium"      
# [19] "Exc_Potassium"     "Exc_Sodium"        "Boron" 


# Correlation panel
panel.cor <- function(x, y){
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- round(cor(x, y), digits=2)
  txt <- paste0("r = ", r)
  cex.cor <- 0.8/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex.cor * r)
}

pairs( vars_trim, gap=0, lower.panel = panel.cor )

m0 <- cca( t(otu_all) ~ 1, vars_trim) # X = community matrix, Y = constraining matrix
# ~1 means fit an intercept only


m1 <- cca( formula = t(otu_all) ~ ., data = vars_trim )


# formula: Model formula, where the left hand side gives the community data matrix, 
# right hand side gives the constraining variables, and conditioning variables can be given within a special function Condition.
# data: data frame containing the variables on the right hand side of the model formula.


m2 <- ordiR2step(m0, scope=formula(m1))


# Df    AIC      F Pr(>F)  
# + Clay  1 292.89 1.2092  0.044 *
#   ---
#   Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# Step: R2.adj= 0.02252948 
# Call: t(otu_all) ~ Organic_carbon + Clay 
# 
# R2.adjusted
# <All variables>      0.04439102
# + Fine_sand          0.02608125
# + Silt               0.02604402
# + Exc_Aluminium      0.02590911
# + Potassium_Colwell  0.02492148
# + Exc_Potassium      0.02445590
# + Gravel             0.02440427
# + Iron               0.02369755
# + Ammonium_nitrogen  0.02339697
# + Copper             0.02336278
# + Exc_Sodium         0.02320543
# + Latitude           0.02316263
# + Boron              0.02271028
# <none>               0.02252948
# + Longitude          0.02231659
# + Nitrate_nitrogen   0.02123196
# + pH_H2O             0.02038397
# + Manganese          0.01945688
# + Zinc               0.01897353
# + Exc_Calcium        0.01887198
# + Texture            0.01706910
# 
# Df    AIC      F Pr(>F)
# + Fine_sand  1 293.64 1.0806  0.278

m2$call

# cca(formula = t(otu_all) ~ Organic_carbon + Clay, data = vars_trim)


m2 <- cca(formula = t(otu_all) ~ Organic_carbon + Clay, data = vars_trim)



row.names(vars_trim)
# [1] "X138408" "X138410" "X138412" "X138418" "X138424" "X138426" "X138428" "X138430" "X138432" "X138434" "X138436" "X138438" "X138444"
# [14] "X138448" "X138450" "X138452" "X138454" "X138456" "X138458" "X138460" "X138462" "X138464" "X138466" "X138468" "X138470" "X138472"

r1.ps@sam_data$group
# [1] Ref   Ref   Ref   30 yr 24 yr 24 yr 38 yr Ref   Ref   38 yr 38 yr Ref   Ref   7 yr  7 yr  Ref   Ref   10 yr 15 yr 10 yr 15 yr 19 yr
# [23] 19 yr 24 yr 15 yr 30 yr
# Levels: 7 yr < 10 yr < 15 yr < 19 yr < 24 yr < 30 yr < 38 yr < Ref

x <- cbind( as.data.frame(row.names(r1.ps@sam_data)), as.data.frame(r1.ps@sam_data$group))
x
# copy, paste, edit the output below to apply colours to the sample labels

# cols.group <- viridis(n=length(levels( r1.ps@sam_data$group)))
# names(cols.group) <- levels(r1.ps@sam_data$group)
# cols.group
# #       7 yr       10 yr       15 yr       19 yr       24 yr       30 yr       38 yr         Ref 
# # "#440154FF" "#46337EFF" "#365C8DFF" "#277F8EFF" "#1FA187FF" "#4AC16DFF" "#9FDA3AFF" "#FDE725FF" 


cols.group.iluka
#     7 yr     10 yr     15 yr     19 yr     24 yr     30 yr     38 yr       Ref 
# "#d53e4f" "#f46d43" "#fdae61" "#e6f598" "#abdda4" "#66c2a5" "#3288bd" "#5e4fa2"

cols.group <- cols.group.iluka

cols.ordi <- c(
  
  #row.names(r1.ps@sam_data) r1.ps@sam_data$group
  "X138408"      ="#5e4fa2"       ,#     Ref
  "X138410"      ="#5e4fa2"       ,#     Ref
  "X138412"      ="#5e4fa2"       ,#     Ref
  "X138418"      ="#66c2a5"       ,#   30 yr
  "X138424"      ="#abdda4"       ,#   24 yr
  "X138426"      ="#abdda4"       ,#   24 yr
  "X138428"      ="#3288bd"       ,#   38 yr
  "X138430"      ="#5e4fa2"       ,#     Ref
  "X138432"      ="#5e4fa2"       ,#     Ref
  "X138434"      ="#3288bd"       ,#   38 yr
  "X138436"      ="#3288bd"       ,#   38 yr
  "X138438"       ="#5e4fa2"      ,#     Ref
  "X138444"       ="#5e4fa2"      ,#     Ref
  "X138448"       ="#d53e4f"      ,#    7 yr
  "X138450"       ="#d53e4f"      ,#    7 yr
  "X138452"       ="#5e4fa2"      ,#     Ref
  "X138454"        ="#5e4fa2"     ,#     Ref
  "X138456"       ="#f46d43"      ,#   10 yr
  "X138458"       ="#fdae61"      ,#   15 yr
  "X138460"       ="#f46d43"      ,#   10 yr
  "X138462"       ="#fdae61"      ,#   15 yr
  "X138464"       ="#e6f598"      ,#   19 yr
  "X138466"       ="#e6f598"      ,#   19 yr
  "X138468"       ="#abdda4"      ,#   24 yr
  "X138470"       ="#fdae61"      ,#   15 yr
  "X138472"       ="#66c2a5"      #   30 yr
  
)

# plot(m2)
# points(x = m2, display = "sites", pch=21, bg = cols.ordi)
# text(m2, "sites", cex=0.7, pos=1)
# 
# legend(x = "topleft", legend=c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref")
#        ,pt.bg=cols.group,pch=21,bty="o",cex=0.8,pt.cex=0.8)


plot(m2, type = "n")
points(m2, display = "spec", pch = 3, cex = 0.7, col = "grey")
points(x = m2, display = "sites", pch=21, bg = cols.ordi)
points(m2, display = "bp", col = "blue")
text(m2, display = "bp", col = "blue", cex = 0.85)
text(m2, "sites", cex=0.7, pos=1)

legend(x = "topright", legend=c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref")
       ,pt.bg=cols.group,pch=21,cex=0.9,pt.cex=0.9, bty = "n") # bty="o",




getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"

dev.print(tiff, filename = paste0(workdir,"/plots/", "CCA-16S-OTUs-and-Soil-vars-Iluka-grey.tiff"),
          width = 18, height = 18, units = "cm", res=600,
          compression = "lzw", type="cairo")


anova(m2, by = "term", permutations = 999)
# Permutation test for cca under reduced model
# Terms added sequentially (first to last)
# Permutation: free
# Number of permutations: 999
# 
# Model: cca(formula = t(otu_all) ~ Organic_carbon + Clay, data = vars_trim)
# Df ChiSquare      F Pr(>F)    
# Organic_carbon  1    0.3637 1.3695  0.001 ***
#   Clay            1    0.3212 1.2092  0.020 *  
#   Residual       23    6.1090                  
# ---
#   Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


# Eneabba:
# organic carbon
# clay content 
# 


df <- dat.temp
names(df)

df$group <- df$year.of.rehab.or.ref
unique(df$group)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000" 
df$group <- factor(df$group, 
                                levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                                labels = c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref"),
                                ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


table( df$group )
#7 yr 10 yr 15 yr 19 yr 24 yr 30 yr 38 yr   Ref 
#   2     2     3     2     3     2     3     9 


names(df)
df <- df[ ,c("Sample_ID","group", "Organic_carbon", "Clay" )]

df.melt <- melt(data = df, id.vars = c("Sample_ID","group"))
levels(df.melt$group) # "7 yr"  "10 yr" "15 yr" "19 yr" "24 yr" "30 yr" "38 yr" "Ref"  

names(df.melt) # "Sample_ID" "group"     "variable"  "value"

unique(df.melt$variable) # Organic_carbon Clay

# var_names <- c(
#   Organic_carbon = "Organic carbon (g C/100g soil)",
#   Clay = "Clay content (% <2 um)"
# )

var_names <- c(
  Organic_carbon = "Organic carbon (mass %)",
  Clay = "Clay content (% <2 \u03BCm)")


p <- ggplot(df.melt, aes(x=group, y=value )) +
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(size=1.5,width = 0.15, alpha=0.4) +
  theme_bw() +
  facet_wrap(facets = vars(variable), nrow = 1, scales = "free_y", 
             strip.position = "left",
             labeller = labeller(variable = var_names)) +
  theme(
    strip.background = element_blank(),
    strip.placement = "outside",
    axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = NULL)
p

grid.text(label = "(B)", x = unit(0.03, "npc") , y = unit(0.94,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Key-soil-variables-from-CCA-Eneabba-B.tiff"), width = 10, height = 7, units = "cm", res=600, compression="lzw",type="cairo")
dev.print(tiff, file = paste0(workdir,"/plots/","Key-soil-variables-from-CCA-Eneabba.tiff"), width = 10, height = 7, units = "cm", res=600, compression="lzw",type="cairo")



#-------------------------


#### Constrained correspondence analysis (CCA) - South32
#-------------------------

# In Ramette 2007: 
# (CCA:) The approach is very similar to that of RDA, except that
# CCA is based on unimodal species-environment relationships
# whereas RDA is based on linear models (ter Braak, 1986).

# Also see Constrained Ordination here:
#http://deneflab.github.io/MicrobeMiseq/demos/mothur_2_phyloseq.html


phy_in <- phy.south32.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54671 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 139 sample variables ]
# tax_table()   Taxonomy Table:    [ 54671 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54671 tips and 54669 internal nodes ]

sample_names(phy_in)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374" "X138376" "X138378" "X138380" "X138382"
# [14] "X138384" "X138386" "X138388" "X138390" "X138392" "X138394" "X138396" "X138398" "X138400" "X138402" "X138404" "X138406"

## identify columns with all same data?
dat.temp <- as( sample_data(phy_in) , "data.frame" )
uniquevals <- list()
for (i in 1:dim(dat.temp)[2]) {
  #i<-30
  vals <- dat.temp[ ,i]
  uniquevals[[i]] <- length(unique(vals, na.rm = TRUE))
  print(paste0("field ",i," has - ",uniquevals[[i]]," - unique values"))
}
sel <- which(unlist(uniquevals) == 1 | is.na(unlist(uniquevals)) ) # qty 100

names(dat.temp)[sel]
# [1] "NCBI_Submission"                                       "NCBI_Sample_Accession"                                 "Organism"                                             
# [4] "Tax_ID"                                                "SampleName_depth"                                      "NCBI_BioProject"                                      
# [7] "Depth..m."                                             "geo_loc..country.subregion."                           "age.of.rehab"                                         
# [10] "Sample_attribution"                                    "Sample_descriptor"                                     "Citation"                                             
# [13] "Funding_source"                                        "Sample_collection_device.method"                       "Sample_material_processing"                           
# [16] "Sample_storage_method"                                 "DNA_extraction_method"                                 "Notes..incl..phys.chem.measurement.method.variation." 
# [19] "Environment..free.living."                             "Env_material..control.vocab.0."                        "Horizon..control.vocab.1."                            
# [22] "General_Env_feature..control.vocab.3."                 "Env_biome..control.vocab.4."                           "Vegetation.Total.cover...."                           
# [25] "Vegetation.Dom..Trees...."                             "Vegetation.Dom..Shrubs...."                            "Vegetation.Dom..Grasses...."                          
# [28] "Slope...."                                             "Slope.Aspect..Direction.or.degrees..e.g...NW.or.315.." "Profile.Position..control.vocab.5."                   
# [31] "Australian.Soil.Classification..control.vocab.6."      "FAO.soil.classification..control.vocab.7."             "Immediate.Previous.Land.Use..control.vocab.2."        
# [34] "Date.since.change.in.Land.Use"                         "Crop.rotation.1yr.since.present"                       "Crop.rotation.2yrs.since.present"                     
# [37] "Crop.rotation.3yrs.since.present"                      "Crop.rotation.4yrs.since.present"                      "Crop.rotation.5yrs.since.present"                     
# [40] "Agrochemical.Additions"                                "Tillage..control.vocab.9."                             "Fire"                                                 
# [43] "fire.intensity.if.known"                               "Flooding"                                              "Extreme.Events"                                       
# [46] "Soil.moisture...."                                     "Water.Holding.capacity"                                "bulk.density..g.cm3.or.kg.m3."                        
# [49] "Microbial.Biomass"                                     "Cation.Exchange.Capacity"                              "Arsenic...90"                                         
# [52] "Scandium"                                              "Vanadium"                                              "Chromium"                                             
# [55] "Cobalt"                                                "Nickel"                                                "Gallium"                                              
# [58] "Germanium"                                             "Arsenic...98"                                          "Selenium"                                             
# [61] "Rubidium"                                              "Strontium"                                             "Yttrium"                                              
# [64] "Zirconium"                                             "Niobium..Columbium."                                   "Molybdenum"                                           
# [67] "Ruthenium"                                             "Rhodium"                                               "Palladium"                                            
# [70] "Silver"                                                "Cadmium"                                               "Tin"                                                  
# [73] "Antimony"                                              "Cesium"                                                "Barium"                                               
# [76] "Lanthanum"                                             "Cerium"                                                "Praseodymium"                                         
# [79] "Neodymium"                                             "Samarium"                                              "Europium"                                             
# [82] "Gadolinium"                                            "Terbium"                                               "Dysprosium"                                           
# [85] "Holmium"                                               "Erbium"                                                "Thulium"                                              
# [88] "Ytterbium"                                             "Lutetium"                                              "Hafnium"                                              
# [91] "Tantalum"                                              "Tungsten.or.Wolfram"                                   "Osmium"                                               
# [94] "Iridium"                                               "Platinum"                                              "Gold"                                                 
# [97] "Lead"                                                  "Thorium"                                               "Uranium"                                              
# [100] "site" 

dat.temp[ ,sel]

## remove these fields - they don't offer any explanatory value
dat.temp2 <- dat.temp[ , -sel ]

### join back to phyloseq object!
identical( row.names(phy_in@sam_data), row.names(dat.temp2) ) # TRUE
sample_data( phy_in ) <- dat.temp2

sample_variables( phy_in )
# [1] "Sample_ID"                           "Date_sampled..YYYY.MM.DD."           "Time_sampled..hh.mm."                "Latitude..decimal.degrees."         
# [5] "Longitude..decimal.degrees."         "Sample_Site..location_description."  "year.of.rehab.or.ref"                "NOTES"                              
# [9] "Broad_land_use..control.vocab.2."    "Detailed_land_use..control.vocab.2." "Temperature..deg.C."                 "Elevation..m."                      
# [13] "Color..control.vocab.10."            "Gravel.........2.0.mm."              "Texture..."                          "Course.Sand......200.2000.µm."      
# [17] "Fine.Sand......20.200.µm."           "Sand...."                            "Silt......2.20.µm."                  "Clay.......2.µm."                   
# [21] "Ammonium.Nitrogen..mg.Kg."           "Nitrate_Nitrogen..mg.Kg."            "Phosphorus_Colwell..mg.Kg."          "Potassium_Colwell..mg.Kg."          
# [25] "Sulphur..mg.Kg."                     "Organic_Carbon...."                  "Conductivity..dS.m."                 "Ph_Level_CaCl2..pH."                
# [29] "pH_Level_H2O..pH."                   "DTPA_Copper..mg.Kg."                 "DTPA_Iron..mg.Kg."                   "DTPA_Manganese..mg.Kg."             
# [33] "DTPA_Zinc..mg.Kg."                   "Exc_Aluminium..meq.100g."            "Exc_Calcium..meq.100g."              "Exc_Magnesium..meq.100g."           
# [37] "Exc_Potassium..meq.100g."            "Exc_Sodium..meq.100g."               "Boron_Hot_CaCl2..mg.Kg."     


soil_vars <- c(
  
  "Latitude..decimal.degrees."     ,    
  "Longitude..decimal.degrees."   ,    
  "Gravel.........2.0.mm."        ,      "Texture..."                  ,        "Course.Sand......200.2000.µm."  ,    
  "Fine.Sand......20.200.µm."     ,      "Sand...."                   ,         "Silt......2.20.µm."            ,      "Clay.......2.µm."                ,   
  "Ammonium.Nitrogen..mg.Kg."     ,      "Nitrate_Nitrogen..mg.Kg."   ,         "Phosphorus_Colwell..mg.Kg."    ,      "Potassium_Colwell..mg.Kg."      ,    
  "Sulphur..mg.Kg."               ,      "Organic_Carbon...."         ,         "Conductivity..dS.m."           ,      "Ph_Level_CaCl2..pH."            ,    
  "pH_Level_H2O..pH."             ,      "DTPA_Copper..mg.Kg."        ,         "DTPA_Iron..mg.Kg."             ,      "DTPA_Manganese..mg.Kg."         ,    
  "DTPA_Zinc..mg.Kg."             ,      "Exc_Aluminium..meq.100g."   ,         "Exc_Calcium..meq.100g."        ,      "Exc_Magnesium..meq.100g."       ,    
  "Exc_Potassium..meq.100g."       ,     "Exc_Sodium..meq.100g."      ,         "Boron_Hot_CaCl2..mg.Kg."     
  
)



## start with evenly sampled data

# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_in, sample.size = min(sample_sums(phy_in)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122

ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 54327 taxa and 25 samples ]
# sample_data() Sample Data:       [ 25 samples by 39 sample variables ]
# tax_table()   Taxonomy Table:    [ 54327 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 54327 tips and 54325 internal nodes ]

r1.ps@sam_data$group <- r1.ps@sam_data$year.of.rehab.or.ref
unique(r1.ps@sam_data$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"

r1.ps@sam_data$group <- factor(r1.ps@sam_data$group,
                               levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                               labels = c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


r1.ps@sam_data


## rescore gravel content classes

unique( r1.ps@sam_data$Gravel.........2.0.mm.)
#  "45-50" "55-60" "75-80"
sel <- which(r1.ps@sam_data$Gravel.........2.0.mm. == "45-50" ) # qty 5
r1.ps@sam_data$Gravel.........2.0.mm.[sel] <- "47.5"
sel <- which(r1.ps@sam_data$Gravel.........2.0.mm. == "55-60" ) # qty 11
r1.ps@sam_data$Gravel.........2.0.mm.[sel] <- "57.5"
sel <- which(r1.ps@sam_data$Gravel.........2.0.mm. == "75-80" ) # qty 9
r1.ps@sam_data$Gravel.........2.0.mm.[sel] <- "77.5"


# change to numeric
class(r1.ps@sam_data$Gravel.........2.0.mm.) # "character"
r1.ps@sam_data$Gravel.........2.0.mm. <- as.numeric(r1.ps@sam_data$Gravel.........2.0.mm.)
summary(r1.ps@sam_data$Gravel.........2.0.mm.)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 47.5    57.5    57.5    62.7    77.5    77.5 


## rescore below detection thresholds to zero
## set values below detection limit to zero

dat.temp <- as( sample_data(r1.ps), "data.frame")
str(dat.temp)

sel <- lapply(dat.temp, function(x) grepl(pattern = "[$<]", x=x))

sel[[1]]

for (i in 1:length(sel)) {
  #i <- 1
  temp1 <- dat.temp[ ,i]
  sel.fix <- which( sel[[i]] == TRUE)
  if (length(sel.fix) >=1 ) { temp1[ sel.fix ] <- "0" }
  dat.temp[ ,i] <- temp1
  print(paste0("completed ",i))
}

dat.temp
str(dat.temp)

## change to numeric data
dat.temp[ , soil_vars] <- lapply(X = dat.temp[ , soil_vars], FUN = as.numeric)
str(dat.temp)
soil_vars
# [1] "Latitude..decimal.degrees."    "Longitude..decimal.degrees."   "Gravel.........2.0.mm."        "Texture..."                    "Course.Sand......200.2000.µm."
# [6] "Fine.Sand......20.200.µm."     "Sand...."                      "Silt......2.20.µm."            "Clay.......2.µm."              "Ammonium.Nitrogen..mg.Kg."    
# [11] "Nitrate_Nitrogen..mg.Kg."      "Phosphorus_Colwell..mg.Kg."    "Potassium_Colwell..mg.Kg."     "Sulphur..mg.Kg."               "Organic_Carbon...."           
# [16] "Conductivity..dS.m."           "Ph_Level_CaCl2..pH."           "pH_Level_H2O..pH."             "DTPA_Copper..mg.Kg."           "DTPA_Iron..mg.Kg."            
# [21] "DTPA_Manganese..mg.Kg."        "DTPA_Zinc..mg.Kg."             "Exc_Aluminium..meq.100g."      "Exc_Calcium..meq.100g."        "Exc_Magnesium..meq.100g."     
# [26] "Exc_Potassium..meq.100g."      "Exc_Sodium..meq.100g."         "Boron_Hot_CaCl2..mg.Kg."      

# rename soil vars

soil_vars_new_names <- c(
  
  "Latitude"  ,  "Longitude" ,  "Gravel"   ,     "Texture"      ,              "Course_sand",
  "Fine_sand"   ,  "Sand"    ,                  "Silt"     ,       "Clay"     ,         "Ammonium_nitrogen"    ,
  "Nitrate_nitrogen"  ,    "Phosphorus_Colwell"   , "Potassium_Colwell"   ,  "Sulphur"      ,         "Organic_carbon"     ,      
  "Conductivity"    ,       "pH_CaCl2"    ,       "pH_H2O"     ,        "Copper"      ,     "Iron"     ,       
  "Manganese"     ,   "Zinc"     ,        "Exc_Aluminium"    ,  "Exc_Calcium"     ,   "Exc_Magnesium"    , 
  "Exc_Potassium"  ,    "Exc_Sodium"   ,      "Boron"      
  
)

length(soil_vars);length(soil_vars_new_names) # 28 28

names(dat.temp[ ,soil_vars])
for (i in 1:length(soil_vars)) {
  #i<-1
  sel.df <- which(names(dat.temp) == soil_vars[i])
  names(dat.temp)[sel.df] <- soil_vars_new_names[i]
  print(paste0("completed ",i))
}

names(dat.temp)
# [1] "Sample_ID"                           "Date_sampled..YYYY.MM.DD."           "Time_sampled..hh.mm."                "Latitude"                           
# [5] "Longitude"                           "Sample_Site..location_description."  "year.of.rehab.or.ref"                "NOTES"                              
# [9] "Broad_land_use..control.vocab.2."    "Detailed_land_use..control.vocab.2." "Temperature..deg.C."                 "Elevation..m."                      
# [13] "Color..control.vocab.10."            "Gravel"                              "Texture"                             "Course_sand"                        
# [17] "Fine_sand"                           "Sand"                                "Silt"                                "Clay"                               
# [21] "Ammonium_nitrogen"                   "Nitrate_nitrogen"                    "Phosphorus_Colwell"                  "Potassium_Colwell"                  
# [25] "Sulphur"                             "Organic_carbon"                      "Conductivity"                        "pH_CaCl2"                           
# [29] "pH_H2O"                              "Copper"                              "Iron"                                "Manganese"                          
# [33] "Zinc"                                "Exc_Aluminium"                       "Exc_Calcium"                         "Exc_Magnesium"                      
# [37] "Exc_Potassium"                       "Exc_Sodium"                          "Boron"                               "group"      

sample_data( r1.ps ) <- dat.temp
r1.ps@sam_data


vars_all <- r1.ps@sam_data[ , soil_vars_new_names ]
str(vars_all)
vars_all <- as(vars_all,"data.frame")
str(vars_all)
class(vars_all)
dim(vars_all) # 25  28
names_vars_all <- names(vars_all)


# scale data
vars_all <- scale(vars_all)
str(vars_all)
class(vars_all) # "matrix" "array" 
vars_all <- as.data.frame(vars_all)
str(vars_all)

otu_all <- r1.ps@otu_table
otu_all <- as.data.frame(otu_all)
str(head(otu_all))
class(otu_all)
dim(otu_all) # 54327    25



## Are there highly correlated predictors?
## Methodology from caret package (see Applied Predictive Modeling text)
## to identify correlated predictors

# Determine correlations between numeric candidate predictors (independent variables)

pairs( vars_all, gap=0 )

correlations <- cor( vars_all  )
correlations
write.csv(x = correlations, file = "south32-soil-vars-all-correlations.csv", row.names=TRUE)

# findCorrelation {caret}: This function searches through a correlation matrix and returns a vector of integers corresponding to columns to remove to reduce pair-wise correlations.
highCorr <- findCorrelation(correlations, cutoff = .75)
length(highCorr) # 13

highCorr
#  10 28 16 15 26 27 13 12 25 24  7  5 18

names( vars_all )[highCorr] #
# [1] "Ammonium_nitrogen"  "Boron"              "Conductivity"       "Organic_carbon"     "Exc_Potassium"      "Exc_Sodium"         "Potassium_Colwell"  "Phosphorus_Colwell"
# [9] "Exc_Magnesium"      "Exc_Calcium"        "Sand"               "Course_sand"        "pH_H2O"  

# inspect correlations table exported to csv > Excel

# - correlates with: 

# "Ammonium_nitrogen" - correlates with: Potassium_Colwell, Organic_carbon, Conductivity, Manganese, Exc_Magnesium, Exc_Potassium, Exc_Sodium, Boron
# "Boron" - correlates with: as above
# "Conductivity" - correlates with: as above
# "Organic_carbon"   - correlates with: as above
# "Exc_Potassium"  - correlates with: Phosphorous_Colwell, Ecx_Calcium, and as above
# "Exc_Sodium"   - correlates with: as above
# "Potassium_Colwell" - correlates with: as above
# "Phosphorus_Colwell" - correlates with: Potassium_Colwell, Organic_carbon, Exc_Calcium
# "Exc_Magnesium"  - correlates with: 
# "Exc_Calcium"  - correlates with: 
# "Sand"   - correlates with: Coarse_sand and negatively with Clay
# "Course_sand"  - correlates with: Sand and negatively with Fine_sand
# "pH_H2O"   - correlates with: pH_CaCl2



## prepare correlation plot
colnames(correlations)
# [1] "Latitude"           "Longitude"          "Gravel"             "Texture"            "Course_sand"        "Fine_sand"          "Sand"               "Silt"              
# [9] "Clay"               "Ammonium_nitrogen"  "Nitrate_nitrogen"   "Phosphorus_Colwell" "Potassium_Colwell"  "Sulphur"            "Organic_carbon"     "Conductivity"      
# [17] "pH_CaCl2"           "pH_H2O"             "Copper"             "Iron"               "Manganese"          "Zinc"               "Exc_Aluminium"      "Exc_Calcium"       
# [25] "Exc_Magnesium"      "Exc_Potassium"      "Exc_Sodium"         "Boron"     
rownames(correlations)
# same as above

corrplot(corr = correlations, method="ellipse", type="upper") #  method="ellipse"  method="circle"

dev.print(tiff, filename = paste0(workdir,"/plots/", "correlation-plot-soil-vars-South32.tiff"),
          width = 16, height = 16, units = "cm", res=600,
          compression = "lzw", type="cairo")


vars_trim <- vars_all[ ,-highCorr]

names(vars_trim)
# [1] "Latitude"         "Longitude"        "Gravel"           "Texture"          "Fine_sand"        "Silt"             "Clay"            
# [8] "Nitrate_nitrogen" "Sulphur"          "pH_CaCl2"         "Copper"           "Iron"             "Manganese"        "Zinc"            
# [15] "Exc_Aluminium"   


# Correlation panel
panel.cor <- function(x, y){
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- round(cor(x, y), digits=2)
  txt <- paste0("r = ", r)
  cex.cor <- 0.8/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex.cor * r)
}

pairs( vars_trim, gap=0, lower.panel = panel.cor )


 
m0 <- cca( t(otu_all) ~ 1, vars_trim) # X = community matrix, Y = constraining matrix
# ~1 means fit an intercept only

#plot(m0)

fig <- ordiplot(m0, labels=TRUE)         # display = sp (species), = sites (=wa), =cn (centroids)
#identify(fig, "sites", cex=0.8) # in this case "species" means locations


 
m1 <- cca( formula = t(otu_all) ~ ., data = vars_trim )


# formula: Model formula, where the left hand side gives the community data matrix, 
# right hand side gives the constraining variables, and conditioning variables can be given within a special function Condition.
# data: data frame containing the variables on the right hand side of the model formula.



m2 <- ordiR2step(m0, scope=formula(m1))


# Step: R2.adj= 0.07069458 
# Call: t(otu_all) ~ Exc_Aluminium + Iron + Latitude 
# 
# R2.adjusted
# <All variables>     0.08512347
# + Texture           0.08018096
# + Nitrate_nitrogen  0.08012615
# + pH_CaCl2          0.07673739
# + Zinc              0.07514439
# + Gravel            0.07388113
# + Longitude         0.07305399
# + Fine_sand         0.07295308
# + Clay              0.07258102
# + Silt              0.07225431
# + Manganese         0.07151621
# + Sulphur           0.07089795
# <none>              0.07069458
# + Copper            0.06615371
# 
# Df    AIC      F Pr(>F)  
# + Texture  1 311.83 1.2057  0.068 .
# ---
#   Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

m2$call

# cca(formula = t(otu_all) ~ Exc_Aluminium + Iron + Latitude, data = vars_trim)

m2 <- cca(formula = t(otu_all) ~ Exc_Aluminium + Iron + Latitude, data = vars_trim)

row.names(vars_trim)
# [1] "X138358" "X138360" "X138362" "X138364" "X138366" "X138368" "X138370" "X138372" "X138374" "X138376" "X138378" "X138380" "X138382" "X138384"
# [15] "X138386" "X138388" "X138390" "X138392" "X138394" "X138396" "X138398" "X138400" "X138402" "X138404" "X138406"

r1.ps@sam_data$group
# [1] 23 yr 23 yr Ref   2 yr  Ref   17 yr 12 yr 23 yr Ref   Ref   Ref   20 yr 8 yr  8 yr  14 yr 20 yr 23 yr 17 yr 8 yr  14 yr 28 yr 28 yr 2 yr  Ref  
# [25] 2 yr 
# Levels: 2 yr < 8 yr < 12 yr < 14 yr < 17 yr < 20 yr < 23 yr < 28 yr < Ref

x <- cbind( as.data.frame(row.names(r1.ps@sam_data)), as.data.frame(r1.ps@sam_data$group))
x
# copy, paste, edit the output below to apply colours to the sample labels

# cols.group <- viridis(n=length(levels( r1.ps@sam_data$group)))
# names(cols.group) <- levels(r1.ps@sam_data$group)
# cols.group
# #        2 yr        8 yr       12 yr       14 yr       17 yr       20 yr       23 yr       28 yr         Ref 
# # "#440154FF" "#472D7BFF" "#3B528BFF" "#2C728EFF" "#21908CFF" "#27AD81FF" "#5DC863FF" "#AADC32FF" "#FDE725FF" 


cols.group.south32
#     2 yr      8 yr     12 yr     14 yr     17 yr     20 yr     23 yr     28 yr       Ref 
# "#9e0142" "#d53e4f" "#f46d43" "#fdae61" "#fee08b" "#e6f598" "#abdda4" "#66c2a5" "#5e4fa2" 

cols.group <- cols.group.south32

cols.ordi <- c(
  
  #   row.names(r1.ps@sam_data) r1.ps@sam_data$group
  "X138358"     ="#abdda4"      ,#     23 yr
  "X138360"     ="#abdda4"       ,#    23 yr
  "X138362"     ="#5e4fa2"       ,#      Ref
  "X138364"     ="#9e0142"       ,#     2 yr
  "X138366"     ="#5e4fa2"       ,#      Ref
  "X138368"     ="#fee08b"       ,#    17 yr
  "X138370"     ="#f46d43"       ,#    12 yr
  "X138372"     ="#abdda4"       ,#    23 yr
  "X138374"       ="#5e4fa2"     ,#      Ref
  "X138376"       ="#5e4fa2"     ,#      Ref
  "X138378"       ="#5e4fa2"     ,#      Ref
  "X138380"       ="#e6f598"     ,#    20 yr
  "X138382"       ="#d53e4f"     ,#     8 yr
  "X138384"       ="#d53e4f"     ,#     8 yr
  "X138386"       ="#fdae61"     ,#    14 yr
  "X138388"       ="#e6f598"     ,#    20 yr
  "X138390"       ="#abdda4"     ,#    23 yr
  "X138392"        ="#fee08b"    ,#    17 yr
  "X138394"        ="#d53e4f"    ,#     8 yr
  "X138396"        ="#fdae61"    ,#    14 yr
  "X138398"       ="#66c2a5"     ,#    28 yr
  "X138400"       ="#66c2a5"     ,#    28 yr
  "X138402"        ="#9e0142"    ,#     2 yr
  "X138404"        ="#5e4fa2"    ,#      Ref
  "X138406"        ="#9e0142"    #     2 yr
  
)

# plot(m2)
# points(x = m2, display = "sites", pch=21, bg = cols.ordi)
# text(m2, "sites", cex=0.7, pos=1)
# 
# legend(x = "bottomleft", legend=c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref")
#        ,pt.bg=cols.group,pch=21,bty="o",cex=0.8,pt.cex=0.8)


plot(m2, type = "n")
points(m2, display = "spec", pch = 3, cex = 0.7, col = "grey")
points(x = m2, display = "sites", pch=21, bg = cols.ordi)
points(m2, display = "bp", col = "blue")
text(m2, display = "bp", col = "blue", cex = 0.85)
text(m2, "sites", cex=0.7, pos=1)

legend(x = "bottomleft", legend=c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref")
       ,pt.bg=cols.group,pch=21,cex=0.9,pt.cex=0.9, bty="n") # ,bty="o"


getwd() # "/Users/lidd0026/WORKSPACE/PROJ/Restoration-Trajectories/modelling"

dev.print(tiff, filename = paste0(workdir,"/plots/", "CCA-16S-OTUs-and-Soil-vars-South32-grey.tiff"),
          width = 24, height = 21, units = "cm", res=600,
          compression = "lzw", type="cairo")


anova(m2, by = "term", permutations = 999)



# South32:
# exchangeable aluminium
# iron (DTPA extraction
# latitude


df <- dat.temp
names(df)

df$group <- df$year.of.rehab.or.ref
unique(df$group)
# "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"
df$group <- factor(df$group, 
                   levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                   labels = c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)

table( df$group )
# 2 yr  8 yr 12 yr 14 yr 17 yr 20 yr 23 yr 28 yr   Ref 
#    3     3     1     2     2     2     4     2     6 


names(df)
# exchangeable aluminium
# iron (DTPA extraction
# latitude

df <- df[ ,c("Sample_ID","group", "Exc_Aluminium", "Iron", "Latitude" )]

df.melt <- melt(data = df, id.vars = c("Sample_ID","group"))
levels(df.melt$group) # "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr" "Ref"   

names(df.melt) # "Sample_ID" "group"     "variable"  "value"

unique(df.melt$variable) # Exc_Aluminium Iron          Latitude     



var_names <- c(
  Exc_Aluminium = "Exc. Aluminium (meq/100g)",
  Iron = "DTPA Iron (mg/kg)",
  Latitude = "Latitude (decimal degrees, S)"
)

p <- ggplot(df.melt, aes(x=group, y=value )) +
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(size=1.5,width = 0.15, alpha=0.4) +
  theme_bw() +
  facet_wrap(facets = vars(variable), nrow = 1, scales = "free_y", 
             strip.position = "left",
             labeller = labeller(variable = var_names)) +
  theme(
    strip.background = element_blank(),
    strip.placement = "outside",
    axis.text.x  = element_text(angle=60, hjust=1, vjust = 1),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()) +
  labs(x = NULL, y = NULL)
p

grid.text(label = "(C)", x = unit(0.03, "npc") , y = unit(0.94,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Key-soil-variables-from-CCA-Worsley-C.tiff"), width = 17, height = 7, units = "cm", res=600, compression="lzw",type="cairo")
dev.print(tiff, file = paste0(workdir,"/plots/","Key-soil-variables-from-CCA-Worsley.tiff"), width = 17, height = 7, units = "cm", res=600, compression="lzw",type="cairo")



#-------------------------


#### Alpha (Shannon) diversity of samples - ASV-level - including merged sample bootstrap resampling
#------------------------

## the function below uses merged-sample bootstrap resampling to estimate the 
## overall site-level alpha diversity.
## merging by group, with rarefying to normalise for sampling effort



## create function to rarefy for initial sample > merge samples by type > rarefy again > calculate shannon's index
# # # # # # # # # # # #
calc_AlphaDiv_in_parallel_any_merge_no <- function(phy_obj, merge_by ) {
  
  # rarefy
  seed <- 123+j
  r16s <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                            rngseed = seed, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)
  
  # merge samples
  merged.r16s <- merge_samples(r16s, group= eval(parse(text= paste0("r16s@sam_data$",merge_by))) )
  
  
  # rarefy back to min sample size for this bootstrap sample - i.e. rarefied-merged-rarefied
  seed <- 234+j
  rmr16s <- rarefy_even_depth(merged.r16s, sample.size = min(sample_sums(phy_obj)),
                              rngseed = seed, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)
  
  # - - - - - - - - - -
  
  ### ALPHA DIVERSITY
  ## calculate Shannon's index
  shan.rmr16s <- plot_richness(rmr16s, measures=c("Shannon"))
  
  # export data
  out <- data.frame(sample=shan.rmr16s$data$samples,shannon=shan.rmr16s$data$value,
                    group=shan.rmr16s$data$samples)
  
  out$eff_no <- exp(out$shannon)
  out$calc_type <- "bootstrap"
  
  return(out)
  
}
# # # # # # # # # # # #


## apply this to each case study - by rehabilitation age

### Alcoa-Huntly

phy_in <- phy.alcoa.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 31746 taxa and 36 samples ]:
# sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
# tax_table()   Taxonomy Table:     [ 31746 taxa by 6 taxonomic ranks ]:
# phy_tree()    Phylogenetic Tree:  [ 31746 tips and 31744 internal nodes ]:
# taxa are rows


table(phy_in@sam_data$`Date since change in Land Use`)
# 1987 1991 1999 2002 2008 2014  N/A 
#    3    3    3    3    3    3   18 


phy_in@sam_data$group <- phy_in@sam_data$`Date since change in Land Use`
sel <- which(phy_in@sam_data$Notes == "adjacent natural") # qty 18
phy_in@sam_data$group[sel] <- "Ref"
unique(phy_in@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

phy_in@sam_data$group <- factor( phy_in@sam_data$group, 
                                 levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                 labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)

table( phy_in@sam_data$group )
# 2   8  14  17  25  29 Ref 
# 3   3   3   3   3   3  18 



min(sample_sums(phy_in)) # 17485
min(taxa_sums(phy_in)) # 2


phy_obj <- phy_in
names(phy_obj@sam_data)

head( phy_obj@sam_data[ ,c("Sample_id","group")] )
# Sample Data:        [ 6 samples by 2 sample variables ]:
#   Sample_id group
# X39041     39041 3    
# X39043     39043 Ref  
# X39045     39045 3    
# X39047     39047 Ref  
# X39049     39049 3    
# X39051     39051 Ref 

length(unique(phy_obj@sam_data$group)) # 7
sort(unique(phy_obj@sam_data$group))
# [1] 2   8   14  17  25  29  Ref
# Levels: 2 < 8 < 14 < 17 < 25 < 29 < Ref

# bootstrap outputs will automatically be stored in a list: b_out

# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_AlphaDiv_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)


length(b_out) # 100
names(b_out[[1]]) # "sample"    "shannon"   "group"     "eff_no"    "calc_type"
dim(b_out[[1]]) # 7  5

b_out[[1]]
# sample  shannon group   eff_no calc_type
# 1      2 7.984210     2 2934.260 bootstrap
# 2      8 7.905697     8 2712.693 bootstrap
# 3     14 8.039976    14 3102.540 bootstrap
# 4     17 7.388585    17 1617.417 bootstrap
# 5     25 7.300135    25 1480.500 bootstrap
# 6     29 7.947411    29 2828.243 bootstrap
# 7    Ref 8.079636   Ref 3228.059 bootstrap

### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  30751
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30751 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 30751 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 30751 tips and 30749 internal nodes ]


shan.r1.ps <- plot_richness(r1.ps, measures=c("Shannon"))
shan.r1.ps
out <- data.frame(sample=shan.r1.ps$data$samples,shannon=shan.r1.ps$data$value,
                  group=shan.r1.ps$data$group )  #
out$eff_no <- exp(out$shannon) # calculate effective no of species
out$calc_type <- "rarefyx1"
str(out)
# 'data.frame':	36 obs. of  5 variables:
#   $ sample   : chr  "X39041" "X39043" "X39045" "X39047" ...
# $ shannon  : num  7.95 7.97 7.75 7.88 7.78 ...
# $ group    : Ord.factor w/ 7 levels "3"<"9"<"15"<"18"<..: 1 7 1 7 1 7 2 7 2 7 ...
# $ eff_no   : num  2845 2906 2329 2632 2389 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


head(r1.ps@sam_data)
metadat <- as(r1.ps@sam_data, "data.frame")
metadat$sample <- row.names(metadat)

metadat <- merge(x = metadat, y = out, by = "sample", all.x = TRUE)

names(metadat)

metadat[ ,c("sample", "FULL_ID", "Location.description", "group.y", "shannon", "eff_no")]


## write out table ...
write.csv(x = metadat[ ,c("sample", "FULL_ID", "Location.description", "group.y", "shannon", "eff_no")], file = "alcoa-alpha-diversity-shannon-effective-no-ASVs.csv")


names(out)        # "sample"    "shannon"   "group"     "eff_no"    "calc_type"

names(b_out[[1]]) #  "sample"    "shannon"   "group"     "eff_no"    "calc_type"
dim(b_out[[1]]) # 7  5

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)
# "sample"    "shannon"   "group"     "eff_no"    "calc_type"


str(temp)
# 'data.frame':	736 obs. of  5 variables:
#   $ sample   : chr  "X39041" "X39043" "X39045" "X39047" ...
# $ shannon  : num  7.95 7.97 7.75 7.88 7.78 ...
# $ group    : Ord.factor w/ 7 levels "2"<"8"<"14"<"17"<..: 1 7 1 7 1 7 2 7 2 7 ...
# $ eff_no   : num  2845 2906 2329 2632 2389 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


levels(temp$group)
# "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

temp$group <- factor(temp$group, 
                     levels = c("2",   "8",   "14",  "17",  "25",  "29",  "Ref"),
                     labels=c("2 yr",   "8 yr",   "14 yr",  "17 yr",  "25 yr",  "29 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "eff_no")


## plot

cols <- cols.group.alcoa



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Effective ASVs (count)") +
  theme(legend.position="none")
  
p

grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Alpha-diversity--Alcoa--A.tiff"), width = 8, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")





### Eneabba-Ilkua

phy_in <- phy.iluka_eneabba.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 31746 taxa and 36 samples ]:
# sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
# tax_table()   Taxonomy Table:     [ 31746 taxa by 6 taxonomic ranks ]:
# phy_tree()    Phylogenetic Tree:  [ 31746 tips and 31744 internal nodes ]:
# taxa are rows


table(phy_in@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 


phy_in@sam_data$group <- phy_in@sam_data$year.of.rehab.or.ref
unique(phy_in@sam_data$group)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000" 
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                                labels = c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref"),
                                ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


table( phy_in@sam_data$group )
#7 yr 10 yr 15 yr 19 yr 24 yr 30 yr 38 yr   Ref 
#   2     2     3     2     3     2     3     9 



min(sample_sums(phy_in)) # 10142
min(taxa_sums(phy_in)) # 1


phy_obj <- phy_in
names(phy_obj@sam_data)

head( phy_obj@sam_data[ ,c("Sample_ID","group")] )
# Sample Data:        [ 6 samples by 2 sample variables ]:
#   Sample_ID          group
# X138408 102.100.100/138408 Ref  
# X138410 102.100.100/138410 Ref  
# X138412 102.100.100/138412 Ref  
# X138418 102.100.100/138418 30 yr
# X138424 102.100.100/138424 24 yr
# X138426 102.100.100/138426 24 yr

length(unique(phy_obj@sam_data$group)) # 8
sort(unique(phy_obj@sam_data$group))
# [1] 7 yr  10 yr 15 yr 19 yr 24 yr 30 yr 38 yr Ref  
# Levels: 7 yr < 10 yr < 15 yr < 19 yr < 24 yr < 30 yr < 38 yr < Ref

# bootstrap outputs will automatically be stored in a list: b_out

# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_AlphaDiv_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)


length(b_out) # 100
names(b_out[[1]]) # "sample"    "shannon"   "group"     "eff_no"    "calc_type"
dim(b_out[[1]]) # 8  5

b_out[[1]]
#   sample  shannon group   eff_no calc_type
# 1   7 yr 7.199884  7 yr 1339.275 bootstrap
# 2  10 yr 7.384699 10 yr 1611.142 bootstrap
# 3  15 yr 7.411158 15 yr 1654.341 bootstrap
# 4  19 yr 7.722033 19 yr 2257.545 bootstrap
# 5  24 yr 7.904702 24 yr 2709.996 bootstrap
# 6  30 yr 7.820954 30 yr 2492.281 bootstrap
# 7  38 yr 7.767626 38 yr 2362.855 bootstrap
# 8    Ref 7.963396   Ref 2873.817 bootstrap


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142

ntaxa(r1.ps) #  27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 27115 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 27115 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 27115 tips and 27114 internal nodes ]:
#   taxa are rows


shan.r1.ps <- plot_richness(r1.ps, measures=c("Shannon"))
shan.r1.ps
out <- data.frame(sample=shan.r1.ps$data$samples,shannon=shan.r1.ps$data$value,
                  group=shan.r1.ps$data$group )  #
out$eff_no <- exp(out$shannon) # calculate effective no of species
out$calc_type <- "rarefyx1"
str(out)
# 'data.frame':	26 obs. of  5 variables:
#   $ sample   : chr  "X138408" "X138410" "X138412" "X138418" ...
# $ shannon  : num  7.48 7.86 7.46 7.68 7.61 ...
# $ group    : Ord.factor w/ 8 levels "7 yr"<"10 yr"<..: 8 8 8 6 5 5 7 8 8 7 ...
# $ eff_no   : num  1771 2598 1742 2174 2024 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


names(out)        # "sample"    "shannon"   "group"     "eff_no"    "calc_type"

names(b_out[[1]]) #  "sample"    "shannon"   "group"     "eff_no"    "calc_type"
dim(b_out[[1]]) # 7  5

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)
# "sample"    "shannon"   "group"     "eff_no"    "calc_type"


str(temp)
# 'data.frame':	826 obs. of  5 variables:
#   $ sample   : chr  "X138408" "X138410" "X138412" "X138418" ...
# $ shannon  : num  7.48 7.86 7.46 7.68 7.61 ...
# $ group    : Ord.factor w/ 8 levels "7 yr"<"10 yr"<..: 8 8 8 6 5 5 7 8 8 7 ...
# $ eff_no   : num  1771 2598 1742 2174 2024 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


levels(temp$group)
# "7 yr"  "10 yr" "15 yr" "19 yr" "24 yr" "30 yr" "38 yr" "Ref"

# temp$group <- factor(temp$group, 
#                      levels = c("3",   "9",   "15",  "18",  "26",  "30",  "Ref"),
#                      labels=c("3 yr",   "9 yr",   "15 yr",  "18 yr",  "26 yr",  "30 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "eff_no")


## plot

cols <- cols.group.iluka



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Effective ASVs (count)") +
  theme(legend.position="none")

p

grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Alpha-diversity--Eneabba--B.tiff"), width = 8, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")





### South32

phy_in <- phy.south32.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 54671 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 54671 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 54671 tips and 54669 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6


phy_in@sam_data$group <- phy_in@sam_data$`year of rehab or ref`
unique(phy_in@sam_data$group)
# "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                                labels = c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


table( phy_in@sam_data$group )
# 2 yr  8 yr 12 yr 14 yr 17 yr 20 yr 23 yr 28 yr   Ref 
#    3     3     1     2     2     2     4     2     6 



min(sample_sums(phy_in)) # 54122
min(taxa_sums(phy_in)) # 2


phy_obj <- phy_in
names(phy_obj@sam_data)

head( phy_obj@sam_data[ ,c("Sample_ID","group")] )
# Sample Data:        [ 6 samples by 2 sample variables ]:
#   Sample_ID          group
# X138358 102.100.100/138358 23 yr
# X138360 102.100.100/138360 23 yr
# X138362 102.100.100/138362 Ref  
# X138364 102.100.100/138364 2 yr 
# X138366 102.100.100/138366 Ref  
# X138368 102.100.100/138368 17 yr


length(unique(phy_obj@sam_data$group)) # 9
sort(unique(phy_obj@sam_data$group))
# [1] 2 yr  8 yr  12 yr 14 yr 17 yr 20 yr 23 yr 28 yr Ref  
# Levels: 2 yr < 8 yr < 12 yr < 14 yr < 17 yr < 20 yr < 23 yr < 28 yr < Ref

# bootstrap outputs will automatically be stored in a list: b_out

# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_AlphaDiv_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)


length(b_out) # 100
names(b_out[[1]]) # "sample"    "shannon"   "group"     "eff_no"    "calc_type"
dim(b_out[[1]]) # 9  5

b_out[[1]]
# sample  shannon group   eff_no calc_type
# 1   2 yr 8.747880  2 yr 6297.326 bootstrap
# 2   8 yr 8.708422  8 yr 6053.684 bootstrap
# 3  12 yr 8.385409 12 yr 4382.649 bootstrap
# 4  14 yr 8.386565 14 yr 4387.721 bootstrap
# 5  17 yr 8.656223 17 yr 5745.794 bootstrap
# 6  20 yr 8.844781 20 yr 6938.085 bootstrap
# 7  23 yr 8.658129 23 yr 5756.752 bootstrap
# 8  28 yr 8.549683 28 yr 5165.114 bootstrap
# 9    Ref 8.764924   Ref 6405.576 bootstrap


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122

ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 54327 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 54327 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 54327 tips and 54325 internal nodes ]:
#   taxa are rows


shan.r1.ps <- plot_richness(r1.ps, measures=c("Shannon"))
shan.r1.ps
out <- data.frame(sample=shan.r1.ps$data$samples,shannon=shan.r1.ps$data$value,
                  group=shan.r1.ps$data$group )  #
out$eff_no <- exp(out$shannon) # calculate effective no of species
out$calc_type <- "rarefyx1"
str(out)
# 'data.frame':	25 obs. of  5 variables:
#   $ sample   : chr  "X138358" "X138360" "X138362" "X138364" ...
# $ shannon  : num  8.52 8.28 8.47 8.35 8.29 ...
# $ group    : Ord.factor w/ 9 levels "2 yr"<"8 yr"<..: 7 7 9 1 9 5 3 7 9 9 ...
# $ eff_no   : num  5036 3958 4781 4242 3977 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


names(out)        # "sample"    "shannon"   "group"     "eff_no"    "calc_type"

names(b_out[[1]]) #  "sample"    "shannon"   "group"     "eff_no"    "calc_type"
dim(b_out[[1]]) # 9  5

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)
# "sample"    "shannon"   "group"     "eff_no"    "calc_type"


str(temp)
# 'data.frame':	925 obs. of  5 variables:
#   $ sample   : chr  "X138358" "X138360" "X138362" "X138364" ...
# $ shannon  : num  8.52 8.28 8.47 8.35 8.29 ...
# $ group    : Ord.factor w/ 9 levels "2 yr"<"8 yr"<..: 7 7 9 1 9 5 3 7 9 9 ...
# $ eff_no   : num  5036 3958 4781 4242 3977 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


levels(temp$group)
# "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr" "Ref"  

# temp$group <- factor(temp$group, 
#                      levels = c("3",   "9",   "15",  "18",  "26",  "30",  "Ref"),
#                      labels=c("3 yr",   "9 yr",   "15 yr",  "18 yr",  "26 yr",  "30 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "eff_no")


## plot

cols <- cols.group.south32



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Effective ASVs (count)") +
  theme(legend.position="none")

p

grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Alpha-diversity--WorsleySouth32--C.tiff"), width = 8, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")



#------------------------


#### Evenness of samples - ASV-level - including merged sample bootstrap resampling
#------------------------

## the function below uses merged-sample bootstrap resampling to estimate the 
## overall site-level alpha diversity (Shannon's diversity) then divides by the natural logarithm of species richness.
## merging by group, with rarefying to normalise for sampling effort


#phy_obj <- phy_in
#merge_by<- "group"
#j<-1

## create function to rarefy for initial sample > merge samples by type > rarefy again > calculate shannon's index
# # # # # # # # # # # #
calc_Evenness_in_parallel_any_merge_no <- function(phy_obj, merge_by ) {
  
  # rarefy
  seed <- 123+j
  r16s <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                            rngseed = seed, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)
  
  # merge samples
  merged.r16s <- merge_samples(r16s, group= eval(parse(text= paste0("r16s@sam_data$",merge_by))) )
  
  
  # rarefy back to min sample size for this bootstrap sample - i.e. rarefied-merged-rarefied
  seed <- 234+j
  rmr16s <- rarefy_even_depth(merged.r16s, sample.size = min(sample_sums(phy_obj)),
                              rngseed = seed, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)
  
  # - - - - - - - - - -
  
  ### ALPHA DIVERSITY
  ## calculate Shannon's index
  shan.rmr16s <- plot_richness(rmr16s, measures=c("Shannon"))
  rich.rmr16s <- plot_richness(rmr16s, measures=c("Observed"))
  #identical(shan.rmr16s$data$samples,rich.rmr16s$data$samples)
  #identical(shan.rmr16s$data$group,rich.rmr16s$data$group)
  
  # export data
  out <- data.frame(sample=shan.rmr16s$data$samples,evenness=shan.rmr16s$data$value/log(rich.rmr16s$data$value),
                    group=shan.rmr16s$data$samples)
  
  out$calc_type <- "bootstrap"
  
  return(out)
  
}
# # # # # # # # # # # #


## apply this to each case study - by rehabilitation age

### Alcoa-Huntly

phy_in <- phy.alcoa.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 31746 taxa and 36 samples ]:
# sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
# tax_table()   Taxonomy Table:     [ 31746 taxa by 6 taxonomic ranks ]:
# phy_tree()    Phylogenetic Tree:  [ 31746 tips and 31744 internal nodes ]:
# taxa are rows


table(phy_in@sam_data$`Date since change in Land Use`)
# 1987 1991 1999 2002 2008 2014  N/A 
#    3    3    3    3    3    3   18 


phy_in@sam_data$group <- phy_in@sam_data$`Date since change in Land Use`
sel <- which(phy_in@sam_data$Notes == "adjacent natural") # qty 18
phy_in@sam_data$group[sel] <- "Ref"
unique(phy_in@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

phy_in@sam_data$group <- factor( phy_in@sam_data$group, 
                                 levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                 labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)

table( phy_in@sam_data$group )
# 2   8  14  17  25  29 Ref 
# 3   3   3   3   3   3  18 


min(sample_sums(phy_in)) # 17485
min(taxa_sums(phy_in)) # 2


phy_obj <- phy_in
names(phy_obj@sam_data)

head( phy_obj@sam_data[ ,c("Sample_id","group")] )
# Sample Data:        [ 6 samples by 2 sample variables ]:
#   Sample_id group
# X39041     39041 2    
# X39043     39043 Ref  
# X39045     39045 2    
# X39047     39047 Ref  
# X39049     39049 2    
# X39051     39051 Ref  

length(unique(phy_obj@sam_data$group)) # 7
sort(unique(phy_obj@sam_data$group))
# [1] 2   8   14  17  25  29  Ref
# Levels: 2 < 8 < 14 < 17 < 25 < 29 < Ref

# bootstrap outputs will automatically be stored in a list: b_out

# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_Evenness_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)


length(b_out) # 100
names(b_out[[1]]) # "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

b_out[[1]]
# sample  evenness group calc_type
# 1      2 0.9342868     2 bootstrap
# 2      8 0.9191228     8 bootstrap
# 3     14 0.9273237    14 bootstrap
# 4     17 0.8854684    17 bootstrap
# 5     25 0.8660495    25 bootstrap
# 6     29 0.9130926    29 bootstrap
# 7    Ref 0.9079074   Ref bootstrap


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  30751
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 30751 taxa and 36 samples ]
# sample_data() Sample Data:       [ 36 samples by 70 sample variables ]
# tax_table()   Taxonomy Table:    [ 30751 taxa by 6 taxonomic ranks ]
# phy_tree()    Phylogenetic Tree: [ 30751 tips and 30749 internal nodes ]


shan.r1.ps <- plot_richness(r1.ps, measures=c("Shannon"))
shan.r1.ps
rich.r1.ps <- plot_richness(r1.ps, measures=c("Observed"))
rich.r1.ps

identical(shan.r1.ps$data$samples,rich.r1.ps$data$samples)
identical(shan.r1.ps$data$group,rich.r1.ps$data$group)

# export data
out <- data.frame(sample=shan.r1.ps$data$samples,evenness=shan.r1.ps$data$value/log(rich.r1.ps$data$value),
                  group=shan.r1.ps$data$group)

out$calc_type <- "rarefyx1"
str(out)


names(out)        # "sample"    "evenness"  "group"     "calc_type"



head(r1.ps@sam_data)
metadat <- as(r1.ps@sam_data, "data.frame")
metadat$sample <- row.names(metadat)

metadat <- merge(x = metadat, y = out, by = "sample", all.x = TRUE)

names(metadat)

metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "evenness")]


## write out table ...
write.csv(x = metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "evenness")], file = "alcoa-evenness-ASVs.csv")




names(b_out[[1]]) #  "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)
# "sample"    "evenness"  "group"     "calc_type"


str(temp)
# 'data.frame':	736 obs. of  4 variables:
#   $ sample   : chr  "X39041" "X39043" "X39045" "X39047" ...
# $ evenness : num  0.924 0.913 0.915 0.905 0.921 ...
# $ group    : Ord.factor w/ 7 levels "2"<"8"<"14"<"17"<..: 1 7 1 7 1 7 2 7 2 7 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


levels(temp$group)
# "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

temp$group <- factor(temp$group, 
                     levels = c("2",   "8",   "14",  "17",  "25",  "29",  "Ref"),
                     labels=c("2 yr",   "8 yr",   "14 yr",  "17 yr",  "25 yr",  "29 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "evenness")


## plot

cols <- cols.group.alcoa



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Evenness (J')") +
  theme(legend.position="none")

p

grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Evenness-ASV-level--Alcoa--A.tiff"), width = 8, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")





### Eneabba-Ilkua

phy_in <- phy.iluka_eneabba.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33636 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33636 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33636 tips and 33634 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 


phy_in@sam_data$group <- phy_in@sam_data$year.of.rehab.or.ref
unique(phy_in@sam_data$group)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000" 
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                                labels = c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref"),
                                ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


table( phy_in@sam_data$group )
#7 yr 10 yr 15 yr 19 yr 24 yr 30 yr 38 yr   Ref 
#   2     2     3     2     3     2     3     9 


min(sample_sums(phy_in)) # 10142
min(taxa_sums(phy_in)) # 1


phy_obj <- phy_in
names(phy_obj@sam_data)

head( phy_obj@sam_data[ ,c("Sample_ID","group")] )
# Sample Data:        [ 6 samples by 2 sample variables ]:
#   Sample_ID          group
# X138408 102.100.100/138408 Ref  
# X138410 102.100.100/138410 Ref  
# X138412 102.100.100/138412 Ref  
# X138418 102.100.100/138418 30 yr
# X138424 102.100.100/138424 24 yr
# X138426 102.100.100/138426 24 yr

length(unique(phy_obj@sam_data$group)) # 8
sort(unique(phy_obj@sam_data$group))
# [1] 7 yr  10 yr 15 yr 19 yr 24 yr 30 yr 38 yr Ref  
# Levels: 7 yr < 10 yr < 15 yr < 19 yr < 24 yr < 30 yr < 38 yr < Ref

# bootstrap outputs will automatically be stored in a list: b_out

# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_Evenness_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)


length(b_out) # 100
names(b_out[[1]]) # "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 8  4

b_out[[1]]
# sample  evenness group calc_type
# 1   7 yr 0.9149213  7 yr bootstrap
# 2  10 yr 0.9237071 10 yr bootstrap
# 3  15 yr 0.9016451 15 yr bootstrap
# 4  19 yr 0.9339995 19 yr bootstrap
# 5  24 yr 0.9441905 24 yr bootstrap
# 6  30 yr 0.9447679 30 yr bootstrap
# 7  38 yr 0.9342442 38 yr bootstrap
# 8    Ref 0.9417974   Ref bootstrap


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 10142

ntaxa(r1.ps) #  27115
sum(sample_sums(r1.ps)) # 263692

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 27115 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 27115 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 27115 tips and 27114 internal nodes ]:
#   taxa are rows

shan.r1.ps <- plot_richness(r1.ps, measures=c("Shannon"))
shan.r1.ps
rich.r1.ps <- plot_richness(r1.ps, measures=c("Observed"))
rich.r1.ps

identical(shan.r1.ps$data$samples,rich.r1.ps$data$samples)
identical(shan.r1.ps$data$group,rich.r1.ps$data$group)

# export data
out <- data.frame(sample=shan.r1.ps$data$samples,evenness=shan.r1.ps$data$value/log(rich.r1.ps$data$value),
                  group=shan.r1.ps$data$group)

out$calc_type <- "rarefyx1"
str(out)
# 'data.frame':	26 obs. of  4 variables:
#   $ sample   : chr  "X138408" "X138410" "X138412" "X138418" ...
# $ evenness : num  0.936 0.937 0.916 0.935 0.923 ...
# $ group    : Ord.factor w/ 8 levels "7 yr"<"10 yr"<..: 8 8 8 6 5 5 7 8 8 7 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


names(out)        # "sample"    "evenness"  "group"     "calc_type"

names(b_out[[1]]) #  "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 8  4

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)


str(temp)
# 'data.frame':	826 obs. of  4 variables:
#   $ sample   : chr  "X138408" "X138410" "X138412" "X138418" ...
# $ evenness : num  0.936 0.937 0.916 0.935 0.923 ...
# $ group    : Ord.factor w/ 8 levels "7 yr"<"10 yr"<..: 8 8 8 6 5 5 7 8 8 7 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


levels(temp$group)
# "7 yr"  "10 yr" "15 yr" "19 yr" "24 yr" "30 yr" "38 yr" "Ref"

# temp$group <- factor(temp$group, 
#                      levels = c("3",   "9",   "15",  "18",  "26",  "30",  "Ref"),
#                      labels=c("3 yr",   "9 yr",   "15 yr",  "18 yr",  "26 yr",  "30 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "evenness")


## plot

cols <- cols.group.iluka



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Evenness (J')") +
  theme(legend.position="none")

p

grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Evenness-ASV-level--Eneabba--B.tiff"), width = 8, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")





### South32

phy_in <- phy.south32.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 54671 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 54671 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 54671 tips and 54669 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6


phy_in@sam_data$group <- phy_in@sam_data$`year of rehab or ref`
unique(phy_in@sam_data$group)
# "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                                labels = c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


table( phy_in@sam_data$group )
# 2 yr  8 yr 12 yr 14 yr 17 yr 20 yr 23 yr 28 yr   Ref 
#    3     3     1     2     2     2     4     2     6 



min(sample_sums(phy_in)) # 54122
min(taxa_sums(phy_in)) # 2


phy_obj <- phy_in
names(phy_obj@sam_data)

head( phy_obj@sam_data[ ,c("Sample_ID","group")] )
# Sample Data:        [ 6 samples by 2 sample variables ]:
#   Sample_ID          group
# X138358 102.100.100/138358 23 yr
# X138360 102.100.100/138360 23 yr
# X138362 102.100.100/138362 Ref  
# X138364 102.100.100/138364 2 yr 
# X138366 102.100.100/138366 Ref  
# X138368 102.100.100/138368 17 yr


length(unique(phy_obj@sam_data$group)) # 9
sort(unique(phy_obj@sam_data$group))
# [1] 2 yr  8 yr  12 yr 14 yr 17 yr 20 yr 23 yr 28 yr Ref  
# Levels: 2 yr < 8 yr < 12 yr < 14 yr < 17 yr < 20 yr < 23 yr < 28 yr < Ref

# bootstrap outputs will automatically be stored in a list: b_out

# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_Evenness_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)


length(b_out) # 100
names(b_out[[1]]) # "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 9  4

b_out[[1]]
# sample  evenness group calc_type
# 1   2 yr 0.9155908  2 yr bootstrap
# 2   8 yr 0.9151008  8 yr bootstrap
# 3  12 yr 0.9078178 12 yr bootstrap
# 4  14 yr 0.8991653 14 yr bootstrap
# 5  17 yr 0.9164167 17 yr bootstrap
# 6  20 yr 0.9234443 20 yr bootstrap
# 7  23 yr 0.9036112 23 yr bootstrap
# 8  28 yr 0.9060438 28 yr bootstrap
# 9    Ref 0.9045748   Ref bootstrap


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 54122

ntaxa(r1.ps) #  54327
sum(sample_sums(r1.ps)) # 1353050

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 54327 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 54327 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 54327 tips and 54325 internal nodes ]:
#   taxa are rows


shan.r1.ps <- plot_richness(r1.ps, measures=c("Shannon"))
shan.r1.ps
rich.r1.ps <- plot_richness(r1.ps, measures=c("Observed"))
rich.r1.ps

identical(shan.r1.ps$data$samples,rich.r1.ps$data$samples)
identical(shan.r1.ps$data$group,rich.r1.ps$data$group)

# export data
out <- data.frame(sample=shan.r1.ps$data$samples,evenness=shan.r1.ps$data$value/log(rich.r1.ps$data$value),
                  group=shan.r1.ps$data$group)

out$calc_type <- "rarefyx1"
str(out)
# 'data.frame':	25 obs. of  4 variables:
#   $ sample   : chr  "X138358" "X138360" "X138362" "X138364" ...
# $ evenness : num  0.895 0.88 0.888 0.887 0.882 ...
# $ group    : Ord.factor w/ 9 levels "2 yr"<"8 yr"<..: 7 7 9 1 9 5 3 7 9 9 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


names(out)        # "sample"    "evenness"  "group"     "calc_type"

names(b_out[[1]]) #  "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 9  4

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)
# "sample"    "evenness"  "group"     "calc_type"


str(temp)
#'data.frame':	925 obs. of  4 variables:
# $ sample   : chr  "X138358" "X138360" "X138362" "X138364" ...
# $ evenness : num  0.895 0.88 0.888 0.887 0.882 ...
# $ group    : Ord.factor w/ 9 levels "2 yr"<"8 yr"<..: 7 7 9 1 9 5 3 7 9 9 ...
# $ calc_type: chr  "rarefyx1" "rarefyx1" "rarefyx1" "rarefyx1" ...


levels(temp$group)
# "2 yr"  "8 yr"  "12 yr" "14 yr" "17 yr" "20 yr" "23 yr" "28 yr" "Ref"  

# temp$group <- factor(temp$group, 
#                      levels = c("3",   "9",   "15",  "18",  "26",  "30",  "Ref"),
#                      labels=c("3 yr",   "9 yr",   "15 yr",  "18 yr",  "26 yr",  "30 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "evenness")


## plot

cols <- cols.group.south32



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Evenness (J')") +
  theme(legend.position="none")

p

grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Evenness-ASV-level--WorsleySouth32--C.tiff"), width = 8, height = 8, units = "cm", res = 600, compression = "lzw",type="cairo")



#------------------------


#### Richness and Evenness - Alcoa only - Phyla - Class - Order - including merged sample bootstrap resampling
#------------------------

## the function below uses merged-sample bootstrap resampling to estimate the 
## overall site-level alpha diversity.
## merging by group, with rarefying to normalise for sampling effort



## create function to rarefy for initial sample > merge samples by type > rarefy again > calculate shannon's index
# # # # # # # # # # # #
calc_SpeciesRichness_in_parallel_any_merge_no <- function(phy_obj, merge_by ) {
  
  # rarefy
  seed <- 123+j
  r16s <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                            rngseed = seed, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)
  
  # merge samples
  merged.r16s <- merge_samples(r16s, group= eval(parse(text= paste0("r16s@sam_data$",merge_by))) )
  
  
  # rarefy back to min sample size for this bootstrap sample - i.e. rarefied-merged-rarefied
  seed <- 234+j
  rmr16s <- rarefy_even_depth(merged.r16s, sample.size = min(sample_sums(phy_obj)),
                              rngseed = seed, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)
  
  # - - - - - - - - - -
  
  ### ALPHA DIVERSITY
  ## calculate Species richness
  alpha.rmr16s <- plot_richness(rmr16s, measures=c("Observed"))
  
  # export data
  out <- data.frame(sample=alpha.rmr16s$data$samples,richness=alpha.rmr16s$data$value,
                    group=alpha.rmr16s$data$samples)
  
  out$calc_type <- "bootstrap"
  
  return(out)
  
}
# # # # # # # # # # # #


## apply this to each case study - by rehabilitation age

### Alcoa-Huntly


### PHYLUM LEVEL


phy_in <- phy.alcoa.withTree.PHYLUM
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33 tips and 32 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$`Date since change in Land Use`)
# 1987 1991 1999 2002 2008 2014  N/A 
#    3    3    3    3    3    3   18 


phy_in@sam_data$group <- phy_in@sam_data$`Date since change in Land Use`
sel <- which(phy_in@sam_data$Notes == "adjacent natural") # qty 18
phy_in@sam_data$group[sel] <- "Ref"
unique(phy_in@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

phy_in@sam_data$group <- factor( phy_in@sam_data$group, 
                                 levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                 labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)

table( phy_in@sam_data$group )
# 2   8  14  17  25  29 Ref 
# 3   3   3   3   3   3  18 



min(sample_sums(phy_in)) # 17485
min(taxa_sums(phy_in)) # 17


phy_obj <- phy_in
names(phy_obj@sam_data)

head( phy_obj@sam_data[ ,c("Sample_id","group")] )
# Sample Data:        [ 6 samples by 2 sample variables ]:
#   Sample_id group
# X39041     39041 2    
# X39043     39043 Ref  
# X39045     39045 2    
# X39047     39047 Ref  
# X39049     39049 2    
# X39051     39051 Ref 

length(unique(phy_obj@sam_data$group)) # 7
sort(unique(phy_obj@sam_data$group))
# [1] 2   8   14  17  25  29  Ref
# Levels: 2 < 8 < 14 < 17 < 25 < 29 < Ref

# bootstrap outputs will automatically be stored in a list: b_out

# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_SpeciesRichness_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)


length(b_out) # 100
names(b_out[[1]]) # "sample"    "richness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

b_out[[1]]
# sample richness group calc_type
# 1      2       25     2 bootstrap
# 2      8       22     8 bootstrap
# 3     14       23    14 bootstrap
# 4     17       21    17 bootstrap
# 5     25       20    25 bootstrap
# 6     29       23    29 bootstrap
# 7    Ref       29   Ref bootstrap


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 7
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  33
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 71 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33 tips and 32 internal nodes ]:
#   taxa are rows


alpha.r1.ps <- plot_richness(r1.ps, measures=c("Observed"))
alpha.r1.ps
out <- data.frame(sample=alpha.r1.ps$data$samples,richness=alpha.r1.ps$data$value,
                  group=alpha.r1.ps$data$group )  #
out$calc_type <- "rarefyx1"
str(out)

names(out)        # "sample"    "richness"  "group"     "calc_type"

head(r1.ps@sam_data)
metadat <- as(r1.ps@sam_data, "data.frame")
metadat$sample <- row.names(metadat)

metadat <- merge(x = metadat, y = out, by = "sample", all.x = TRUE)

names(metadat)

metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "richness")]


## write out table ...
write.csv(x = metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "richness")], file = "alcoa-richness-PHYLUM.csv")


names(b_out[[1]]) #  "sample"    "richness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)


str(temp)


levels(temp$group)
# "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

temp$group <- factor(temp$group, 
                     levels = c("2",   "8",   "14",  "17",  "25",  "29",  "Ref"),
                     labels=c("2 yr",   "8 yr",   "14 yr",  "17 yr",  "25 yr",  "29 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "richness")


## plot

cols <- cols.group.alcoa



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Richness (count)") +
  theme(legend.position="none")

p

grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Phyla-Richness--Alcoa--A.tiff"), width = 8, height = 7, units = "cm", res = 600, compression = "lzw",type="cairo")



## re-use phy_obj from Phyla level
## re-use function calc_Evenness_in_parallel_any_merge_no() defined above
## re-use r1.ps derived above


# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_Evenness_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)

length(b_out) # 100
names(b_out[[1]]) # "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

b_out[[1]]


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# # rarefy #1
# seed <- 123
# r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
#                            rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
# 
# min(taxa_sums(r1.ps)) # 1
# sample_sums(r1.ps) # all 17485
# 
# ntaxa(r1.ps) #  30751
# sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 71 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33 tips and 32 internal nodes ]:
#   taxa are rows

shan.r1.ps <- plot_richness(r1.ps, measures=c("Shannon"))
shan.r1.ps
rich.r1.ps <- plot_richness(r1.ps, measures=c("Observed"))
rich.r1.ps

identical(shan.r1.ps$data$samples,rich.r1.ps$data$samples)
identical(shan.r1.ps$data$group,rich.r1.ps$data$group)

# export data
out <- data.frame(sample=shan.r1.ps$data$samples,evenness=shan.r1.ps$data$value/log(rich.r1.ps$data$value),
                  group=shan.r1.ps$data$group)

out$calc_type <- "rarefyx1"
str(out)


names(out)        # "sample"    "evenness"  "group"     "calc_type"


head(r1.ps@sam_data)
metadat <- as(r1.ps@sam_data, "data.frame")
metadat$sample <- row.names(metadat)

metadat <- merge(x = metadat, y = out, by = "sample", all.x = TRUE)

names(metadat)

metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "evenness")]


## write out table ...
write.csv(x = metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "evenness")], file = "alcoa-evenness-PHYLUM.csv")


names(b_out[[1]]) #  "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)
# "sample"    "evenness"  "group"     "calc_type"


str(temp)


levels(temp$group)
# "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

temp$group <- factor(temp$group, 
                     levels = c("2",   "8",   "14",  "17",  "25",  "29",  "Ref"),
                     labels=c("2 yr",   "8 yr",   "14 yr",  "17 yr",  "25 yr",  "29 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "evenness")


## plot

cols <- cols.group.alcoa



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Evenness (J')") +
  theme(legend.position="none")

p

grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Phyla-Evenness--Alcoa--B.tiff"), width = 8, height = 7, units = "cm", res = 600, compression = "lzw",type="cairo")



### CLASS LEVEL



phy_in <- phy.alcoa.withTree.CLASS
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 110 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 110 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 110 tips and 109 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$`Date since change in Land Use`)
# 1987 1991 1999 2002 2008 2014  N/A 
#    3    3    3    3    3    3   18 


phy_in@sam_data$group <- phy_in@sam_data$`Date since change in Land Use`
sel <- which(phy_in@sam_data$Notes == "adjacent natural") # qty 18
phy_in@sam_data$group[sel] <- "Ref"
unique(phy_in@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

phy_in@sam_data$group <- factor( phy_in@sam_data$group, 
                                 levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                 labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)

table( phy_in@sam_data$group )
# 2   8  14  17  25  29 Ref 
# 3   3   3   3   3   3  18 



min(sample_sums(phy_in)) # 17485
min(taxa_sums(phy_in)) # 3


phy_obj <- phy_in
names(phy_obj@sam_data)

head( phy_obj@sam_data[ ,c("Sample_id","group")] )
# Sample Data:        [ 6 samples by 2 sample variables ]:
#   Sample_id group
# X39041     39041 2    
# X39043     39043 Ref  
# X39045     39045 2    
# X39047     39047 Ref  
# X39049     39049 2    
# X39051     39051 Ref 

length(unique(phy_obj@sam_data$group)) # 7
sort(unique(phy_obj@sam_data$group))
# [1] 2   8   14  17  25  29  Ref
# Levels: 2 < 8 < 14 < 17 < 25 < 29 < Ref

# bootstrap outputs will automatically be stored in a list: b_out

# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_SpeciesRichness_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)


length(b_out) # 100
names(b_out[[1]]) # "sample"    "richness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

b_out[[1]]
# sample richness group calc_type
# 1      2       86     2 bootstrap
# 2      8       75     8 bootstrap
# 3     14       69    14 bootstrap
# 4     17       55    17 bootstrap
# 5     25       59    25 bootstrap
# 6     29       65    29 bootstrap
# 7    Ref       74   Ref bootstrap


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  109
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 109 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 71 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 109 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 109 tips and 108 internal nodes ]:
#   taxa are rows


alpha.r1.ps <- plot_richness(r1.ps, measures=c("Observed"))
alpha.r1.ps
out <- data.frame(sample=alpha.r1.ps$data$samples,richness=alpha.r1.ps$data$value,
                  group=alpha.r1.ps$data$group )  #
out$calc_type <- "rarefyx1"
str(out)

names(out)        # "sample"    "richness"  "group"     "calc_type"



head(r1.ps@sam_data)
metadat <- as(r1.ps@sam_data, "data.frame")
metadat$sample <- row.names(metadat)

metadat <- merge(x = metadat, y = out, by = "sample", all.x = TRUE)

names(metadat)

metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "richness")]


## write out table ...
write.csv(x = metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "richness")], file = "alcoa-richness-CLASS.csv")




names(b_out[[1]]) #  "sample"    "richness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)


str(temp)


levels(temp$group)
# "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

temp$group <- factor(temp$group, 
                     levels = c("2",   "8",   "14",  "17",  "25",  "29",  "Ref"),
                     labels=c("2 yr",   "8 yr",   "14 yr",  "17 yr",  "25 yr",  "29 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "richness")


## plot

cols <- cols.group.alcoa



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Richness (count)") +
  theme(legend.position="none")

p

grid.text(label = "(C)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Class-Richness--Alcoa--C.tiff"), width = 8, height = 7, units = "cm", res = 600, compression = "lzw",type="cairo")



## re-use phy_obj from Phyla level
## re-use function calc_Evenness_in_parallel_any_merge_no() defined above
## re-use r1.ps derived above


# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_Evenness_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)

length(b_out) # 100
names(b_out[[1]]) # "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

b_out[[1]]


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# # rarefy #1
# seed <- 123
# r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
#                            rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
# 
# min(taxa_sums(r1.ps)) # 1
# sample_sums(r1.ps) # all 17485
# 
# ntaxa(r1.ps) #  30751
# sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 109 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 71 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 109 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 109 tips and 108 internal nodes ]:
#   taxa are rows

shan.r1.ps <- plot_richness(r1.ps, measures=c("Shannon"))
shan.r1.ps
rich.r1.ps <- plot_richness(r1.ps, measures=c("Observed"))
rich.r1.ps

identical(shan.r1.ps$data$samples,rich.r1.ps$data$samples)
identical(shan.r1.ps$data$group,rich.r1.ps$data$group)

# export data
out <- data.frame(sample=shan.r1.ps$data$samples,evenness=shan.r1.ps$data$value/log(rich.r1.ps$data$value),
                  group=shan.r1.ps$data$group)

out$calc_type <- "rarefyx1"
str(out)

names(out)        # "sample"    "evenness"  "group"     "calc_type"


head(r1.ps@sam_data)
metadat <- as(r1.ps@sam_data, "data.frame")
metadat$sample <- row.names(metadat)

metadat <- merge(x = metadat, y = out, by = "sample", all.x = TRUE)

names(metadat)

metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "evenness")]


## write out table ...
write.csv(x = metadat[ ,c("sample", "FULL_ID","lat.....in.decimal.degrees","long.in.decimal.degrees", "Location.description", "group.y", "evenness")], file = "alcoa-evenness-CLASS.csv")



names(b_out[[1]]) #  "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)
# "sample"    "evenness"  "group"     "calc_type"


str(temp)


levels(temp$group)
# "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

temp$group <- factor(temp$group, 
                     levels = c("2",   "8",   "14",  "17",  "25",  "29",  "Ref"),
                     labels=c("2 yr",   "8 yr",   "14 yr",  "17 yr",  "25 yr",  "29 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "evenness")


## plot

cols <- cols.group.alcoa



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Evenness (J')") +
  theme(legend.position="none")

p

grid.text(label = "(D)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Class-Evenness--Alcoa--D.tiff"), width = 8, height = 7, units = "cm", res = 600, compression = "lzw",type="cairo")




### ORDER LEVEL



phy_in <- phy.alcoa.withTree.ORDER
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 266 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 266 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 266 tips and 265 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$`Date since change in Land Use`)
# 1987 1991 1999 2002 2008 2014  N/A 
#    3    3    3    3    3    3   18 


phy_in@sam_data$group <- phy_in@sam_data$`Date since change in Land Use`
sel <- which(phy_in@sam_data$Notes == "adjacent natural") # qty 18
phy_in@sam_data$group[sel] <- "Ref"
unique(phy_in@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

phy_in@sam_data$group <- factor( phy_in@sam_data$group, 
                                 levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                 labels = c("2","8","14","17","25","29","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)

table( phy_in@sam_data$group )
# 2   8  14  17  25  29 Ref 
# 3   3   3   3   3   3  18 



min(sample_sums(phy_in)) # 17485
min(taxa_sums(phy_in)) # 2


phy_obj <- phy_in
names(phy_obj@sam_data)

head( phy_obj@sam_data[ ,c("Sample_id","group")] )
# Sample Data:        [ 6 samples by 2 sample variables ]:
#   Sample_id group
# X39041     39041 2    
# X39043     39043 Ref  
# X39045     39045 2    
# X39047     39047 Ref  
# X39049     39049 2    
# X39051     39051 Ref 

length(unique(phy_obj@sam_data$group)) # 7
sort(unique(phy_obj@sam_data$group))
# [1] 2   8   14  17  25  29  Ref
# Levels: 2 < 8 < 14 < 17 < 25 < 29 < Ref

# bootstrap outputs will automatically be stored in a list: b_out

# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_SpeciesRichness_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)


length(b_out) # 100
names(b_out[[1]]) # "sample"    "richness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

b_out[[1]]
# sample richness group calc_type
# 1      2      192     2 bootstrap
# 2      8      165     8 bootstrap
# 3     14      171    14 bootstrap
# 4     17      132    17 bootstrap
# 5     25      149    25 bootstrap
# 6     29      165    29 bootstrap
# 7    Ref      166   Ref bootstrap


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# rarefy #1
seed <- 123
r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
                           rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)

min(taxa_sums(r1.ps)) # 1
sample_sums(r1.ps) # all 17485

ntaxa(r1.ps) #  265
sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 265 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 71 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 265 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 265 tips and 264 internal nodes ]:
#   taxa are rows


alpha.r1.ps <- plot_richness(r1.ps, measures=c("Observed"))
alpha.r1.ps
out <- data.frame(sample=alpha.r1.ps$data$samples,richness=alpha.r1.ps$data$value,
                  group=alpha.r1.ps$data$group )  #
out$calc_type <- "rarefyx1"
str(out)

names(out)        # "sample"    "richness"  "group"     "calc_type"

names(b_out[[1]]) #  "sample"    "richness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)


str(temp)


levels(temp$group)
# "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

temp$group <- factor(temp$group, 
                     levels = c("2",   "8",   "14",  "17",  "25",  "29",  "Ref"),
                     labels=c("2 yr",   "8 yr",   "14 yr",  "17 yr",  "25 yr",  "29 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "richness")


## plot

cols <- cols.group.alcoa



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Richness (count)") +
  theme(legend.position="none")

p

grid.text(label = "(E)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Order-Richness--Alcoa--E.tiff"), width = 8, height = 7, units = "cm", res = 600, compression = "lzw",type="cairo")



## re-use phy_obj from Phyla level
## re-use function calc_Evenness_in_parallel_any_merge_no() defined above
## re-use r1.ps derived above


# number of bootstrap resamples
B <- 100

# iterate in parallel
cl<-makeCluster( detectCores()-1 ) # detectCores()-1
registerDoParallel(cl) # remember to stopcluster()

b_out <- foreach(j=1:B, .packages=c('phyloseq')) %dopar%
  calc_Evenness_in_parallel_any_merge_no(phy_obj=phy_obj, merge_by="group" )

stopCluster(cl)

length(b_out) # 100
names(b_out[[1]]) # "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

b_out[[1]]


### 1st calculate diversity from one rarefying step
### then append bootstrap-derived uncertainty...(?)


# # rarefy #1
# seed <- 123
# r1.ps <- rarefy_even_depth(phy_obj, sample.size = min(sample_sums(phy_obj)),
#                            rngseed = seed, replace = FALSE, trimOTUs = TRUE, verbose = TRUE)
# 
# min(taxa_sums(r1.ps)) # 1
# sample_sums(r1.ps) # all 17485
# 
# ntaxa(r1.ps) #  30751
# sum(sample_sums(r1.ps)) # 629460

r1.ps
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 265 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 71 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 265 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 265 tips and 264 internal nodes ]:
#   taxa are rows

shan.r1.ps <- plot_richness(r1.ps, measures=c("Shannon"))
shan.r1.ps
rich.r1.ps <- plot_richness(r1.ps, measures=c("Observed"))
rich.r1.ps

identical(shan.r1.ps$data$samples,rich.r1.ps$data$samples)
identical(shan.r1.ps$data$group,rich.r1.ps$data$group)

# export data
out <- data.frame(sample=shan.r1.ps$data$samples,evenness=shan.r1.ps$data$value/log(rich.r1.ps$data$value),
                  group=shan.r1.ps$data$group)

out$calc_type <- "rarefyx1"
str(out)


names(out)        # "sample"    "evenness"  "group"     "calc_type"

names(b_out[[1]]) #  "sample"    "evenness"  "group"     "calc_type"
dim(b_out[[1]]) # 7  4

temp <- out

for (j in 1:B) {
  temp <- rbind(temp,b_out[[j]])
}

head(temp)
temp[1:50, ]
tail(temp)

names(temp)
# "sample"    "evenness"  "group"     "calc_type"


str(temp)


levels(temp$group)
#  "2"   "8"   "14"  "17"  "25"  "29"  "Ref"

temp$group <- factor(temp$group, 
                     levels = c("2",   "8",   "14",  "17",  "25",  "29",  "Ref"),
                     labels=c("2 yr",   "8 yr",   "14 yr",  "17 yr",  "25 yr",  "29 yr",  "Ref"), ordered = TRUE)


melt.out <- melt(temp,id.vars = c("group","calc_type"), measure.vars = "evenness")


## plot

cols <- cols.group.alcoa



# plot
p <- ggplot(data=melt.out, aes(x=group, value)) + 
  #ggtitle("b") +
  geom_violin(data = melt.out[ which(melt.out$calc_type == "bootstrap"), ], aes(color = group) ) +
  scale_colour_manual(values = cols) +
  #geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="gray20", shape=1  ) +
  geom_point(data = melt.out[ which(melt.out$calc_type == "rarefyx1"), ], color="grey40", size=1 ) +
  theme_bw() +
  #theme_classic() +
  #theme(axis.text.x  = element_text(angle=90, vjust=0.5) ) +
  theme(axis.text.x  = element_text(angle=60,hjust=1, vjust=1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  labs(x = NULL, y = "Evenness (J')") +
  theme(legend.position="none")

p

grid.text(label = "(F)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Order-Evenness--Alcoa--F.tiff"), width = 8, height = 7, units = "cm", res = 600, compression = "lzw",type="cairo")


#------------------------


#### Taxa relative abundance heat maps - Alcoa-Huntly - Phylum, Class, Order
#------------------------

### PHYLUM LEVEL


phy_in <- phy.alcoa.withTree.PHYLUM
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33 tips and 32 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$`Date since change in Land Use`)
# 1987 1991 1999 2002 2008 2014  N/A 
#    3    3    3    3    3    3   18 


phy_in@sam_data$group <- phy_in@sam_data$`Date since change in Land Use`
sel <- which(phy_in@sam_data$Notes == "adjacent natural") # qty 18
phy_in@sam_data$group[sel] <- "Ref"
unique(phy_in@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

phy_in@sam_data$group <- factor( phy_in@sam_data$group, 
                                 levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                 labels = c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)


table( phy_in@sam_data$group )
# 2 yr  8 yr 14 yr 17 yr 25 yr 29 yr   Ref 
#    3     3     3     3     3     3    18 



# convert to % relative abundance

rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
rel_abun
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 71 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33 tips and 32 internal nodes ]:
#   taxa are rows

rel_abun@sam_data$group
as.data.frame(rel_abun@sam_data)[ ,c("Sample_id", "group")]

rel_abun@sam_data$Rehabilitation_age <- rel_abun@sam_data$group
rel_abun@sam_data$Rehabilitation_age_sample_ID <- paste0(rel_abun@sam_data$Rehabilitation_age," (X",rel_abun@sam_data$Sample_id,")")

#same order as factor for group
samp_order <- sample_names(rel_abun)[order(rel_abun@sam_data$group)]
samp_order

rank_names(rel_abun) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

# remname phyla
rel_abun@tax_table[ ,"Phylum"] <- gsub(pattern = "p__", replacement = "", x = rel_abun@tax_table[ ,"Phylum"])

# p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age", taxa.label = "Phylum",
#                             sample.order = samp_order )
p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age_sample_ID", taxa.label = "Phylum",
                            sample.order = samp_order )
pp <- p + guides(fill=guide_legend(title= "Relative\nabundance (%)" ))

pp

#grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Phylum--Alcoa--A.tiff"), width = 20, height = 12, units = "cm", res = 400, compression = "lzw",type="cairo")







phy_in <- phy.alcoa.withTree.CLASS
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 110 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 110 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 110 tips and 109 internal nodes ]:
#   taxa are rows

sel <- which(phy_in@tax_table[ ,"Class"] != "c__Unclassified") # 94 taxa are assigned to a Class


table(phy_in@sam_data$`Date since change in Land Use`)
# 1987 1991 1999 2002 2008 2014  N/A 
#    3    3    3    3    3    3   18 


phy_in@sam_data$group <- phy_in@sam_data$`Date since change in Land Use`
sel <- which(phy_in@sam_data$Notes == "adjacent natural") # qty 18
phy_in@sam_data$group[sel] <- "Ref"
unique(phy_in@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

phy_in@sam_data$group <- factor( phy_in@sam_data$group, 
                                 levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                 labels = c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)

table( phy_in@sam_data$group )
# 2 yr  8 yr 14 yr 17 yr 25 yr 29 yr   Ref 
# 3     3     3     3     3     3    18 



# convert to % relative abundance

rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
rel_abun
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 110 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 71 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 110 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 110 tips and 109 internal nodes ]:
#   taxa are rows

rel_abun@sam_data$group
as.data.frame(rel_abun@sam_data)[ ,c("Sample_id", "group")]

rel_abun@sam_data$Rehabilitation_age <- rel_abun@sam_data$group
rel_abun@sam_data$Rehabilitation_age_sample_ID <- paste0(rel_abun@sam_data$Rehabilitation_age," (X",rel_abun@sam_data$Sample_id,")")


#same order as factor for group
samp_order <- sample_names(rel_abun)[order(rel_abun@sam_data$group)]
samp_order

rank_names(rel_abun) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"


## get a better label for taxa (incl unclassified)

## remove prefixes

tax <- as.data.frame(rel_abun@tax_table)
class(tax)  # "data.frame"
taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__")
names(tax) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

for (i in 1:length(names(tax))) {
  #i<-1
  tax[ , names(tax)[i] ] <- gsub(pattern = taxa_prefix[i], replacement = "", x = tax[ , names(tax)[i] ])
  print(paste0("completed ",i))
}

tax

this_rank <- "Class"

labels <- tax[ , this_rank]

## go to next available taxa
## establish better label for Unclassified

ranks <- c("Genus","Family","Order","Class","Phylum","Kingdom")

idx_start <- which(ranks == this_rank)

for (i in 1:length(labels)) {
  #i<-4
  idx_rank <- idx_start
  
  # if "Unclassified"
  if (labels[i] == "Unclassified") {
    this_taxaname <- "Unclassified"
    while (this_taxaname == "Unclassified") {
      idx_rank <- idx_rank +1
      get_rank <- ranks[idx_rank]
      this_taxaname <- as.character( tax[ i , get_rank ] )
    }
    
    if (this_rank=="Genus") {
      this_fullname <- paste0( this_taxaname," spp." )
    } else {
      this_fullname <- paste0("Unclassified (",get_rank,": ",this_taxaname,")")
    }
    #labels[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    labels[i] <- this_fullname
  #   
  #   # if Species is available  
  # } else {
  #   
  #   this_Genus <- sub(pattern="g__", x=tab$Genus[i], replacement="")
  #   this_Species <- sub(pattern="s__", x=tab$Species[i], replacement="")
  #   this_fullname <- paste0(this_Genus," ",this_Species)
  #   tab$label[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
  # }
  # 
  # # add OTUId
  # tab$OTU[i] <- paste0( row.names(tab)[i])
  # print(paste0("completed ",i))
  }
}

tax[ , this_rank]
labels
tax[ , this_rank] <- labels

rel_abun@tax_table <- tax_table(as.matrix(tax))
this_rank # "Class"
rel_abun@tax_table[ , this_rank]


# p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age", taxa.label = "Class",
#                             sample.order = samp_order )
p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age_sample_ID", taxa.label = "Class",
                            sample.order = samp_order )
pp <- p + guides(fill=guide_legend(title= "Relative\nabundance (%)" ))

pp

#grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Class--Huntly.tiff"), width = 20, height = 24, units = "cm", res = 400, compression = "lzw",type="cairo")

ntaxa(rel_abun) # 110
length(grep(pattern = "Unclassified", x = rel_abun@tax_table[ ,this_rank])) # 16






phy_in <- phy.alcoa.withTree.ORDER
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 266 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 70 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 266 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 266 tips and 265 internal nodes ]:
#   taxa are rows

sel <- which(phy_in@tax_table[ ,"Order"] != "o__Unclassified") # 202 taxa are assigned to an Order


table(phy_in@sam_data$`Date since change in Land Use`)
# 1987 1991 1999 2002 2008 2014  N/A 
#    3    3    3    3    3    3   18 


phy_in@sam_data$group <- phy_in@sam_data$`Date since change in Land Use`
sel <- which(phy_in@sam_data$Notes == "adjacent natural") # qty 18
phy_in@sam_data$group[sel] <- "Ref"
unique(phy_in@sam_data$group)
# "2014" "Ref"  "2008" "1991" "1987" "2002" "1999"

phy_in@sam_data$group <- factor( phy_in@sam_data$group, 
                                 levels = c("2014","2008","2002","1999","1991","1987","Ref"),
                                 labels = c("2 yr","8 yr","14 yr","17 yr","25 yr","29 yr","Ref"), ordered=TRUE)
# Alcoa (sampled in 2016; rehab 2-29 yr old)

table( phy_in@sam_data$group )
# 2 yr  8 yr 14 yr 17 yr 25 yr 29 yr   Ref 
# 3     3     3     3     3     3    18 



# convert to % relative abundance

rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
rel_abun
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 266 taxa and 36 samples ]:
#   sample_data() Sample Data:        [ 36 samples by 71 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 266 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 266 tips and 265 internal nodes ]:
#   taxa are rows

rel_abun@sam_data$group
as.data.frame(rel_abun@sam_data)[ ,c("Sample_id", "group")]

rel_abun@sam_data$Rehabilitation_age <- rel_abun@sam_data$group
rel_abun@sam_data$Rehabilitation_age_sample_ID <- paste0(rel_abun@sam_data$Rehabilitation_age," (X",rel_abun@sam_data$Sample_id,")")

#same order as factor for group
samp_order <- sample_names(rel_abun)[order(rel_abun@sam_data$group)]
samp_order

rank_names(rel_abun) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"


## get a better label for taxa (incl unclassified)

## remove prefixes

tax <- as.data.frame(rel_abun@tax_table)
class(tax)  # "data.frame"
taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__")
names(tax) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

for (i in 1:length(names(tax))) {
  #i<-1
  tax[ , names(tax)[i] ] <- gsub(pattern = taxa_prefix[i], replacement = "", x = tax[ , names(tax)[i] ])
  print(paste0("completed ",i))
}

tax

this_rank <- "Order"

labels <- tax[ , this_rank]

## go to next available taxa
## establish better label for Unclassified

ranks <- c("Genus","Family","Order","Class","Phylum","Kingdom")

idx_start <- which(ranks == this_rank)

for (i in 1:length(labels)) {
  #i<-1
  idx_rank <- idx_start
  
  # if "Unclassified"
  if (labels[i] == "Unclassified") {
    this_taxaname <- "Unclassified"
    while (this_taxaname == "Unclassified") {
      idx_rank <- idx_rank +1
      get_rank <- ranks[idx_rank]
      this_taxaname <- as.character( tax[ i , get_rank ] )
    }
    
    if (this_rank=="Genus") {
      this_fullname <- paste0( this_taxaname," spp." )
    } else {
      this_fullname <- paste0("Unclassified (",get_rank,": ",this_taxaname,")")
    }
    #labels[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    labels[i] <- this_fullname
    #   
    #   # if Species is available  
    # } else {
    #   
    #   this_Genus <- sub(pattern="g__", x=tab$Genus[i], replacement="")
    #   this_Species <- sub(pattern="s__", x=tab$Species[i], replacement="")
    #   this_fullname <- paste0(this_Genus," ",this_Species)
    #   tab$label[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    # }
    # 
    # # add OTUId
    # tab$OTU[i] <- paste0( row.names(tab)[i])
    # print(paste0("completed ",i))
  }
}

tax[ , this_rank]
labels
tax[ , this_rank] <- labels

rel_abun@tax_table <- tax_table(as.matrix(tax))
temp <- rel_abun
#rel_abun <- temp
dim(tax) # 266   6
ntaxa(rel_abun) # 266

length(grep(pattern = "Unclassified", x = rel_abun@tax_table[ ,this_rank])) # 64


# this plot has to many rows to display effectively in one panel
# therefore splitting it up into two panels

# part 1
rel_abun <- prune_taxa(taxa = taxa_names(rel_abun)[1:132], x = rel_abun )
# part 2
rel_abun <- prune_taxa(taxa = taxa_names(rel_abun)[133:266], x = rel_abun )


rel_abun@tax_table

this_rank # "Order"
rel_abun@tax_table[ , this_rank]

rank_names(rel_abun)
taxa_names(rel_abun)

# p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age", taxa.label = "Order",
#                             sample.order = samp_order )
p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age_sample_ID", taxa.label = "Order",
                            sample.order = samp_order )
pp <- p + guides(fill=guide_legend(title= "Relative\nabundance (%)" ))

pp

grid.text(label = "(Part 1)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Order-Part1-Huntly.tiff"), width = 20, height = 28, units = "cm", res = 400, compression = "lzw",type="cairo")

grid.text(label = "(Part 2)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Order-Part2-Huntly.tiff"), width = 20, height = 28, units = "cm", res = 400, compression = "lzw",type="cairo")



#------------------------


#### Taxa relative abundance heat maps - Iluka-Eneabba - Phylum, Class, Order
#------------------------

### PHYLUM LEVEL


phy_in <- phy.iluka_eneabba.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33636 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33636 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33636 tips and 33634 internal nodes ]:
#   taxa are rows

rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"  

phy_in <- phyloseq::tax_glom(physeq = phy_in, taxrank = "Phylum")
phy.iluka_eneabba.withTree.PHYLUM <- phy_in

phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 28 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 28 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 28 tips and 27 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 


phy_in@sam_data$group <- phy_in@sam_data$year.of.rehab.or.ref
unique(phy_in@sam_data$group)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000" 
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                                labels = c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref"),
                                ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


table( phy_in@sam_data$group )
#7 yr 10 yr 15 yr 19 yr 24 yr 30 yr 38 yr   Ref 
#   2     2     3     2     3     2     3     9 




# convert to % relative abundance

rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
rel_abun
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 28 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 28 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 28 tips and 27 internal nodes ]:
#   taxa are rows

rel_abun@sam_data$group
as.data.frame(rel_abun@sam_data)[ ,c("Sample_ID", "group")]

rel_abun@sam_data$Rehabilitation_age <- rel_abun@sam_data$group
rel_abun@sam_data$Rehabilitation_age_sample_ID <- paste0(rel_abun@sam_data$Rehabilitation_age," (",sample_names(rel_abun),")")

#same order as factor for group
samp_order <- sample_names(rel_abun)[order(rel_abun@sam_data$group)]
samp_order

rank_names(rel_abun) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

# remname phyla
rel_abun@tax_table[ ,"Phylum"] <- gsub(pattern = "p__", replacement = "", x = rel_abun@tax_table[ ,"Phylum"])

# plot_heatmap(rel_abun)
# names(rel_abun@sam_data)
# plot_heatmap(rel_abun, taxa.label = "Phylum", sample.order = samp_order )

# p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age", taxa.label = "Phylum",
#                             sample.order = samp_order )
p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age_sample_ID", taxa.label = "Phylum",
                            sample.order = samp_order )
pp <- p + guides(fill=guide_legend(title= "Relative\nabundance (%)" ))

pp

#grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Phylum--Iluka-Eneabba.tiff"), width = 20, height = 14, units = "cm", res = 400, compression = "lzw",type="cairo")





## CLASS
phy_in <- phy.iluka_eneabba.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33636 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33636 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33636 tips and 33634 internal nodes ]:
#   taxa are rows

rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"  

phy_in <- phyloseq::tax_glom(physeq = phy_in, taxrank = "Class")
phy.iluka_eneabba.withTree.CLASS <- phy_in

#phy_in <- phy.iluka_eneabba.withTree.CLASS

phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 76 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 76 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 76 tips and 75 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 


phy_in@sam_data$group <- phy_in@sam_data$year.of.rehab.or.ref
unique(phy_in@sam_data$group)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000" 
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                                labels = c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref"),
                                ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


table( phy_in@sam_data$group )
#7 yr 10 yr 15 yr 19 yr 24 yr 30 yr 38 yr   Ref 
#   2     2     3     2     3     2     3     9 



# convert to % relative abundance

rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
rel_abun
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 76 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 76 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 76 tips and 75 internal nodes ]:
#   taxa are rows

rel_abun@sam_data$group
as.data.frame(rel_abun@sam_data)[ ,c("Sample_ID", "group")]

rel_abun@sam_data$Rehabilitation_age <- rel_abun@sam_data$group
rel_abun@sam_data$Rehabilitation_age_sample_ID <- paste0(rel_abun@sam_data$Rehabilitation_age," (",sample_names(rel_abun),")")


#same order as factor for group
samp_order <- sample_names(rel_abun)[order(rel_abun@sam_data$group)]
samp_order

rank_names(rel_abun) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"


## get a better label for taxa (incl unclassified)

## remove prefixes

tax <- as.data.frame(rel_abun@tax_table)
class(tax)  # "data.frame"
taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__")
names(tax) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

for (i in 1:length(names(tax))) {
  #i<-1
  tax[ , names(tax)[i] ] <- gsub(pattern = taxa_prefix[i], replacement = "", x = tax[ , names(tax)[i] ])
  print(paste0("completed ",i))
}

tax

this_rank <- "Class"

labels <- tax[ , this_rank]

## go to next available taxa
## establish better label for Unclassified

ranks <- c("Genus","Family","Order","Class","Phylum","Kingdom")

idx_start <- which(ranks == this_rank)

for (i in 1:length(labels)) {
  #i<-4
  idx_rank <- idx_start
  
  # if "Unclassified"
  if (labels[i] == "Unclassified") {
    this_taxaname <- "Unclassified"
    while (this_taxaname == "Unclassified") {
      idx_rank <- idx_rank +1
      get_rank <- ranks[idx_rank]
      this_taxaname <- as.character( tax[ i , get_rank ] )
    }
    
    if (this_rank=="Genus") {
      this_fullname <- paste0( this_taxaname," spp." )
    } else {
      this_fullname <- paste0("Unclassified (",get_rank,": ",this_taxaname,")")
    }
    #labels[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    labels[i] <- this_fullname
    #   
    #   # if Species is available  
    # } else {
    #   
    #   this_Genus <- sub(pattern="g__", x=tab$Genus[i], replacement="")
    #   this_Species <- sub(pattern="s__", x=tab$Species[i], replacement="")
    #   this_fullname <- paste0(this_Genus," ",this_Species)
    #   tab$label[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    # }
    # 
    # # add OTUId
    # tab$OTU[i] <- paste0( row.names(tab)[i])
    # print(paste0("completed ",i))
  }
}

tax[ , this_rank]
labels
tax[ , this_rank] <- labels

rel_abun@tax_table <- tax_table(as.matrix(tax))
this_rank # "Class"
rel_abun@tax_table[ , this_rank]


# p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age", taxa.label = "Class",
#                             sample.order = samp_order )
p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age_sample_ID", taxa.label = "Class",
                            sample.order = samp_order )
pp <- p + guides(fill=guide_legend(title= "Relative\nabundance (%)" ))

pp

#grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Class--Eneabba.tiff"), width = 20, height = 24, units = "cm", res = 400, compression = "lzw",type="cairo")

ntaxa(rel_abun) # 76
length(grep(pattern = "Unclassified", x = rel_abun@tax_table[ ,this_rank])) # 13




## ORDER
phy_in <- phy.iluka_eneabba.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 33636 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 33636 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 33636 tips and 33634 internal nodes ]:
#   taxa are rows

rank_names(phy_in) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"  

phy_in <- phyloseq::tax_glom(physeq = phy_in, taxrank = "Order")
phy.iluka_eneabba.withTree.ORDER <- phy_in

phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 194 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 194 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 194 tips and 193 internal nodes ]:
#   taxa are rows


table(phy_in@sam_data$year.of.rehab.or.ref)
# 1981 1989 1995 2000 2004 2009 2012  REF 
#    3    2    3    2    3    2    2    9 


phy_in@sam_data$group <- phy_in@sam_data$year.of.rehab.or.ref
unique(phy_in@sam_data$group)
# "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000" 
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                                labels = c("7 yr", "10 yr", "15 yr", "19 yr", "24 yr", "30 yr", "38 yr", "Ref"),
                                ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)


table( phy_in@sam_data$group )
#7 yr 10 yr 15 yr 19 yr 24 yr 30 yr 38 yr   Ref 
#   2     2     3     2     3     2     3     9 



# convert to % relative abundance

rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
rel_abun
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 194 taxa and 26 samples ]:
#   sample_data() Sample Data:        [ 26 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 194 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 194 tips and 193 internal nodes ]:
#   taxa are rows

rel_abun@sam_data$group
as.data.frame(rel_abun@sam_data)[ ,c("Sample_id", "group")]

rel_abun@sam_data$Rehabilitation_age <- rel_abun@sam_data$group
rel_abun@sam_data$Rehabilitation_age_sample_ID <- paste0(rel_abun@sam_data$Rehabilitation_age," (",sample_names(rel_abun),")")

#same order as factor for group
samp_order <- sample_names(rel_abun)[order(rel_abun@sam_data$group)]
samp_order

rank_names(rel_abun) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"


## get a better label for taxa (incl unclassified)

## remove prefixes

tax <- as.data.frame(rel_abun@tax_table)
class(tax)  # "data.frame"
taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__")
names(tax) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

for (i in 1:length(names(tax))) {
  #i<-1
  tax[ , names(tax)[i] ] <- gsub(pattern = taxa_prefix[i], replacement = "", x = tax[ , names(tax)[i] ])
  print(paste0("completed ",i))
}

tax

this_rank <- "Order"

labels <- tax[ , this_rank]

## go to next available taxa
## establish better label for Unclassified

ranks <- c("Genus","Family","Order","Class","Phylum","Kingdom")

idx_start <- which(ranks == this_rank)

for (i in 1:length(labels)) {
  #i<-1
  idx_rank <- idx_start
  
  # if "Unclassified"
  if (labels[i] == "Unclassified") {
    this_taxaname <- "Unclassified"
    while (this_taxaname == "Unclassified") {
      idx_rank <- idx_rank +1
      get_rank <- ranks[idx_rank]
      this_taxaname <- as.character( tax[ i , get_rank ] )
    }
    
    if (this_rank=="Genus") {
      this_fullname <- paste0( this_taxaname," spp." )
    } else {
      this_fullname <- paste0("Unclassified (",get_rank,": ",this_taxaname,")")
    }
    #labels[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    labels[i] <- this_fullname
    #   
    #   # if Species is available  
    # } else {
    #   
    #   this_Genus <- sub(pattern="g__", x=tab$Genus[i], replacement="")
    #   this_Species <- sub(pattern="s__", x=tab$Species[i], replacement="")
    #   this_fullname <- paste0(this_Genus," ",this_Species)
    #   tab$label[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    # }
    # 
    # # add OTUId
    # tab$OTU[i] <- paste0( row.names(tab)[i])
    # print(paste0("completed ",i))
  }
}

tax[ , this_rank]
labels
tax[ , this_rank] <- labels

rel_abun@tax_table <- tax_table(as.matrix(tax))
temp <- rel_abun
#rel_abun <- temp
dim(tax) # 194   6
ntaxa(rel_abun) # 194

length(grep(pattern = "Unclassified", x = rel_abun@tax_table[ ,this_rank])) # 43


# this plot has to many rows to display effectively in one panel
# therefore splitting it up into two panels

# part 1
rel_abun <- prune_taxa(taxa = taxa_names(rel_abun)[1:100], x = rel_abun )
# part 2
rel_abun <- prune_taxa(taxa = taxa_names(rel_abun)[101:194], x = rel_abun )


rel_abun@tax_table

this_rank # "Order"
rel_abun@tax_table[ , this_rank]

rank_names(rel_abun)
taxa_names(rel_abun)

# p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age", taxa.label = "Order",
#                             sample.order = samp_order )
p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age_sample_ID", taxa.label = "Order",
                            sample.order = samp_order )
pp <- p + guides(fill=guide_legend(title= "Relative\nabundance (%)" ))

pp

grid.text(label = "(Part 1)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Order-Part1-Eneabba.tiff"), width = 20, height = 28, units = "cm", res = 400, compression = "lzw",type="cairo")

grid.text(label = "(Part 2)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Order-Part2-Eneabba.tiff"), width = 20, height = 27, units = "cm", res = 400, compression = "lzw",type="cairo")



#------------------------


#### Taxa relative abundance heat maps - Worsley-South32 - Phylum, Class, Order
#------------------------

### PHYLUM LEVEL


phy_in <- phy.south32.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 54671 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 54671 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 54671 tips and 54669 internal nodes ]:
#   taxa are rows

phy_in <- tax_glom(physeq = phy_in, taxrank = "Phylum")

phy.south32.withTree.PHYLUM <- phy_in

table(phy_in@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6


phy_in@sam_data$group <- phy_in@sam_data$`year of rehab or ref`
unique(phy_in@sam_data$group)
# "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                                labels = c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


table( phy_in@sam_data$group )
# 2 yr  8 yr 12 yr 14 yr 17 yr 20 yr 23 yr 28 yr   Ref 
#    3     3     1     2     2     2     4     2     6 




# convert to % relative abundance

rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
rel_abun
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 35 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 35 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 35 tips and 34 internal nodes ]:
#   taxa are rows


rel_abun@sam_data$group
as.data.frame(rel_abun@sam_data)[ ,c("Sample_ID", "group")]

rel_abun@sam_data$Rehabilitation_age <- rel_abun@sam_data$group
rel_abun@sam_data$Rehabilitation_age_sample_ID <- paste0(rel_abun@sam_data$Rehabilitation_age," (",sample_names(rel_abun),")")


#same order as factor for group
samp_order <- sample_names(rel_abun)[order(rel_abun@sam_data$group)]
samp_order

rank_names(rel_abun) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

# remname phyla
rel_abun@tax_table[ ,"Phylum"] <- gsub(pattern = "p__", replacement = "", x = rel_abun@tax_table[ ,"Phylum"])

# p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age", taxa.label = "Phylum",
#                             sample.order = samp_order )
p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age_sample_ID", taxa.label = "Phylum",
                            sample.order = samp_order )
pp <- p + guides(fill=guide_legend(title= "Relative\nabundance (%)" ))

pp

#grid.text(label = "(A)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Phylum--South32-Worsley.tiff"), width = 20, height = 16, units = "cm", res = 400, compression = "lzw",type="cairo")





## CLASS

phy_in <- phy.south32.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 54671 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 54671 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 54671 tips and 54669 internal nodes ]:
#   taxa are rows

phy_in <- tax_glom(physeq = phy_in, taxrank = "Class")

phy.south32.withTree.CLASS <- phy_in

sel <- which(phy_in@tax_table[ ,"Class"] != "c__Unclassified") # 94 taxa are assigned to a Class


table(phy_in@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6


phy_in@sam_data$group <- phy_in@sam_data$`year of rehab or ref`
unique(phy_in@sam_data$group)
# "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                                labels = c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


table( phy_in@sam_data$group )
# 2 yr  8 yr 12 yr 14 yr 17 yr 20 yr 23 yr 28 yr   Ref 
#    3     3     1     2     2     2     4     2     6 




# convert to % relative abundance

rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
rel_abun
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 110 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 110 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 110 tips and 109 internal nodes ]:
#   taxa are rows

rel_abun@sam_data$group
as.data.frame(rel_abun@sam_data)[ ,c("Sample_ID", "group")]

rel_abun@sam_data$Rehabilitation_age <- rel_abun@sam_data$group
rel_abun@sam_data$Rehabilitation_age_sample_ID <- paste0(rel_abun@sam_data$Rehabilitation_age," (",sample_names(rel_abun),")")

#same order as factor for group
samp_order <- sample_names(rel_abun)[order(rel_abun@sam_data$group)]
samp_order

rank_names(rel_abun) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"


## get a better label for taxa (incl unclassified)

## remove prefixes

tax <- as.data.frame(rel_abun@tax_table)
class(tax)  # "data.frame"
taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__")
names(tax) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

for (i in 1:length(names(tax))) {
  #i<-1
  tax[ , names(tax)[i] ] <- gsub(pattern = taxa_prefix[i], replacement = "", x = tax[ , names(tax)[i] ])
  print(paste0("completed ",i))
}

tax

this_rank <- "Class"

labels <- tax[ , this_rank]

## go to next available taxa
## establish better label for Unclassified

ranks <- c("Genus","Family","Order","Class","Phylum","Kingdom")

idx_start <- which(ranks == this_rank)

for (i in 1:length(labels)) {
  #i<-4
  idx_rank <- idx_start
  
  # if "Unclassified"
  if (labels[i] == "Unclassified") {
    this_taxaname <- "Unclassified"
    while (this_taxaname == "Unclassified") {
      idx_rank <- idx_rank +1
      get_rank <- ranks[idx_rank]
      this_taxaname <- as.character( tax[ i , get_rank ] )
    }
    
    if (this_rank=="Genus") {
      this_fullname <- paste0( this_taxaname," spp." )
    } else {
      this_fullname <- paste0("Unclassified (",get_rank,": ",this_taxaname,")")
    }
    #labels[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    labels[i] <- this_fullname
    #   
    #   # if Species is available  
    # } else {
    #   
    #   this_Genus <- sub(pattern="g__", x=tab$Genus[i], replacement="")
    #   this_Species <- sub(pattern="s__", x=tab$Species[i], replacement="")
    #   this_fullname <- paste0(this_Genus," ",this_Species)
    #   tab$label[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    # }
    # 
    # # add OTUId
    # tab$OTU[i] <- paste0( row.names(tab)[i])
    # print(paste0("completed ",i))
  }
}

tax[ , this_rank]
labels
tax[ , this_rank] <- labels

rel_abun@tax_table <- tax_table(as.matrix(tax))
this_rank # "Class"
rel_abun@tax_table[ , this_rank]


# p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age", taxa.label = "Class",
#                             sample.order = samp_order )
p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age_sample_ID", taxa.label = "Class",
                            sample.order = samp_order )
pp <- p + guides(fill=guide_legend(title= "Relative\nabundance (%)" ))

pp

#grid.text(label = "(B)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Class--Worsley.tiff"), width = 20, height = 24, units = "cm", res = 400, compression = "lzw",type="cairo")

ntaxa(rel_abun) # 110
length(grep(pattern = "Unclassified", x = rel_abun@tax_table[ ,this_rank])) # 16





## ORDER

phy_in <- phy.south32.withTree
phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 54671 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 54671 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 54671 tips and 54669 internal nodes ]:
#   taxa are rows

phy_in <- tax_glom(physeq = phy_in, taxrank = "Order")
phy.south32.withTree.ORDER <- phy_in

phy_in
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 244 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 139 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 244 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 244 tips and 243 internal nodes ]:
#   taxa are rows

sel <- which(phy_in@tax_table[ ,"Order"] != "o__Unclassified") # 185 taxa are assigned to an Order


table(phy_in@sam_data$`year of rehab or ref`)
# 1991 1996 1999 2002 2005 2007 2011 2017  REF 
#    2    4    2    2    2    1    3    3    6


phy_in@sam_data$group <- phy_in@sam_data$`year of rehab or ref`
unique(phy_in@sam_data$group)
# "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"
phy_in@sam_data$group <- factor(phy_in@sam_data$group, 
                                levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                                labels = c("2 yr","8 yr","12 yr","14 yr","17 yr","20 yr","23 yr","28 yr","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)


table( phy_in@sam_data$group )
# 2 yr  8 yr 12 yr 14 yr 17 yr 20 yr 23 yr 28 yr   Ref 
#    3     3     1     2     2     2     4     2     6 




# convert to % relative abundance

rel_abun <- transform_sample_counts(phy_in, function(x) 100*x / sum(x) )
rel_abun
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 244 taxa and 25 samples ]:
#   sample_data() Sample Data:        [ 25 samples by 140 sample variables ]:
#   tax_table()   Taxonomy Table:     [ 244 taxa by 6 taxonomic ranks ]:
#   phy_tree()    Phylogenetic Tree:  [ 244 tips and 243 internal nodes ]:
#   taxa are rows

rel_abun@sam_data$group
as.data.frame(rel_abun@sam_data)[ ,c("Sample_ID", "group")]

rel_abun@sam_data$Rehabilitation_age <- rel_abun@sam_data$group
rel_abun@sam_data$Rehabilitation_age_sample_ID <- paste0(rel_abun@sam_data$Rehabilitation_age," (",sample_names(rel_abun),")")

#same order as factor for group
samp_order <- sample_names(rel_abun)[order(rel_abun@sam_data$group)]
samp_order

rank_names(rel_abun) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"


## get a better label for taxa (incl unclassified)

## remove prefixes

tax <- as.data.frame(rel_abun@tax_table)
class(tax)  # "data.frame"
taxa_prefix <- c("k__", "p__", "c__", "o__", "f__", "g__")
names(tax) # "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"

for (i in 1:length(names(tax))) {
  #i<-1
  tax[ , names(tax)[i] ] <- gsub(pattern = taxa_prefix[i], replacement = "", x = tax[ , names(tax)[i] ])
  print(paste0("completed ",i))
}

tax

this_rank <- "Order"

labels <- tax[ , this_rank]

## go to next available taxa
## establish better label for Unclassified

ranks <- c("Genus","Family","Order","Class","Phylum","Kingdom")

idx_start <- which(ranks == this_rank)

for (i in 1:length(labels)) {
  #i<-1
  idx_rank <- idx_start
  
  # if "Unclassified"
  if (labels[i] == "Unclassified") {
    this_taxaname <- "Unclassified"
    while (this_taxaname == "Unclassified") {
      idx_rank <- idx_rank +1
      get_rank <- ranks[idx_rank]
      this_taxaname <- as.character( tax[ i , get_rank ] )
    }
    
    if (this_rank=="Genus") {
      this_fullname <- paste0( this_taxaname," spp." )
    } else {
      this_fullname <- paste0("Unclassified (",get_rank,": ",this_taxaname,")")
    }
    #labels[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    labels[i] <- this_fullname
    #   
    #   # if Species is available  
    # } else {
    #   
    #   this_Genus <- sub(pattern="g__", x=tab$Genus[i], replacement="")
    #   this_Species <- sub(pattern="s__", x=tab$Species[i], replacement="")
    #   this_fullname <- paste0(this_Genus," ",this_Species)
    #   tab$label[i] <- paste0(sub(pattern = "p__",replacement = "", tab$Phylum[i]),": ",this_fullname)
    # }
    # 
    # # add OTUId
    # tab$OTU[i] <- paste0( row.names(tab)[i])
    # print(paste0("completed ",i))
  }
}

tax[ , this_rank]
labels
tax[ , this_rank] <- labels

rel_abun@tax_table <- tax_table(as.matrix(tax))
temp <- rel_abun
#rel_abun <- temp
dim(tax) # 244   6
ntaxa(rel_abun) # 244

length(grep(pattern = "Unclassified", x = rel_abun@tax_table[ ,this_rank])) # 59


# this plot has to many rows to display effectively in one panel
# therefore splitting it up into two panels

# part 1
rel_abun <- prune_taxa(taxa = taxa_names(rel_abun)[1:122], x = rel_abun )
# part 2
rel_abun <- prune_taxa(taxa = taxa_names(rel_abun)[123:244], x = rel_abun )


rel_abun@tax_table

this_rank # "Order"
rel_abun@tax_table[ , this_rank]

rank_names(rel_abun)
taxa_names(rel_abun)

# p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age", taxa.label = "Order",
#                             sample.order = samp_order )
p <- phyloseq::plot_heatmap(rel_abun, sample.label = "Rehabilitation_age_sample_ID", taxa.label = "Order",
                            sample.order = samp_order )
pp <- p + guides(fill=guide_legend(title= "Relative\nabundance (%)" ))

pp

grid.text(label = "(Part 1)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Order-Part1-Worsley.tiff"), width = 20, height = 28, units = "cm", res = 400, compression = "lzw",type="cairo")

grid.text(label = "(Part 2)", x = unit(0.0375, "npc") , y = unit(0.97,"npc"), gp=gpar(fontsize=12, fontface="bold") )
dev.print(tiff, file = paste0(workdir,"/plots/","Heatmap-Order-Part2-Worsley.tiff"), width = 20, height = 28, units = "cm", res = 400, compression = "lzw",type="cairo")



#------------------------



#### Site data
#-------------------------

library(sf); packageVersion("sf") #‘0.9.6’

# define projection WGS84
WGS84<-"+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0"


## Alcoa
site_data <- as(phy.alcoa@sam_data, "data.frame")
class(site_data) # "data.frame"
str(site_data)

site_data$longitude_GDA94 <- site_data$long.in.decimal.degrees 
# convert latitude to negative value (as this is South)
site_data$lat.....in.decimal.degrees <- -site_data$lat.....in.decimal.degrees
site_data$latitude_GDA94 <- site_data$lat.....in.decimal.degrees

site.shp <- site_data
coordinates(site.shp) <- ~ long.in.decimal.degrees + lat.....in.decimal.degrees   # of form: ~x+y
site.shp@proj4string # NA
site.shp@proj4string <- CRS(WGS84)
names(site_data)

head(site.shp@data)
# 
#        longitude_GDA94 latitude_GDA94 group
# X39041        116.1425       32.51827  3 yr
# X39043        116.1424       32.51761   Ref
# X39045        116.1837       32.52046  3 yr
# X39047        116.1843       32.51976   Ref
# X39049        116.1719       32.52362  3 yr
# X39051        116.1718       32.52270   Ref


df <- site.shp@data
df$group <- df$Date.since.change.in.Land.Use
unique(df$group) # "2014" "N/A"  "2008" "1991" "1987" "2002" "1999"
df$group <- factor(df$group, 
                   levels = c("2014","2008","2002","1999","1991","1987","N/A"),
                   labels = c("3 yr","9 yr","15 yr","18 yr","26 yr","30 yr","Ref"), ordered=TRUE)
# Alcoa (sampled in 2017; rehab 3-30 yr old)

site.shp@data <- df

plot(site.shp, col = site.shp@data$group)

#site.sf <- st_as_sf(site.shp)

min_x <- min(df$longitude_GDA94)
max_x <- max(df$longitude_GDA94)
min_y <- min(df$latitude_GDA94)
max_y <- max(df$latitude_GDA94)

mid_x <- min_x + 0.5*(max_x - min_x)
mid_y <- min_y + 0.5*(max_y - min_y)

min_x;mid_x;max_x
# [1] 116.0363
# [1] 116.1302
# [1] 116.2241

min_y;mid_y;max_y
# [1] -32.62009
# [1] -32.56885
# [1] -32.51761

mid_x; mid_y
# [1] 116.1302
# [1] -32.56885

write.csv(x = data.frame(site="Huntly-Alcoa",mid_x_long=mid_x,mid_y_lay= mid_y), file = "Alcoa-mid-coords.csv")

names(df)
# ...
# [69] "longitude_GDA94"                                       "latitude_GDA94"                                       
# [71] "group" 
write.csv(x = df[ ,c("Sample_id","longitude_GDA94","latitude_GDA94" , "group" ) ], file = "Alcoa-all-site-coords.csv")





### Iluka-Enneabba

site_data <- as(phy.iluka_eneabba@sam_data, "data.frame")
class(site_data) # "data.frame"
str(site_data)

site_data$longitude_GDA94 <- site_data$Longitude..decimal.degrees.
# convert latitude to negative value (as this is South)
site_data$Latitude..decimal.degrees. <- -site_data$Latitude..decimal.degrees.
site_data$latitude_GDA94 <- site_data$Latitude..decimal.degrees.

site.shp <- site_data
coordinates(site.shp) <- ~ Longitude..decimal.degrees. + Latitude..decimal.degrees.   # of form: ~x+y
site.shp@proj4string # NA
site.shp@proj4string <- CRS(WGS84)
names(site_data)

head(site.shp@data)
# longitude_GDA94 latitude_GDA94
# X138408        115.3068      -29.85854
# X138410        115.3077      -29.85744
# X138412        115.3045      -29.90079
# X138418        115.2935      -29.91182
# X138424        115.2825      -29.92179
# X138426        115.2799      -29.92265


df <- site.shp@data
df$group <- df$year.of.rehab.or.ref
unique(df$group) # "REF"  "1989" "1995" "1981" "2012" "2009" "2004" "2000"
df$group <- factor(df$group, 
                   levels = c("2012","2009","2004","2000","1995","1989","1981","REF"),
                   labels = c("7", "10", "15", "19", "24", "30", "38", "Ref"), ordered=TRUE)
# Iluka-Eneabba (sampled in 2019; rehab 7-38 yr old)
                   

site.shp@data <- df

plot(site.shp, col = site.shp@data$group)

#site.sf <- st_as_sf(site.shp)

min_x <- min(df$longitude_GDA94)
max_x <- max(df$longitude_GDA94)
min_y <- min(df$latitude_GDA94)
max_y <- max(df$latitude_GDA94)

mid_x <- min_x + 0.5*(max_x - min_x)
mid_y <- min_y + 0.5*(max_y - min_y)

min_x;mid_x;max_x
# [1] 115.2545
# [1] 115.2811
# [1] 115.3077

min_y;mid_y;max_y
# [1] -29.94226
# [1] -29.89962
# [1] -29.85698

mid_x; mid_y
# [1] 115.2811
# [1] -29.89962

write.csv(x = data.frame(site="Eneabba-Iluka",mid_x_long=mid_x,mid_y_lay= mid_y), file = "Iluka-Eneabba-mid-coords.csv")

names(df)
# ...
# [69] "longitude_GDA94"                                       "latitude_GDA94"                                       
# [71] "group" 
write.csv(x = df[ ,c("Sample_ID","longitude_GDA94","latitude_GDA94" , "group" ) ], file = "Iluka-Eneabba-all-site-coords.csv")





### South32

site_data <- as(phy.south32@sam_data, "data.frame")
class(site_data) # "data.frame"
str(site_data)

site_data$longitude_GDA94 <- site_data$Longitude..decimal.degrees.
# convert latitude to negative value (as this is South)
site_data$Latitude..decimal.degrees. <- -site_data$Latitude..decimal.degrees.
site_data$latitude_GDA94 <- site_data$Latitude..decimal.degrees.

site.shp <- site_data
coordinates(site.shp) <- ~ Longitude..decimal.degrees. + Latitude..decimal.degrees.   # of form: ~x+y
site.shp@proj4string # NA
site.shp@proj4string <- CRS(WGS84)
names(site_data)

head(site.shp@data)
# longitude_GDA94 latitude_GDA94
# X138358        116.4657      -32.92270
# X138360        116.4717      -32.92646
# X138362        116.4738      -32.92846
# X138364        116.4561      -32.92999
# X138366        116.4555      -32.93049
# X138368        116.4539      -32.91135


df <- site.shp@data
df$group <- df$year.of.rehab.or.ref
unique(df$group) # "1996" "REF"  "2017" "2002" "2007" "1999" "2011" "2005" "1991"
df$group <- factor(df$group, 
                   levels = c("2017","2011","2007","2005","2002","1999","1996","1991","REF"),
                   labels = c("2","8","12","14","17","20","23","28","Ref"), ordered=TRUE)
# South32 (sampled in 2019; rehab 2-28 yr old)

site.shp@data <- df

plot(site.shp, col = site.shp@data$group)

#site.sf <- st_as_sf(site.shp)

min_x <- min(df$longitude_GDA94)
max_x <- max(df$longitude_GDA94)
min_y <- min(df$latitude_GDA94)
max_y <- max(df$latitude_GDA94)

mid_x <- min_x + 0.5*(max_x - min_x)
mid_y <- min_y + 0.5*(max_y - min_y)

min_x;mid_x;max_x
# [1] 116.4278
# [1] 116.4574
# [1] 116.487

min_y;mid_y;max_y
# [1] -32.99188
# [1] -32.93971
# [1] -32.88755

mid_x; mid_y
# [1] 116.4574
# [1] -32.93971

write.csv(x = data.frame(site="Worsley-South32",mid_x_long=mid_x,mid_y_lay= mid_y), file = "Worsley-South32-mid-coords.csv")

names(df)
# ...
# [69] "longitude_GDA94"                                       "latitude_GDA94"                                       
# [71] "group" 
write.csv(x = df[ ,c("Sample_ID","longitude_GDA94","latitude_GDA94" , "group" ) ], file = "Worsley-South32-all-site-coords.csv")




#-------------------------

### END of code

